<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Preprocess Short Reads — preprocessReads • QuasR</title>


<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous" />

<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script>

<!-- bootstrap-toc -->
<link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous" />
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script>

<!-- headroom.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>




<meta property="og:title" content="Preprocess Short Reads — preprocessReads" />
<meta property="og:description" content="Truncate sequences, remove parts matching to adapters and filter out low 
quality or low complexity sequences from (compressed) 'fasta' or 'fastq' files." />




<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->



  </head>

  <body data-spy="scroll" data-target="#toc">
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">QuasR</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">1.33.3</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../articles/QuasR.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Preprocess Short Reads</h1>
    
    <div class="hidden name"><code>preprocessReads.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>Truncate sequences, remove parts matching to adapters and filter out low 
quality or low complexity sequences from (compressed) 'fasta' or 'fastq' files.</p>
    </div>

    <pre class="usage"><span class='fu'>preprocessReads</span><span class='op'>(</span>
  <span class='va'>filename</span>,
  outputFilename <span class='op'>=</span> <span class='cn'>NULL</span>,
  filenameMate <span class='op'>=</span> <span class='cn'>NULL</span>,
  outputFilenameMate <span class='op'>=</span> <span class='cn'>NULL</span>,
  truncateStartBases <span class='op'>=</span> <span class='cn'>NULL</span>,
  truncateEndBases <span class='op'>=</span> <span class='cn'>NULL</span>,
  Lpattern <span class='op'>=</span> <span class='st'>""</span>,
  Rpattern <span class='op'>=</span> <span class='st'>""</span>,
  max.Lmismatch <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/rep.html'>rep</a></span><span class='op'>(</span><span class='fl'>0</span><span class='op'>:</span><span class='fl'>2</span>, <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='fl'>6</span>, <span class='fl'>3</span>, <span class='fl'>100</span><span class='op'>)</span><span class='op'>)</span>,
  max.Rmismatch <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/rep.html'>rep</a></span><span class='op'>(</span><span class='fl'>0</span><span class='op'>:</span><span class='fl'>2</span>, <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='fl'>6</span>, <span class='fl'>3</span>, <span class='fl'>100</span><span class='op'>)</span><span class='op'>)</span>,
  with.Lindels <span class='op'>=</span> <span class='cn'>FALSE</span>,
  with.Rindels <span class='op'>=</span> <span class='cn'>FALSE</span>,
  minLength <span class='op'>=</span> <span class='fl'>14L</span>,
  nBases <span class='op'>=</span> <span class='fl'>2L</span>,
  complexity <span class='op'>=</span> <span class='cn'>NULL</span>,
  nrec <span class='op'>=</span> <span class='fl'>1000000L</span>,
  clObj <span class='op'>=</span> <span class='cn'>NULL</span>
<span class='op'>)</span></pre>

    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>filename</th>
      <td><p>the name(s) of the input sequence file(s).</p></td>
    </tr>
    <tr>
      <th>outputFilename</th>
      <td><p>the name(s) of the output sequence file(s).</p></td>
    </tr>
    <tr>
      <th>filenameMate</th>
      <td><p>for paired-end experiments, the name(s) of the
input sequence file(s) containing the second read (mate) of each pair.</p></td>
    </tr>
    <tr>
      <th>outputFilenameMate</th>
      <td><p>for paired-end experiments, the name(s) of the
output sequence file(s) containing the second read (mate) of each pair.</p></td>
    </tr>
    <tr>
      <th>truncateStartBases</th>
      <td><p>integer(1): the number of bases to be truncated 
(removed) from the beginning of each sequence.</p></td>
    </tr>
    <tr>
      <th>truncateEndBases</th>
      <td><p>integer(1): the number of bases to be truncated 
(removed) from the end of each sequence.</p></td>
    </tr>
    <tr>
      <th>Lpattern</th>
      <td><p>character(1): the left (5'-end) adapter sequence.</p></td>
    </tr>
    <tr>
      <th>Rpattern</th>
      <td><p>character(1): the right (3'-end) adapter sequence.</p></td>
    </tr>
    <tr>
      <th>max.Lmismatch</th>
      <td><p>mismatch tolerance when searching for matches of 
<code>Lpattern</code> (see &#8216;Details&#8217;).</p></td>
    </tr>
    <tr>
      <th>max.Rmismatch</th>
      <td><p>mismatch tolerance when searching for matches of 
<code>Rpattern</code> (see &#8216;Details&#8217;).</p></td>
    </tr>
    <tr>
      <th>with.Lindels</th>
      <td><p>if <code>TRUE</code>, indels are allowed in the alignments of 
the suffixes of <code>Lpattern</code> with the subject, at its beginning 
(see &#8216;Details&#8217;).</p></td>
    </tr>
    <tr>
      <th>with.Rindels</th>
      <td><p>same as <code>with.Lindels</code> but for alignments of the 
prefixes of <code>Rpattern</code> with the subject, at its end (see 
&#8216;Details&#8217;).</p></td>
    </tr>
    <tr>
      <th>minLength</th>
      <td><p>integer(1): the minimal allowed sequence length.</p></td>
    </tr>
    <tr>
      <th>nBases</th>
      <td><p>integer(1): the maximal number of Ns allowed per sequence.</p></td>
    </tr>
    <tr>
      <th>complexity</th>
      <td><p><code>NULL</code> (default) or numeric(1): If not <code>NULL</code>,
the minimal sequence complexity, as a fraction of the average complexity
in the human genome (~3.9bits). For example, <code>complexity = 0.5</code> will
filter out sequences that do not have at least half the complexity of the
human genome. See &#8216;Details&#8217; on how the complexity is calculated.</p></td>
    </tr>
    <tr>
      <th>nrec</th>
      <td><p>integer(1): the number of sequence records to read at a time.</p></td>
    </tr>
    <tr>
      <th>clObj</th>
      <td><p>a cluster object to be used for parallel processing of multiple 
files (see &#8216;Details&#8217;).</p></td>
    </tr>
    </table>

    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p>A matrix with summary statistics on the processed sequences, containing:</p><ul>
<li><p>One column per input file (or pair of input files for paired-end
                                   experiments).</p></li>
<li><p>The number of sequences or sequence pairs in rows:</p><ul>
<li><p><code>totalSequences</code> - the total number in the input</p></li>
<li><p><code>matchTo5pAdaptor</code> - matching to <code>Lpattern</code></p></li>
<li><p><code>matchTo3pAdaptor</code> - matching to <code>Rpattern</code></p></li>
<li><p><code>tooShort</code> - shorter than <code>minLength</code></p></li>
<li><p><code>tooManyN</code> - more than <code>nBases</code> Ns</p></li>
<li><p><code>lowComplexity</code> - relative complexity below <code>complexity</code></p></li>
<li><p><code>totalPassed</code> - the number of sequences/sequence pairs
      that pass all filtering criteria and were written to the output file(s).</p></li>
</ul></li>
</ul>

    <h2 class="hasAnchor" id="details"><a class="anchor" href="#details"></a>Details</h2>

    <p>Sequence files can be in fasta or fastq format, and can be compressed by
either gzip, bzip2 or xz (extensions .gz, .bz2 or .xz). Multiple files
can be processed by a single call to <code>preprocessReads</code>; in that
case all sequence file vectors must have identical lengths.</p>
<p><code>nrec</code> can be used to limit the memory usage when processing
large input files. <code>preprocessReads</code> iteratively loads chunks of
<code>nrec</code> sequences from the input until all data been processed.</p>
<p>Sequence pairs from paired-end experiments can be processed by
specifying pairs of input and output files (<code>filenameMate</code> and
<code>outputFilenameMate</code> arguments). In that case, it is assumed that
pairs appear in the same order in the two input files, and only pairs
in which both reads pass all filtering criteria are written to the
output files, maintaining the consistent ordering.</p>
<p>If output files are compressed, the processed sequences are first
written to temporary files (created in the same directory as the final
output file), and the output files are generated at the end by compressing
the temporary files.</p>
<p>For the trimming of left and/or right flanking sequences (adapters) from
sequence reads, the <code><a href='https://rdrr.io/pkg/Biostrings/man/trimLRPatterns.html'>trimLRPatterns</a></code> function
from package <span class="pkg">Biostrings</span> is used, and the arguments <code>Lpattern</code>,
<code>Rpattern</code>, <code>max.Lmismatch</code>, <code>max.Rmismatch</code>,
<code>with.Lindels</code> and <code>with.Rindels</code> are used in the call to
<code>trimLRPatterns</code>. <code>Lfixed</code> and <code>Rfixed</code> arguments
of <code>trimLRPatterns</code> are set to <code>TRUE</code>, thus only fixed
patterns (without IUPAC codes for ambigous bases) can be
used. Currently, trimming of adapters is only supported for single read
experiments.</p>
<p>Sequence complexity (\(H\)) is calculated based on the dinucleotide
composition using the formula (Shannon entropy): $$H = -\sum_i {f_i \log_2 f_i},$$
where \(f_i\) is the fraction of dinucleotide \(i\) from all
dinucleotides in the sequence. Sequence reads that fulfill the condition
\(H/H_r \ge c\) are retained (not filtered out), where \(H_r =
3.908\) is the reference complexity in bits obtained from the human
genome, and \(c\) is the value given to the argument <code>complexity</code>.</p>
<p>If an object that inherits from class <code>cluster</code> is provided to
the <code>clObj</code> argument, for example an object returned by
<code><a href='https://rdrr.io/r/parallel/makeCluster.html'>makeCluster</a></code> from package <span class="pkg">parallel</span>,
multiple files will be processed in parallel using
<code><a href='https://rdrr.io/r/parallel/clusterApply.html'>clusterMap</a></code> from package <span class="pkg">parallel</span>.</p>
    <h2 class="hasAnchor" id="see-also"><a class="anchor" href="#see-also"></a>See also</h2>

    <div class='dont-index'><p><code><a href='https://rdrr.io/pkg/Biostrings/man/trimLRPatterns.html'>trimLRPatterns</a></code> from package <span class="pkg">Biostrings</span>,
<code><a href='https://rdrr.io/r/parallel/makeCluster.html'>makeCluster</a></code> from package <span class="pkg">parallel</span></p></div>
    <h2 class="hasAnchor" id="author"><a class="anchor" href="#author"></a>Author</h2>

    <p>Anita Lerch, Dimos Gaidatzis and Michael Stadler</p>

    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <pre class="examples"><div class='input'><span class='co'># sample files</span>
<span class='va'>infiles</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/system.file.html'>system.file</a></span><span class='op'>(</span>package<span class='op'>=</span><span class='st'>"QuasR"</span>, <span class='st'>"extdata"</span>,
                       <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='st'>"rna_1_1.fq.bz2"</span>,<span class='st'>"rna_1_2.fq.bz2"</span><span class='op'>)</span><span class='op'>)</span>
<span class='va'>outfiles</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/paste.html'>paste</a></span><span class='op'>(</span><span class='fu'><a href='https://rdrr.io/r/base/tempfile.html'>tempfile</a></span><span class='op'>(</span>pattern<span class='op'>=</span><span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='st'>"output_1_"</span>,<span class='st'>"output_2_"</span><span class='op'>)</span><span class='op'>)</span>,<span class='st'>".fastq"</span>,sep<span class='op'>=</span><span class='st'>""</span><span class='op'>)</span>
<span class='co'># single read example</span>
<span class='fu'>preprocessReads</span><span class='op'>(</span><span class='va'>infiles</span>, <span class='va'>outfiles</span>, nBases<span class='op'>=</span><span class='fl'>0</span>, complexity<span class='op'>=</span><span class='fl'>0.6</span><span class='op'>)</span>
</div><div class='output co'>#&gt; <span class='message'>  filtering /Users/runner/work/_temp/Library/QuasR/extdata/rna_1_1.fq.bz2</span></div><div class='output co'>#&gt; <span class='message'>  filtering /Users/runner/work/_temp/Library/QuasR/extdata/rna_1_2.fq.bz2</span></div><div class='output co'>#&gt;                  rna_1_1.fq.bz2 rna_1_2.fq.bz2
#&gt; totalSequences             3002           3002
#&gt; matchTo5pAdapter              0              0
#&gt; matchTo3pAdapter              0              0
#&gt; tooShort                      0              0
#&gt; tooManyN                      3              0
#&gt; lowComplexity                 0              0
#&gt; totalPassed                2999           3002</div><div class='input'><span class='fu'><a href='https://rdrr.io/r/base/unlink.html'>unlink</a></span><span class='op'>(</span><span class='va'>outfiles</span><span class='op'>)</span>
<span class='co'># paired-end example</span>
<span class='fu'>preprocessReads</span><span class='op'>(</span>filename<span class='op'>=</span><span class='va'>infiles</span><span class='op'>[</span><span class='fl'>1</span><span class='op'>]</span>,
                outputFilename<span class='op'>=</span><span class='va'>outfiles</span><span class='op'>[</span><span class='fl'>1</span><span class='op'>]</span>,
                filenameMate<span class='op'>=</span><span class='va'>infiles</span><span class='op'>[</span><span class='fl'>2</span><span class='op'>]</span>,
                outputFilenameMate<span class='op'>=</span><span class='va'>outfiles</span><span class='op'>[</span><span class='fl'>2</span><span class='op'>]</span>,
                nBases<span class='op'>=</span><span class='fl'>0</span>, complexity<span class='op'>=</span><span class='fl'>0.6</span><span class='op'>)</span>
</div><div class='output co'>#&gt; <span class='message'>  filtering /Users/runner/work/_temp/Library/QuasR/extdata/rna_1_1.fq.bz2 and</span>
#&gt; <span class='message'>    /Users/runner/work/_temp/Library/QuasR/extdata/rna_1_2.fq.bz2</span></div><div class='output co'>#&gt;                  rna_1_1.fq.bz2:rna_1_2.fq.bz2
#&gt; totalSequences                            3002
#&gt; matchTo5pAdapter                            NA
#&gt; matchTo3pAdapter                            NA
#&gt; tooShort                                     0
#&gt; tooManyN                                     3
#&gt; lowComplexity                                0
#&gt; totalPassed                               2999</div><div class='input'><span class='fu'><a href='https://rdrr.io/r/base/unlink.html'>unlink</a></span><span class='op'>(</span><span class='va'>outfiles</span><span class='op'>)</span>

</div></pre>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top">
      <h2 data-toc-skip>Contents</h2>
    </nav>
  </div>
</div>


      <footer>
      <div class="copyright">
  <p>Developed by Anita Lerch, Charlotte Soneson, Dimos Gaidatzis, Michael Stadler.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
   </div>

  


  </body>
</html>


