<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Quantify alignments — qCount • QuasR</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="Quantify alignments — qCount"><meta property="og:description" content="Quantify alignments from sequencing data."><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">QuasR</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.35.1</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../articles/QuasR.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
      </ul><ul class="nav navbar-nav navbar-right"></ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Quantify alignments</h1>
    
    <div class="hidden name"><code>qCount.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>Quantify alignments from sequencing data.</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="fu">qCount</span><span class="op">(</span>
  <span class="va">proj</span>,
  <span class="va">query</span>,
  reportLevel <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="cn">NULL</span>, <span class="st">"gene"</span>, <span class="st">"exon"</span>, <span class="st">"promoter"</span>, <span class="st">"junction"</span><span class="op">)</span>,
  selectReadPosition <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"start"</span>, <span class="st">"end"</span><span class="op">)</span>,
  shift <span class="op">=</span> <span class="fl">0L</span>,
  orientation <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"any"</span>, <span class="st">"same"</span>, <span class="st">"opposite"</span><span class="op">)</span>,
  useRead <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"any"</span>, <span class="st">"first"</span>, <span class="st">"last"</span><span class="op">)</span>,
  auxiliaryName <span class="op">=</span> <span class="cn">NULL</span>,
  mask <span class="op">=</span> <span class="cn">NULL</span>,
  collapseBySample <span class="op">=</span> <span class="cn">TRUE</span>,
  includeSpliced <span class="op">=</span> <span class="cn">TRUE</span>,
  includeSecondary <span class="op">=</span> <span class="cn">TRUE</span>,
  mapqMin <span class="op">=</span> <span class="fl">0L</span>,
  mapqMax <span class="op">=</span> <span class="fl">255L</span>,
  absIsizeMin <span class="op">=</span> <span class="cn">NULL</span>,
  absIsizeMax <span class="op">=</span> <span class="cn">NULL</span>,
  maxInsertSize <span class="op">=</span> <span class="fl">500L</span>,
  clObj <span class="op">=</span> <span class="cn">NULL</span>
<span class="op">)</span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>
    <dl><dt>proj</dt>
<dd><p>A <code><a href="qProject-class.html">qProject</a></code> object representing a
sequencing experiment as returned by <code><a href="qAlign.html">qAlign</a></code></p></dd>
<dt>query</dt>
<dd><p>An object of type <code><a href="https://rdrr.io/pkg/GenomicRanges/man/GRanges-class.html" class="external-link">GRanges</a></code>, 
<code><a href="https://rdrr.io/pkg/GenomicRanges/man/GRangesList-class.html" class="external-link">GRangesList</a></code> or 
<code><a href="https://rdrr.io/pkg/GenomicFeatures/man/TxDb-class.html" class="external-link">TxDb</a></code> with the regions to be
quantified. The type of <code>query</code> will determine the mode of
quantification (see ‘Details’). For
<code>reportLevel="junction"</code>, <code>query</code> is ignored and can also
be <code>NULL</code>.</p></dd>
<dt>reportLevel</dt>
<dd><p>Level of quantification (<code>query</code> of type
<code>TxDb</code> or <code>NULL</code>), one of</p><ul><li><p><code>gene</code> (default): one value per gene</p></li>
<li><p><code>exon</code>: one value per exon</p></li>
<li><p><code>promoter</code>: one value per promoter</p></li>
<li><p><code>junction</code>: one count per detected exon-exon junction</p></li>
</ul></dd>
<dt>selectReadPosition</dt>
<dd><p>defines the part of the alignment that has
to be contained within a query region to produce an overlap (see
‘Details’). Possible values are:</p><ul><li><p><code>start</code> (default): start of the alignment</p></li>
<li><p><code>end</code>: end of the alignment</p></li>
</ul></dd>
<dt>shift</dt>
<dd><p>controls the shifting alignments towards their 3'-end before
quantification. <code>shift</code> can be one of:</p><ul><li><p>an “integer” vector of the same length as the
  number of alignment files</p></li>
<li><p>a single “integer” value</p></li>
<li><p>the character string <code>"halfInsert"</code> (only available for
  paired-end experiments)</p></li>
</ul><p>The default of <code>0</code> will not shift any alignments.</p></dd>
<dt>orientation</dt>
<dd><p>sets the required orientation of the alignments relative 
to the query region in order to be counted, one of:</p><ul><li><p><code>any</code> (default): count alignment on the same and opposite strand</p></li>
<li><p><code>same</code> : count only alignment on the same strand</p></li>
<li><p><code>opposite</code> : count only alignment on the opposite strand</p></li>
</ul></dd>
<dt>useRead</dt>
<dd><p>For paired-end experiments, selects the read mate whose
alignments should be counted, one of:</p><ul><li><p><code>any</code> (default): count all alignments</p></li>
<li><p><code>first</code> : count only alignments from the first read</p></li>
<li><p><code>last</code> : count only alignments from the last read</p></li>
</ul></dd>
<dt>auxiliaryName</dt>
<dd><p>Which bam files to use in an experiments with 
auxiliary alignments (see ‘Details’).</p></dd>
<dt>mask</dt>
<dd><p>If not <code>NULL</code>, a <code><a href="https://rdrr.io/pkg/GenomicRanges/man/GRanges-class.html" class="external-link">GRanges</a></code>
object with reference regions to be masked, i.e. excluded from the
quantification, such as unmappable or highly repetitive regions (see
‘Details’).</p></dd>
<dt>collapseBySample</dt>
<dd><p>If <code>TRUE</code> (the default), sum alignment
counts from bam files with the same sample name.</p></dd>
<dt>includeSpliced</dt>
<dd><p>If <code>TRUE</code> (the default), include spliced
alignments when counting. A spliced alignment is defined as an
alignment with a gap in the read of at least 60 bases.</p></dd>
<dt>includeSecondary</dt>
<dd><p>If <code>TRUE</code> (the default), include alignments
with the secondary bit (0x0100) set in the <code>FLAG</code> when counting.</p></dd>
<dt>mapqMin</dt>
<dd><p>Minimal mapping quality of alignments to be included when
counting (mapping quality must be greater than or equal to
<code>mapqMin</code>). Valid values are between 0 and 255. The default (0)
will include all alignments.</p></dd>
<dt>mapqMax</dt>
<dd><p>Maximal mapping quality of alignments to be included when
counting (mapping quality must be less than or equal to <code>mapqMax</code>).
Valid values are between 0 and 255. The default (255) will include
all alignments.</p></dd>
<dt>absIsizeMin</dt>
<dd><p>For paired-end experiments, minimal absolute insert
size (TLEN field in SAM Spec v1.4) of alignments to be included when
counting. Valid values are greater than 0 or <code>NULL</code> (default),
which will not apply any minimum insert size filtering.</p></dd>
<dt>absIsizeMax</dt>
<dd><p>For paired-end experiments, maximal absolute insert
size (TLEN field in SAM Spec v1.4) of alignments to be included when
counting. Valid values are greater than 0 or <code>NULL</code> (default),
which will not apply any maximum insert size filtering.</p></dd>
<dt>maxInsertSize</dt>
<dd><p>Maximal fragment size of the paired-end experiment. 
This parameter is used if <code>shift="halfInsert"</code> and will
ensure that query regions are made wide enough to emcompass all
alignment pairs whose mid falls into the query region. The default
value is <code>500</code> bases.</p></dd>
<dt>clObj</dt>
<dd><p>A cluster object to be used for parallel processing (see
‘Details’).</p></dd>
</dl></div>
    <div id="value">
    <h2>Value</h2>
    <p>A <code>matrix</code> with effective query regions width in the first
column, and alignment counts in subsequent columns, or a
<code>GRanges</code> object if <code>reportLevel="junction"</code>.
The effective query region width returned as first column in the
matrix is calculated by the number of unique, non-masked bases in the
reference sequence that contributed to the count of this query
name (irrespective if the bases were covered by alignments or not).
An effective width of zero indicates that the region was fully
masked and will have zero counts in all samples.
The alignment counts in the matrix are contained from column two
onwards. For projects with allele-specific quantification, i.e. if a
file with single nucleotide polymorphisms was supplied to the
<code>snpFile</code> argument of <code><a href="qAlign.html">qAlign</a></code>, there will be
three columns per bam file (number of alignments for Reference, 
Unknown and Alternative genotypes, with suffixed _R, _U and
_A). Otherwise there is a single columns per bam file.
If <code>collapseBySample</code>=<code>TRUE</code>, groups of bam files with identical 
sample name are combined by summing their alignment counts.
For <code>reportLevel="junction"</code>, the return value is a
<code>GRanges</code> object. The start and end coordinates correspond to the
first and last base in each detected intron. Plus- and minus-strand
alignments are quantified separately, so that in an unstranded RNA-seq
experiment, the same intron may be represented twice; once for each
strand. The counts for each sample are contained in the <code>mcols</code>of the <code>GRanges</code> object.</p>
    </div>
    <div id="details">
    <h2>Details</h2>
    <p><code>qCount</code> is used to count alignments in each sample from a
<code>qProject</code> object. The features to be quantified, together with
the mode of quantification, are specified by the <code>query</code>
argument, which is one of:</p><ul><li><p><code><a href="https://rdrr.io/pkg/GenomicRanges/man/GRanges-class.html" class="external-link">GRanges</a></code>: Overlapping alignments
  are counted separately for each coordinate region. If multiple
  regions have identical names, their counts will be summed, counting
  each alignment only once even if it overlaps more than one of these
  regions. Alignments may be counted more than once if they overlap
  multiple regions that have different names.
  This mode is for example used to quantify ChIP-seq alignments in
  promoter regions, or gene expression levels in an RNA-seq experiment
  (using a <code>query</code> with exon regions named by gene).</p></li>
<li><p><code><a href="https://rdrr.io/pkg/GenomicRanges/man/GRangesList-class.html" class="external-link">GRangesList</a></code>: Alignments are
  counted and summed for each list element in <code>query</code> if they
  overlap with any of the regions contained in the list element. The
  order of the list elements defines a hierarchy for quantification:
  Alignment will only be counted for the first element (the one with
  the lowest index in <code>query</code>) that they overlap, but not for any
  potential further list elements containing overlapping regions.
  This mode can be used to hierarchically and uniquely count (assign)
  each alignment to a one of several groups of regions (the elements
  in <code>query</code>), for example to estimate the fractions of different
  classes of RNA in an RNA-seq experiment (rRNA, tRNA, snRNA, snoRNA,
  mRNA, etc.)</p></li>
<li><p><code><a href="https://rdrr.io/pkg/GenomicFeatures/man/TxDb-class.html" class="external-link">TxDb</a></code>: Used to extract
  regions from annotation and report alignment counts depending on the
  value of <code>reportLevel</code>. If <code>reportLevel="exon"</code>,
  alignments overlapping each exon in <code>query</code> are counted.
  If <code>reportLevel="gene"</code>, alignment counts for all exons of a
  gene will be summed, counting each alignment only once even if it
  overlaps multiple annotated exons of a gene. These are useful to
  calculate exon or gene expression levels in RNA-seq experiments
  based on the annotation in a <code>TxDb</code> object. If
  <code>reportLevel="promoter"</code>, the <code>promoters</code> function from package
  <span class="pkg">GenomicFeatures</span> is used with default arguments to extract
  promoter regions around transcript start sites, e.g. to quantify
  alignments inf a ChIP-seq experiment.</p></li>
<li><p>any of the above or <code>NULL</code> for
  <code>reportLevel="junction"</code>: The <code>query</code> argument is ignored
  if <code>reportLevel</code> is set to <code>"junction"</code>, and <code>qCount</code>
  will count the number of alignments supporting each exon-exon
  junction detected in any of the samples in <code>proj</code>. The
  arguments <code>selectReadPosition</code>, <code>shift</code>,
  <code>orientation</code>, <code>useRead</code> and <code>mask</code> will have no
  effect in this quantification mode.</p></li>
</ul><p>The additional arguments allow to fine-tune the quantification:</p>
<p><code>selectReadPosition</code> defines the part of the alignment that has
to be contained within a query region for an overlap. The values
<code>start</code> (default) and <code>end</code> refer to the biological start
(5'-end) and end (3'-end) of the alignment. For example, the
<code>start</code> of an alignment on the plus strand is its leftmost
(lowest) base, and the <code>end</code> of an alignment on the minus strand
is also the leftmost base.</p>
<p><code>shift</code> allows on-the-fly shifting of alignments towards their
3'-end prior to overlap determination and counting. This can be
helpful to increase resolution of ChIP-seq experiments by moving
alignments by half the immuno-precipitated fragment size towards the
middle of fragments. <code>shift</code> is either an “integer” vector
with one value per alignment file in <code>proj</code>, or a single
“integer” value, in which case all alignment files will be
shifted by the same value. For paired-end experiments, it can be
alternatively set to "halfInsert", which will estimate the true
fragment size from the distance between aligned read pairs and shift
the alignments accordingly.</p>
<p><code>orientation</code> controls the interpretation of alignment strand
when counting, relative to the strand of the query region. <code>any</code>
will count all overlapping alignments, irrespective of the alignment
strand (e.g. used in an unstranded RNA-seq experiment). <code>same</code>
will only count the alignments on the same strand as the query region
(e.g. in a stranded RNA-seq experiment that generates reads from the 
sense strand), and <code>opposite</code> will only
count the alignments on the opposite strand from the query region
(e.g. to quantify anti-sense transcription in a stranded RNA-seq
experiment).</p>
<p><code>includeSpliced</code> and <code>includeSecondary</code> can be used to
include or exclude spliced or secondary alignments,
respectively. <code>mapqMin</code> and <code>mapqMax</code> allow to select alignments
based on their mapping qualities. <code>mapqMin</code> and <code>mapqMax</code> can
take integer values between 0 and 255 and equal to
\(-10 log_{10} Pr(\textnormal{mapping position is wrong})\), rounded to the nearest
integer. A value 255 indicates that the mapping quality is not available.</p>
<p>In paired-end experiments, <code>useRead</code> allows to quantify either
all alignments (<code>useRead="any"</code>), or only the first
(<code>useRead="first"</code>) or last (<code>useRead="last"</code>) read from a
read pair or read group. Note that for <code>useRead="any"</code> (the
default), an alignment pair that is fully contained within a query
region will contribute two counts to the value of that
region. <code>absIsizeMin</code> and <code>absIsizeMax</code> can be used to
select alignments based on their insert size (TLEN field in SAM Spec
v1.4).</p>
<p><code>auxiliaryName</code> selects the reference sequence for which
alignments should be quantified. <code>NULL</code> (the default) will
select alignments against the genome. If set to a character string
that matches one of the auxiliary target names (as specified in
the <code>auxiliaryFile</code> argument of <code><a href="qAlign.html">qAlign</a></code>), the
corresponding alignments will be counted.</p>
<p><code>mask</code> can be used to specify a
<code><a href="https://rdrr.io/pkg/GenomicRanges/man/GRanges-class.html" class="external-link">GRanges</a></code> object with regions in the
reference sequence to be excluded from quantification. The regions
will be considered unstranded (<code>strand="*"</code>). Alignments that
overlap with a region in <code>mask</code> will not be counted. Masking may
reduce the effective width of query regions reported by <code>qCount</code>,
even down to zero for regions that are fully contained in <code>mask</code>.</p>
<p>If <code>clObj</code> is set to an object that inherits from class
<code>cluster</code>, for example an object returned by
<code><a href="https://rdrr.io/r/parallel/makeCluster.html" class="external-link">makeCluster</a></code> from package <span class="pkg">parallel</span>, the
quantification task is split into multiple chunks and processed in
parallel using <code><a href="https://rdrr.io/r/parallel/clusterApply.html" class="external-link">clusterMap</a></code>. Currently, not all
tasks will be efficiently parallelized: For example, a single query
region and a single (group of) bam files will not be split into
multiple chunks.</p>
    </div>
    <div id="see-also">
    <h2>See also</h2>
    <div class="dont-index"><p><code><a href="qAlign.html">qAlign</a></code>,
<code><a href="qProject-class.html">qProject</a></code>,
<code><a href="https://rdrr.io/r/parallel/makeCluster.html" class="external-link">makeCluster</a></code> from package <span class="pkg">parallel</span></p></div>
    </div>
    <div id="author">
    <h2>Author</h2>
    <p>Anita Lerch, Dimos Gaidatzis and Michael Stadler</p>
    </div>

    <div id="ref-examples">
    <h2>Examples</h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://bioconductor.org/packages/GenomicRanges" class="external-link">GenomicRanges</a></span><span class="op">)</span></span>
<span class="r-in"><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://bioconductor.org/packages/Biostrings" class="external-link">Biostrings</a></span><span class="op">)</span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Loading required package: XVector</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Attaching package: ‘Biostrings’</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> The following object is masked from ‘package:base’:</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     strsplit</span>
<span class="r-in"><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://bioconductor.org/packages/Rsamtools" class="external-link">Rsamtools</a></span><span class="op">)</span></span>
<span class="r-in"></span>
<span class="r-in"><span class="co"># copy example data to current working directory</span></span>
<span class="r-in"><span class="fu"><a href="https://rdrr.io/r/base/files.html" class="external-link">file.copy</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span>package<span class="op">=</span><span class="st">"QuasR"</span>, <span class="st">"extdata"</span><span class="op">)</span>, <span class="st">"."</span>, recursive<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] TRUE</span>
<span class="r-in"></span>
<span class="r-in"><span class="co"># load genome sequence</span></span>
<span class="r-in"><span class="va">genomeFile</span> <span class="op">&lt;-</span> <span class="st">"extdata/hg19sub.fa"</span></span>
<span class="r-in"><span class="va">gseq</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/Biostrings/man/XStringSet-io.html" class="external-link">readDNAStringSet</a></span><span class="op">(</span><span class="va">genomeFile</span><span class="op">)</span></span>
<span class="r-in"><span class="va">chrRegions</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/GenomicRanges/man/GRanges-class.html" class="external-link">GRanges</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">gseq</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/pkg/IRanges/man/IRanges-constructor.html" class="external-link">IRanges</a></span><span class="op">(</span>start<span class="op">=</span><span class="fl">1</span>,width<span class="op">=</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/start.html" class="external-link">width</a></span><span class="op">(</span><span class="va">gseq</span><span class="op">)</span>,names<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">gseq</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span class="r-in"></span>
<span class="r-in"><span class="co"># create alignments (paired-end experiment)</span></span>
<span class="r-in"><span class="va">sampleFile</span> <span class="op">&lt;-</span> <span class="st">"extdata/samples_rna_paired.txt"</span></span>
<span class="r-in"><span class="va">proj</span> <span class="op">&lt;-</span> <span class="fu"><a href="qAlign.html">qAlign</a></span><span class="op">(</span><span class="va">sampleFile</span>, <span class="va">genomeFile</span>, splicedAlignment<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Creating .fai file for: /private/var/folders/24/8k48jl6d249_n_qfxwsl6xvm0000gn/T/RtmpyA7fDu/fileb5714eb11350/reference/extdata/hg19sub.fa</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> alignment files missing - need to:</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     create alignment index for the genome</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     create 2 genomic alignment(s)</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Creating an Rbowtie index for /private/var/folders/24/8k48jl6d249_n_qfxwsl6xvm0000gn/T/RtmpyA7fDu/fileb5714eb11350/reference/extdata/hg19sub.fa</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Finished creating index</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Testing the compute nodes...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> OK</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Loading QuasR on the compute nodes...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> preparing to run on 1 nodes...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> done</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Available cores:</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Mac-1647262101956.local: 1</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Performing genomic alignments for 2 samples. See progress in the log file:</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> /private/var/folders/24/8k48jl6d249_n_qfxwsl6xvm0000gn/T/RtmpyA7fDu/fileb5714eb11350/reference/QuasR_log_b57143d06ab2.txt</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Genomic alignments have been created successfully</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-in"></span>
<span class="r-in"><span class="co"># count reads using a "GRanges" query</span></span>
<span class="r-in"><span class="fu">qCount</span><span class="op">(</span><span class="va">proj</span>, query<span class="op">=</span><span class="va">chrRegions</span><span class="op">)</span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> counting alignments...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> done</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>      width Sample1 Sample2</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> chr1 40000     850    1464</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> chr2 10000    2924    2224</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> chr3 45000    2228    2312</span>
<span class="r-in"><span class="fu">qCount</span><span class="op">(</span><span class="va">proj</span>, query<span class="op">=</span><span class="va">chrRegions</span>, useRead<span class="op">=</span><span class="st">"first"</span><span class="op">)</span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> counting alignments...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> done</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>      width Sample1 Sample2</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> chr1 40000     425     732</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> chr2 10000    1462    1112</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> chr3 45000    1114    1156</span>
<span class="r-in"></span>
<span class="r-in"><span class="co"># hierarchical counting using a "GRangesList" query</span></span>
<span class="r-in"><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">rtracklayer</span><span class="op">)</span></span>
<span class="r-in"><span class="va">annotationFile</span> <span class="op">&lt;-</span> <span class="st">"extdata/hg19sub_annotation.gtf"</span></span>
<span class="r-in"><span class="va">gtfRegions</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/rtracklayer/man/GFFFile-class.html" class="external-link">import.gff</a></span><span class="op">(</span><span class="va">annotationFile</span>, format<span class="op">=</span><span class="st">"gtf"</span>, feature.type<span class="op">=</span><span class="st">"exon"</span><span class="op">)</span></span>
<span class="r-in"><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">gtfRegions</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Vector-class.html" class="external-link">mcols</a></span><span class="op">(</span><span class="va">gtfRegions</span><span class="op">)</span><span class="op">$</span><span class="va">source</span></span>
<span class="r-in"><span class="va">gtfRegionList</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/splitAsList.html" class="external-link">split</a></span><span class="op">(</span><span class="va">gtfRegions</span>, <span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">gtfRegions</span><span class="op">)</span><span class="op">)</span></span>
<span class="r-in"><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">gtfRegionList</span><span class="op">)</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "lincRNA"        "protein_coding" "snoRNA"        </span>
<span class="r-in"></span>
<span class="r-in"><span class="va">res3</span> <span class="op">&lt;-</span> <span class="fu">qCount</span><span class="op">(</span><span class="va">proj</span>, <span class="va">gtfRegionList</span><span class="op">)</span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> hierarchically removing redundancy from 'query'...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> done</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> counting alignments...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> done</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> collapsing counts by query name...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> done</span>
<span class="r-in"><span class="va">res3</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                width Sample1 Sample2</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> lincRNA         1104       0       0</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> protein_coding 21576    5876    5919</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> snoRNA           104       0       0</span>
<span class="r-in"></span>
<span class="r-in"><span class="co"># gene expression levels using a "TxDb" query</span></span>
<span class="r-in"><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="st"><a href="https://bioconductor.org/packages/GenomicFeatures" class="external-link">"GenomicFeatures"</a></span><span class="op">)</span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Loading required package: AnnotationDbi</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Loading required package: Biobase</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Welcome to Bioconductor</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     Vignettes contain introductory material; view with</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     'browseVignettes()'. To cite Bioconductor, see</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     'citation("Biobase")', and for packages 'citation("pkgname")'.</span>
<span class="r-in"><span class="va">genomeRegion</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/Rsamtools/man/scanFa.html" class="external-link">scanFaIndex</a></span><span class="op">(</span><span class="va">genomeFile</span><span class="op">)</span></span>
<span class="r-in"><span class="va">chrominfo</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span>chrom<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/character.html" class="external-link">as.character</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/GenomeInfoDb/man/seqinfo.html" class="external-link">seqnames</a></span><span class="op">(</span><span class="va">genomeRegion</span><span class="op">)</span><span class="op">)</span>,</span>
<span class="r-in">                        length<span class="op">=</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/start.html" class="external-link">end</a></span><span class="op">(</span><span class="va">genomeRegion</span><span class="op">)</span>,</span>
<span class="r-in">                        is_circular<span class="op">=</span><span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Vector-class.html" class="external-link">rep</a></span><span class="op">(</span><span class="cn">FALSE</span>, <span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">genomeRegion</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span class="r-in"><span class="va">txdb</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/GenomicFeatures/man/makeTxDbFromGFF.html" class="external-link">makeTxDbFromGFF</a></span><span class="op">(</span><span class="va">annotationFile</span>, </span>
<span class="r-in">                        format<span class="op">=</span><span class="st">"gtf"</span>, </span>
<span class="r-in">                        chrominfo<span class="op">=</span><span class="va">chrominfo</span>,</span>
<span class="r-in">                        dataSource<span class="op">=</span><span class="st">"Ensembl modified"</span>,</span>
<span class="r-in">                        organism<span class="op">=</span><span class="st">"Homo sapiens"</span><span class="op">)</span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Import genomic features from the file as a GRanges object ... </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> OK</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Prepare the 'metadata' data frame ... </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> OK</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Make the TxDb object ... </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> OK</span>
<span class="r-in"></span>
<span class="r-in"><span class="va">res4</span> <span class="op">&lt;-</span> <span class="fu">qCount</span><span class="op">(</span><span class="va">proj</span>, <span class="va">txdb</span>, reportLevel<span class="op">=</span><span class="st">"gene"</span><span class="op">)</span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> extracting gene regions from TxDb...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> done</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> counting alignments...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> done</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> collapsing counts by query name...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> done</span>
<span class="r-in"><span class="va">res4</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                 width Sample1 Sample2</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> ENSG00000078808  4697     708    1076</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> ENSG00000134075   589    1201    1322</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> ENSG00000134086  4213     282     295</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> ENSG00000171863  5583    2922    2224</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> ENSG00000176022  2793      62     344</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> ENSG00000186827  1721      37       8</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> ENSG00000186891  1370      26       2</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> ENSG00000238345    98       0       0</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> ENSG00000238642   104       0       0</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> ENSG00000247886  1104       0       0</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> ENSG00000252531   163       6       2</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> ENSG00000254999  1233    1839    1970</span>
<span class="r-in"></span>
<span class="r-in"><span class="co"># exon-exon junctions</span></span>
<span class="r-in"><span class="va">res5</span> <span class="op">&lt;-</span> <span class="fu">qCount</span><span class="op">(</span><span class="va">proj</span>, <span class="cn">NULL</span>, reportLevel<span class="op">=</span><span class="st">"junction"</span><span class="op">)</span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> counting junctions...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> done</span>
<span class="r-in"><span class="va">res5</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> GRanges object with 52 ranges and 2 metadata columns:</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>        seqnames      ranges strand |   Sample1   Sample2</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>           &lt;Rle&gt;   &lt;IRanges&gt;  &lt;Rle&gt; | &lt;numeric&gt; &lt;numeric&gt;</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    [1]     chr1 12213-12321      + |         1         0</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    [2]     chr1 12213-12321      - |         2         0</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    [3]     chr1 13085-13371      - |         1         0</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    [4]     chr1 18069-18837      + |         8         7</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    [5]     chr1 18069-18837      - |         6        12</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    ...      ...         ...    ... .       ...       ...</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   [48]     chr1 14166-14362      - |         0         1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   [49]     chr1 19009-19148      - |         0         2</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   [50]     chr1 29327-32271      - |         0         2</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   [51]     chr1   4617-4778      - |         0         1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   [52]     chr3   2504-5589      + |         0         2</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   -------</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   seqinfo: 3 sequences from an unspecified genome; no seqlengths</span>
<span class="r-in"></span>
</code></pre></div>
    </div>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Anita Lerch, Charlotte Soneson, Dimos Gaidatzis, Michael Stadler.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.2.</p>
</div>

      </footer></div>

  


  

  </body></html>

