<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Quantify DNA methylation — qMeth • QuasR</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="Quantify DNA methylation — qMeth"><meta property="og:description" content="Quantify methylation of cytosines from bisulfite sequencing data."><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">QuasR</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.39.2</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../articles/QuasR.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
      </ul><ul class="nav navbar-nav navbar-right"></ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Quantify DNA methylation</h1>
    
    <div class="hidden name"><code>qMeth.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>Quantify methylation of cytosines from bisulfite sequencing data.</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">qMeth</span><span class="op">(</span></span>
<span>  <span class="va">proj</span>,</span>
<span>  query <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  reportLevel <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"C"</span>, <span class="st">"alignment"</span><span class="op">)</span>,</span>
<span>  mode <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"CpGcomb"</span>, <span class="st">"CpG"</span>, <span class="st">"allC"</span>, <span class="st">"var"</span><span class="op">)</span>,</span>
<span>  collapseBySample <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  collapseByQueryRegion <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  asGRanges <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  mask <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  reference <span class="op">=</span> <span class="st">"genome"</span>,</span>
<span>  keepZero <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  mapqMin <span class="op">=</span> <span class="fl">0L</span>,</span>
<span>  mapqMax <span class="op">=</span> <span class="fl">255L</span>,</span>
<span>  clObj <span class="op">=</span> <span class="cn">NULL</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>
    <dl><dt>proj</dt>
<dd><p>A <code>qProject</code> object from a bisulfite sequencing experiment</p></dd>


<dt>query</dt>
<dd><p>A <code>GRanges</code> object with the regions to be
quantified. If <code>NULL</code>, all available target sequences (e.g. the
whole genome) will be analyzed. Available target sequences are
extracted from the header of the first bam file.</p></dd>


<dt>reportLevel</dt>
<dd><p>Report results combined for C's
(<code>reportLevel</code>=“C”), the default) or individually for single
alignments (<code>reportLevel</code>=“alignment”). The latter imposes
further restrictions on some arguments (see ‘Details’).</p></dd>


<dt>mode</dt>
<dd><p>Cytosine quantification mode, one of:</p><ul><li><p><code>CpGcomb</code> : only C's in CpG context (strands combined)</p></li>
<li><p><code>CpG</code> : only C's in CpG context (strands separate)</p></li>
<li><p><code>allC</code> : all C's (strands separate)</p></li>
<li><p><code>var</code> : variant detection (all C's, strands separate)</p></li>
</ul><p><code>CpGcomb</code> is the default.</p></dd>


<dt>collapseBySample</dt>
<dd><p>If <code>TRUE</code>, combine (sum) counts from
bamfiles with the same sample name.</p></dd>


<dt>collapseByQueryRegion</dt>
<dd><p>If <code>TRUE</code>, combine (sum) counts for
all cytosines contained in the same query region.</p></dd>


<dt>asGRanges</dt>
<dd><p>If <code>TRUE</code>, return results as a <code>GRanges</code> object;
if <code>FALSE</code>, the results are returned as a <code>data.frame</code>.</p></dd>


<dt>mask</dt>
<dd><p>An optional <code>GRanges</code> object with genomic regions to
be masked, i.e. excluded from the analysis (e.g. unmappable regions).</p></dd>


<dt>reference</dt>
<dd><p>Source of bam files; can be either “genome”
(then the alignments against the genome are used) or the name of an
auxiliary target sequence (then alignments against this target
sequence will be used). The auxiliary name must correspond to the
name contained in the auxiliary file refered by the
<code>auxiliaryFile</code> argument of <code><a href="qAlign.html">qAlign</a></code>.</p></dd>


<dt>keepZero</dt>
<dd><p>If <code>FALSE</code>, only cytosines covered by at least
one alignment will be returned; <code>keepZero</code> must be <code>TRUE</code>
if multiple samples have the same sample name and
<code>collapseBySample</code> is <code>TRUE</code>.</p></dd>


<dt>mapqMin</dt>
<dd><p>Minimal mapping quality of alignments to be included when
counting (mapping quality must be greater than or equal to
<code>mapqMin</code>). Valid values are between 0 and 255. The default (0)
will include all alignments.</p></dd>


<dt>mapqMax</dt>
<dd><p>Maximal mapping quality of alignments to be included when
counting (mapping quality must be less than or equal to <code>mapqMax</code>).
Valid values are between 0 and 255. The default (255) will include
all alignments.</p></dd>


<dt>clObj</dt>
<dd><p>A cluster object to be used for parallel processing of
multiple files (see ‘Details’).</p></dd>

</dl></div>
    <div id="value">
    <h2>Value</h2>
    

<p>For <code>reportLevel</code>=“C”, a <code>GRanges</code> object if
<code>asGRanges</code>=<code>TRUE</code>, otherwise a <code>data.frame</code>.</p>


<p>Each row contains the coordinates of individual cytosines for
<code>collapseByQueryRegion</code>=<code>FALSE</code> or query regions
for <code>collapseByQueryRegion</code>=<code>TRUE</code>.</p>


<p>In addition to the coordinates columns (or <code>seqnames</code>,
<code>ranges</code> and <code>strand</code> slots for <code>GRanges</code> objects),
each row contains per bam file:</p>


<p>Two values (total and methylated events, with suffixes _T and _M), or
if the <code>qProject</code> object was created including a SNP table,
six values (total and methylated events for Reference, Unknown and
Alternative genotypes, with suffixed _TR, _TU, _TA, _MR, _MU and _MA).
In the latter case, C's or CpG's that overlap with SNPs in the table
are removed.</p>


<p>If <code>collapseBySample</code>=<code>TRUE</code>, groups of bam files with
identical sample name are combined (summed) and will be represented by
a single set of total and methylated count columns.</p>


<p>If <code>mode</code>=“var”, the _T and _M columns correspond to total
and matching alignments overlapping the guanine paired to the cytosine.</p>


<p>For <code>reportLevel</code>=“alignment”, a <code>list</code> with one
element per bam file or sample (depending on <code>collapseBySample</code>). 
Each list element is another list with the elements:</p><ul><li><p><code>aid</code>: character vector with unique alignment identifiers</p></li>
<li><p><code>Cid</code>: integer vector with genomic coordinate of C base</p></li>
<li><p><code>strand</code>: character vector with the strand of the C base</p></li>
<li><p><code>meth</code>: integer vector with methylation state for
  alignment and C defined by <code>aid</code> and <code>Cid</code>. The values are
  1 for methylated or 0 for unmethylated states.</p></li>
</ul></div>
    <div id="details">
    <h2>Details</h2>
    <p><code>qMeth</code> can be used on a <code>qProject</code> object from a bisulfite
sequencing experiment (sequencing of bisulfite-converted DNA), such as
the one returned by <code><a href="qAlign.html">qAlign</a></code> when its parameter
<code>bisulfite</code> is set to a different value than “no”.</p>
<p><code>qMeth</code> quantifies DNA methylation by counting total and
methylated events for individual cytosines, using the alignments that
have been generated in converted (three-letter) sequence space for
example by <code><a href="qAlign.html">qAlign</a></code>. A methylated event corresponds 
to a C/C match in the alignment, an unmethylated event to a T/C mismatch
(or G/G matches and A/G mismatches on the opposite strand). For paired-end
samples, the part of the left fragment alignment that overlaps
with the right fragment alignment is ignored, preventing the
use of redundant information coming from the same molecule.</p>
<p>Both directed (<code>bisulfite</code>=“dir”) and undirected
(<code>bisulfite</code>=“undir”) experimental protocols are supported
by <code><a href="qAlign.html">qAlign</a></code> and <code>qMeth</code>.</p>
<p>By default, results are returned per C nucleotide. If
<code>reportLevel</code>=“alignment”, results are reported separately
for individual alignments. In that case, <code>query</code> has to be a
<code>GRanges</code> object with exactly one region, <code>mode</code> has to be
either “CpG” or “allC”, the arguments
<code>collapseByQueryRegion</code>, <code>asGRanges</code>, <code>mask</code> and
<code>keepZero</code> have no effect and allele-specific projects are
treated in the same way as normal (non-allele specific) projects.</p>
<p>Using the parameter <code>mode</code>, quantification can be limited to
cytosines in CpG context, and counts obtained for the two cytosines on
opposite strands within a single CpG can be combined (summed).</p>
<p>The quantification of methylation for all cytosines in the query region(s)
(<code>mode</code>=“allC”) should be done with care, especially for
large query regions, as the return value may require a large amount of
memory.</p>
<p>If <code>mode</code> is set to “var”, <code>qMeth</code> only counts reads
from the strand opposite of the cytosine and reports total and
matching alignments. For a position identical to the reference
sequence, only matches (and very few sequencing errors) are
expected, independent on the methylation state of the cytosine. A
reduced fraction of alignments matching the reference are indicative
of sequence variations in the sequenced sample.</p>
<p><code>mapqMin</code> and <code>mapqMax</code> allow to select alignments
based on their mapping qualities. <code>mapqMin</code> and <code>mapqMax</code> can
take integer values between 0 and 255 and equal to
\(-10 log_{10} Pr(\textnormal{mapping position is wrong})\), rounded to the nearest
integer. A value 255 indicates that the mapping quality is not available.</p>
<p>If an object that inherits from class <code>cluster</code> is provided to
the <code>clObj</code> argument, for example an object returned by
<code><a href="https://rdrr.io/r/parallel/makeCluster.html" class="external-link">makeCluster</a></code> from package <span class="pkg">parallel</span>,
the quantification task is split into multiple chunks and processed in
parallel using <code><a href="https://rdrr.io/r/parallel/clusterApply.html" class="external-link">clusterApplyLB</a></code> from package
<span class="pkg">parallel</span>. Not all tasks will be efficiently parallelized: For
example, a single query region and a single (group of) bam files will
not be split into multiple chunks.</p>
    </div>
    <div id="see-also">
    <h2>See also</h2>
    <div class="dont-index"><p><code><a href="qAlign.html">qAlign</a></code>,
<code><a href="https://rdrr.io/r/parallel/makeCluster.html" class="external-link">makeCluster</a></code> from package <span class="pkg">parallel</span></p></div>
    </div>
    <div id="author">
    <h2>Author</h2>
    <p>Anita Lerch, Dimos Gaidatzis and Michael Stadler</p>
    </div>

    <div id="ref-examples">
    <h2>Examples</h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="co"># copy example data to current working directory</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/base/files.html" class="external-link">file.copy</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span>package<span class="op">=</span><span class="st">"QuasR"</span>, <span class="st">"extdata"</span><span class="op">)</span>, <span class="st">"."</span>, recursive<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] TRUE</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># create alignments</span></span></span>
<span class="r-in"><span><span class="va">sampleFile</span> <span class="op">&lt;-</span> <span class="st">"extdata/samples_bis_single.txt"</span></span></span>
<span class="r-in"><span><span class="va">genomeFile</span> <span class="op">&lt;-</span> <span class="st">"extdata/hg19sub.fa"</span></span></span>
<span class="r-in"><span><span class="va">proj</span> <span class="op">&lt;-</span> <span class="fu"><a href="qAlign.html">qAlign</a></span><span class="op">(</span><span class="va">sampleFile</span>, <span class="va">genomeFile</span>, bisulfite<span class="op">=</span><span class="st">"dir"</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> alignment files missing - need to:</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     create alignment index for the genome</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     create 1 genomic alignment(s)</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Creating an RbowtieCtoT index for /private/var/folders/24/8k48jl6d249_n_qfxwsl6xvm0000gn/T/Rtmp73ILzI/file105c124dabb43/reference/extdata/hg19sub.fa</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Finished creating index</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Testing the compute nodes...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> OK</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Loading QuasR on the compute nodes...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> preparing to run on 1 nodes...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> done</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Available cores:</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Mac-1674836252750.local: 1</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Performing genomic alignments for 1 samples. See progress in the log file:</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> /private/var/folders/24/8k48jl6d249_n_qfxwsl6xvm0000gn/T/Rtmp73ILzI/file105c124dabb43/reference/QuasR_log_105c17c84f229.txt</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Genomic alignments have been created successfully</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-in"><span><span class="va">proj</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Project: qProject</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  Options   : maxHits         : 1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>              paired          : no</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>              splicedAlignment: FALSE</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>              bisulfite       : dir</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>              snpFile         : none</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>              geneAnnotation  : none</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  Aligner   : Rbowtie v1.39.0 (parameters: -k 2 --best --strata -v 2)</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  Genome    : /private/var/folders/24/8k48jl6d249_n_qfxwsl6x.../hg19sub.fa (file)</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  Reads     : 1 file, 1 sample (fasta format):</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    1. bis_1_1.fa.bz2  Sample1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  Genome alignments: directory: same as reads</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    1. bis_1_1_105c17c36a19c.bam</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  Aux. alignments: none</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># calculate methylation states</span></span></span>
<span class="r-in"><span><span class="va">meth</span> <span class="op">&lt;-</span> <span class="fu">qMeth</span><span class="op">(</span><span class="va">proj</span>, mode<span class="op">=</span><span class="st">"CpGcomb"</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">meth</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> GRanges object with 3110 ranges and 2 metadata columns:</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>          seqnames      ranges strand | Sample1_T Sample1_M</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>             &lt;Rle&gt;   &lt;IRanges&gt;  &lt;Rle&gt; | &lt;integer&gt; &lt;integer&gt;</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>      [1]     chr1       19-20      * |         1         1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>      [2]     chr1       21-22      * |         1         1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>      [3]     chr1       54-55      * |         3         1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>      [4]     chr1       57-58      * |         3         0</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>      [5]     chr1       80-81      * |         6         5</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>      ...      ...         ...    ... .       ...       ...</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   [3106]     chr3 44957-44958      * |         8         7</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   [3107]     chr3 44977-44978      * |         5         3</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   [3108]     chr3 44981-44982      * |         4         3</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   [3109]     chr3 44989-44990      * |         1         1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   [3110]     chr3 44993-44994      * |         1         1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   -------</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   seqinfo: 3 sequences from an unspecified genome</span>
<span class="r-in"><span></span></span>
</code></pre></div>
    </div>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Anita Lerch, Charlotte Soneson, Dimos Gaidatzis, Michael Stadler.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer></div>

  


  

  </body></html>

