<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Quantify alignments by relative position — qProfile • QuasR</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="Quantify alignments by relative position — qProfile"><meta property="og:description" content="Quantify alignments from sequencing data, relative to their position in 
query regions."><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">QuasR</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.43.1</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../articles/QuasR.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
      </ul><ul class="nav navbar-nav navbar-right"></ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Quantify alignments by relative position</h1>
    
    <div class="hidden name"><code>qProfile.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>Quantify alignments from sequencing data, relative to their position in 
query regions.</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">qProfile</span><span class="op">(</span></span>
<span>  <span class="va">proj</span>,</span>
<span>  <span class="va">query</span>,</span>
<span>  upstream <span class="op">=</span> <span class="fl">1000</span>,</span>
<span>  downstream <span class="op">=</span> <span class="va">upstream</span>,</span>
<span>  selectReadPosition <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"start"</span>, <span class="st">"end"</span><span class="op">)</span>,</span>
<span>  shift <span class="op">=</span> <span class="fl">0L</span>,</span>
<span>  orientation <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"any"</span>, <span class="st">"same"</span>, <span class="st">"opposite"</span><span class="op">)</span>,</span>
<span>  useRead <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"any"</span>, <span class="st">"first"</span>, <span class="st">"last"</span><span class="op">)</span>,</span>
<span>  auxiliaryName <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  mask <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  collapseBySample <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  includeSpliced <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  includeSecondary <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  mapqMin <span class="op">=</span> <span class="fl">0L</span>,</span>
<span>  mapqMax <span class="op">=</span> <span class="fl">255L</span>,</span>
<span>  absIsizeMin <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  absIsizeMax <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  maxInsertSize <span class="op">=</span> <span class="fl">500L</span>,</span>
<span>  binSize <span class="op">=</span> <span class="fl">1L</span>,</span>
<span>  clObj <span class="op">=</span> <span class="cn">NULL</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>
    <dl><dt>proj</dt>
<dd><p>A <code><a href="qProject-class.html">qProject</a></code> object representing a 
sequencing experiment as returned by <code><a href="qAlign.html">qAlign</a></code></p></dd>


<dt>query</dt>
<dd><p>An object of type <code><a href="https://rdrr.io/pkg/GenomicRanges/man/GRanges-class.html" class="external-link">GRanges</a></code>
with the regions to be profiled. All regions in <code>query</code> will be
anchored at their biological start position (<code>start(query)</code> for
regions on strand “+” or “*”, <code>end(query)</code> for
regions on strand “-”). This position will become position zero 
in the return value.</p></dd>


<dt>upstream</dt>
<dd><p>An “integer” vector of length one or the same 
length as <code>query</code> indicating the number of bases upstream of the 
anchor position to include in the profile.</p></dd>


<dt>downstream</dt>
<dd><p>An “integer” vector of length one or the same 
length as <code>query</code> indicating the number of bases downstream of the
anchor position to include in the profile.</p></dd>


<dt>selectReadPosition</dt>
<dd><p>defines the part of the alignment that has to be 
contained within a query region to produce an overlap (see Details), and 
that is used to calculate the relative position within the query region. 
Possible values are:</p><ul><li><p><code>start</code> (default): start of the alignment</p></li>
<li><p><code>end</code>: end of the alignment</p></li>
</ul></dd>


<dt>shift</dt>
<dd><p>controls the shifting alignments towards their 3'-end before 
quantification. <code>shift</code> can be one of:</p><ul><li><p>an “integer” vector of the same length as the
  number of alignment files</p></li>
<li><p>a single “integer” value</p></li>
<li><p>the character string <code>"halfInsert"</code> (only available for
  paired-end experiments)</p></li>
</ul><p>The default of <code>0</code> will not shift any alignments.</p></dd>


<dt>orientation</dt>
<dd><p>sets the required orientation of the alignments relative 
to the query region in order to be counted, one of:</p><ul><li><p><code>any</code> (default): count alignment on the same and opposite strand</p></li>
<li><p><code>same</code> : count only alignment on the same strand</p></li>
<li><p><code>opposite</code> : count only alignment on the opposite strand</p></li>
</ul></dd>


<dt>useRead</dt>
<dd><p>For paired-end experiments, selects the read mate whose 
alignments should be counted, one of:</p><ul><li><p><code>any</code> (default): count all alignments</p></li>
<li><p><code>first</code> : count only alignments from the first read</p></li>
<li><p><code>last</code> : count only alignments from the last read</p></li>
</ul></dd>


<dt>auxiliaryName</dt>
<dd><p>Which bam files to use in an experiments with 
auxiliary alignments (see Details).</p></dd>


<dt>mask</dt>
<dd><p>If not <code>NULL</code>, a <code><a href="https://rdrr.io/pkg/GenomicRanges/man/GRanges-class.html" class="external-link">GRanges</a></code> 
object with reference regions to be masked, i.e. excluded from the 
quantification, such as unmappable or highly repetitive regions (see 
Details).</p></dd>


<dt>collapseBySample</dt>
<dd><p>If <code>TRUE</code> (the default), sum alignment 
counts from bam files with the same sample name.</p></dd>


<dt>includeSpliced</dt>
<dd><p>If <code>TRUE</code> (the default), include spliced 
alignments when counting. A spliced alignment is defined as an 
alignment with a gap in the read of at least 60 bases.</p></dd>


<dt>includeSecondary</dt>
<dd><p>If <code>TRUE</code> (the default), include alignments 
with the secondary bit (0x0100) set in the <code>FLAG</code> when counting.</p></dd>


<dt>mapqMin</dt>
<dd><p>Minimal mapping quality of alignments to be included when 
counting (mapping quality must be greater than or equal to <code>mapqMin</code>). 
Valid values are between 0 and 255. The default (0) will include all 
alignments.</p></dd>


<dt>mapqMax</dt>
<dd><p>Maximal mapping quality of alignments to be included when 
counting (mapping quality must be less than or equal to <code>mapqMax</code>).
Valid values are between 0 and 255. The default (255) will include all 
alignments.</p></dd>


<dt>absIsizeMin</dt>
<dd><p>For paired-end experiments, minimal absolute insert 
size (TLEN field in SAM Spec v1.4) of alignments to be included when 
counting. Valid values are greater than 0 or <code>NULL</code> (default), 
which will not apply any minimum insert size filtering.</p></dd>


<dt>absIsizeMax</dt>
<dd><p>For paired-end experiments, maximal absolute insert 
size (TLEN field in SAM Spec v1.4) of alignments to be included when 
counting. Valid values are greater than 0 or <code>NULL</code> (default), 
which will not apply any maximum insert size filtering.</p></dd>


<dt>maxInsertSize</dt>
<dd><p>Maximal fragment size of the paired-end experiment.
This parameter is used if <code>shift="halfInsert"</code> and will ensure that 
query regions are made wide enough to emcompass all alignment pairs whose 
mid falls into the query region. The default value is <code>500</code> bases.</p></dd>


<dt>binSize</dt>
<dd><p>Numeric scalar giving the size of bins (must be an odd number).
The default value (<code>1</code>) gives back counts for single bases. Otherwise,
alignments are counted in adjacent, non-overlapping windows of size 
<code>binSize</code> that tile the interval defined by <code>upstream</code> and 
<code>downstream</code>.</p></dd>


<dt>clObj</dt>
<dd><p>A cluster object to be used for parallel processing (see 
‘Details’).</p></dd>

</dl></div>
    <div id="value">
    <h2>Value</h2>
    

<p>A <code>list</code> of matrices with <code>length(unique(names(query)))</code> rows
with profile names, and <code>max(upstream)+max(downstream)+1</code> columns
indicating relative position (for <code>binsize=1</code>).</p>


<p>For <code>binSize</code> values greater than 1, the number of columns corresponds to
the number of bins (tiles), namely 
<code>ceiling(max(upstream)/binSize)+ceiling(max(downstream)/binSize)</code>.
A middle bin of size <code>binSize</code> is always positioned centered at the anchor
of each region. Additional bins are positioned upstream and downstream, adjacent
to that middle bin, in order to include at least <code>upstream</code> and
<code>downstream</code> bases, respectively (potentially more in order to fill the 
first and last bins).</p>


<p>The relative positions are given as column names (for <code>binSize &gt; 1</code></p>


<p>they refer to the bin mid). In that case, the bins are "right-open". For
example, if <code>binSize = 10</code>, the bin with the midpoint "-50" contains
counts for the alignments in [-55,-45).</p>


<p>The first list element is called “coverage” and contains, for each
profile and relative position, the number of overlapping regions
that contributed to the profile.</p>


<p>Subsequent list elements contain the alignment counts for individual
sequence files (<code>collapseBySample=FALSE</code>) or samples 
(<code>collapseBySample=TRUE</code>) in <code>proj</code>.</p>


<p>For projects with allele-specific quantification, i.e. if a file with
single nucleotide polymorphisms was supplied to the <code>snpFile</code></p>


<p>argument of <code><a href="qAlign.html">qAlign</a></code>, there will be three rows
instead of one row with counts per unique region name, with numbers
of alignments for Reference, Unknown and Alternative genotypes
(suffixed _R, _U and _A).</p>
    </div>
    <div id="details">
    <h2>Details</h2>
    <p><code>qProfile</code> is used to count alignments in each sample from a 
<code>qProject</code> object, relative to their position in query regions.</p>
<p>Most arguments are identical to the ones of <code><a href="qCount.html">qCount</a></code>.</p>
<p>The <code>query</code> argument is a <code><a href="https://rdrr.io/pkg/GenomicRanges/man/GRanges-class.html" class="external-link">GRanges</a></code>
object that defines the regions for the profile. All regions in
<code>query</code> will be aligned to one another at their anchor position,
which corresponds to their biological start position (<code>start(query)</code>
for regions on strand “+” or “*”, <code>end(query)</code> for 
regions on strand “-”).</p>
<p>This anchor position will be extended (with regard to strand) by
the number of bases specified by <code>upstream</code> and <code>downstream</code>.
In the return value, the anchor position will be at position zero.</p>
<p>If <code>binSize</code> is greater than one, <code>upstream</code> and <code>downstream</code>
will be slightly increased in order to include the complete first and last
bins of <code>binSize</code> bases.</p>
<p>Regions with identical names in <code>names{query}</code> will be summed, and
profiles will be padded with zeros to accomodate the length of all profiles.</p>
    </div>
    <div id="see-also">
    <h2>See also</h2>
    <div class="dont-index"><p><code><a href="qCount.html">qCount</a></code>,
<code><a href="qAlign.html">qAlign</a></code>,
<code><a href="qProject-class.html">qProject</a></code>,
<code><a href="https://rdrr.io/r/parallel/makeCluster.html" class="external-link">makeCluster</a></code> from package <span class="pkg">parallel</span></p></div>
    </div>
    <div id="author">
    <h2>Author</h2>
    <p>Anita Lerch, Dimos Gaidatzis and Michael Stadler</p>
    </div>

    <div id="ref-examples">
    <h2>Examples</h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="co"># copy example data to current working directory</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/base/files.html" class="external-link">file.copy</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span>package<span class="op">=</span><span class="st">"QuasR"</span>, <span class="st">"extdata"</span><span class="op">)</span>, <span class="st">"."</span>, recursive<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] TRUE</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># create alignments (single-end experiment)</span></span></span>
<span class="r-in"><span><span class="va">genomeFile</span> <span class="op">&lt;-</span> <span class="st">"extdata/hg19sub.fa"</span></span></span>
<span class="r-in"><span><span class="va">sampleFile</span> <span class="op">&lt;-</span> <span class="st">"extdata/samples_chip_single.txt"</span></span></span>
<span class="r-in"><span><span class="va">proj</span> <span class="op">&lt;-</span> <span class="fu"><a href="qAlign.html">qAlign</a></span><span class="op">(</span><span class="va">sampleFile</span>, <span class="va">genomeFile</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> all necessary alignment files found</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># load transcript start site coordinates</span></span></span>
<span class="r-in"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">rtracklayer</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">annotationFile</span> <span class="op">&lt;-</span> <span class="st">"extdata/hg19sub_annotation.gtf"</span></span></span>
<span class="r-in"><span><span class="va">tssRegions</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/rtracklayer/man/GFFFile-class.html" class="external-link">import.gff</a></span><span class="op">(</span><span class="va">annotationFile</span>, format<span class="op">=</span><span class="st">"gtf"</span>,</span></span>
<span class="r-in"><span>                         feature.type<span class="op">=</span><span class="st">"start_codon"</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># obtain a combined TSS profile</span></span></span>
<span class="r-in"><span><span class="va">pr1</span> <span class="op">&lt;-</span> <span class="fu">qProfile</span><span class="op">(</span><span class="va">proj</span>, <span class="va">tssRegions</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> profiling alignments...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> done</span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="va">pr1</span>, <span class="va">dim</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> $coverage</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1]    1 2001</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> $Sample1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1]    1 2001</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> $Sample2</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1]    1 2001</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="va">pr1</span>, <span class="st">"["</span>, , <span class="fl">1</span><span class="op">:</span><span class="fl">5</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> $coverage</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> -1000  -999  -998  -997  -996 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    19    19    19    19    19 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> $Sample1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> -1000  -999  -998  -997  -996 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     0     0     1     0     0 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> $Sample2</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> -1000  -999  -998  -997  -996 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     1     1     1     2     0 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="va">prComb</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/do.call.html" class="external-link">do.call</a></span><span class="op">(</span><span class="st">"+"</span>, <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="va">pr1</span><span class="op">[</span><span class="op">-</span><span class="fl">1</span><span class="op">]</span>, <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="va">x</span><span class="op">/</span><span class="va">pr1</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/graphics/barplot.html" class="external-link">barplot</a></span><span class="op">(</span><span class="va">prComb</span>, xlab<span class="op">=</span><span class="st">"Position"</span>, ylab<span class="op">=</span><span class="st">"Mean no. of alignments"</span><span class="op">)</span></span></span>
<span class="r-plt img"><img src="qProfile-1.png" alt="" width="700" height="433"></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># obtain TSS profiles for individual regions</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">tssRegions</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Vector-class.html" class="external-link">mcols</a></span><span class="op">(</span><span class="va">tssRegions</span><span class="op">)</span><span class="op">$</span><span class="va">transcript_id</span></span></span>
<span class="r-in"><span><span class="va">pr2</span> <span class="op">&lt;-</span> <span class="fu">qProfile</span><span class="op">(</span><span class="va">proj</span>, <span class="va">tssRegions</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> profiling alignments...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> done</span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="va">pr2</span>, <span class="va">dim</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> $coverage</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1]   19 2001</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> $Sample1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1]   19 2001</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> $Sample2</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1]   19 2001</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="va">pr2</span>, <span class="st">"["</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">3</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">5</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> $coverage</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                 -1000 -999 -998 -997 -996</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> ENST00000328596     1    1    1    1    1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> ENST00000379265     1    1    1    1    1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> ENST00000379268     1    1    1    1    1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> $Sample1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                 -1000 -999 -998 -997 -996</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> ENST00000328596     0    0    0    0    0</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> ENST00000379265     0    0    0    0    0</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> ENST00000379268     0    0    0    0    0</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> $Sample2</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                 -1000 -999 -998 -997 -996</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> ENST00000328596     0    0    0    0    0</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> ENST00000379265     0    0    0    0    0</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> ENST00000379268     0    0    0    0    0</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-in"><span></span></span>
</code></pre></div>
    </div>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Anita Lerch, Adam Alexander Thil SMITH, Charlotte Soneson, Dimos Gaidatzis, Michael Stadler.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer></div>

  


  

  </body></html>

