<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>qProject objects — qProject-class • QuasR</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="qProject objects — qProject-class"><meta property="og:description" content="The qProject class is a container for the meta-data (e.g. sample
names, paths and names of sequence and alignment files) associated
with a high-throughput sequencing experiment analyzed with QuasR."><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">QuasR</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.41.2</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../articles/QuasR.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
      </ul><ul class="nav navbar-nav navbar-right"></ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>qProject objects</h1>
    
    <div class="hidden name"><code>qProject-class.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>The qProject class is a container for the meta-data (e.g. sample
names, paths and names of sequence and alignment files) associated
with a high-throughput sequencing experiment analyzed with <code>QuasR</code>.</p>
    </div>


    <div id="details">
    <h2>Details</h2>
    <p>The qProject class is returned by <a href="qAlign.html">qAlign</a> and stores all
information on a high-throughput sequencing experiment analyzed with
<code>QuasR</code>. qProject objects can be conveniently passed to
‘q’-functions (function name starting with the letter
‘q’). The information is stored in the following slots:</p><dl><dt><code>reads</code></dt>
<dd><p>a 'data.frame' with sequence read files.</p></dd>

  <dt><code>reads_md5subsum</code></dt>
<dd><p>a 'data.frame' with fingerprints for
  sequence read files.</p></dd>

  <dt><code>alignments</code></dt>
<dd><p>a 'data.frame' with alignment files.</p></dd>

  <dt><code>samplesFormat</code></dt>
<dd><p>a 'character(1)' specifying the format
  of input files.</p></dd>

  <dt><code>genome</code></dt>
<dd><p>a 'character(1)' specifying the reference genome.</p></dd>

  <dt><code>genomeFormat</code></dt>
<dd><p>a 'character(1)' specifying the format of
  the reference genome.</p></dd>

  <dt><code>aux</code></dt>
<dd><p>a 'data.frame' with auxiliary reference sequence files.</p></dd>

  <dt><code>auxAlignments</code></dt>
<dd><p>a 'data.frame' with alignment files for
  auxiliary reference sequence files.</p></dd>

  <dt><code>aligner</code></dt>
<dd><p>a 'character(1)' specifying the aligner.</p></dd>

  <dt><code>maxHits</code></dt>
<dd><p>a 'numeric(1)' specifying the maximum number
  of alignments per sequence.</p></dd>

  <dt><code>paired</code></dt>
<dd><p>a 'character(1)' specifying the paired-type;
  one of "no", "fr", "rf", "ff".</p></dd>

  <dt><code>splicedAlignment</code></dt>
<dd><p>a 'logical(1)'; <code>TRUE</code> when
  performing spliced-alignments.</p></dd>

  <dt><code>snpFile</code></dt>
<dd><p>a 'character(1)' with a file name containing
  SNP information.</p></dd>

  <dt><code>bisulfite</code></dt>
<dd><p>a 'character(1)' defining the bisulfite
  type; one of "no", "dir", "undir".</p></dd>

  <dt><code>alignmentParameter</code></dt>
<dd><p>a 'character(1)' with aligner
  command line parameters.</p></dd>

  <dt><code>projectName</code></dt>
<dd><p>a 'character(1)' with the project name.</p></dd>

  <dt><code>alignmentsDir</code></dt>
<dd><p>a 'character(1)' with the directory to
  be used to store alignment files.</p></dd>

  <dt><code>lib.loc</code></dt>
<dd><p>a 'character(1)' with the library directory to
  use for installing of alignment index packages.</p></dd>

  <dt><code>cacheDir</code></dt>
<dd><p>a 'character(1)' with a directory to use for
  temporary files.</p></dd>

  <dt><code>alnModeID</code></dt>
<dd><p>a 'character(1)' used internally to indicate
  the alignment mode.</p></dd>


</dl></div>
    <div id="accessors">
    <h2>Accessors</h2>
    

<p>In the following code snippets, <code>x</code> is a qProject object.</p><dl><dt><code>length(x)</code></dt>
<dd><p>: Gets the number of input files.</p></dd>

  <dt><code>genome(x)</code></dt>
<dd><p>: Gets the reference genome as a 'character(1)'.
  The type of genome is stored as an attribute in
  <code>attr(genome(x),"genomeFormat")</code>: "BSgenome" indicates that
  <code>genome(x)</code> refers to the name of a BSgenome package, "file"
  indicates that it contains the path and filename of a genome in
  FASTA format.</p></dd>

  <dt><code>auxiliaries(x)</code></dt>
<dd><p>: Gets a <code>data.frame</code> with auxiliary
  target sequences, with one row per auxiliary target, and columns
  "FileName" and "AuxName".</p></dd>

  <dt><code>alignments(x)</code></dt>
<dd><p>: Gets a list with two elements "genome" and
  "aux". <code>alignments(x)$genome</code> contains a <code>data.frame</code>
  with <code>length(x)</code> rows and the columns "FileName" (containing
  the path to bam files with genomic alignments) and
  "SampleName". <code>alignments(x)$aux</code> contains a
  <code>data.frame</code> with one row per auxiliary target sequence (with
  auxiliary names as row names), and <code>length(x)</code> columns.</p></dd>


</dl></div>
    <div id="subsetting">
    <h2>Subsetting</h2>
    

<p>In the following code snippets, <code>x</code> is a qProject object.</p><dl><dt><code>x[i]</code></dt>
<dd><p>: Get <code>qProject</code> object instance with <code>i</code>
  input files, where <code>i</code> can be an NA-free logical, numeric, or
  character vector.</p></dd>


</dl></div>
    <div id="see-also">
    <h2>See also</h2>
    <div class="dont-index"><p><a href="qAlign.html">qAlign</a></p></div>
    </div>
    <div id="author">
    <h2>Author</h2>
    <p>Anita Lerch, Dimos Gaidatzis and Michael Stadler</p>
    </div>

    <div id="ref-examples">
    <h2>Examples</h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="co"># copy example data to current working directory</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/base/files.html" class="external-link">file.copy</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span>package<span class="op">=</span><span class="st">"QuasR"</span>, <span class="st">"extdata"</span><span class="op">)</span>, <span class="st">"."</span>, recursive<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] TRUE</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># create alignments</span></span></span>
<span class="r-in"><span><span class="va">sampleFile</span> <span class="op">&lt;-</span> <span class="st">"extdata/samples_chip_single.txt"</span></span></span>
<span class="r-in"><span><span class="va">genomeFile</span> <span class="op">&lt;-</span> <span class="st">"extdata/hg19sub.fa"</span></span></span>
<span class="r-in"><span><span class="va">auxFile</span> <span class="op">&lt;-</span> <span class="st">"extdata/auxiliaries.txt"</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="va">proj</span> <span class="op">&lt;-</span> <span class="fu"><a href="qAlign.html">qAlign</a></span><span class="op">(</span><span class="va">sampleFile</span>, <span class="va">genomeFile</span>, auxiliaryFile<span class="op">=</span><span class="va">auxFile</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> alignment files missing - need to:</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     create 2 auxiliary alignment(s)</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Creating an Rbowtie index for /private/var/folders/24/8k48jl6d249_n_qfxwsl6xvm0000gn/T/RtmpuUw7QE/file7d91c507f93/reference/extdata/NC_001422.1.fa</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Finished creating index</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Testing the compute nodes...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> OK</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Loading QuasR on the compute nodes...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> preparing to run on 1 nodes...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> done</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Available cores:</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> nodeNames</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Mac-1688038232051.local </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                       1 </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Performing auxiliary alignments for 2 samples. See progress in the log file:</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> /private/var/folders/24/8k48jl6d249_n_qfxwsl6xvm0000gn/T/RtmpuUw7QE/file7d91c507f93/reference/QuasR_log_7d917a1fdb6b.txt</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Auxiliary alignments have been created successfully</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-in"><span><span class="va">proj</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Project: qProject</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  Options   : maxHits         : 1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>              paired          : no</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>              splicedAlignment: FALSE</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>              bisulfite       : no</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>              snpFile         : none</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>              geneAnnotation  : none</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  Aligner   : Rbowtie v1.41.0 (parameters: -m 1 --best --strata)</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  Genome    : /private/var/folders/24/8k48jl6d249_n_qfxwsl6x.../hg19sub.fa (file)</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  Reads     : 2 files, 2 samples (fastq format):</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    1. chip_1_1.fq.bz2  Sample1 (phred33)</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    2. chip_2_1.fq.bz2  Sample2 (phred33)</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  Genome alignments: directory: same as reads</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    1. chip_1_1_7d913b618788.bam</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    2. chip_2_1_7d919102d.bam   </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  Aux. alignments: 1 file, directory: same as reads</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    a. /private/var/folders/24/8k48jl6d249_n_qfxwsl6xv.../NC_001422.1.fa  phiX174</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>      1. chip_1_1_7d91417e8c70.bam</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>      2. chip_2_1_7d915b362ea7.bam</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># alignment statistics using a qProject</span></span></span>
<span class="r-in"><span><span class="fu"><a href="alignmentStats.html">alignmentStats</a></span><span class="op">(</span><span class="va">proj</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                 seqlength mapped unmapped</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Sample1:genome      95000   2339      258</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Sample2:genome      95000   3609      505</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Sample1:phiX174      5386    251        7</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Sample2:phiX174      5386    493       12</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># alignment statistics using bam files</span></span></span>
<span class="r-in"><span><span class="fu"><a href="alignmentStats.html">alignmentStats</a></span><span class="op">(</span><span class="fu">alignments</span><span class="op">(</span><span class="va">proj</span><span class="op">)</span><span class="op">$</span><span class="va">genome</span><span class="op">$</span><span class="va">FileName</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                           seqlength mapped unmapped</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> chip_1_1_7d913b618788.bam     95000   2339      258</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> chip_2_1_7d919102d.bam        95000   3609      505</span>
<span class="r-in"><span><span class="fu"><a href="alignmentStats.html">alignmentStats</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/unlist.html" class="external-link">unlist</a></span><span class="op">(</span><span class="fu">alignments</span><span class="op">(</span><span class="va">proj</span><span class="op">)</span><span class="op">$</span><span class="va">aux</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                           seqlength mapped unmapped</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> chip_1_1_7d91417e8c70.bam      5386    251        0</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> chip_2_1_7d915b362ea7.bam      5386    493        0</span>
<span class="r-in"><span></span></span>
</code></pre></div>
    </div>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Anita Lerch, Adam Alexander Thil SMITH, Charlotte Soneson, Dimos Gaidatzis, Michael Stadler.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer></div>

  


  

  </body></html>

