<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>QuasR Quality Control Report — qQCReport • QuasR</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="QuasR Quality Control Report — qQCReport"><meta property="og:description" content="Generate quality control plots for a qProject object or a vector of 
fasta/fastq/bam files. The available plots vary depending on the types of 
available input (fasta, fastq, bam files or qProject object; 
paired-end or single-end)."><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">QuasR</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.43.1</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../articles/QuasR.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
      </ul><ul class="nav navbar-nav navbar-right"></ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>QuasR Quality Control Report</h1>
    
    <div class="hidden name"><code>qQCReport.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>Generate quality control plots for a <code>qProject</code> object or a vector of 
fasta/fastq/bam files. The available plots vary depending on the types of 
available input (fasta, fastq, bam files or <code>qProject</code> object; 
paired-end or single-end).</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">qQCReport</span><span class="op">(</span></span>
<span>  <span class="va">input</span>,</span>
<span>  pdfFilename <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  chunkSize <span class="op">=</span> <span class="fl">1000000L</span>,</span>
<span>  useSampleNames <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  clObj <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  a4layout <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  <span class="va">...</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>
    <dl><dt>input</dt>
<dd><p>A vector of files or a <code>qProject</code> object as returned by 
<code>qAlign</code>.</p></dd>


<dt>pdfFilename</dt>
<dd><p>The path and name of a pdf file to store the report. 
If <code>NULL</code>, the quality control plots will be generated in separate 
plotting windows on the standard graphical device.</p></dd>


<dt>chunkSize</dt>
<dd><p>The number of sequences, sequence pairs (for paired-end 
data) or alignments that will be sampled from each data file to collect 
quality statistics.</p></dd>


<dt>useSampleNames</dt>
<dd><p>If TRUE, the plots will be labelled using the sample 
names instead of the file names. Sample names are obtained from the 
<code>qProject</code> object, or from <code>names(input)</code> if <code>input</code> is a 
named vector of file names. Please not that if there are multiple files 
for the same sample, the sample names will not be unique.</p></dd>


<dt>clObj</dt>
<dd><p>A cluster object to be used for parallel processing of multiple 
input files.</p></dd>


<dt>a4layout</dt>
<dd><p>A logical scalar. If TRUE, the output of mapping rate and 
uniqueness plots will be adjusted for a4 format devices.</p></dd>


<dt>...</dt>
<dd><p>Additional arguments that will be passed to the functions 
generating the individual quality control plots, see ‘Details’.</p></dd>

</dl></div>
    <div id="value">
    <h2>Value</h2>
    

<p>The function is called for its side effect of generating quality
control plots. It invisibly returns a list with components that contain the 
data used to generate each of the QC plots. Available components are 
(depending on input data, see ‘Details’):</p><ul><li><p><em>qualByCycle</em>: quality score boxplot</p></li>
<li><p><em>nuclByCycle</em>: nucleotide frequency plot</p></li>
<li><p><em>duplicated</em>: duplication level plot</p></li>
<li><p><em>mappings</em>: mapping statistics barplot</p></li>
<li><p><em>uniqueness</em>: library complexity barplot</p></li>
<li><p><em>errorsByCycle</em>: mismatch frequency plot</p></li>
<li><p><em>mismatchTypes</em>: mismatch type plot</p></li>
<li><p><em>fragDistribution</em>: fragment size distribution plot</p></li>
</ul></div>
    <div id="details">
    <h2>Details</h2>
    <p>This function generates quality control plots for all input files or the 
sequence and alignment files contained in a <code>qProject</code> object, 
allowing assessment of the quality of a sequencing experiment. 
<code>qQCReport</code> uses functionality from the <span class="pkg">ShortRead</span> package to 
collect quality data, and visualizes the results similarly as the 
‘FastQC’ quality control tool from Simon Andrews (see 
‘References’ below). It is recommended to create PDF reports 
(<code>pdfFilename</code> argument), for which the plot layouts have been optimised.</p>
<p>Some plots will only be generated if the necessary information is available 
(e.g. base qualities in fastq sequence files).</p>
<p>The currently available plot types are:</p><dl><dt><em>Quality score boxplot</em></dt>
<dd><p>shows the distribution of
        base quality values as a box plot for each position in the input
        sequence. The background color (green, orange or red)
        indicates ranges of high, intermediate and low qualities.</p></dd>

  <dt><em>Nucleotide frequency</em></dt>
<dd><p>plot shows the frequency of A, C,
        G, T and N bases by position in the read.</p></dd>

  <dt><em>Duplication level</em></dt>
<dd><p>plot shows for each sample the
        fraction of reads observed at different duplication levels
        (e.g. once, two-times, three-times, etc.). In addition, the most
        frequent sequences are listed.</p></dd>

  <dt><em>Mapping statistics</em></dt>
<dd><p>shows fractions of reads that were
        (un)mappable to the reference genome.</p></dd>

  <dt><em>Library complexity</em></dt>
<dd><p>shows fractions of unique
        read(-pair) alignment positions, as a measure of the complexity in
        the sequencing library. Please note that this measure is not
        independent from the total number of reads in a library, and is best
        compared between libraries of similar sizes.</p></dd>

  <dt><em>Mismatch frequency</em></dt>
<dd><p>shows the frequency and position
        (relative to the read sequence) of mismatches in the alignments
        against the reference genome.</p></dd>

  <dt><em>Mismatch types</em></dt>
<dd><p>shows the frequency of
        read bases that caused mismatches in the alignments to the
        reference genome, separately for each genome base.</p></dd>

  <dt><em>Fragment size</em></dt>
<dd><p>shows the distribution of fragment sizes
        inferred from aligned read pairs.</p></dd>


</dl><p>One approach to assess the quality of a sample is to compare its
control plots to the ones from other samples and search for relative
differences. Special quality measures are expected for certain types
of experiments: A genomic re-sequencing sample with an
overrepresentation of T bases may be suspicious, while such a
nucleotide bias is normal for a directed bisulfite-sequencing sample.</p>
<p>Additional arguments can be passed to the internal functions that
generate the individual quality control plots using <code>...{}</code>:</p><dl><dt><code>lmat</code>:</dt>
<dd><p>a matrix (e.g. <code>matrix(1:12, ncol=2)</code>) used
        by an internal call to the <code>layout</code> function to specify the
        positioning of multiple plot panels on a device page. Individual panels
        correspond to different samples.</p></dd>

  <dt><code>breaks</code>:</dt>
<dd><p>a numerical vector
        (e.g. <code>c(1:10)</code>) defining the bins used by
        the ‘Duplication level’ plot.</p></dd>


</dl></div>
    <div id="references">
    <h2>References</h2>
    <p>FastQC quality control tool at 
<a href="http://www.bioinformatics.bbsrc.ac.uk/projects/fastqc/" class="external-link">http://www.bioinformatics.bbsrc.ac.uk/projects/fastqc/</a></p>
    </div>
    <div id="see-also">
    <h2>See also</h2>
    <div class="dont-index"><p><code><a href="qProject-class.html">qProject</a></code>, <code><a href="qAlign.html">qAlign</a></code>,
<code>ShortRead</code> package</p></div>
    </div>
    <div id="author">
    <h2>Author</h2>
    <p>Anita Lerch, Dimos Gaidatzis and Michael Stadler</p>
    </div>

    <div id="ref-examples">
    <h2>Examples</h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="co"># copy example data to current working directory</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/base/files.html" class="external-link">file.copy</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span>package<span class="op">=</span><span class="st">"QuasR"</span>, <span class="st">"extdata"</span><span class="op">)</span>, <span class="st">"."</span>, recursive<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] TRUE</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># create alignments</span></span></span>
<span class="r-in"><span><span class="va">sampleFile</span> <span class="op">&lt;-</span> <span class="st">"extdata/samples_chip_single.txt"</span></span></span>
<span class="r-in"><span><span class="va">genomeFile</span> <span class="op">&lt;-</span> <span class="st">"extdata/hg19sub.fa"</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="va">proj</span> <span class="op">&lt;-</span> <span class="fu"><a href="qAlign.html">qAlign</a></span><span class="op">(</span><span class="va">sampleFile</span>, <span class="va">genomeFile</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> all necessary alignment files found</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># create quality control report</span></span></span>
<span class="r-in"><span><span class="fu">qQCReport</span><span class="op">(</span><span class="va">proj</span>, pdfFilename<span class="op">=</span><span class="st">"qc_report.pdf"</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> collecting quality control data</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> creating QC plots</span>
<span class="r-in"><span></span></span>
</code></pre></div>
    </div>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Anita Lerch, Adam Alexander Thil SMITH, Charlotte Soneson, Dimos Gaidatzis, Michael Stadler.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer></div>

  


  

  </body></html>

