<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>An introduction to QuasR • QuasR</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="An introduction to QuasR">
<meta property="og:description" content="QuasR">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">QuasR</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.43.1</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../articles/QuasR.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right"></ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>An introduction to QuasR</h1>
                        <h4 data-toc-skip class="author">Michael
Stadler</h4>
                        <h4 data-toc-skip class="author">Dimos
Gaidatzis</h4>
                        <h4 data-toc-skip class="author">Charlotte
Soneson</h4>
                        <h4 data-toc-skip class="author">Anita
Lerch</h4>
            
            <h4 data-toc-skip class="date">01 November, 2023</h4>
      
      
      <div class="hidden name"><code>QuasR.Rmd</code></div>

    </div>

    
    
<!--
\DeclareRobustCommand{\QuasRlogo}{\raisebox{-8mm}{\includegraphics[width=45mm]{QuasR-logo.pdf}}}% use \DeclareRobustCommand instead of \newcommand to avoid premature expansion of \QuasRlogo within by 'titling' package
-->
<p><img src="QuasR-logo.png" style="position:absolute;top:0px;right:0px;"></p>
<!--

-->
<div class="section level2">
<h2 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<p>The <em><a href="https://bioconductor.org/packages/3.19/QuasR" class="external-link">QuasR</a></em>
package (short for <i>Qu</i>antify and <i>a</i>nnotate <i>s</i>hort
reads in <i>R</i>) integrates the functionality of several
<strong>R</strong> packages (such as <em><a href="https://bioconductor.org/packages/3.19/IRanges" class="external-link">IRanges</a></em>
<span class="citation">(Lawrence et al. 2013)</span> and <em><a href="https://bioconductor.org/packages/3.19/Rsamtools" class="external-link">Rsamtools</a></em>)
and external software (e.g. <code>bowtie</code>, through the <em><a href="https://bioconductor.org/packages/3.19/Rbowtie" class="external-link">Rbowtie</a></em>
package, and <code>HISAT2</code>, through the <em><a href="https://bioconductor.org/packages/3.19/Rhisat2" class="external-link">Rhisat2</a></em>
package). The package aims to cover the whole analysis workflow of
typical high throughput sequencing experiments, starting from the raw
sequence reads, over pre-processing and alignment, up to quantification.
A single <strong>R</strong> script can contain all steps of a complete
analysis, making it simple to document, reproduce or share the workflow
containing all relevant details.</p>
<p>The current <em><a href="https://bioconductor.org/packages/3.19/QuasR" class="external-link">QuasR</a></em>
release supports the analysis of single read and paired-end ChIP-seq
(chromatin immuno-precipitation combined with sequencing), RNA-seq (gene
expression profiling by sequencing of RNA) and Bis-seq (measurement of
DNA methylation by sequencing of bisulfite-converted genomic DNA)
experiments. It has been successfully used with data from Illumina, 454
Life Technologies and SOLiD sequencers, the latter by using bam files
created externally of <em><a href="https://bioconductor.org/packages/3.19/QuasR" class="external-link">QuasR</a></em>.</p>
</div>
<div class="section level2">
<h2 id="preliminaries">Preliminaries<a class="anchor" aria-label="anchor" href="#preliminaries"></a>
</h2>
<div class="section level3">
<h3 id="citing-quasr">Citing <em><a href="https://bioconductor.org/packages/3.19/QuasR" class="external-link">QuasR</a></em><a class="anchor" aria-label="anchor" href="#citing-quasr"></a>
</h3>
<p>If you use <em><a href="https://bioconductor.org/packages/3.19/QuasR" class="external-link">QuasR</a></em> <span class="citation">(Gaidatzis et al. 2015)</span> in your work, you can
cite it as follows:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/citation.html" class="external-link">citation</a></span><span class="op">(</span><span class="st">"QuasR"</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Please use the QuasR reference below to cite the software itself. If</span></span>
<span><span class="co">## you were using qAlign with Rbowtie as aligner, it can be cited as</span></span>
<span><span class="co">## Langmead et al. (2009) (unspliced alignments) or Au et al. (2010)</span></span>
<span><span class="co">## (spliced alignments). If you were using qAlign with Rhisat2 as aligner,</span></span>
<span><span class="co">## it can be cited as Kim et al. (2015).</span></span>
<span><span class="co">## </span></span>
<span><span class="co">##   Gaidatzis D, Lerch A, Hahne F, Stadler MB. QuasR: Quantification and</span></span>
<span><span class="co">##   annotation of short reads in R. Bioinformatics 31(7):1130-1132</span></span>
<span><span class="co">##   (2015).</span></span>
<span><span class="co">## </span></span>
<span><span class="co">##   Langmead B, Trapnell C, Pop M, Salzberg SL. Ultrafast and</span></span>
<span><span class="co">##   memory-efficient alignment of short DNA sequences to the human</span></span>
<span><span class="co">##   genome. Genome Biology 10(3):R25 (2009).</span></span>
<span><span class="co">## </span></span>
<span><span class="co">##   Au KF, Jiang H, Lin L, Xing Y, Wong WH. Detection of splice junctions</span></span>
<span><span class="co">##   from paired-end RNA-seq data by SpliceMap. Nucleic Acids Research,</span></span>
<span><span class="co">##   38(14):4570-8 (2010).</span></span>
<span><span class="co">## </span></span>
<span><span class="co">##   Kim D, Langmead B, Salzberg SL. HISAT: a fast spliced aligner with</span></span>
<span><span class="co">##   low memory requirements. Nat Methods, 12(4):357-60 (2015).</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## This free open-source software implements academic research by the</span></span>
<span><span class="co">## authors and co-workers. If you use it, please support the project by</span></span>
<span><span class="co">## citing the appropriate journal articles.</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## To see these entries in BibTeX format, use 'print(&lt;citation&gt;,</span></span>
<span><span class="co">## bibtex=TRUE)', 'toBibtex(.)', or set</span></span>
<span><span class="co">## 'options(citation.bibtex.max=999)'.</span></span></code></pre>
</div>
<div class="section level3">
<h3 id="installation">Installation<a class="anchor" aria-label="anchor" href="#installation"></a>
</h3>
<p><em><a href="https://bioconductor.org/packages/3.19/QuasR" class="external-link">QuasR</a></em> is a
package for the <strong>R</strong> computing environment and it is
assumed that you have already installed <strong>R</strong>. See the
<strong>R</strong> project at (<a href="http://www.r-project.org" class="external-link uri">http://www.r-project.org</a>). To install the latest version
of <em><a href="https://bioconductor.org/packages/3.19/QuasR" class="external-link">QuasR</a></em>, you
will need to be using the latest version of <strong>R</strong>. <em><a href="https://bioconductor.org/packages/3.19/QuasR" class="external-link">QuasR</a></em> is
part of the Bioconductor project at (<a href="http://www.bioconductor.org" class="external-link uri">http://www.bioconductor.org</a>). To get <em><a href="https://bioconductor.org/packages/3.19/QuasR" class="external-link">QuasR</a></em>
together with its dependencies you can use</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">require</a></span><span class="op">(</span><span class="st"><a href="https://bioconductor.github.io/BiocManager/" class="external-link">"BiocManager"</a></span><span class="op">)</span><span class="op">)</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"BiocManager"</span><span class="op">)</span></span>
<span><span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://bioconductor.github.io/BiocManager/reference/install.html" class="external-link">install</a></span><span class="op">(</span><span class="st">"QuasR"</span><span class="op">)</span></span></code></pre></div>
<p>Bioconductor works on a 6-monthly official release cycle. As with
other Bioconductor packages, there are always two versions of <em><a href="https://bioconductor.org/packages/3.19/QuasR" class="external-link">QuasR</a></em>. Most
users will use the current official release version, which will be
installed by <code><a href="https://bioconductor.github.io/BiocManager/reference/install.html" class="external-link">BiocManager::install</a></code> if you are using the
current release version of <strong>R</strong>. There is also a
development version of <em><a href="https://bioconductor.org/packages/3.19/QuasR" class="external-link">QuasR</a></em> that
includes new features due for the next official release. The development
version will be installed if you are using the development version of
<strong>Bioconductor</strong> (see <code>version = "devel"</code> in
<em><a href="https://bioconductor.org/packages/3.19/BiocManager" class="external-link">BiocManager</a></em>).
The official release version always has an even second number (for
example 1.20.1), whereas the developmental version has an odd second
number (for example 1.21.4).</p>
</div>
<div class="section level3">
<h3 id="loading-of-quasr-and-other-required-packages">Loading of <em>QuasR</em> and other required packages<a class="anchor" aria-label="anchor" href="#loading-of-quasr-and-other-required-packages"></a>
</h3>
<p>In order to run the code examples in this vignette, the <em><a href="https://bioconductor.org/packages/3.19/QuasR" class="external-link">QuasR</a></em>
package and a few additional packages need to be loaded:</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/message.html" class="external-link">suppressPackageStartupMessages</a></span><span class="op">(</span><span class="op">{</span></span>
<span>    <span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">QuasR</span><span class="op">)</span></span>
<span>    <span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://bioconductor.org/packages/BSgenome" class="external-link">BSgenome</a></span><span class="op">)</span></span>
<span>    <span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://bioconductor.org/packages/Rsamtools" class="external-link">Rsamtools</a></span><span class="op">)</span></span>
<span>    <span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">rtracklayer</span><span class="op">)</span></span>
<span>    <span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://bioconductor.org/packages/GenomicFeatures" class="external-link">GenomicFeatures</a></span><span class="op">)</span></span>
<span>    <span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/ivanek/Gviz" class="external-link">Gviz</a></span><span class="op">)</span></span>
<span><span class="op">}</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="how-to-get-help">How to get help<a class="anchor" aria-label="anchor" href="#how-to-get-help"></a>
</h3>
<p>Most questions about <em><a href="https://bioconductor.org/packages/3.19/QuasR" class="external-link">QuasR</a></em> will
hopefully be answered by the documentation or references. If you’ve run
into a question which isn’t addressed by the documentation, or you’ve
found a conflict between the documentation and software, then there is
an active support community which can offer help.</p>
<p>The authors of the package (maintainer: Michael Stadler <a href="mailto:michael.stadler@fmi.ch" class="email">michael.stadler@fmi.ch</a>) always appreciate receiving
reports of bugs in the package functions or in the documentation. The
same goes for well-considered suggestions for improvements.</p>
<p>Any other questions or problems concerning <em><a href="https://bioconductor.org/packages/3.19/QuasR" class="external-link">QuasR</a></em>
should be posted to the Bioconductor support site (<a href="https://support.bioconductor.org" class="external-link uri">https://support.bioconductor.org</a>). Users posting to the
support site for the first time should read the helpful posting guide at
(<a href="https://support.bioconductor.org/info/faq/" class="external-link uri">https://support.bioconductor.org/info/faq/</a>). Note that
each function in <em><a href="https://bioconductor.org/packages/3.19/QuasR" class="external-link">QuasR</a></em> has
it’s own help page, as described in the section @ref(introToR). Posting
etiquette requires that you read the relevant help page carefully before
posting a problem to the site.</p>
</div>
</div>
<div class="section level2">
<h2 id="quick-start">Quick Start<a class="anchor" aria-label="anchor" href="#quick-start"></a>
</h2>
<div class="section level3">
<h3 id="introToR">A brief introduction to <strong>R</strong><a class="anchor" aria-label="anchor" href="#introToR"></a>
</h3>
<p>If you already use <strong>R</strong> and know about its interface,
just skip this section and continue with section
@ref(sampleQuasRsession).</p>
<p>The structure of this vignette and in particular this section is
based on the excellent user guide of the <em><a href="https://bioconductor.org/packages/3.19/limma" class="external-link">limma</a></em>
package, which we would like to hereby acknowledge. <strong>R</strong>
is a program for statistical computing. It is a command-driven language
meaning that you have to type commands into it rather than pointing and
clicking using a mouse. In this guide it will be assumed that you have
successfully downloaded and installed <strong>R</strong> from (<a href="http://www.r-project.org" class="external-link uri">http://www.r-project.org</a>) as well as <em><a href="https://bioconductor.org/packages/3.19/QuasR" class="external-link">QuasR</a></em> (see
section @ref(installation)). A good way to get started is to type</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/help.start.html" class="external-link">help.start</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p>at the <strong>R</strong> prompt or, if you’re using
<strong>R</strong> for Windows, to follow the drop-down menu items
<em>Help</em> <span class="math inline">\(\succ\)</span> <em>Html
help</em>. Following the links <em>Packages</em> <span class="math inline">\(\succ\)</span> <em><a href="https://bioconductor.org/packages/3.19/QuasR" class="external-link">QuasR</a></em> from
the html help page will lead you to the contents page of help topics for
functions in <em><a href="https://bioconductor.org/packages/3.19/QuasR" class="external-link">QuasR</a></em>.</p>
<p>Before you can use any <em><a href="https://bioconductor.org/packages/3.19/QuasR" class="external-link">QuasR</a></em>
commands you have to load the package by typing</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">QuasR</span><span class="op">)</span></span></code></pre></div>
<p>at the <strong>R</strong> prompt. You can get help on any function in
any loaded package by typing <code>?</code> and the function name at the
<strong>R</strong> prompt, for example</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="op">?</span><span class="va">preprocessReads</span></span></code></pre></div>
<p>or equivalently</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/help.html" class="external-link">help</a></span><span class="op">(</span><span class="st">"preprocessReads"</span><span class="op">)</span></span></code></pre></div>
<p>for detailed help on the <code>preprocessReads</code> function. The
function help page is especially useful to learn about its arguments and
its return value.</p>
<p>Working with <strong>R</strong> usually means creating and
transforming <em>objects</em>. Objects might include data sets,
variables, functions, anything at all. For example</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">x</span> <span class="op">&lt;-</span> <span class="fl">2</span></span></code></pre></div>
<p>will create a variable <code>x</code> and will assign it the value 2.
At any stage of your <strong>R</strong> session you can type</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/AnnotationDbi/man/Bimap-envirAPI.html" class="external-link">ls</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p>to get a list of all the objects currently existing in your session.
You can display an object by typing its name on the prompt. The
following displays the object <code>x</code>:</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">x</span></span></code></pre></div>
<p>We hope that you can use <em><a href="https://bioconductor.org/packages/3.19/QuasR" class="external-link">QuasR</a></em>
without having to spend a lot of time learning about the
<strong>R</strong> language itself but a little knowledge in this
direction will be very helpful, especially when you want to do something
not explicitly provided for in <em><a href="https://bioconductor.org/packages/3.19/QuasR" class="external-link">QuasR</a></em> or in
the other Bioconductor packages. For more details about the
<strong>R</strong> language see <em>An Introduction to R</em> which is
available from the online help, or one of the many great online
resources, like the documentation at <a href="https://www.r-project.org/" class="external-link">r-project.org</a>, the growing list of
free books at <a href="https://bioconductor.org/books/release/" class="external-link">bioconductor.org</a>, or
the books from <a href="https://rstudio.com/resources/books/" class="external-link">rstudio.com</a> (many of
which are also available for free). For more background on using
<strong>R</strong> for statistical analysis see <span class="citation">(Dalgaard 2002)</span>.</p>
</div>
<div class="section level3">
<h3 id="sampleQuasRsession">Sample <em>QuasR</em> session<a class="anchor" aria-label="anchor" href="#sampleQuasRsession"></a>
</h3>
<p>This is a quick overview of what an analysis could look like for
users preferring to jump right into an analysis. The example uses data
that is provided with the <em><a href="https://bioconductor.org/packages/3.19/QuasR" class="external-link">QuasR</a></em>
package, which is first copied to the current working directory, into a
subfolder called <code>"extdata"</code>:</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/files.html" class="external-link">file.copy</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span>package <span class="op">=</span> <span class="st">"QuasR"</span>, <span class="st">"extdata"</span><span class="op">)</span>, <span class="st">"."</span>, recursive <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] TRUE</span></span></code></pre>
<p>The sequence files to be analyzed are listed in
<code>sampleFile</code> (see section @ref(sampleFile) for details). The
sequence reads will be aligned using <code>bowtie</code> <span class="citation">(Langmead et al. 2009)</span> (from the <em><a href="https://bioconductor.org/packages/3.19/Rbowtie" class="external-link">Rbowtie</a></em>
package <span class="citation">(Hahne, Lerch, and Stadler 2012)</span>)
to a small reference genome (consisting of three short segments from the
hg19 human genome assembly, available in full for example in the <em><a href="https://bioconductor.org/packages/3.19/BSgenome.Hsapiens.UCSC.hg19" class="external-link">BSgenome.Hsapiens.UCSC.hg19</a></em>
package). Information on selecting an appropriate reference genome is
summarized in section @ref(genome).</p>
<p>Make sure that you have sufficient disk space, both in your
<strong>R</strong> temporary directory (<code><a href="https://rdrr.io/r/base/tempfile.html" class="external-link">tempdir()</a></code>) as well
as to store the resulting alignments (see section @ref(qAlign)).</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sampleFile</span> <span class="op">&lt;-</span> <span class="st">"extdata/samples_chip_single.txt"</span></span>
<span><span class="va">genomeFile</span> <span class="op">&lt;-</span> <span class="st">"extdata/hg19sub.fa"</span></span>
<span></span>
<span><span class="va">proj</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/qAlign.html">qAlign</a></span><span class="op">(</span><span class="va">sampleFile</span>, <span class="va">genomeFile</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Creating .fai file for: /Users/runner/work/QuasR/QuasR/vignettes/extdata/hg19sub.fa</span></span></code></pre>
<pre><code><span><span class="co">## alignment files missing - need to:</span></span></code></pre>
<pre><code><span><span class="co">##     create alignment index for the genome</span></span></code></pre>
<pre><code><span><span class="co">##     create 2 genomic alignment(s)</span></span></code></pre>
<pre><code><span><span class="co">## Creating an Rbowtie index for /Users/runner/work/QuasR/QuasR/vignettes/extdata/hg19sub.fa</span></span></code></pre>
<pre><code><span><span class="co">## Finished creating index</span></span></code></pre>
<pre><code><span><span class="co">## Testing the compute nodes...OK</span></span>
<span><span class="co">## Loading QuasR on the compute nodes...preparing to run on 1 nodes...done</span></span>
<span><span class="co">## Available cores:</span></span>
<span><span class="co">## Mac-1698842773251.local: 1</span></span>
<span><span class="co">## Performing genomic alignments for 2 samples. See progress in the log file:</span></span>
<span><span class="co">## /Users/runner/work/QuasR/QuasR/vignettes/QuasR_log_10c5343ff93dc.txt</span></span>
<span><span class="co">## Genomic alignments have been created successfully</span></span></code></pre>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">proj</span></span></code></pre></div>
<pre><code><span><span class="co">## Project: qProject</span></span>
<span><span class="co">##  Options   : maxHits         : 1</span></span>
<span><span class="co">##              paired          : no</span></span>
<span><span class="co">##              splicedAlignment: FALSE</span></span>
<span><span class="co">##              bisulfite       : no</span></span>
<span><span class="co">##              snpFile         : none</span></span>
<span><span class="co">##              geneAnnotation  : none</span></span>
<span><span class="co">##  Aligner   : Rbowtie v1.43.0 (parameters: -m 1 --best --strata)</span></span>
<span><span class="co">##  Genome    : /Users/runner/work/QuasR/QuasR/vignettes/extdata/hg19sub.fa (file)</span></span>
<span><span class="co">## </span></span>
<span><span class="co">##  Reads     : 2 files, 2 samples (fastq format):</span></span>
<span><span class="co">##    1. chip_1_1.fq.bz2  Sample1 (phred33)</span></span>
<span><span class="co">##    2. chip_2_1.fq.bz2  Sample2 (phred33)</span></span>
<span><span class="co">## </span></span>
<span><span class="co">##  Genome alignments: directory: same as reads</span></span>
<span><span class="co">##    1. chip_1_1_10c5340447364.bam</span></span>
<span><span class="co">##    2. chip_2_1_10c534df3cb32.bam</span></span>
<span><span class="co">## </span></span>
<span><span class="co">##  Aux. alignments: none</span></span></code></pre>
<p>The <code>proj</code> object keeps track of all the information of a
sequencing experiment, for example where sequence and alignment files
are stored, and what aligner and reference genome was used to generate
the alignments.</p>
<p>Now that the alignments have been generated, further analyses can be
performed. A quality control report is saved to the
<code>"extdata/qc_report.pdf"</code> file using the
<code>qQCReport</code> function.</p>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/qQCReport.html">qQCReport</a></span><span class="op">(</span><span class="va">proj</span>, <span class="st">"extdata/qc_report.pdf"</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## collecting quality control data</span></span></code></pre>
<pre><code><span><span class="co">## creating QC plots</span></span></code></pre>
<p>The number of alignments per promoter region is quantified using
<code>qCount</code>. Genomic coordinates for promoter regions are
imported from a gtf file (<code>annotFile</code>) into the
<code>GRanges</code>-object with the name <code>promReg</code>:</p>
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">rtracklayer</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://bioconductor.org/packages/GenomicFeatures" class="external-link">GenomicFeatures</a></span><span class="op">)</span></span>
<span><span class="va">annotFile</span> <span class="op">&lt;-</span> <span class="st">"extdata/hg19sub_annotation.gtf"</span></span>
<span><span class="va">txStart</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/rtracklayer/man/GFFFile-class.html" class="external-link">import.gff</a></span><span class="op">(</span><span class="va">annotFile</span>, format <span class="op">=</span> <span class="st">"gtf"</span>, feature.type <span class="op">=</span> <span class="st">"start_codon"</span><span class="op">)</span></span>
<span><span class="va">promReg</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/IRanges/man/intra-range-methods.html" class="external-link">promoters</a></span><span class="op">(</span><span class="va">txStart</span>, upstream <span class="op">=</span> <span class="fl">500</span>, downstream <span class="op">=</span> <span class="fl">500</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">promReg</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Vector-class.html" class="external-link">mcols</a></span><span class="op">(</span><span class="va">promReg</span><span class="op">)</span><span class="op">$</span><span class="va">transcript_name</span></span>
<span></span>
<span><span class="va">promCounts</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/qCount.html">qCount</a></span><span class="op">(</span><span class="va">proj</span>, query <span class="op">=</span> <span class="va">promReg</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## counting alignments...done</span></span></code></pre>
<div class="sourceCode" id="cb29"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">promCounts</span></span></code></pre></div>
<pre><code><span><span class="co">##                width Sample1 Sample2</span></span>
<span><span class="co">## TNFRSF18-003    1000      20       4</span></span>
<span><span class="co">## TNFRSF18-002    1000      20       4</span></span>
<span><span class="co">## TNFRSF18-001    1000      20       4</span></span>
<span><span class="co">## TNFRSF4-001     1000       5       2</span></span>
<span><span class="co">## SDF4-007        1000       8       2</span></span>
<span><span class="co">## SDF4-001        1000       8       2</span></span>
<span><span class="co">## SDF4-002        1000       8       2</span></span>
<span><span class="co">## SDF4-201        1000       8       2</span></span>
<span><span class="co">## B3GALT6-001     1000      25     274</span></span>
<span><span class="co">## RPS7-001        1000     121     731</span></span>
<span><span class="co">## RPS7-008        1000     121     731</span></span>
<span><span class="co">## RPS7-009        1000     121     731</span></span>
<span><span class="co">## RPS7-005        1000     121     731</span></span>
<span><span class="co">## C3orf10-201     1000     176     496</span></span>
<span><span class="co">## C3orf10-001     1000     176     496</span></span>
<span><span class="co">## AC034193.1-201  1000       5       2</span></span>
<span><span class="co">## VHL-001         1000      61     336</span></span>
<span><span class="co">## VHL-002         1000      61     336</span></span>
<span><span class="co">## VHL-201         1000      61     336</span></span></code></pre>
</div>
</div>
<div class="section level2">
<h2 id="quasr-overview">
<em>QuasR</em> Overview<a class="anchor" aria-label="anchor" href="#quasr-overview"></a>
</h2>
<p>The following scheme shows the major components of <em><a href="https://bioconductor.org/packages/3.19/QuasR" class="external-link">QuasR</a></em> and
their relationships: <img src="QuasR-scheme.png"></p>
<p><em><a href="https://bioconductor.org/packages/3.19/QuasR" class="external-link">QuasR</a></em> works
with data (sequences and alignments, reference genome, etc.) that are
stored as files on your storage (the gray cylinder on the lower left of
Figure above, see section @ref(fileStorageLocations) for details on
storage locations). <em><a href="https://bioconductor.org/packages/3.19/QuasR" class="external-link">QuasR</a></em> does
not need a database management system, or these files to be named and
organized according to a specific scheme.</p>
<p>In order to keep track of directory paths during an analysis, <em><a href="https://bioconductor.org/packages/3.19/QuasR" class="external-link">QuasR</a></em> makes
use of a <code>qProject</code> object that is returned by the
<code>qAlign</code> function, which at the minimum requires two inputs:
the name of a samples text file (see section @ref(sampleFile) for
details), and the reference genome for the alignments (see section
@ref(genome)).</p>
<p>The <code>qProject</code> object is the main argument passed to
subsequent functions such as <code>qQCReport</code> and
<code>qCount</code>. The <code>qProject</code> object contains all
necessary information on the current project and eliminates the need to
repeatedly enter the same information. All functions that work on
<code>qProject</code> objects can be recognized by their names starting
with the letter <em>q</em>.</p>
<p>Read quantification (apart from quantification of methylation which
has its own function <code>qMeth</code>) is done using the
<code>qCount</code> function: It counts the alignments in regions of
interest (e.g. promoters, genes, exons, etc.) and produces a count table
(regions in rows, samples in columns) for further visualization and
analysis. The count table can also be used as input to a statistical
analysis using packages such as <em><a href="https://bioconductor.org/packages/3.19/edgeR" class="external-link">edgeR</a></em> <span class="citation">(Robinson, McCarthy, and Smyth 2010)</span>, <em><a href="https://bioconductor.org/packages/3.19/DESeq" class="external-link">DESeq</a></em> <span class="citation">(Anders and Huber 2010)</span>, <em><a href="https://bioconductor.org/packages/3.19/DESeq2" class="external-link">DESeq2</a></em>
<span class="citation">(Love, Huber, and Anders 2014)</span>, <em><a href="https://bioconductor.org/packages/3.19/TCC" class="external-link">TCC</a></em> <span class="citation">(Sun et al. 2013)</span>, <em><a href="https://bioconductor.org/packages/3.19/DEXSeq" class="external-link">DEXSeq</a></em>
<span class="citation">(Anders, Reyes, and Huber 2012)</span> or <em><a href="https://bioconductor.org/packages/3.19/baySeq" class="external-link">baySeq</a></em>
<span class="citation">(Hardcastle and Kelly 2010)</span>.</p>
<p>In summary, a typical <em><a href="https://bioconductor.org/packages/3.19/QuasR" class="external-link">QuasR</a></em>
analysis consists of the following steps (some of them are
optional):</p>
<ul>
<li>
<code>preprocessReads</code> (optional): Remove adapters from start
or end of reads, filter out reads of low quality, short length or low
complexity (section @ref(preProcessing)).<br>
</li>
<li>Prepare <em>samples file</em>: List sequence files or alignments,
provide sample names (section @ref(sampleFile)).<br>
</li>
<li>Prepare <em>auxiliary file</em> (optional): List additional
reference sequences for alignment of reads not matching the reference
genome (section @ref(auxiliaryFile)).<br>
</li>
<li>
<code>qAlign</code>: Create <code>qProject</code> object and specify
project parameters. Also download BSgenome package, create aligner
indices and align reads if not already existing (section
@ref(qAlign)).<br>
</li>
<li>
<code>qQCReport</code> (optional): Create quality control report
with plots on sequence qualities and alignment statistics (section
@ref(qQCReport)).<br>
</li>
<li>
<code>qExportWig</code> (optional): Export genomic alignments as
wiggle tracks for genome browser visualization (section
@ref(qExportWig)).<br>
</li>
<li>
<code>qCount</code>: Quantify alignments in regions of interest
(section @ref(qCount)).</li>
</ul>
<p>Recurrent example tasks that may be part of any typical analysis are
described in section @ref(exampleTasks). Example workflows for specific
experiment types (ChIP-seq, RNA-seq and Bis-seq) are described in
section @ref(exampleWorkflows).</p>
<div class="section level3">
<h3 id="fileStorageLocations">File storage locations<a class="anchor" aria-label="anchor" href="#fileStorageLocations"></a>
</h3>
<p>Apart from <code>qExportWig</code> and <code>qQCReport</code>, which
generate wig files and pdf reports, <code>qAlign</code> is the only
function in <em><a href="https://bioconductor.org/packages/3.19/QuasR" class="external-link">QuasR</a></em> that
stores files on the disk (see section @ref(qAlign) for details). All
files generated by <code>qAlign</code> are listed here by type, together
with their default location and how locations can be changed.</p>
<ul>
<li>
<em>Temporary files</em> (default: <code><a href="https://rdrr.io/r/base/tempfile.html" class="external-link">tempdir()</a></code>):
Temporary files include reference genomes in <code>fasta</code> format,
decompressed input sequence files, and temporary alignments in text
format, and can require a large amount of disk space. By default, these
files will be written to the temporary directory of the
<strong>R</strong> process (as reported by <code><a href="https://rdrr.io/r/base/tempfile.html" class="external-link">tempdir()</a></code>). If
using <code>clObj</code> for parallel processing, this may be the
<code><a href="https://rdrr.io/r/base/tempfile.html" class="external-link">tempdir()</a></code> from the cluster node(s). An alternative location
can be set using the <code>TMPDIR</code> environment variable (see
<code><a href="https://rdrr.io/r/base/tempfile.html" class="external-link">?tempdir</a></code>).<br>
</li>
<li>
<em>Alignment files (<code>bam</code> format)</em> (default: same
directory as the input sequence files): Alignments against reference
genome and auxiliary targets are stored in <code>bam</code> format in
the same directory that also contains the input sequence file (listed in
<code>sampleFile</code>). Please note that if the input sequence file
corresponds to a symbolic link, <em><a href="https://bioconductor.org/packages/3.19/QuasR" class="external-link">QuasR</a></em> will
follow the link and use the directory of the original file instead. An
alternative directory can be specified with the
<code>alignmentsDir</code> argument from <code>qAlign</code>, which will
store all <code>bam</code> files in that directory even if the input
sequence files are located in different directories.<br>
</li>
<li>
<em>Alignment index files</em> (default: depends on
<code>genome</code> and <code>snpFile</code> arguments): Many alignment
tools including <code>bowtie</code> require an index of the reference
sequence to perform alignments. If necessary, <code>qAlign</code> will
build this index automatically and store it in a default location that
depends on the <code>genome</code> argument:
<ul>
<li>
<code>BSgenome</code>: If <code>genome</code> is the name of a
<em><a href="https://bioconductor.org/packages/3.19/BSgenome" class="external-link">BSgenome</a></em>
package (such as <code>"BSgenome.Hsapiens.UCSC.hg19"</code>), the index
will be stored as a new <strong>R</strong> package in the default
library path (as reported by <code>.libPaths()[1]</code>, see
<code><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">?install.packages</a></code> for details), or alternatively in the
library specified by the <code>lib.loc</code> argument of
<code>qAlign</code>. The name of this index package will be the name of
the original <em><a href="https://bioconductor.org/packages/3.19/BSgenome" class="external-link">BSgenome</a></em>
package with a suffix for the index type, for example
<code>"BSgenome.Hsapiens.UCSC.hg19.Rbowtie"</code>.<br>
</li>
<li>
<code>fasta</code>: If <code>genome</code> refers to a reference
genome file in <code>fasta</code> format, the index will be stored in a
subdirectory at the same location. Similarly, the indices for files
listed in <code>auxiliaryFile</code> are store at the location of these
files. For example, the <code>Rbowtie</code> index for the genome at
<code>"./genome/mm9.fa"</code> is stored in
<code>"./genome/mm9.fa.Rbowtie"</code>.<br>
</li>
<li>
<em>Allele-specific analysis</em>: A special case is the
allele-specific analysis, where reference and alternative alleles listed
in <code>snpFile</code> (e.g. <code>"./mySNPs.tab"</code>) are injected
into the <code>genome</code>
(e.g. <code>"BSgenome.Mmusculus.UCSC.mm9"</code>) to create two variant
genomes to be indexed. These indices are saved at the location of the
<code>snpFile</code> in a directory named after <code>snpFile</code>,
<code>genome</code> and the index type
(e.g. <code>"./mySNPs.tab.BSgenome.Mmusculus.UCSC.mm9.A.fa.Rbowtie"</code>).</li>
</ul>
</li>
</ul>
</div>
</div>
<div class="section level2">
<h2 id="exampleTasks">Example tasks<a class="anchor" aria-label="anchor" href="#exampleTasks"></a>
</h2>
<div class="section level3">
<h3 id="sampleFile">Create a sample file<a class="anchor" aria-label="anchor" href="#sampleFile"></a>
</h3>
<p>The sample file is a tab-delimited text file with two or three
columns. The first row contains the column names: For a single read
experiment, these are ‘FileName’ and ‘SampleName’; for a paired-end
experiment, these are ‘FileName1’, ‘FileName2’ and ‘SampleName’. If the
first row does not contain the correctly spelled column names, <em><a href="https://bioconductor.org/packages/3.19/QuasR" class="external-link">QuasR</a></em> will
not accept the samples file. Subsequent rows contain the input sequence
files.</p>
<p>Here are examples of such sample files for a single read
experiment:</p>
<pre>
FileName    SampleName
chip_1_1.fq.bz2 Sample1
chip_2_1.fq.bz2 Sample2
</pre>
<!--
------------------- | -------------
`FileName`          | `SampleName`  
`chip_1_1.fq.bz2`   | `Sample1`  
`chip_2_1.fq.bz2`   | `Sample2`  
-->
<p>and for a paired-end experiment:</p>
<pre>
FileName1   FileName2   SampleName
rna_1_1.fq.bz2  rna_1_2.fq.bz2  Sample1
rna_2_1.fq.bz2  rna_2_2.fq.bz2  Sample2
</pre>
<!--
`FileName1`        | `FileName2`        | `SampleName`  
`rna_1_1.fq.bz2`   | `rna_1_2.fq.bz2`   | `Sample1`  
`rna_2_1.fq.bz2`   | `rna_2_2.fq.bz2`   | `Sample2`  
-->
<p>These example files are also contained in the <em><a href="https://bioconductor.org/packages/3.19/QuasR" class="external-link">QuasR</a></em>
package and may be used as templates. The path of the files can be
determined using:</p>
<div class="sourceCode" id="cb31"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sampleFile1</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span>package<span class="op">=</span><span class="st">"QuasR"</span>, <span class="st">"extdata"</span>,</span>
<span>                           <span class="st">"samples_chip_single.txt"</span><span class="op">)</span></span>
<span><span class="va">sampleFile2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span>package<span class="op">=</span><span class="st">"QuasR"</span>, <span class="st">"extdata"</span>,</span>
<span>                           <span class="st">"samples_rna_paired.txt"</span><span class="op">)</span></span></code></pre></div>
<p>The columns <em>FileName</em> for single-read, or <em>FileName1</em>
and <em>FileName2</em> for paired-end experiments contain paths and
names to files containing the sequence data. The paths can be absolute
or relative to the location of the sample file. This allows combining
files from different directories in a single analysis. For each input
sequence file, <code>qAlign</code> will create one alignment file and by
default store it in the same directory as the sequence file. Already
existing alignment files with identical parameters will not be
re-created, so that it is easy to reuse the same sequence files in
multiple projects without unnecessarily copying sequence files or
recreating alignments.</p>
<p>The <em>SampleName</em> column contains sample names for each
sequence file. The same name can be used on several lines to indicate
multiple sequence files that belong to the same sample
(<code>qCount</code> can use this information to automatically combine
counts for one sample from multiple files).</p>
<p>Three file formats are supported for input files (but cannot be mixed
within a single sample file):</p>
<ul>
<li>
<strong>fasta</strong> files have names that end with ‘.fa’, ‘.fna’
or ‘.fasta’. They contain only sequences (and no base qualities) and
will thus by default be aligned on the basis of mismatches (the best
alignment is the one with fewest mismatches).<br>
</li>
<li>
<strong>fastq</strong> files have names that end with ‘.fq’ or
‘.fastq’. They contain sequences and corresponding base qualities and
will be aligned by default using these qualities. The encoding scheme of
base qualities is automatically detected for each individual fastq
file.<br>
</li>
<li>
<strong>bam</strong> files have names that end with ‘.bam’. They can
be used if the sequence reads have already been aligned outside of
<em><a href="https://bioconductor.org/packages/3.19/QuasR" class="external-link">QuasR</a></em>, and
<em><a href="https://bioconductor.org/packages/3.19/QuasR" class="external-link">QuasR</a></em> will
only be used for downstream analysis based on the alignments contained
in the <code>bam</code> files. This makes it possible to use alignment
tools that are not available within <em><a href="https://bioconductor.org/packages/3.19/QuasR" class="external-link">QuasR</a></em>, but
making use of this option comes with a risk and should only be used by
experienced users. For example, it cannot be guaranteed any more that
certain assumptions made by <code>qCount</code> are fulfilled by the
external aligner (see below). When using external <code>bam</code>
files, we recommend to use files which contain only one alignment per
read. This may also include multi-hit reads, for which one of the
alignments is randomly selected. This allows <em><a href="https://bioconductor.org/packages/3.19/QuasR" class="external-link">QuasR</a></em> to
count the total number of reads by counting the total number of
alignments. Furthermore, if the <code>bam</code> files also contain the
unmapped reads, <em><a href="https://bioconductor.org/packages/3.19/QuasR" class="external-link">QuasR</a></em> will
be able to calculate the fraction of mapped reads. For bisulfite samples
we require ungapped alignments stored in unpaired or paired <em>ff</em>
orientation (even if the input reads are <em>fr</em>). For
allele-specific <code>bam</code> files, <em><a href="https://bioconductor.org/packages/3.19/QuasR" class="external-link">QuasR</a></em>
requires an additional tag for each alignment called <code>XV</code> of
type <code>A</code> (printable character) with the possible values
<code>R</code> (Reference), <code>U</code> (Unknown) and <code>A</code>
(Alternative).</li>
</ul>
<p><strong>fasta</strong> and <strong>fastq</strong> files can be
compressed with gzip, bzip2 or xz (file extensions ‘.gz’, ‘.bz2’ or
‘xz’, respectively) and will be automatically decompressed when
necessary.</p>
<div class="section level4">
<h4 id="working-only-with-bam-files-after-performing-alignments">Working only with <code>bam</code> files after performing
alignments<a class="anchor" aria-label="anchor" href="#working-only-with-bam-files-after-performing-alignments"></a>
</h4>
<p>Once alignments have been created, most analyses will only require
the <code>bam</code> files and will not access the original raw sequence
files anymore. However, re-creating a <code>qProject</code> object by a
later identical call to <code>qAlign</code> will still need access to
the raw sequences to verify consistency between raw data and alignments.
It may be desirable to remove this dependency, for example to archive or
move away the raw sequence files and to reclaim used disk space.</p>
<p>This can be achieved using the following procedure involving two
sequential calls to <code>qAlign</code>. First, <code>qAlign</code> is
called with the orignial sample file (<code>sampleFile1</code>) that
lists the raw sequence files, and subsequently with a second sample file
(<code>sampleFile2</code>) that lists the <code>bam</code> files
generated in the first call. Such a second sample file can be easily
generated given the <code>qProject</code> object (<code>proj1</code>)
returned by the first call:</p>
<div class="sourceCode" id="cb32"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sampleFile1</span> <span class="op">&lt;-</span> <span class="st">"samples_fastq.txt"</span></span>
<span><span class="va">sampleFile2</span> <span class="op">&lt;-</span> <span class="st">"samples_bam.txt"</span></span>
<span></span>
<span><span class="va">proj1</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/qAlign.html">qAlign</a></span><span class="op">(</span><span class="va">sampleFile1</span>, <span class="va">genomeFile</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/write.table.html" class="external-link">write.table</a></span><span class="op">(</span><span class="fu"><a href="../reference/qProject-class.html">alignments</a></span><span class="op">(</span><span class="va">proj1</span><span class="op">)</span><span class="op">$</span><span class="va">genome</span>, <span class="va">sampleFile2</span>, sep<span class="op">=</span><span class="st">"\t"</span>, row.names<span class="op">=</span><span class="cn">FALSE</span><span class="op">)</span></span>
<span></span>
<span><span class="va">proj2</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/qAlign.html">qAlign</a></span><span class="op">(</span><span class="va">sampleFile2</span>, <span class="va">genomeFile</span><span class="op">)</span></span></code></pre></div>
<p>The analysis can now be exclusively based on the <code>bam</code>
files using <code>sampleFile2</code> and <code>proj2</code>.</p>
</div>
<div class="section level4">
<h4 id="consistency-of-samples-within-a-project">Consistency of samples within a project<a class="anchor" aria-label="anchor" href="#consistency-of-samples-within-a-project"></a>
</h4>
<p>The sample file implicitly defines the type of samples contained in
the project: <em>single read</em> or <em>paired-end read</em>, sequences
<em>with</em> or <em>without</em> qualities. This type will have a
profound impact on the downstream analysis. For example, it controls
whether alignments will be performed in single or paired-end mode,
either with or without base qualities. That will also determine
availability of certain options for quality control and quantification
in <code>qQCReport</code> and <code>qCount</code>. For consistency, it
is therefore required that all samples within a project have the same
type; it is not possible to mix both single and paired-end read samples,
or <strong>fasta</strong> and <strong>fastq</strong> files in a single
project (sample file). If necessary, it may be possible to analyse
different types of files in separate <em><a href="https://bioconductor.org/packages/3.19/QuasR" class="external-link">QuasR</a></em>
projects and combine the derived results at the end.</p>
</div>
</div>
<div class="section level3">
<h3 id="auxiliaryFile">Create an auxiliary file (optional)<a class="anchor" aria-label="anchor" href="#auxiliaryFile"></a>
</h3>
<p>By default <em><a href="https://bioconductor.org/packages/3.19/QuasR" class="external-link">QuasR</a></em>
aligns reads only to the reference genome. However, it may be
interesting to align non-matching reads to further targets, for example
to identify contamination from vectors or a different species, or in
order to quantify spike-in material not contained in the reference
genome. In <em><a href="https://bioconductor.org/packages/3.19/QuasR" class="external-link">QuasR</a></em>, such
supplementary reference files are called <em>auxiliary</em> references
and can be specified to <code>qAlign</code> using the
<code>auxiliaryFile</code> argument (see section @ref(qAlign) for
details). The format of the auxiliary file is similar to the one of the
sample file described in section @ref(sampleFile): It contains two
columns with column names ‘FileName’ and ‘AuxName’ in the first row.
Additional rows contain names and files of one or several auxiliary
references in <code>fasta</code> format.</p>
<p>An example auxiliary file looks like this:</p>
<pre>
FileName    AuxName
NC_001422.1.fa  phiX174
</pre>
<!--
`FileName`        | `AuxName`  
`NC_001422.1.fa`  | `phiX174`  
-->
<p>and is available from your <em><a href="https://bioconductor.org/packages/3.19/QuasR" class="external-link">QuasR</a></em>
installation at</p>
<div class="sourceCode" id="cb33"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">auxFile</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span>package <span class="op">=</span> <span class="st">"QuasR"</span>, <span class="st">"extdata"</span>, <span class="st">"auxiliaries.txt"</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="genome">Select the reference genome<a class="anchor" aria-label="anchor" href="#genome"></a>
</h3>
<p>Sequence reads are primarily aligned against the reference genome
(see section @ref(redundancy) on how to choose a suitable reference
assembly). If necessary, <em><a href="https://bioconductor.org/packages/3.19/QuasR" class="external-link">QuasR</a></em> will
create an aligner index for the genome. The reference genome can be
provided in one of two different formats:</p>
<ul>
<li>
<strong>a string</strong>, referring to the name of a <em><a href="https://bioconductor.org/packages/3.19/BSgenome" class="external-link">BSgenome</a></em>
package:</li>
</ul>
<div class="sourceCode" id="cb34"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/BSgenome/man/available.genomes.html" class="external-link">available.genomes</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## 'getOption("repos")' replaces Bioconductor standard repositories, see</span></span>
<span><span class="co">## 'help("repositories", package = "BiocManager")' for details.</span></span>
<span><span class="co">## Replacement repositories:</span></span>
<span><span class="co">##     CRAN: https://cran.rstudio.com</span></span></code></pre>
<pre><code><span><span class="co">##   [1] "BSgenome.Alyrata.JGI.v1"                           </span></span>
<span><span class="co">##   [2] "BSgenome.Amellifera.BeeBase.assembly4"             </span></span>
<span><span class="co">##   [3] "BSgenome.Amellifera.NCBI.AmelHAv3.1"               </span></span>
<span><span class="co">##   [4] "BSgenome.Amellifera.UCSC.apiMel2"                  </span></span>
<span><span class="co">##   [5] "BSgenome.Amellifera.UCSC.apiMel2.masked"           </span></span>
<span><span class="co">##   [6] "BSgenome.Aofficinalis.NCBI.V1"                     </span></span>
<span><span class="co">##   [7] "BSgenome.Athaliana.TAIR.04232008"                  </span></span>
<span><span class="co">##   [8] "BSgenome.Athaliana.TAIR.TAIR9"                     </span></span>
<span><span class="co">##   [9] "BSgenome.Btaurus.UCSC.bosTau3"                     </span></span>
<span><span class="co">##  [10] "BSgenome.Btaurus.UCSC.bosTau3.masked"              </span></span>
<span><span class="co">##  [11] "BSgenome.Btaurus.UCSC.bosTau4"                     </span></span>
<span><span class="co">##  [12] "BSgenome.Btaurus.UCSC.bosTau4.masked"              </span></span>
<span><span class="co">##  [13] "BSgenome.Btaurus.UCSC.bosTau6"                     </span></span>
<span><span class="co">##  [14] "BSgenome.Btaurus.UCSC.bosTau6.masked"              </span></span>
<span><span class="co">##  [15] "BSgenome.Btaurus.UCSC.bosTau8"                     </span></span>
<span><span class="co">##  [16] "BSgenome.Btaurus.UCSC.bosTau9"                     </span></span>
<span><span class="co">##  [17] "BSgenome.Btaurus.UCSC.bosTau9.masked"              </span></span>
<span><span class="co">##  [18] "BSgenome.Carietinum.NCBI.v1"                       </span></span>
<span><span class="co">##  [19] "BSgenome.Celegans.UCSC.ce10"                       </span></span>
<span><span class="co">##  [20] "BSgenome.Celegans.UCSC.ce11"                       </span></span>
<span><span class="co">##  [21] "BSgenome.Celegans.UCSC.ce2"                        </span></span>
<span><span class="co">##  [22] "BSgenome.Celegans.UCSC.ce6"                        </span></span>
<span><span class="co">##  [23] "BSgenome.Cfamiliaris.UCSC.canFam2"                 </span></span>
<span><span class="co">##  [24] "BSgenome.Cfamiliaris.UCSC.canFam2.masked"          </span></span>
<span><span class="co">##  [25] "BSgenome.Cfamiliaris.UCSC.canFam3"                 </span></span>
<span><span class="co">##  [26] "BSgenome.Cfamiliaris.UCSC.canFam3.masked"          </span></span>
<span><span class="co">##  [27] "BSgenome.Cjacchus.UCSC.calJac3"                    </span></span>
<span><span class="co">##  [28] "BSgenome.Cjacchus.UCSC.calJac4"                    </span></span>
<span><span class="co">##  [29] "BSgenome.CneoformansVarGrubiiKN99.NCBI.ASM221672v1"</span></span>
<span><span class="co">##  [30] "BSgenome.Creinhardtii.JGI.v5.6"                    </span></span>
<span><span class="co">##  [31] "BSgenome.Dmelanogaster.UCSC.dm2"                   </span></span>
<span><span class="co">##  [32] "BSgenome.Dmelanogaster.UCSC.dm2.masked"            </span></span>
<span><span class="co">##  [33] "BSgenome.Dmelanogaster.UCSC.dm3"                   </span></span>
<span><span class="co">##  [34] "BSgenome.Dmelanogaster.UCSC.dm3.masked"            </span></span>
<span><span class="co">##  [35] "BSgenome.Dmelanogaster.UCSC.dm6"                   </span></span>
<span><span class="co">##  [36] "BSgenome.Drerio.UCSC.danRer10"                     </span></span>
<span><span class="co">##  [37] "BSgenome.Drerio.UCSC.danRer11"                     </span></span>
<span><span class="co">##  [38] "BSgenome.Drerio.UCSC.danRer5"                      </span></span>
<span><span class="co">##  [39] "BSgenome.Drerio.UCSC.danRer5.masked"               </span></span>
<span><span class="co">##  [40] "BSgenome.Drerio.UCSC.danRer6"                      </span></span>
<span><span class="co">##  [41] "BSgenome.Drerio.UCSC.danRer6.masked"               </span></span>
<span><span class="co">##  [42] "BSgenome.Drerio.UCSC.danRer7"                      </span></span>
<span><span class="co">##  [43] "BSgenome.Drerio.UCSC.danRer7.masked"               </span></span>
<span><span class="co">##  [44] "BSgenome.Dvirilis.Ensembl.dvircaf1"                </span></span>
<span><span class="co">##  [45] "BSgenome.Ecoli.NCBI.20080805"                      </span></span>
<span><span class="co">##  [46] "BSgenome.Gaculeatus.UCSC.gasAcu1"                  </span></span>
<span><span class="co">##  [47] "BSgenome.Gaculeatus.UCSC.gasAcu1.masked"           </span></span>
<span><span class="co">##  [48] "BSgenome.Ggallus.UCSC.galGal3"                     </span></span>
<span><span class="co">##  [49] "BSgenome.Ggallus.UCSC.galGal3.masked"              </span></span>
<span><span class="co">##  [50] "BSgenome.Ggallus.UCSC.galGal4"                     </span></span>
<span><span class="co">##  [51] "BSgenome.Ggallus.UCSC.galGal4.masked"              </span></span>
<span><span class="co">##  [52] "BSgenome.Ggallus.UCSC.galGal5"                     </span></span>
<span><span class="co">##  [53] "BSgenome.Ggallus.UCSC.galGal6"                     </span></span>
<span><span class="co">##  [54] "BSgenome.Gmax.NCBI.Gmv40"                          </span></span>
<span><span class="co">##  [55] "BSgenome.Hsapiens.1000genomes.hs37d5"              </span></span>
<span><span class="co">##  [56] "BSgenome.Hsapiens.NCBI.GRCh38"                     </span></span>
<span><span class="co">##  [57] "BSgenome.Hsapiens.NCBI.T2T.CHM13v2.0"              </span></span>
<span><span class="co">##  [58] "BSgenome.Hsapiens.UCSC.hg17"                       </span></span>
<span><span class="co">##  [59] "BSgenome.Hsapiens.UCSC.hg17.masked"                </span></span>
<span><span class="co">##  [60] "BSgenome.Hsapiens.UCSC.hg18"                       </span></span>
<span><span class="co">##  [61] "BSgenome.Hsapiens.UCSC.hg18.masked"                </span></span>
<span><span class="co">##  [62] "BSgenome.Hsapiens.UCSC.hg19"                       </span></span>
<span><span class="co">##  [63] "BSgenome.Hsapiens.UCSC.hg19.masked"                </span></span>
<span><span class="co">##  [64] "BSgenome.Hsapiens.UCSC.hg38"                       </span></span>
<span><span class="co">##  [65] "BSgenome.Hsapiens.UCSC.hg38.dbSNP151.major"        </span></span>
<span><span class="co">##  [66] "BSgenome.Hsapiens.UCSC.hg38.dbSNP151.minor"        </span></span>
<span><span class="co">##  [67] "BSgenome.Hsapiens.UCSC.hg38.masked"                </span></span>
<span><span class="co">##  [68] "BSgenome.Hsapiens.UCSC.hs1"                        </span></span>
<span><span class="co">##  [69] "BSgenome.Mdomestica.UCSC.monDom5"                  </span></span>
<span><span class="co">##  [70] "BSgenome.Mfascicularis.NCBI.5.0"                   </span></span>
<span><span class="co">##  [71] "BSgenome.Mfascicularis.NCBI.6.0"                   </span></span>
<span><span class="co">##  [72] "BSgenome.Mfuro.UCSC.musFur1"                       </span></span>
<span><span class="co">##  [73] "BSgenome.Mmulatta.UCSC.rheMac10"                   </span></span>
<span><span class="co">##  [74] "BSgenome.Mmulatta.UCSC.rheMac2"                    </span></span>
<span><span class="co">##  [75] "BSgenome.Mmulatta.UCSC.rheMac2.masked"             </span></span>
<span><span class="co">##  [76] "BSgenome.Mmulatta.UCSC.rheMac3"                    </span></span>
<span><span class="co">##  [77] "BSgenome.Mmulatta.UCSC.rheMac3.masked"             </span></span>
<span><span class="co">##  [78] "BSgenome.Mmulatta.UCSC.rheMac8"                    </span></span>
<span><span class="co">##  [79] "BSgenome.Mmusculus.UCSC.mm10"                      </span></span>
<span><span class="co">##  [80] "BSgenome.Mmusculus.UCSC.mm10.masked"               </span></span>
<span><span class="co">##  [81] "BSgenome.Mmusculus.UCSC.mm39"                      </span></span>
<span><span class="co">##  [82] "BSgenome.Mmusculus.UCSC.mm8"                       </span></span>
<span><span class="co">##  [83] "BSgenome.Mmusculus.UCSC.mm8.masked"                </span></span>
<span><span class="co">##  [84] "BSgenome.Mmusculus.UCSC.mm9"                       </span></span>
<span><span class="co">##  [85] "BSgenome.Mmusculus.UCSC.mm9.masked"                </span></span>
<span><span class="co">##  [86] "BSgenome.Osativa.MSU.MSU7"                         </span></span>
<span><span class="co">##  [87] "BSgenome.Ppaniscus.UCSC.panPan1"                   </span></span>
<span><span class="co">##  [88] "BSgenome.Ppaniscus.UCSC.panPan2"                   </span></span>
<span><span class="co">##  [89] "BSgenome.Ptroglodytes.UCSC.panTro2"                </span></span>
<span><span class="co">##  [90] "BSgenome.Ptroglodytes.UCSC.panTro2.masked"         </span></span>
<span><span class="co">##  [91] "BSgenome.Ptroglodytes.UCSC.panTro3"                </span></span>
<span><span class="co">##  [92] "BSgenome.Ptroglodytes.UCSC.panTro3.masked"         </span></span>
<span><span class="co">##  [93] "BSgenome.Ptroglodytes.UCSC.panTro5"                </span></span>
<span><span class="co">##  [94] "BSgenome.Ptroglodytes.UCSC.panTro6"                </span></span>
<span><span class="co">##  [95] "BSgenome.Rnorvegicus.UCSC.rn4"                     </span></span>
<span><span class="co">##  [96] "BSgenome.Rnorvegicus.UCSC.rn4.masked"              </span></span>
<span><span class="co">##  [97] "BSgenome.Rnorvegicus.UCSC.rn5"                     </span></span>
<span><span class="co">##  [98] "BSgenome.Rnorvegicus.UCSC.rn5.masked"              </span></span>
<span><span class="co">##  [99] "BSgenome.Rnorvegicus.UCSC.rn6"                     </span></span>
<span><span class="co">## [100] "BSgenome.Rnorvegicus.UCSC.rn7"                     </span></span>
<span><span class="co">## [101] "BSgenome.Scerevisiae.UCSC.sacCer1"                 </span></span>
<span><span class="co">## [102] "BSgenome.Scerevisiae.UCSC.sacCer2"                 </span></span>
<span><span class="co">## [103] "BSgenome.Scerevisiae.UCSC.sacCer3"                 </span></span>
<span><span class="co">## [104] "BSgenome.Sscrofa.UCSC.susScr11"                    </span></span>
<span><span class="co">## [105] "BSgenome.Sscrofa.UCSC.susScr3"                     </span></span>
<span><span class="co">## [106] "BSgenome.Sscrofa.UCSC.susScr3.masked"              </span></span>
<span><span class="co">## [107] "BSgenome.Tgondii.ToxoDB.7.0"                       </span></span>
<span><span class="co">## [108] "BSgenome.Tguttata.UCSC.taeGut1"                    </span></span>
<span><span class="co">## [109] "BSgenome.Tguttata.UCSC.taeGut1.masked"             </span></span>
<span><span class="co">## [110] "BSgenome.Tguttata.UCSC.taeGut2"                    </span></span>
<span><span class="co">## [111] "BSgenome.Vvinifera.URGI.IGGP12Xv0"                 </span></span>
<span><span class="co">## [112] "BSgenome.Vvinifera.URGI.IGGP12Xv2"                 </span></span>
<span><span class="co">## [113] "BSgenome.Vvinifera.URGI.IGGP8X"</span></span></code></pre>
<div class="sourceCode" id="cb37"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">genomeName</span> <span class="op">&lt;-</span> <span class="st">"BSgenome.Hsapiens.UCSC.hg19"</span></span></code></pre></div>
<p>In this example, the BSgenome package
<code>"BSgenome.Hsapiens.UCSC.hg19"</code> refers to an unmasked genome;
alignment index and alignments will be performed on the full unmasked
genome sequence (recommended). If using a masked genome
(e.g. <code>"BSgenome.Hsapiens.UCSC.hg19.masked"</code>), masked regions
will be replaced with <code>"N"</code> bases, and this hard-masked
version of the genome will be used for creating the alignment index and
further alignments. Please also see section @ref(redundancy) for
potential issues with redundant sequences contained in the reference
genome, e.g. in <em><a href="https://bioconductor.org/packages/3.19/BSgenome.Hsapiens.UCSC.hg19" class="external-link">BSgenome.Hsapiens.UCSC.hg19</a></em>
or <em><a href="https://bioconductor.org/packages/3.19/BSgenome.Hsapiens.UCSC.hg38" class="external-link">BSgenome.Hsapiens.UCSC.hg38</a></em>.</p>
<ul>
<li>
<strong>a file name</strong>, referring to a sequence file
containing one or several reference sequences (e.g. chromosomes) in
<code>fasta</code> format:</li>
</ul>
<div class="sourceCode" id="cb38"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">genomeFile</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span>package<span class="op">=</span><span class="st">"QuasR"</span>, <span class="st">"extdata"</span>, <span class="st">"hg19sub.fa"</span><span class="op">)</span></span></code></pre></div>
<div class="section level4">
<h4 id="redundancy">Choosing a suitable (non-redundant) reference genome<a class="anchor" aria-label="anchor" href="#redundancy"></a>
</h4>
<p>For some organisms, several versions of the genome assembly exist.
These differ for example in whether or not they include alternative
variants for sequences that are variable within the species. This may
lead to redundant sequences in the assembly, and thus reads mapping to
such sequences being wrongly classified as “multi-mapping”, and
comparing data aligned to different assembly version may give rise to
incorrect results. A nice summary of this issue is provided in this <a href="https://lh3.github.io/2017/11/13/which-human-reference-genome-to-use" class="external-link">blog
post</a> from Heng Li.</p>
<p>The <em><a href="https://bioconductor.org/packages/3.19/BSgenome" class="external-link">BSgenome</a></em>
packages <em><a href="https://bioconductor.org/packages/3.19/BSgenome.Hsapiens.UCSC.hg19" class="external-link">BSgenome.Hsapiens.UCSC.hg19</a></em>
(versions newer than 1.4.0 from Bioconductor 3.10) and <em><a href="https://bioconductor.org/packages/3.19/BSgenome.Hsapiens.UCSC.hg38" class="external-link">BSgenome.Hsapiens.UCSC.hg38</a></em>
(all versions) do contain such redundant sequences and are therefore not
ideal references for alignment of human data. Specific “analysis set” or
“primary assembly” versions of the assembly should be used instead (see
the before-mentioned <a href="https://lh3.github.io/2017/11/13/which-human-reference-genome-to-use" class="external-link">blog
post</a> for details).</p>
<p>When using a <em><a href="https://bioconductor.org/packages/3.19/BSgenome" class="external-link">BSgenome</a></em>
reference, <em><a href="https://bioconductor.org/packages/3.19/QuasR" class="external-link">QuasR</a></em> will
check in the <code>qAlign</code> function whether the chromosome names
and lengths contained in the header of any pre-existing bam files are
identical to the ones provided by the genome and warn if this is not the
case.</p>
</div>
</div>
<div class="section level3">
<h3 id="preProcessing">Sequence data pre-processing<a class="anchor" aria-label="anchor" href="#preProcessing"></a>
</h3>
<p>The <code>preprocessReads</code> function can be used to prepare the
input sequence files prior to alignment. The function takes one or
several sequence files (or pairs of files for a paired-end experiment)
in <code>fasta</code> or <code>fastq</code> format as input and produces
the same number of output files with the processed reads.</p>
<p>In the following example, we truncate the reads by removing the three
bases from the 3’-end (the right side), remove the adapter sequence
<code>AAAAAAAAAA</code> from the 5’-end (the left side) and filter out
reads that, after truncation and adapter removal, are shorter than 14
bases or contain more than 2 <code>N</code> bases:</p>
<div class="sourceCode" id="cb39"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">td</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/tempfile.html" class="external-link">tempdir</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="va">infiles</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span>package <span class="op">=</span> <span class="st">"QuasR"</span>, <span class="st">"extdata"</span>,</span>
<span>                       <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"rna_1_1.fq.bz2"</span>,<span class="st">"rna_2_1.fq.bz2"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">outfiles</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">td</span>, <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/path.html" class="external-link">basename</a></span><span class="op">(</span><span class="va">infiles</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">res</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/preprocessReads.html">preprocessReads</a></span><span class="op">(</span>filename <span class="op">=</span> <span class="va">infiles</span>,</span>
<span>                       outputFilename <span class="op">=</span> <span class="va">outfiles</span>,</span>
<span>                       truncateEndBases <span class="op">=</span> <span class="fl">3</span>,</span>
<span>                       Lpattern <span class="op">=</span> <span class="st">"AAAAAAAAAA"</span>,</span>
<span>                       minLength <span class="op">=</span> <span class="fl">14</span>, </span>
<span>                       nBases <span class="op">=</span> <span class="fl">2</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##   filtering /Users/runner/work/_temp/Library/QuasR/extdata/rna_1_1.fq.bz2</span></span></code></pre>
<pre><code><span><span class="co">##   filtering /Users/runner/work/_temp/Library/QuasR/extdata/rna_2_1.fq.bz2</span></span></code></pre>
<div class="sourceCode" id="cb42"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">res</span></span></code></pre></div>
<pre><code><span><span class="co">##                  rna_1_1.fq.bz2 rna_2_1.fq.bz2</span></span>
<span><span class="co">## totalSequences             3002           3000</span></span>
<span><span class="co">## matchTo5pAdapter            466            463</span></span>
<span><span class="co">## matchTo3pAdapter              0              0</span></span>
<span><span class="co">## tooShort                    107             91</span></span>
<span><span class="co">## tooManyN                      0              0</span></span>
<span><span class="co">## lowComplexity                 0              0</span></span>
<span><span class="co">## totalPassed                2895           2909</span></span></code></pre>
<div class="sourceCode" id="cb44"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/unlink.html" class="external-link">unlink</a></span><span class="op">(</span><span class="va">outfiles</span><span class="op">)</span></span></code></pre></div>
<p><code>preprocessReads</code> returns a matrix with a summary of the
pre-processing. The matrix contains one column per (pair of) input
sequence files, and contains the total number of reads
(<code>totalSequences</code>), the number of reads that matched to the
five prime or three prime adapters (<code>matchTo5pAdapter</code> and
<code>matchTo3pAdapter</code>), the number of reads that were too short
(<code>tooShort</code>), contained too many non-base characters
(<code>tooManyN</code>) or were of low sequence complexity
(<code>lowComplexity</code>, deactivated by default). Finally, the
number of reads that passed the filtering steps is reported in the last
row (<code>totalPassed</code>).</p>
<p>In the example below we process paired-end reads, removing all pairs
with one or several <code>N</code> bases. Even if only one sequence in a
pair fulfills the filtering criteria, both reads in the pair are
removed, thereby preserving the matching order of the sequences in the
two files:</p>
<div class="sourceCode" id="cb45"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">td</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/tempfile.html" class="external-link">tempdir</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="va">infiles1</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span>package <span class="op">=</span> <span class="st">"QuasR"</span>, <span class="st">"extdata"</span>, <span class="st">"rna_1_1.fq.bz2"</span><span class="op">)</span></span>
<span><span class="va">infiles2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span>package <span class="op">=</span> <span class="st">"QuasR"</span>, <span class="st">"extdata"</span>, <span class="st">"rna_1_2.fq.bz2"</span><span class="op">)</span></span>
<span><span class="va">outfiles1</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">td</span>, <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/path.html" class="external-link">basename</a></span><span class="op">(</span><span class="va">infiles1</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">outfiles2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">td</span>, <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/path.html" class="external-link">basename</a></span><span class="op">(</span><span class="va">infiles2</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">res</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/preprocessReads.html">preprocessReads</a></span><span class="op">(</span>filename <span class="op">=</span> <span class="va">infiles1</span>,</span>
<span>                       filenameMate <span class="op">=</span> <span class="va">infiles2</span>,</span>
<span>                       outputFilename <span class="op">=</span> <span class="va">outfiles1</span>,</span>
<span>                       outputFilenameMate <span class="op">=</span> <span class="va">outfiles2</span>,</span>
<span>                       nBases <span class="op">=</span> <span class="fl">0</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##   filtering /Users/runner/work/_temp/Library/QuasR/extdata/rna_1_1.fq.bz2 and</span></span>
<span><span class="co">##     /Users/runner/work/_temp/Library/QuasR/extdata/rna_1_2.fq.bz2</span></span></code></pre>
<div class="sourceCode" id="cb47"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">res</span></span></code></pre></div>
<pre><code><span><span class="co">##                  rna_1_1.fq.bz2:rna_1_2.fq.bz2</span></span>
<span><span class="co">## totalSequences                            3002</span></span>
<span><span class="co">## matchTo5pAdapter                            NA</span></span>
<span><span class="co">## matchTo3pAdapter                            NA</span></span>
<span><span class="co">## tooShort                                     0</span></span>
<span><span class="co">## tooManyN                                     3</span></span>
<span><span class="co">## lowComplexity                                0</span></span>
<span><span class="co">## totalPassed                               2999</span></span></code></pre>
<p>More details on the <code>preprocessReads</code> function can be
found in the function documentation (see <code><a href="../reference/preprocessReads.html">?preprocessReads</a></code>)
or in the section @ref(preprocessReads).</p>
</div>
</div>
<div class="section level2">
<h2 id="exampleWorkflows">Example workflows<a class="anchor" aria-label="anchor" href="#exampleWorkflows"></a>
</h2>
<div class="section level3">
<h3 id="ChIPseq">ChIP-seq: Measuring protein-DNA binding and chromatin
modifications<a class="anchor" aria-label="anchor" href="#ChIPseq"></a>
</h3>
<p>Here we show an exemplary single-end ChIP-seq workflow using a small
number of reads from a histone 3 lysine 4 trimethyl (H3K4me3) ChIP-seq
experiment. This histone modification is known to locate to genomic
regions with a high density of CpG dinucleotides (so called CpG
islands); about 60% of mammalian genes have such a CpG island close to
their transcript start site. All necessary files are included in the
<em><a href="https://bioconductor.org/packages/3.19/QuasR" class="external-link">QuasR</a></em>
package, and we start the example workflow by copying those files into
the current working directly, into a subfolder called
<code>"extdata"</code>:</p>
<div class="sourceCode" id="cb49"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/files.html" class="external-link">file.copy</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span>package <span class="op">=</span> <span class="st">"QuasR"</span>, <span class="st">"extdata"</span><span class="op">)</span>, <span class="st">"."</span>, recursive <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] TRUE</span></span></code></pre>
<div class="section level4">
<h4 id="align-reads-using-the-qalign-function">Align reads using the <code>qAlign</code> function<a class="anchor" aria-label="anchor" href="#align-reads-using-the-qalign-function"></a>
</h4>
<p>We assume that the sequence reads have already been pre-processed as
described in section @ref(preProcessing). Also, a sample file (section
@ref(sampleFile)) that lists all sequence files to be analyzed has been
prepared. A <code>fasta</code> file with the reference genome
sequence(s) is also available (section @ref(genome)), as well as an
auxiliary file for alignment of reads that failed to match the reference
genome (section @ref(auxiliaryFile)).</p>
<p>By default, newly generated <code>bam</code> files will be stored at
the location of the input sequence files, which should be writable and
have sufficient capacity (an alternative location can be specified using
the <code>alignmentsDir</code> argument). Make also sure that you have
sufficient temporary disk space for intermediate files in
<code><a href="https://rdrr.io/r/base/tempfile.html" class="external-link">tempdir()</a></code> (see section @ref(qAlign)). We start by aligning
the reads using <code>qAlign</code>:</p>
<div class="sourceCode" id="cb51"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sampleFile</span> <span class="op">&lt;-</span> <span class="st">"extdata/samples_chip_single.txt"</span></span>
<span><span class="va">auxFile</span> <span class="op">&lt;-</span> <span class="st">"extdata/auxiliaries.txt"</span></span>
<span><span class="va">genomeFile</span> <span class="op">&lt;-</span> <span class="st">"extdata/hg19sub.fa"</span></span>
<span></span>
<span><span class="va">proj1</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/qAlign.html">qAlign</a></span><span class="op">(</span><span class="va">sampleFile</span>, genome <span class="op">=</span> <span class="va">genomeFile</span>, auxiliaryFile <span class="op">=</span> <span class="va">auxFile</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## alignment files missing - need to:</span></span></code></pre>
<pre><code><span><span class="co">##     create 2 auxiliary alignment(s)</span></span></code></pre>
<pre><code><span><span class="co">## Creating an Rbowtie index for /Users/runner/work/QuasR/QuasR/vignettes/extdata/NC_001422.1.fa</span></span></code></pre>
<pre><code><span><span class="co">## Finished creating index</span></span></code></pre>
<pre><code><span><span class="co">## Testing the compute nodes...OK</span></span>
<span><span class="co">## Loading QuasR on the compute nodes...preparing to run on 1 nodes...done</span></span>
<span><span class="co">## Available cores:</span></span></code></pre>
<pre><code><span><span class="co">## nodeNames</span></span>
<span><span class="co">## Mac-1698842773251.local </span></span>
<span><span class="co">##                       1</span></span></code></pre>
<pre><code><span><span class="co">## Performing auxiliary alignments for 2 samples. See progress in the log file:</span></span>
<span><span class="co">## /Users/runner/work/QuasR/QuasR/vignettes/QuasR_log_10c532a58fe9c.txt</span></span>
<span><span class="co">## Auxiliary alignments have been created successfully</span></span></code></pre>
<div class="sourceCode" id="cb59"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">proj1</span></span></code></pre></div>
<pre><code><span><span class="co">## Project: qProject</span></span>
<span><span class="co">##  Options   : maxHits         : 1</span></span>
<span><span class="co">##              paired          : no</span></span>
<span><span class="co">##              splicedAlignment: FALSE</span></span>
<span><span class="co">##              bisulfite       : no</span></span>
<span><span class="co">##              snpFile         : none</span></span>
<span><span class="co">##              geneAnnotation  : none</span></span>
<span><span class="co">##  Aligner   : Rbowtie v1.43.0 (parameters: -m 1 --best --strata)</span></span>
<span><span class="co">##  Genome    : /Users/runner/work/QuasR/QuasR/vignettes/extdata/hg19sub.fa (file)</span></span>
<span><span class="co">## </span></span>
<span><span class="co">##  Reads     : 2 files, 2 samples (fastq format):</span></span>
<span><span class="co">##    1. chip_1_1.fq.bz2  Sample1 (phred33)</span></span>
<span><span class="co">##    2. chip_2_1.fq.bz2  Sample2 (phred33)</span></span>
<span><span class="co">## </span></span>
<span><span class="co">##  Genome alignments: directory: same as reads</span></span>
<span><span class="co">##    1. chip_1_1_10c5340447364.bam</span></span>
<span><span class="co">##    2. chip_2_1_10c534df3cb32.bam</span></span>
<span><span class="co">## </span></span>
<span><span class="co">##  Aux. alignments: 1 file, directory: same as reads</span></span>
<span><span class="co">##    a. /Users/runner/work/QuasR/QuasR/vignettes/extdata/NC_001422.1.fa    phiX174</span></span>
<span><span class="co">##      1. chip_1_1_10c5338b3c97c.bam</span></span>
<span><span class="co">##      2. chip_2_1_10c53237108f9.bam</span></span></code></pre>
<p><code>qAlign</code> will build alignment indices if they do not yet
exist (by default, if the genome and auxiliary sequences are given in
the form of <code>fasta</code> files, they will be stored in the same
folder). The <code>qProject</code> object (<code>proj1</code>) returned
by <code>qAlign</code> now contains all information about the ChIP-seq
experiment: the (optional) project name, the project options, aligner
package, reference genome, and at the bottom the sequence and alignment
files. For each input sequence file, there will be one <code>bam</code>
file with alignments against the reference genome, and one for each
auxiliary target sequence with alignments of reads without genome hits.
Our <code>auxFile</code> contains a single auxiliary target sequence, so
we expect two <code>bam</code> files per input sequence file:</p>
<div class="sourceCode" id="cb61"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/list.files.html" class="external-link">list.files</a></span><span class="op">(</span><span class="st">"extdata"</span>, pattern <span class="op">=</span> <span class="st">".bam$"</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "chip_1_1_10c5338b3c97c.bam"    "chip_1_1_10c5340447364.bam"   </span></span>
<span><span class="co">## [3] "chip_2_1_10c53237108f9.bam"    "chip_2_1_10c534df3cb32.bam"   </span></span>
<span><span class="co">## [5] "phiX_paired_withSecondary.bam"</span></span></code></pre>
<p>The <code>bam</code> file names consist of the base name of the
sequence file with an added random string. The random suffix makes sure
that newly generated alignment files do not overwrite existing ones, for
example of the same reads aligned against an alternative reference
genome. Each alignment file is accompanied by two additional files with
suffixes <code>.bai</code> and <code>.txt</code>:</p>
<div class="sourceCode" id="cb63"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/list.files.html" class="external-link">list.files</a></span><span class="op">(</span><span class="st">"extdata"</span>, pattern <span class="op">=</span> <span class="st">"^chip_1_1_"</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">3</span><span class="op">]</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "chip_1_1_10c5338b3c97c.bam"     "chip_1_1_10c5338b3c97c.bam.bai"</span></span>
<span><span class="co">## [3] "chip_1_1_10c5338b3c97c.bam.txt"</span></span></code></pre>
<p>The <code>.bai</code> file is the <code>bam</code> index used for
fast access by genomic coordinate. The <code>.txt</code> file contains
all the parameters used to generate the corresponding <code>bam</code>
file. Before new alignments are generated, <code>qAlign</code> will look
for available <code>.txt</code> files in default locations (the
directory containing the input sequence file, or the value of
<code>alignmentsDir</code>), and read their contents to determine if a
compatible <code>bam</code> file already exists. A compatible
<code>bam</code> file is one with the same reads and genome, aligned
using the same aligner and identical alignment parameters. If a
compatible <code>bam</code> file is not found, or the <code>.txt</code>
file is missing, <code>qAlign</code> will generate a new
<code>bam</code> file. It is therefore recommended not to delete the
<code>.txt</code> file - without it, the corresponding <code>bam</code>
file will become unusable for <em><a href="https://bioconductor.org/packages/3.19/QuasR" class="external-link">QuasR</a></em>.</p>
</div>
<div class="section level4">
<h4 id="create-a-quality-control-report">Create a quality control report<a class="anchor" aria-label="anchor" href="#create-a-quality-control-report"></a>
</h4>
<p><em><a href="https://bioconductor.org/packages/3.19/QuasR" class="external-link">QuasR</a></em> can
produce a quality control report in the form of a series of diagnostic
plots with details on sequences and alignments (see QuasR scheme figure
above). The plots are generated by calling the <code>qQCReport</code>
function with the <code>qProject</code> object as argument.
<code>qQCReport</code> uses <em><a href="https://bioconductor.org/packages/3.19/ShortRead" class="external-link">ShortRead</a></em>
<span class="citation">(Morgan et al. 2009)</span> internally to obtain
some of the quality metrics, and some of the plots are inspired by the
FastQC quality control tool by Simon Andrews (<a href="http://www.bioinformatics.bbsrc.ac.uk/projects/fastqc/" class="external-link uri">http://www.bioinformatics.bbsrc.ac.uk/projects/fastqc/</a>).
The plots will be stored into a multipage PDF document defined by the
<code>pdfFilename</code> argument, or else shown as individual plot
windows on the current graphics device. In order to keep the running
time reasonably short, some quality metrics are obtained from a random
sub-sample of the sequences or alignments.</p>
<pre><code><span><span class="co">## collecting quality control data</span></span></code></pre>
<pre><code><span><span class="co">## creating QC plots</span></span></code></pre>
<div class="sourceCode" id="cb67"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/qQCReport.html">qQCReport</a></span><span class="op">(</span><span class="va">proj1</span>, pdfFilename <span class="op">=</span> <span class="st">"extdata/qc_report.pdf"</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## collecting quality control data</span></span></code></pre>
<pre><code><span><span class="co">## creating QC plots</span></span></code></pre>
<p>Currently available plots are described in section @ref(qQCReport)
and following.</p>
</div>
<div class="section level4">
<h4 id="alignment-statistics">Alignment statistics<a class="anchor" aria-label="anchor" href="#alignment-statistics"></a>
</h4>
<p>The <code>alignmentStats</code> gets the number of (un-)mapped reads
for each sequence file in a <code>qProject</code> object, by reading the
<code>bam</code> file indices, and returns them as a
<code>data.frame</code>. The function also works for arguments of type
<code>character</code> with one or several <code>bam</code> file names
(for details see section @ref(alignmentStats)).</p>
<div class="sourceCode" id="cb70"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/alignmentStats.html">alignmentStats</a></span><span class="op">(</span><span class="va">proj1</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##                 seqlength mapped unmapped</span></span>
<span><span class="co">## Sample1:genome      95000   2339      258</span></span>
<span><span class="co">## Sample2:genome      95000   3609      505</span></span>
<span><span class="co">## Sample1:phiX174      5386    251        7</span></span>
<span><span class="co">## Sample2:phiX174      5386    493       12</span></span></code></pre>
</div>
<div class="section level4">
<h4 id="export-genome-wig-file-from-alignments">Export genome wig file from alignments<a class="anchor" aria-label="anchor" href="#export-genome-wig-file-from-alignments"></a>
</h4>
<p>For visualization in a genome browser, alignment coverage along the
genome can be exported to a (compressed) wig file using the
<code>qExportWig</code> function. The created fixedStep wig file (see
(<a href="http://genome.ucsc.edu/goldenPath/help/wiggle.html" class="external-link uri">http://genome.ucsc.edu/goldenPath/help/wiggle.html</a>) for
details on the wig format) will contain one track per sample in the
<code>qProject</code> object. The resolution is defined using the
<code>binsize</code> argument, and if <code>scaling</code> is set to
<code>TRUE</code>, read counts per bin are scaled by the total number of
aligned reads in each sample to improve comparability:</p>
<div class="sourceCode" id="cb72"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/qExportWig.html">qExportWig</a></span><span class="op">(</span><span class="va">proj1</span>, binsize <span class="op">=</span> <span class="fl">100L</span>, scaling <span class="op">=</span> <span class="cn">TRUE</span>, collapseBySample <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## collecting mapping statistics for scaling...done</span></span>
<span><span class="co">## start creating wig files...</span></span>
<span><span class="co">##   Sample1.wig.gz (Sample1)</span></span>
<span><span class="co">##   Sample2.wig.gz (Sample2)</span></span>
<span><span class="co">## done</span></span></code></pre>
</div>
<div class="section level4">
<h4 id="count-alignments-using-qcount">Count alignments using <code>qCount</code><a class="anchor" aria-label="anchor" href="#count-alignments-using-qcount"></a>
</h4>
<p>Alignments are quantified using <code>qCount</code>, for example
using a <code>GRanges</code> object as a query. In our H3K4me3 ChIP-seq
example, we expect the reads to occur around the transcript start site
of genes. We can therefore construct suitable query regions using
genomic intervals around the start sites of known genes. In the code
below, this is achieved with help from the <em><a href="https://bioconductor.org/packages/3.19/GenomicFeatures" class="external-link">GenomicFeatures</a></em>
package: We first create a <code>TxDb</code> object from a
<code>.gtf</code> file with gene annotation. With the
<code>promoters</code> function, we can then create the
<code>GRanges</code> object with regions to be quantified. Finally,
because most genes consist of multiple overlapping transcripts, we
select the first transcript for each gene:</p>
<div class="sourceCode" id="cb74"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://bioconductor.org/packages/GenomicFeatures" class="external-link">GenomicFeatures</a></span><span class="op">)</span></span>
<span><span class="va">annotFile</span> <span class="op">&lt;-</span> <span class="st">"extdata/hg19sub_annotation.gtf"</span></span>
<span><span class="va">chrLen</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/Rsamtools/man/scanFa.html" class="external-link">scanFaIndex</a></span><span class="op">(</span><span class="va">genomeFile</span><span class="op">)</span></span>
<span><span class="va">chrominfo</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span>chrom <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/character.html" class="external-link">as.character</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/GenomeInfoDb/man/seqinfo.html" class="external-link">seqnames</a></span><span class="op">(</span><span class="va">chrLen</span><span class="op">)</span><span class="op">)</span>,</span>
<span>                        length <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/start.html" class="external-link">width</a></span><span class="op">(</span><span class="va">chrLen</span><span class="op">)</span>,</span>
<span>                        is_circular <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="cn">FALSE</span>, <span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">chrLen</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">txdb</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/GenomicFeatures/man/makeTxDbFromGFF.html" class="external-link">makeTxDbFromGFF</a></span><span class="op">(</span>file <span class="op">=</span> <span class="va">annotFile</span>, format <span class="op">=</span> <span class="st">"gtf"</span>,</span>
<span>                        chrominfo <span class="op">=</span> <span class="va">chrominfo</span>,</span>
<span>                        dataSource <span class="op">=</span> <span class="st">"Ensembl"</span>,</span>
<span>                        organism <span class="op">=</span> <span class="st">"Homo sapiens"</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Import genomic features from the file as a GRanges object ... OK</span></span>
<span><span class="co">## Prepare the 'metadata' data frame ... OK</span></span>
<span><span class="co">## Make the TxDb object ... OK</span></span></code></pre>
<div class="sourceCode" id="cb76"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">promReg</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/IRanges/man/intra-range-methods.html" class="external-link">promoters</a></span><span class="op">(</span><span class="va">txdb</span>, upstream <span class="op">=</span> <span class="fl">1000</span>, downstream <span class="op">=</span> <span class="fl">500</span>,</span>
<span>                     columns <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"gene_id"</span>,<span class="st">"tx_id"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">gnId</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">vapply</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Vector-class.html" class="external-link">mcols</a></span><span class="op">(</span><span class="va">promReg</span><span class="op">)</span><span class="op">$</span><span class="va">gene_id</span>,</span>
<span>               FUN <span class="op">=</span> <span class="va">paste</span>, FUN.VALUE <span class="op">=</span> <span class="st">""</span>,</span>
<span>               collapse <span class="op">=</span> <span class="st">","</span><span class="op">)</span></span>
<span><span class="va">promRegSel</span> <span class="op">&lt;-</span> <span class="va">promReg</span><span class="op">[</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/match.html" class="external-link">match</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="va">gnId</span><span class="op">)</span>, <span class="va">gnId</span><span class="op">)</span> <span class="op">]</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">promRegSel</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="va">gnId</span><span class="op">)</span></span>
<span><span class="va">promRegSel</span></span></code></pre></div>
<pre><code><span><span class="co">## GRanges object with 12 ranges and 2 metadata columns:</span></span>
<span><span class="co">##                   seqnames      ranges strand |         gene_id     tx_id</span></span>
<span><span class="co">##                      &lt;Rle&gt;   &lt;IRanges&gt;  &lt;Rle&gt; | &lt;CharacterList&gt; &lt;integer&gt;</span></span>
<span><span class="co">##   ENSG00000176022     chr1 31629-33128      + | ENSG00000176022         1</span></span>
<span><span class="co">##   ENSG00000186891     chr1   6452-7951      - | ENSG00000186891         2</span></span>
<span><span class="co">##   ENSG00000186827     chr1 14013-15512      - | ENSG00000186827         6</span></span>
<span><span class="co">##   ENSG00000078808     chr1 31882-33381      - | ENSG00000078808         9</span></span>
<span><span class="co">##   ENSG00000171863     chr2   1795-3294      + | ENSG00000171863        17</span></span>
<span><span class="co">##               ...      ...         ...    ... .             ...       ...</span></span>
<span><span class="co">##   ENSG00000254999     chr3   1276-2775      + | ENSG00000254999        28</span></span>
<span><span class="co">##   ENSG00000238642     chr3 19069-20568      + | ENSG00000238642        30</span></span>
<span><span class="co">##   ENSG00000134086     chr3 26692-28191      + | ENSG00000134086        31</span></span>
<span><span class="co">##   ENSG00000238345     chr3 26834-28333      + | ENSG00000238345        32</span></span>
<span><span class="co">##   ENSG00000134075     chr3 13102-14601      - | ENSG00000134075        36</span></span>
<span><span class="co">##   -------</span></span>
<span><span class="co">##   seqinfo: 3 sequences from an unspecified genome</span></span></code></pre>
<p>Using <code>promRegSel</code> object as query, we can now count the
alignment per sample in each of the promoter windows.</p>
<div class="sourceCode" id="cb78"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">cnt</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/qCount.html">qCount</a></span><span class="op">(</span><span class="va">proj1</span>, <span class="va">promRegSel</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## counting alignments...done</span></span></code></pre>
<div class="sourceCode" id="cb80"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">cnt</span></span></code></pre></div>
<pre><code><span><span class="co">##                 width Sample1 Sample2</span></span>
<span><span class="co">## ENSG00000176022  1500     157     701</span></span>
<span><span class="co">## ENSG00000186891  1500      22       5</span></span>
<span><span class="co">## ENSG00000186827  1500      10       3</span></span>
<span><span class="co">## ENSG00000078808  1500      73     558</span></span>
<span><span class="co">## ENSG00000171863  1500      94     339</span></span>
<span><span class="co">## ENSG00000252531  1500      59       9</span></span>
<span><span class="co">## ENSG00000247886  1500     172     971</span></span>
<span><span class="co">## ENSG00000254999  1500     137     389</span></span>
<span><span class="co">## ENSG00000238642  1500       8       3</span></span>
<span><span class="co">## ENSG00000134086  1500       9      18</span></span>
<span><span class="co">## ENSG00000238345  1500      13      25</span></span>
<span><span class="co">## ENSG00000134075  1500       7       3</span></span></code></pre>
<p>The counts returned by <code>qCount</code> are the raw number of
alignments per sample and region, without any normalization for the
query region length, or the total number of aligned reads in a sample.
As expected, we can find H3K4me3 signal at promoters of a subset of the
genes with CpG island promoters, which we can visualize with help of the
<em><a href="https://bioconductor.org/packages/3.19/Gviz" class="external-link">Gviz</a></em>
package:</p>
<div class="sourceCode" id="cb82"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">gr1</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocIO/man/IO.html" class="external-link">import</a></span><span class="op">(</span><span class="st">"Sample1.wig.gz"</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Warning in asMethod(object): NAs introduced by coercion</span></span></code></pre>
<div class="sourceCode" id="cb84"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">gr2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocIO/man/IO.html" class="external-link">import</a></span><span class="op">(</span><span class="st">"Sample2.wig.gz"</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Warning in asMethod(object): NAs introduced by coercion</span></span></code></pre>
<div class="sourceCode" id="cb86"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/ivanek/Gviz" class="external-link">Gviz</a></span><span class="op">)</span></span>
<span><span class="va">axisTrack</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/Gviz/man/GenomeAxisTrack-class.html" class="external-link">GenomeAxisTrack</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="va">dTrack1</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/Gviz/man/DataTrack-class.html" class="external-link">DataTrack</a></span><span class="op">(</span>range <span class="op">=</span> <span class="va">gr1</span>, name <span class="op">=</span> <span class="st">"Sample 1"</span>, type <span class="op">=</span> <span class="st">"h"</span><span class="op">)</span></span>
<span><span class="va">dTrack2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/Gviz/man/DataTrack-class.html" class="external-link">DataTrack</a></span><span class="op">(</span>range <span class="op">=</span> <span class="va">gr2</span>, name <span class="op">=</span> <span class="st">"Sample 2"</span>, type <span class="op">=</span> <span class="st">"h"</span><span class="op">)</span></span>
<span><span class="va">txTrack</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/Gviz/man/GeneRegionTrack-class.html" class="external-link">GeneRegionTrack</a></span><span class="op">(</span><span class="va">txdb</span>, name <span class="op">=</span> <span class="st">"Transcripts"</span>, showId <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/Gviz/man/plotTracks.html" class="external-link">plotTracks</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="va">axisTrack</span>, <span class="va">dTrack1</span>, <span class="va">dTrack2</span>, <span class="va">txTrack</span><span class="op">)</span>,</span>
<span>           chromosome <span class="op">=</span> <span class="st">"chr3"</span>, extend.left <span class="op">=</span> <span class="fl">1000</span><span class="op">)</span></span></code></pre></div>
<p><img src="QuasR_files/figure-html/ChIP_visualize-1.png" width="768"></p>
</div>
<div class="section level4">
<h4 id="create-a-genomic-profile-for-a-set-of-regions-using-qprofile">Create a genomic profile for a set of regions using
<code>qProfile</code><a class="anchor" aria-label="anchor" href="#create-a-genomic-profile-for-a-set-of-regions-using-qprofile"></a>
</h4>
<p>Given a set of anchor positions in the genome, <code>qProfile</code>
calculates the number of nearby alignments relative to the anchor
position, for example to generate a average profile. The neighborhood
around anchor positions can be specified by the <code>upstream</code>
and <code>downstream</code> argument. Alignments that are upstream of an
anchor position will have a negative relative position, and downstream
alignments a positive. The anchor positions are all aligned at position
zero in the return value.</p>
<p>Anchor positions will be provided to <code>qProfile</code> using the
<code>query</code> argument, which takes a <code>GRanges</code> object.
The anchor positions correspond to <code><a href="https://rdrr.io/r/stats/start.html" class="external-link">start()</a></code> for regions on
<code>+</code> or <code>*</code> strands, and to <code><a href="https://rdrr.io/r/stats/start.html" class="external-link">end()</a></code> for
regions on the <code>-</code> strand. As mentioned above, we expect
H3K4me3 ChIP-seq alignments to be enriched around the transcript start
site of genes. We can therefore construct a suitable <code>query</code>
object from the start sites of known genes. In the code below, start
sites (<code>start_codon</code>) are imported from a <code>.gtf</code>
file with the help of the <em><a href="https://bioconductor.org/packages/3.19/rtracklayer" class="external-link">rtracklayer</a></em>
package. In addition, <code>strand</code> and <code>gene_name</code> are
also selected for import. Duplicated start sites, e.g. from genes with
multiple transcripts, are removed. Finally, all regions are given the
name <code>TSS</code>, because <code>qProfile</code> combines regions
with identical names into a single profile.</p>
<div class="sourceCode" id="cb87"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">rtracklayer</span><span class="op">)</span></span>
<span><span class="va">annotationFile</span> <span class="op">&lt;-</span> <span class="st">"extdata/hg19sub_annotation.gtf"</span></span>
<span><span class="va">tssRegions</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/rtracklayer/man/GFFFile-class.html" class="external-link">import.gff</a></span><span class="op">(</span><span class="va">annotationFile</span>, format <span class="op">=</span> <span class="st">"gtf"</span>,</span>
<span>                         feature.type <span class="op">=</span> <span class="st">"start_codon"</span>,</span>
<span>                         colnames <span class="op">=</span> <span class="st">"gene_name"</span><span class="op">)</span></span>
<span><span class="va">tssRegions</span> <span class="op">&lt;-</span> <span class="va">tssRegions</span><span class="op">[</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/duplicated.html" class="external-link">duplicated</a></span><span class="op">(</span><span class="va">tssRegions</span><span class="op">)</span><span class="op">]</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">tssRegions</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="st">"TSS"</span>, <span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">tssRegions</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">tssRegions</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## GRanges object with 6 ranges and 1 metadata column:</span></span>
<span><span class="co">##       seqnames      ranges strand |   gene_name</span></span>
<span><span class="co">##          &lt;Rle&gt;   &lt;IRanges&gt;  &lt;Rle&gt; | &lt;character&gt;</span></span>
<span><span class="co">##   TSS     chr1   6949-6951      - |    TNFRSF18</span></span>
<span><span class="co">##   TSS     chr1 14505-14507      - |     TNFRSF4</span></span>
<span><span class="co">##   TSS     chr1 29171-29173      - |        SDF4</span></span>
<span><span class="co">##   TSS     chr1 32659-32661      + |     B3GALT6</span></span>
<span><span class="co">##   TSS     chr2   3200-3202      + |        RPS7</span></span>
<span><span class="co">##   TSS     chr3   2386-2388      + |     C3orf10</span></span>
<span><span class="co">##   -------</span></span>
<span><span class="co">##   seqinfo: 3 sequences from an unspecified genome; no seqlengths</span></span></code></pre>
<p>Alignments around the <code>tssRegions</code> coordinates are counted
in a window defined by the <code>upstream</code> and
<code>downstream</code> arguments, which specify the number of bases to
include around each anchor position. For <code>query</code> regions on
<code>+</code> or <code>*</code> strands, upstream refers to the left
side of the anchor position (lower coordinates), while for regions on
the <code>-</code> strand, upstream refers to the right side (higher
coordinates). The following example creates separate profiles for
alignments on the <em>same</em> and on the <em>opposite</em> strand of
the regions in <code>query</code>.</p>
<div class="sourceCode" id="cb89"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">prS</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/qProfile.html">qProfile</a></span><span class="op">(</span><span class="va">proj1</span>, <span class="va">tssRegions</span>, upstream <span class="op">=</span> <span class="fl">3000</span>, downstream <span class="op">=</span> <span class="fl">3000</span>, </span>
<span>                orientation <span class="op">=</span> <span class="st">"same"</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## profiling alignments...done</span></span></code></pre>
<div class="sourceCode" id="cb91"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">prO</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/qProfile.html">qProfile</a></span><span class="op">(</span><span class="va">proj1</span>, <span class="va">tssRegions</span>, upstream <span class="op">=</span> <span class="fl">3000</span>, downstream <span class="op">=</span> <span class="fl">3000</span>, </span>
<span>                orientation <span class="op">=</span> <span class="st">"opposite"</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## profiling alignments...done</span></span></code></pre>
<div class="sourceCode" id="cb93"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="va">prS</span>, <span class="st">"["</span>, , <span class="fl">1</span><span class="op">:</span><span class="fl">10</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## $coverage</span></span>
<span><span class="co">## -3000 -2999 -2998 -2997 -2996 -2995 -2994 -2993 -2992 -2991 </span></span>
<span><span class="co">##     8     8     8     8     8     8     8     8     8     8 </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## $Sample1</span></span>
<span><span class="co">## -3000 -2999 -2998 -2997 -2996 -2995 -2994 -2993 -2992 -2991 </span></span>
<span><span class="co">##     1     0     0     0     0     0     0     0     0     0 </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## $Sample2</span></span>
<span><span class="co">## -3000 -2999 -2998 -2997 -2996 -2995 -2994 -2993 -2992 -2991 </span></span>
<span><span class="co">##     0     0     0     2     0     0     1     1     1     0</span></span></code></pre>
<p>The counts returned by <code>qProfile</code> are the raw number of
alignments per sample and position, without any normalization for the
number of query regions or the total number of alignments in a sample
per position. To obtain the average number of alignments, we divide the
alignment counts by the number of <code>query</code> regions that
covered a given relative position around the anchor sites. This coverage
is available as the first element in the return value. The shift between
<em>same</em> and <em>opposite</em> strand alignments is indicative for
the average length of the sequenced ChIP fragments.</p>
<div class="sourceCode" id="cb95"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">prCombS</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/do.call.html" class="external-link">do.call</a></span><span class="op">(</span><span class="st">"+"</span>, <span class="va">prS</span><span class="op">[</span><span class="op">-</span><span class="fl">1</span><span class="op">]</span><span class="op">)</span> <span class="op">/</span> <span class="va">prS</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span></span>
<span><span class="va">prCombO</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/do.call.html" class="external-link">do.call</a></span><span class="op">(</span><span class="st">"+"</span>, <span class="va">prO</span><span class="op">[</span><span class="op">-</span><span class="fl">1</span><span class="op">]</span><span class="op">)</span> <span class="op">/</span> <span class="va">prO</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/numeric.html" class="external-link">as.numeric</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/row_colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">prCombS</span><span class="op">)</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/stats/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">prCombS</span><span class="op">[</span><span class="fl">1</span>,<span class="op">]</span>, <span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="fl">1</span><span class="op">/</span><span class="fl">100</span>,<span class="fl">100</span><span class="op">)</span><span class="op">)</span>, </span>
<span>     type <span class="op">=</span> <span class="st">'l'</span>, xlab <span class="op">=</span> <span class="st">"Position relative to TSS"</span>,</span>
<span>     ylab <span class="op">=</span> <span class="st">"Mean no. of alignments"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/lines.html" class="external-link">lines</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/numeric.html" class="external-link">as.numeric</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/row_colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">prCombO</span><span class="op">)</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/stats/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">prCombO</span><span class="op">[</span><span class="fl">1</span>,<span class="op">]</span>, <span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="fl">1</span><span class="op">/</span><span class="fl">100</span>,<span class="fl">100</span><span class="op">)</span><span class="op">)</span>, </span>
<span>      type <span class="op">=</span> <span class="st">'l'</span>, col <span class="op">=</span> <span class="st">"red"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/legend.html" class="external-link">legend</a></span><span class="op">(</span>title <span class="op">=</span> <span class="st">"strand"</span>, legend <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"same as query"</span>,<span class="st">"opposite of query"</span><span class="op">)</span>, </span>
<span>       x <span class="op">=</span> <span class="st">"topleft"</span>, col <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"black"</span>,<span class="st">"red"</span><span class="op">)</span>,</span>
<span>       lwd <span class="op">=</span> <span class="fl">1.5</span>, bty <span class="op">=</span> <span class="st">"n"</span>, title.adj <span class="op">=</span> <span class="fl">0.1</span><span class="op">)</span></span></code></pre></div>
<p><img src="QuasR_files/figure-html/ChIP_visualizeProfile-1.png" width="768"></p>
</div>
<div class="section level4">
<h4 id="using-a-bsgenome-package-as-reference-genome">Using a <em><a href="https://bioconductor.org/packages/3.19/BSgenome" class="external-link">BSgenome</a></em>
package as reference genome<a class="anchor" aria-label="anchor" href="#using-a-bsgenome-package-as-reference-genome"></a>
</h4>
<p><em><a href="https://bioconductor.org/packages/3.19/QuasR" class="external-link">QuasR</a></em> also
allows using of <em><a href="https://bioconductor.org/packages/3.19/BSgenome" class="external-link">BSgenome</a></em>
packages instead of a <code>fasta</code> file as reference genome (see
section @ref(genome)). To use a <em><a href="https://bioconductor.org/packages/3.19/BSgenome" class="external-link">BSgenome</a></em>,
the <code>genome</code> argument of <code>qAlign</code> is set to a
string matching the name of a <em><a href="https://bioconductor.org/packages/3.19/BSgenome" class="external-link">BSgenome</a></em>
package, for example <code>"BSgenome.Hsapiens.UCSC.hg19"</code>. If that
package is not already installed, <code>qAlign</code> will abort with an
informative message describing how to install the package using
<code><a href="https://bioconductor.github.io/BiocManager/reference/install.html" class="external-link">BiocManager::install</a></code>. The corresponding alignment index
will be saved as a new package, named after the original <em><a href="https://bioconductor.org/packages/3.19/BSgenome" class="external-link">BSgenome</a></em>
package and the aligner used to build the index, for example
<code>BSgenome.Hsapiens.UCSC.hg19.Rbowtie</code>.</p>
<p>The code example below illustrates the use of a <em><a href="https://bioconductor.org/packages/3.19/BSgenome" class="external-link">BSgenome</a></em>
reference genome for the same example data as above. Running it for the
first time will take a few hours in order to build the aligner index,
but subsequent uses of the same reference genome will reuse the existing
index and immediately start alignments:</p>
<div class="sourceCode" id="cb96"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/files.html" class="external-link">file.copy</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span>package<span class="op">=</span><span class="st">"QuasR"</span>, <span class="st">"extdata"</span><span class="op">)</span>, <span class="st">"."</span>, recursive<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span></span>
<span></span>
<span><span class="va">sampleFile</span> <span class="op">&lt;-</span> <span class="st">"extdata/samples_chip_single.txt"</span></span>
<span><span class="va">auxFile</span> <span class="op">&lt;-</span> <span class="st">"extdata/auxiliaries.txt"</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/BSgenome/man/available.genomes.html" class="external-link">available.genomes</a></span><span class="op">(</span><span class="op">)</span> <span class="co"># list available genomes</span></span>
<span><span class="va">genomeName</span> <span class="op">&lt;-</span> <span class="st">"BSgenome.Hsapiens.UCSC.hg19"</span></span>
<span></span>
<span><span class="va">proj1</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/qAlign.html">qAlign</a></span><span class="op">(</span><span class="va">sampleFile</span>, genome<span class="op">=</span><span class="va">genomeName</span>, auxiliaryFile<span class="op">=</span><span class="va">auxFile</span><span class="op">)</span></span>
<span><span class="va">proj1</span></span></code></pre></div>
</div>
</div>
<div class="section level3">
<h3 id="RNAseq">RNA-seq: Gene expression profiling<a class="anchor" aria-label="anchor" href="#RNAseq"></a>
</h3>
<p>In <em><a href="https://bioconductor.org/packages/3.19/QuasR" class="external-link">QuasR</a></em>, an
analysis workflow for an RNA-seq dataset is very similar to the one
described above for a ChIP-seq experiment. The major difference is that
here reads are aligned using
<code>qAlign(..., splicedAlignment=TRUE, aligner="Rhisat2")</code>,
which will cause <code>qAlign</code> to align reads with the
<code>HISAT2</code> aligner <span class="citation">(Kim, Langmead, and
Salzberg 2015)</span> (via the <em><a href="https://bioconductor.org/packages/3.19/Rhisat2" class="external-link">Rhisat2</a></em>
package), rather than with <code>bowtie</code> <span class="citation">(Langmead et al. 2009)</span>. Before the <em><a href="https://bioconductor.org/packages/3.19/Rhisat2" class="external-link">Rhisat2</a></em>
package was available (introduced in Bioconductor 3.9),
<code>qAlign(... splicedAlignment=TRUE)</code> aligned reads using
<code>SpliceMap</code> <span class="citation">(Au et al. 2010)</span>,
which is not recommended now but still possible in order to reproduce
old results. Spliced paired-end alignments are also supported; the
<code>splicedAlignment</code> argument can be freely combined with the
<code>paired</code> argument. In addition, HISAT2 also allows the
specification of known splice sites, which can help in the read
alignment. This is done by specifying the argument
<code>geneAnnotation</code> in <code><a href="../reference/qAlign.html">qAlign()</a></code>, to either a
<code>.gtf</code> file or a <code>sqlite</code> database generated by
exporting a <code>TxDb</code> object.</p>
<div class="section level4">
<h4 id="spliced-alignment-of-rna-seq-reads">Spliced alignment of RNA-seq reads<a class="anchor" aria-label="anchor" href="#spliced-alignment-of-rna-seq-reads"></a>
</h4>
<p>We start the example workflow by copying the example data files into
the current working directly, into a subfolder called
<code>"extdata"</code>, and then create spliced alignments using
<code>qAlign</code>:</p>
<div class="sourceCode" id="cb97"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/files.html" class="external-link">file.copy</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span>package <span class="op">=</span> <span class="st">"QuasR"</span>, <span class="st">"extdata"</span><span class="op">)</span>, <span class="st">"."</span>, recursive <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] TRUE</span></span></code></pre>
<div class="sourceCode" id="cb99"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sampleFile</span> <span class="op">&lt;-</span> <span class="st">"extdata/samples_rna_paired.txt"</span></span>
<span><span class="va">genomeFile</span> <span class="op">&lt;-</span> <span class="st">"extdata/hg19sub.fa"</span></span>
<span></span>
<span><span class="va">proj2</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/qAlign.html">qAlign</a></span><span class="op">(</span><span class="va">sampleFile</span>, genome <span class="op">=</span> <span class="va">genomeFile</span>,</span>
<span>                splicedAlignment <span class="op">=</span> <span class="cn">TRUE</span>, aligner <span class="op">=</span> <span class="st">"Rhisat2"</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## alignment files missing - need to:</span></span></code></pre>
<pre><code><span><span class="co">##     create alignment index for the genome</span></span></code></pre>
<pre><code><span><span class="co">##     create 2 genomic alignment(s)</span></span></code></pre>
<pre><code><span><span class="co">## Creating an Rhisat2 index for /Users/runner/work/QuasR/QuasR/vignettes/extdata/hg19sub.fa</span></span></code></pre>
<pre><code><span><span class="co">## Finished creating index</span></span></code></pre>
<pre><code><span><span class="co">## Testing the compute nodes...OK</span></span>
<span><span class="co">## Loading QuasR on the compute nodes...preparing to run on 1 nodes...done</span></span>
<span><span class="co">## Available cores:</span></span>
<span><span class="co">## Mac-1698842773251.local: 1</span></span>
<span><span class="co">## Performing genomic alignments for 2 samples. See progress in the log file:</span></span>
<span><span class="co">## /Users/runner/work/QuasR/QuasR/vignettes/QuasR_log_10c531d326c7c.txt</span></span>
<span><span class="co">## Genomic alignments have been created successfully</span></span></code></pre>
<div class="sourceCode" id="cb106"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">proj2</span></span></code></pre></div>
<pre><code><span><span class="co">## Project: qProject</span></span>
<span><span class="co">##  Options   : maxHits         : 1</span></span>
<span><span class="co">##              paired          : fr</span></span>
<span><span class="co">##              splicedAlignment: TRUE</span></span>
<span><span class="co">##              bisulfite       : no</span></span>
<span><span class="co">##              snpFile         : none</span></span>
<span><span class="co">##              geneAnnotation  : none</span></span>
<span><span class="co">##  Aligner   : Rhisat2 v1.19.0 (parameters: -k 2)</span></span>
<span><span class="co">##  Genome    : /Users/runner/work/QuasR/QuasR/vignettes/extdata/hg19sub.fa (file)</span></span>
<span><span class="co">## </span></span>
<span><span class="co">##  Reads     : 2 pairs of files, 2 samples (fastq format):</span></span>
<span><span class="co">##    1. rna_1_1.fq.bz2  rna_1_2.fq.bz2  Sample1 (phred33)</span></span>
<span><span class="co">##    2. rna_2_1.fq.bz2  rna_2_2.fq.bz2  Sample2 (phred33)</span></span>
<span><span class="co">## </span></span>
<span><span class="co">##  Genome alignments: directory: same as reads</span></span>
<span><span class="co">##    1. rna_1_1_10c5359704fdd.bam</span></span>
<span><span class="co">##    2. rna_2_1_10c535c8b640a.bam</span></span>
<span><span class="co">## </span></span>
<span><span class="co">##  Aux. alignments: none</span></span></code></pre>
<p>Aligning the reads with <code>splicedAlignment=TRUE</code> will allow
to also align reads that cross exon junctions, and thus have a large
deletion (the intron) relative to the reference genome.</p>
<div class="sourceCode" id="cb108"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">proj2unspl</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/qAlign.html">qAlign</a></span><span class="op">(</span><span class="va">sampleFile</span>, genome <span class="op">=</span> <span class="va">genomeFile</span>,</span>
<span>                     splicedAlignment <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## alignment files missing - need to:</span></span></code></pre>
<pre><code><span><span class="co">##     create 2 genomic alignment(s)</span></span></code></pre>
<pre><code><span><span class="co">## Testing the compute nodes...OK</span></span>
<span><span class="co">## Loading QuasR on the compute nodes...preparing to run on 1 nodes...done</span></span>
<span><span class="co">## Available cores:</span></span>
<span><span class="co">## Mac-1698842773251.local: 1</span></span>
<span><span class="co">## Performing genomic alignments for 2 samples. See progress in the log file:</span></span>
<span><span class="co">## /Users/runner/work/QuasR/QuasR/vignettes/QuasR_log_10c534961e69.txt</span></span>
<span><span class="co">## Genomic alignments have been created successfully</span></span></code></pre>
<div class="sourceCode" id="cb112"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/alignmentStats.html">alignmentStats</a></span><span class="op">(</span><span class="va">proj2</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##                seqlength mapped unmapped</span></span>
<span><span class="co">## Sample1:genome     95000   5998        6</span></span>
<span><span class="co">## Sample2:genome     95000   5998        2</span></span></code></pre>
<div class="sourceCode" id="cb114"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/alignmentStats.html">alignmentStats</a></span><span class="op">(</span><span class="va">proj2unspl</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##                seqlength mapped unmapped</span></span>
<span><span class="co">## Sample1:genome     95000   2258     3746</span></span>
<span><span class="co">## Sample2:genome     95000   2652     3348</span></span></code></pre>
</div>
<div class="section level4">
<h4 id="quantification-of-gene-and-exon-expression">Quantification of gene and exon expression<a class="anchor" aria-label="anchor" href="#quantification-of-gene-and-exon-expression"></a>
</h4>
<p>As with ChIP-seq experiments, <code>qCount</code> is used to quantify
alignments. For quantification of gene or exon expression levels,
<code>qCount</code> can be called with a query of type
<code>TxDb</code>, such as the one we constructed in the ChIP-seq
workflow above from a <code>.gtf</code> file. The argument
<code>reportLevel</code> can be used to control if annotated exonic
regions should be quantified independently
(<code>reportLevel="exon"</code>) or non-redundantly combined per gene
(<code>reportLevel="gene"</code>):</p>
<div class="sourceCode" id="cb116"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">geneLevels</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/qCount.html">qCount</a></span><span class="op">(</span><span class="va">proj2</span>, <span class="va">txdb</span>, reportLevel <span class="op">=</span> <span class="st">"gene"</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## extracting gene regions from TxDb...done</span></span>
<span><span class="co">## counting alignments...done</span></span>
<span><span class="co">## collapsing counts by query name...done</span></span></code></pre>
<div class="sourceCode" id="cb118"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">exonLevels</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/qCount.html">qCount</a></span><span class="op">(</span><span class="va">proj2</span>, <span class="va">txdb</span>, reportLevel <span class="op">=</span> <span class="st">"exon"</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## extracting exon regions from TxDb...done</span></span>
<span><span class="co">## counting alignments...done</span></span></code></pre>
<div class="sourceCode" id="cb120"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">geneLevels</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##                 width Sample1 Sample2</span></span>
<span><span class="co">## ENSG00000078808  4697     710    1083</span></span>
<span><span class="co">## ENSG00000134075   589    1173    1303</span></span>
<span><span class="co">## ENSG00000134086  4213     279     295</span></span>
<span><span class="co">## ENSG00000171863  5583    2924    2224</span></span>
<span><span class="co">## ENSG00000176022  2793      62     344</span></span>
<span><span class="co">## ENSG00000186827  1721      37       8</span></span></code></pre>
<div class="sourceCode" id="cb122"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">exonLevels</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##    width Sample1 Sample2</span></span>
<span><span class="co">## 1   2793      62     344</span></span>
<span><span class="co">## 10   187       3       0</span></span>
<span><span class="co">## 11   307       3       0</span></span>
<span><span class="co">## 12   300      11       2</span></span>
<span><span class="co">## 13   493      19       2</span></span>
<span><span class="co">## 14   129       7       0</span></span></code></pre>
</div>
<div class="section level4">
<h4 id="calculation-of-rpkm-expression-values">Calculation of RPKM expression values<a class="anchor" aria-label="anchor" href="#calculation-of-rpkm-expression-values"></a>
</h4>
<p>The values returned by <code>qCount</code> are the number of
alignments. Sometimes it is required to normalize for the length of
query regions, or the size of the libraries. For example, gene
expression levels in the form of <em>RPKM</em> values (reads per
kilobase of transcript and million mapped reads) can be obtained as
follows:</p>
<div class="sourceCode" id="cb124"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">geneRPKM</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/t.html" class="external-link">t</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/t.html" class="external-link">t</a></span><span class="op">(</span><span class="va">geneLevels</span><span class="op">[</span>,<span class="op">-</span><span class="fl">1</span><span class="op">]</span> <span class="op">/</span> <span class="va">geneLevels</span><span class="op">[</span>,<span class="fl">1</span><span class="op">]</span> <span class="op">*</span> <span class="fl">1000</span><span class="op">)</span></span>
<span>              <span class="op">/</span> <span class="fu"><a href="https://rdrr.io/r/base/colSums.html" class="external-link">colSums</a></span><span class="op">(</span><span class="va">geneLevels</span><span class="op">[</span>,<span class="op">-</span><span class="fl">1</span><span class="op">]</span><span class="op">)</span> <span class="op">*</span> <span class="fl">1e6</span><span class="op">)</span></span>
<span><span class="va">geneRPKM</span></span></code></pre></div>
<pre><code><span><span class="co">##                 Sample1 Sample2</span></span>
<span><span class="co">## ENSG00000078808   21350   31786</span></span>
<span><span class="co">## ENSG00000134075  281287  304966</span></span>
<span><span class="co">## ENSG00000134086    9354    9653</span></span>
<span><span class="co">## ENSG00000171863   73974   54915</span></span>
<span><span class="co">## ENSG00000176022    3135   16979</span></span>
<span><span class="co">## ENSG00000186827    3037     641</span></span>
<span><span class="co">## ENSG00000186891    2681     201</span></span>
<span><span class="co">## ENSG00000238345       0       0</span></span>
<span><span class="co">## ENSG00000238642       0       0</span></span>
<span><span class="co">## ENSG00000247886       0       0</span></span>
<span><span class="co">## ENSG00000252531    6066    1691</span></span>
<span><span class="co">## ENSG00000254999  213296  222826</span></span></code></pre>
<p>Please note the RPKM values in our example are higher than what you
would usually get for a real RNA-seq dataset. The values here are
artificially scaled up because our example data contains reads only for
a small number of genes.</p>
</div>
<div class="section level4">
<h4 id="analysis-of-alternative-splicing-quantification-of-exon-exon-junctions">Analysis of alternative splicing: Quantification of exon-exon
junctions<a class="anchor" aria-label="anchor" href="#analysis-of-alternative-splicing-quantification-of-exon-exon-junctions"></a>
</h4>
<p>Exon-exon junctions can be quantified by setting
<code>reportLevel="junction"</code>. In this case, <code>qCount</code>
will ignore the <code>query</code> argument and scan all alignments for
any detected splices, which are returned as a <code>GRanges</code>
object: The region start and end coordinates correspond to the first and
last bases of the intron, and the counts are returned in the
<code><a href="https://rdrr.io/pkg/S4Vectors/man/Vector-class.html" class="external-link">mcols()</a></code> of the <code>GRanges</code> object. Alignments that
are identically spliced but reside on opposite strands will be
quantified separately. In an unstranded RNA-seq experiment, this may
give rise to two separate counts for the same intron, one each for the
supporting alignments on plus and minus strands.</p>
<div class="sourceCode" id="cb126"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">exonJunctions</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/qCount.html">qCount</a></span><span class="op">(</span><span class="va">proj2</span>, <span class="cn">NULL</span>, reportLevel <span class="op">=</span> <span class="st">"junction"</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## counting junctions...done</span></span></code></pre>
<div class="sourceCode" id="cb128"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">exonJunctions</span></span></code></pre></div>
<pre><code><span><span class="co">## GRanges object with 46 ranges and 2 metadata columns:</span></span>
<span><span class="co">##        seqnames      ranges strand |   Sample1   Sample2</span></span>
<span><span class="co">##           &lt;Rle&gt;   &lt;IRanges&gt;  &lt;Rle&gt; | &lt;numeric&gt; &lt;numeric&gt;</span></span>
<span><span class="co">##    [1]     chr1 12213-12321      + |         3         0</span></span>
<span><span class="co">##    [2]     chr1 13085-13371      - |         1         0</span></span>
<span><span class="co">##    [3]     chr1 18069-18837      + |         9        16</span></span>
<span><span class="co">##    [4]     chr1 18069-18837      - |         7         4</span></span>
<span><span class="co">##    [5]     chr1 18185-18837      - |         2         0</span></span>
<span><span class="co">##    ...      ...         ...    ... .       ...       ...</span></span>
<span><span class="co">##   [42]     chr1 14166-14362      + |         0         1</span></span>
<span><span class="co">##   [43]     chr1 19308-23623      - |         0         2</span></span>
<span><span class="co">##   [44]     chr1 29327-32271      + |         0         2</span></span>
<span><span class="co">##   [45]     chr1 29327-32271      - |         0         1</span></span>
<span><span class="co">##   [46]     chr3   2504-5589      - |         0         3</span></span>
<span><span class="co">##   -------</span></span>
<span><span class="co">##   seqinfo: 3 sequences from an unspecified genome; no seqlengths</span></span></code></pre>
<p>About half of the exon-exon junctions detected in this sample dataset
correspond to known introns; they tend to be the ones with higher
coverage:</p>
<div class="sourceCode" id="cb130"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">knownIntrons</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/unlist.html" class="external-link">unlist</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/GenomicFeatures/man/transcriptsBy.html" class="external-link">intronsByTranscript</a></span><span class="op">(</span><span class="va">txdb</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">isKnown</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/IRanges/man/findOverlaps-methods.html" class="external-link">overlapsAny</a></span><span class="op">(</span><span class="va">exonJunctions</span>, <span class="va">knownIntrons</span>, type <span class="op">=</span> <span class="st">"equal"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/table.html" class="external-link">table</a></span><span class="op">(</span><span class="va">isKnown</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## isKnown</span></span>
<span><span class="co">## FALSE  TRUE </span></span>
<span><span class="co">##    25    21</span></span></code></pre>
<div class="sourceCode" id="cb132"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/tapply.html" class="external-link">tapply</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/colSums.html" class="external-link">rowSums</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">as.matrix</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Vector-class.html" class="external-link">mcols</a></span><span class="op">(</span><span class="va">exonJunctions</span><span class="op">)</span><span class="op">)</span><span class="op">)</span>,</span>
<span>       <span class="va">isKnown</span>, <span class="va">summary</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## $`FALSE`</span></span>
<span><span class="co">##    Min. 1st Qu.  Median    Mean 3rd Qu.    Max. </span></span>
<span><span class="co">##       1       3       7      47      31     342 </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## $`TRUE`</span></span>
<span><span class="co">##    Min. 1st Qu.  Median    Mean 3rd Qu.    Max. </span></span>
<span><span class="co">##     1.0     2.0    16.0    50.7    91.0   210.0</span></span></code></pre>
<p>When quantifying exon junctions, only spliced alignments will be
included in the quantification. It is also possible to only include
unspliced alignments in the quantification, for example when counting
exon body alignments that complement the exon junction alignments. This
can be done using the <code>includeSpliced</code> argument from
<code>qCount</code>:</p>
<div class="sourceCode" id="cb134"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">exonBodyLevels</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/qCount.html">qCount</a></span><span class="op">(</span><span class="va">proj2</span>, <span class="va">txdb</span>, reportLevel <span class="op">=</span> <span class="st">"exon"</span>,</span>
<span>                         includeSpliced <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## extracting exon regions from TxDb...done</span></span>
<span><span class="co">## counting alignments...done</span></span></code></pre>
<div class="sourceCode" id="cb136"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/summary.html" class="external-link">summary</a></span><span class="op">(</span><span class="va">exonLevels</span> <span class="op">-</span> <span class="va">exonBodyLevels</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##      width      Sample1       Sample2   </span></span>
<span><span class="co">##  Min.   :0   Min.   :  0   Min.   :  0  </span></span>
<span><span class="co">##  1st Qu.:0   1st Qu.:  0   1st Qu.:  0  </span></span>
<span><span class="co">##  Median :0   Median :  3   Median :  1  </span></span>
<span><span class="co">##  Mean   :0   Mean   : 42   Mean   : 35  </span></span>
<span><span class="co">##  3rd Qu.:0   3rd Qu.: 48   3rd Qu.: 50  </span></span>
<span><span class="co">##  Max.   :0   Max.   :819   Max.   :650</span></span></code></pre>
<pre><code><span><span class="co">## collecting quality control data</span></span></code></pre>
<pre><code><span><span class="co">## creating QC plots</span></span></code></pre>
</div>
</div>
<div class="section level3">
<h3 id="smrna-seq-small-rna-and-mirna-expression-profiling">smRNA-seq: small RNA and miRNA expression profiling<a class="anchor" aria-label="anchor" href="#smrna-seq-small-rna-and-mirna-expression-profiling"></a>
</h3>
<p>Expression profiling of miRNAs differs only slightly from the
profiling of mRNAs. There are a few details that need special care,
which are outlined in this section.</p>
<div class="section level4">
<h4 id="preprocessing-of-small-rna-mirna-reads">Preprocessing of small RNA (miRNA) reads<a class="anchor" aria-label="anchor" href="#preprocessing-of-small-rna-mirna-reads"></a>
</h4>
<p>Again, we start the example workflow by copying the example data
files into the current working directly, into a subfolder called
<code>"extdata"</code>.</p>
<div class="sourceCode" id="cb140"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/files.html" class="external-link">file.copy</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span>package <span class="op">=</span> <span class="st">"QuasR"</span>, <span class="st">"extdata"</span><span class="op">)</span>, <span class="st">"."</span>, recursive <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] TRUE</span></span></code></pre>
<p>As a next step, we need to remove library adapter sequences from
short RNA reads. Most sequencing experiments generate reads that are
longer than the average length of a miRNA (22nt). Therefore, the read
sequence will run through the miRNA into the library adapter sequence
and would not match when aligned in full to the reference genome.</p>
<p>We can remove those adapter sequences using
<code>preprocessReads</code> (see section @ref(preprocessReads) for more
details), which for each input sequence file will generate an output
sequence file containing appropriately truncated sequences. In the
example below, we get the input sequence filenames from
<code>sampleFile</code>, and also prepare an updated
<code>sampleFile2</code> that refers to newly generated processed
sequence files:</p>
<div class="sourceCode" id="cb142"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># prepare sample file with processed reads filenames</span></span>
<span><span class="va">sampleFile</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"samples_mirna.txt"</span><span class="op">)</span></span>
<span><span class="va">sampleFile</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "extdata/samples_mirna.txt"</span></span></code></pre>
<div class="sourceCode" id="cb144"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sampleFile2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/grep.html" class="external-link">sub</a></span><span class="op">(</span><span class="st">".txt"</span>, <span class="st">"_processed.txt"</span>, <span class="va">sampleFile</span><span class="op">)</span></span>
<span><span class="va">sampleFile2</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "extdata/samples_mirna_processed.txt"</span></span></code></pre>
<div class="sourceCode" id="cb146"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">tab</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html" class="external-link">read.delim</a></span><span class="op">(</span><span class="va">sampleFile</span>, header <span class="op">=</span> <span class="cn">TRUE</span>, as.is <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="va">tab</span></span></code></pre></div>
<pre><code><span><span class="co">##     FileName SampleName</span></span>
<span><span class="co">## 1 mirna_1.fa     miRNAs</span></span></code></pre>
<div class="sourceCode" id="cb148"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">tab2</span> <span class="op">&lt;-</span> <span class="va">tab</span></span>
<span><span class="va">tab2</span><span class="op">$</span><span class="va">FileName</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/grep.html" class="external-link">sub</a></span><span class="op">(</span><span class="st">".fa"</span>, <span class="st">"_processed.fa"</span>, <span class="va">tab</span><span class="op">$</span><span class="va">FileName</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/write.table.html" class="external-link">write.table</a></span><span class="op">(</span><span class="va">tab2</span>, <span class="va">sampleFile2</span>, sep <span class="op">=</span> <span class="st">"\t"</span>, quote <span class="op">=</span> <span class="cn">FALSE</span>, row.names <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="va">tab2</span></span></code></pre></div>
<pre><code><span><span class="co">##               FileName SampleName</span></span>
<span><span class="co">## 1 mirna_1_processed.fa     miRNAs</span></span></code></pre>
<div class="sourceCode" id="cb150"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># remove adapters</span></span>
<span><span class="va">oldwd</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/getwd.html" class="external-link">setwd</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/path.html" class="external-link">dirname</a></span><span class="op">(</span><span class="va">sampleFile</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">res</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/preprocessReads.html">preprocessReads</a></span><span class="op">(</span><span class="va">tab</span><span class="op">$</span><span class="va">FileName</span>,</span>
<span>                       <span class="va">tab2</span><span class="op">$</span><span class="va">FileName</span>,</span>
<span>                       Rpattern <span class="op">=</span> <span class="st">"TGGAATTCTCGGGTGCCAAGG"</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##   filtering mirna_1.fa</span></span></code></pre>
<div class="sourceCode" id="cb152"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">res</span></span></code></pre></div>
<pre><code><span><span class="co">##                  mirna_1.fa</span></span>
<span><span class="co">## totalSequences         1000</span></span>
<span><span class="co">## matchTo5pAdapter          0</span></span>
<span><span class="co">## matchTo3pAdapter       1000</span></span>
<span><span class="co">## tooShort                  0</span></span>
<span><span class="co">## tooManyN                  0</span></span>
<span><span class="co">## lowComplexity             0</span></span>
<span><span class="co">## totalPassed            1000</span></span></code></pre>
<div class="sourceCode" id="cb154"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/getwd.html" class="external-link">setwd</a></span><span class="op">(</span><span class="va">oldwd</span><span class="op">)</span></span></code></pre></div>
<p>The miRNA reads in <code>mirna_1.fa</code> are by the way synthetic
sequences and do not correspond to any existing miRNAs. As you can see
above from the return value of <code>preprocessReads</code>, all reads
matched to the 3’-adapter and were therefore truncated, reducing their
length to roughly the expected 22nt:</p>
<div class="sourceCode" id="cb155"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># get read lengths</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://bioconductor.org/packages/Biostrings" class="external-link">Biostrings</a></span><span class="op">)</span></span>
<span><span class="va">oldwd</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/getwd.html" class="external-link">setwd</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/path.html" class="external-link">dirname</a></span><span class="op">(</span><span class="va">sampleFile</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">lens</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/Biostrings/man/XStringSet-io.html" class="external-link">fasta.seqlengths</a></span><span class="op">(</span><span class="va">tab</span><span class="op">$</span><span class="va">FileName</span>, nrec <span class="op">=</span> <span class="fl">1e5</span><span class="op">)</span></span>
<span><span class="va">lens2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/Biostrings/man/XStringSet-io.html" class="external-link">fasta.seqlengths</a></span><span class="op">(</span><span class="va">tab2</span><span class="op">$</span><span class="va">FileName</span>, nrec <span class="op">=</span> <span class="fl">1e5</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/getwd.html" class="external-link">setwd</a></span><span class="op">(</span><span class="va">oldwd</span><span class="op">)</span></span>
<span><span class="co"># plot length distribution</span></span>
<span><span class="va">lensTab</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/cbind.html" class="external-link">rbind</a></span><span class="op">(</span>raw <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/tabulate.html" class="external-link">tabulate</a></span><span class="op">(</span><span class="va">lens</span>, <span class="fl">50</span><span class="op">)</span>,</span>
<span>                 processed <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/tabulate.html" class="external-link">tabulate</a></span><span class="op">(</span><span class="va">lens2</span>, <span class="fl">50</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/row_colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">lensTab</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fl">1</span><span class="op">:</span><span class="fl">50</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/barplot.html" class="external-link">barplot</a></span><span class="op">(</span><span class="va">lensTab</span><span class="op">/</span><span class="fu"><a href="https://rdrr.io/r/base/colSums.html" class="external-link">rowSums</a></span><span class="op">(</span><span class="va">lensTab</span><span class="op">)</span><span class="op">*</span><span class="fl">100</span>,</span>
<span>        xlab <span class="op">=</span> <span class="st">"Read length (nt)"</span>, ylab <span class="op">=</span> <span class="st">"Percent of reads"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/legend.html" class="external-link">legend</a></span><span class="op">(</span>x <span class="op">=</span> <span class="st">"topleft"</span>, bty <span class="op">=</span> <span class="st">"n"</span>, fill <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/grDevices/gray.colors.html" class="external-link">gray.colors</a></span><span class="op">(</span><span class="fl">2</span><span class="op">)</span>,</span>
<span>       legend <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/row_colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">lensTab</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="QuasR_files/figure-html/miRNA_lengthes-1.png" width="576"></p>
</div>
<div class="section level4">
<h4 id="alignment-of-small-rna-mirna-reads">Alignment of small RNA (miRNA) reads<a class="anchor" aria-label="anchor" href="#alignment-of-small-rna-mirna-reads"></a>
</h4>
<p>Next, we create alignments using <code>qAlign</code>. In contrast to
the general RNA-seq workflow (section @ref(RNAseq)), alignment time can
be reduced by using the default unspliced alignment
(<code>splicedAlignment=FALSE</code>). Importantly, we need to set
<code>maxHits=50</code> or similar to also align reads that perfectly
match the genome multiple times. This is required because of the miRNAs
that are encoded by multiple genes. Reads from such miRNAs would not be
aligned and thus their expression would be underestimated if using the
default <code>maxHits=1</code>. An example of such a multiply-encoded
miRNA is mmu-miR-669a-5p, which has twelve exact copies in the mm10
genome assembly according to mirBase19.</p>
<div class="sourceCode" id="cb156"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">proj3</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/qAlign.html">qAlign</a></span><span class="op">(</span><span class="va">sampleFile2</span>, <span class="va">genomeFile</span>, maxHits <span class="op">=</span> <span class="fl">50</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## alignment files missing - need to:</span></span></code></pre>
<pre><code><span><span class="co">##     create 1 genomic alignment(s)</span></span></code></pre>
<pre><code><span><span class="co">## Testing the compute nodes...OK</span></span>
<span><span class="co">## Loading QuasR on the compute nodes...preparing to run on 1 nodes...done</span></span>
<span><span class="co">## Available cores:</span></span>
<span><span class="co">## Mac-1698842773251.local: 1</span></span>
<span><span class="co">## Performing genomic alignments for 1 samples. See progress in the log file:</span></span>
<span><span class="co">## /Users/runner/work/QuasR/QuasR/vignettes/QuasR_log_10c5346620246.txt</span></span>
<span><span class="co">## Genomic alignments have been created successfully</span></span></code></pre>
<div class="sourceCode" id="cb160"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/alignmentStats.html">alignmentStats</a></span><span class="op">(</span><span class="va">proj3</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##               seqlength mapped unmapped</span></span>
<span><span class="co">## miRNAs:genome     95000   1000        0</span></span></code></pre>
<p>A more detailed picture of the experiments’ quality can be obtained
using <code>qQCReport(proj3, "qcreport.pdf")</code> or similar (see also
section @ref(qQCReport)).</p>
</div>
<div class="section level4">
<h4 id="quantification-of-small-rna-mirna-reads">Quantification of small RNA (miRNA) reads<a class="anchor" aria-label="anchor" href="#quantification-of-small-rna-mirna-reads"></a>
</h4>
<p>As with other experiment types, miRNAs are quantified using
<code>qCount</code>. For this purpose, we first construct a query
<code>GRanges</code> object with the genomic locations of mature miRNAs.
The locations can be obtained from the <em><a href="https://bioconductor.org/packages/3.19/mirbase.db" class="external-link">mirbase.db</a></em>
package, or directly from the species-specific gff files provided by the
mirBase database (e.g. (<a href="ftp://mirbase.org/pub/mirbase/19/genomes/mmu.gff3" class="uri">ftp://mirbase.org/pub/mirbase/19/genomes/mmu.gff3</a>)). For
the purpose of this example, the <em><a href="https://bioconductor.org/packages/3.19/QuasR" class="external-link">QuasR</a></em>
package provides a small gff file (<code>"mirbaseXX_qsr.gff3"</code>)
that is formatted as the ones available from mirBase. The gff file
contains both the locations of pre-miRNAs (hairpin precursors), as well
as mature miRNAs. The two can be discriminated by their
<code>"type"</code>:</p>
<div class="sourceCode" id="cb162"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">mirs</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocIO/man/IO.html" class="external-link">import</a></span><span class="op">(</span><span class="st">"extdata/mirbaseXX_qsr.gff3"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">mirs</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">mirs</span><span class="op">$</span><span class="va">Name</span></span>
<span><span class="va">preMirs</span> <span class="op">&lt;-</span> <span class="va">mirs</span><span class="op">[</span> <span class="va">mirs</span><span class="op">$</span><span class="va">type</span> <span class="op">==</span> <span class="st">"miRNA_primary_transcript"</span> <span class="op">]</span></span>
<span><span class="va">matureMirs</span> <span class="op">&lt;-</span> <span class="va">mirs</span><span class="op">[</span> <span class="va">mirs</span><span class="op">$</span><span class="va">type</span> <span class="op">==</span> <span class="st">"miRNA"</span> <span class="op">]</span></span></code></pre></div>
<p>Please note that the name attribute of the <code>GRanges</code>
object must be set appropriately, so that <code>qCount</code> can
identify a single mature miRNA sequence that is encoded by multiple loci
(see below) by their identical names. In this example, there are no
multiply-encoded mature miRNAs, but in a real sample, you can detect
them for example with <code>table(names(mirs))</code>.</p>
<p>The <code>preMirs</code> and <code>matureMirs</code> could now be
used as <code>query</code> in <code>qCount</code>. In practise however,
miRNA seem to not always be processed with high accuracy. Many miRNA
reads that start one or two bases earlier or later can be observed in
real data, and also their length may vary for a few bases. This is the
case for the synthetic miRNAs used in this example, whose lengthes and
start positions have been sampled from a read data set:</p>
<div class="sourceCode" id="cb163"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://bioconductor.org/packages/Rsamtools" class="external-link">Rsamtools</a></span><span class="op">)</span></span>
<span><span class="va">alns</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/Rsamtools/man/scanBam.html" class="external-link">scanBam</a></span><span class="op">(</span><span class="fu"><a href="../reference/qProject-class.html">alignments</a></span><span class="op">(</span><span class="va">proj3</span><span class="op">)</span><span class="op">$</span><span class="va">genome</span><span class="op">$</span><span class="va">FileName</span>,</span>
<span>                param <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/Rsamtools/man/ScanBamParam-class.html" class="external-link">ScanBamParam</a></span><span class="op">(</span>what <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/Rsamtools/man/ScanBamParam-class.html" class="external-link">scanBamWhat</a></span><span class="op">(</span><span class="op">)</span>,</span>
<span>                                     which <span class="op">=</span> <span class="va">preMirs</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span><span class="op">)</span><span class="op">)</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span></span>
<span><span class="va">alnsIR</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/IRanges/man/IRanges-constructor.html" class="external-link">IRanges</a></span><span class="op">(</span>start <span class="op">=</span> <span class="va">alns</span><span class="op">$</span><span class="va">pos</span> <span class="op">-</span> <span class="fu"><a href="https://rdrr.io/r/stats/start.html" class="external-link">start</a></span><span class="op">(</span><span class="va">preMirs</span><span class="op">)</span>, width <span class="op">=</span> <span class="va">alns</span><span class="op">$</span><span class="va">qwidth</span><span class="op">)</span></span>
<span><span class="va">mp</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/graphics/barplot.html" class="external-link">barplot</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/as.vector.html" class="external-link">as.vector</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/IRanges/man/coverage-methods.html" class="external-link">coverage</a></span><span class="op">(</span><span class="va">alnsIR</span><span class="op">)</span><span class="op">)</span>, names.arg <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq_len</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/Extremes.html" class="external-link">max</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/stats/start.html" class="external-link">end</a></span><span class="op">(</span><span class="va">alnsIR</span><span class="op">)</span><span class="op">)</span><span class="op">)</span>,</span>
<span>              xlab <span class="op">=</span> <span class="st">"Relative position in pre-miRNA"</span>,</span>
<span>              ylab <span class="op">=</span> <span class="st">"Alignment coverage"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/rect.html" class="external-link">rect</a></span><span class="op">(</span>xleft <span class="op">=</span> <span class="va">mp</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/stats/start.html" class="external-link">start</a></span><span class="op">(</span><span class="va">matureMirs</span><span class="op">)</span> <span class="op">-</span> <span class="fu"><a href="https://rdrr.io/r/stats/start.html" class="external-link">start</a></span><span class="op">(</span><span class="va">preMirs</span><span class="op">)</span> <span class="op">+</span> <span class="fl">1</span>,<span class="fl">1</span><span class="op">]</span>,</span>
<span>     ybottom <span class="op">=</span> <span class="op">-</span><span class="fu"><a href="https://rdrr.io/r/graphics/par.html" class="external-link">par</a></span><span class="op">(</span><span class="st">'cxy'</span><span class="op">)</span><span class="op">[</span><span class="fl">2</span><span class="op">]</span>,</span>
<span>     xright <span class="op">=</span> <span class="va">mp</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/stats/start.html" class="external-link">end</a></span><span class="op">(</span><span class="va">matureMirs</span><span class="op">)</span> <span class="op">-</span> <span class="fu"><a href="https://rdrr.io/r/stats/start.html" class="external-link">start</a></span><span class="op">(</span><span class="va">preMirs</span><span class="op">)</span> <span class="op">+</span> <span class="fl">1</span>,<span class="fl">1</span><span class="op">]</span>,</span>
<span>     ytop <span class="op">=</span> <span class="fl">0</span>, col <span class="op">=</span> <span class="st">"#CCAA0088"</span>, border <span class="op">=</span> <span class="cn">NA</span>, xpd <span class="op">=</span> <span class="cn">NA</span><span class="op">)</span></span></code></pre></div>
<p><img src="QuasR_files/figure-html/miRNA_coverage-1.png" width="576"></p>
<p>By default, <code>qCount</code> will count alignments that have their
5’-end within the query region (see <code>selectReadPosition</code>
argument). The 5’-end correspond to the lower (left) coordinate for
alignments on the plus strand, and to the higher (right) coordinate for
alignments on the minus strand. In order not to miss miRNAs that have a
couple of extra or missing bases, we therefore construct a query window
around the 5’-end of each mature miRNA, by adding three bases up- and
downstream:</p>
<div class="sourceCode" id="cb164"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">matureMirsExtended</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/IRanges/man/intra-range-methods.html" class="external-link">resize</a></span><span class="op">(</span><span class="va">matureMirs</span>, width <span class="op">=</span> <span class="fl">1L</span>, fix <span class="op">=</span> <span class="st">"start"</span><span class="op">)</span> <span class="op">+</span> <span class="fl">3L</span></span></code></pre></div>
<p>The resulting extended query is then used to quantify mature miRNAs.
Multiple-encoded miRNAs will be represented by multiple ranges in
<code>matureMirs</code> and <code>matureMirsExtended</code>, which have
identical names. <code>qCount</code> will automatically sum all
alignments from any of those regions and return a single number per
sample and unique miRNA name.</p>
<div class="sourceCode" id="cb165"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># quantify mature miRNAs</span></span>
<span><span class="va">cnt</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/qCount.html">qCount</a></span><span class="op">(</span><span class="va">proj3</span>, <span class="va">matureMirsExtended</span>, orientation <span class="op">=</span> <span class="st">"same"</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## counting alignments...done</span></span></code></pre>
<div class="sourceCode" id="cb167"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">cnt</span></span></code></pre></div>
<pre><code><span><span class="co">##                 width miRNAs</span></span>
<span><span class="co">## qsr-miR-9876-5p     7     13</span></span>
<span><span class="co">## qsr-miR-9876-3p     7    984</span></span></code></pre>
<div class="sourceCode" id="cb169"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># quantify pre-miRNAs</span></span>
<span><span class="va">cnt</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/qCount.html">qCount</a></span><span class="op">(</span><span class="va">proj3</span>, <span class="va">preMirs</span>, orientation <span class="op">=</span> <span class="st">"same"</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## counting alignments...done</span></span></code></pre>
<div class="sourceCode" id="cb171"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">cnt</span></span></code></pre></div>
<pre><code><span><span class="co">##              width miRNAs</span></span>
<span><span class="co">## qsr-mir-9876    75   1000</span></span></code></pre>
</div>
</div>
<div class="section level3">
<h3 id="bis-seq-measuring-dna-methylation">Bis-seq: Measuring DNA methylation<a class="anchor" aria-label="anchor" href="#bis-seq-measuring-dna-methylation"></a>
</h3>
<p>Sequencing of bisulfite-converted genomic DNA allows detection of
methylated cytosines, which in mammalian genomes typically occur in the
context of CpG dinucleotides. The treatment of DNA with bisulfite
induces deamination of non-methylated cytosines, converting them to
uracils. Sequencing and aligning of such bisulfite-converted DNA results
in C-to-T mismatches. Both alignment of converted reads, as well as the
interpretation of the alignments for calculation of methylation levels
require specific approaches and are supported in <em><a href="https://bioconductor.org/packages/3.19/QuasR" class="external-link">QuasR</a></em> by
<code>qAlign</code> (<code>bisulfite</code> argument, section
@ref(qAlign)) and <code>qMeth</code> (section @ref(qMeth)),
respectively.</p>
<p>We start the analysis by copying the example data files into the
current working directly, into a subfolder called
<code>"extdata"</code>. Then, bisulfite-specific alignment is selected
in <code>qAlign</code> by setting <code>bisulfite</code> to
<code>"dir"</code> for a directional experiment, or to
<code>"undir"</code> for an undirectional Bis-seq experiment:</p>
<div class="sourceCode" id="cb173"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/files.html" class="external-link">file.copy</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span>package <span class="op">=</span> <span class="st">"QuasR"</span>, <span class="st">"extdata"</span><span class="op">)</span>, <span class="st">"."</span>, recursive <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] TRUE</span></span></code></pre>
<div class="sourceCode" id="cb175"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sampleFile</span> <span class="op">&lt;-</span> <span class="st">"extdata/samples_bis_single.txt"</span></span>
<span><span class="va">genomeFile</span> <span class="op">&lt;-</span> <span class="st">"extdata/hg19sub.fa"</span></span>
<span></span>
<span><span class="va">proj4</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/qAlign.html">qAlign</a></span><span class="op">(</span><span class="va">sampleFile</span>, <span class="va">genomeFile</span>, bisulfite <span class="op">=</span> <span class="st">"dir"</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## alignment files missing - need to:</span></span></code></pre>
<pre><code><span><span class="co">##     create alignment index for the genome</span></span></code></pre>
<pre><code><span><span class="co">##     create 1 genomic alignment(s)</span></span></code></pre>
<pre><code><span><span class="co">## Creating an RbowtieCtoT index for /Users/runner/work/QuasR/QuasR/vignettes/extdata/hg19sub.fa</span></span></code></pre>
<pre><code><span><span class="co">## Finished creating index</span></span></code></pre>
<pre><code><span><span class="co">## Testing the compute nodes...OK</span></span>
<span><span class="co">## Loading QuasR on the compute nodes...preparing to run on 1 nodes...done</span></span>
<span><span class="co">## Available cores:</span></span>
<span><span class="co">## Mac-1698842773251.local: 1</span></span>
<span><span class="co">## Performing genomic alignments for 1 samples. See progress in the log file:</span></span>
<span><span class="co">## /Users/runner/work/QuasR/QuasR/vignettes/QuasR_log_10c537425151b.txt</span></span>
<span><span class="co">## Genomic alignments have been created successfully</span></span></code></pre>
<div class="sourceCode" id="cb182"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">proj4</span></span></code></pre></div>
<pre><code><span><span class="co">## Project: qProject</span></span>
<span><span class="co">##  Options   : maxHits         : 1</span></span>
<span><span class="co">##              paired          : no</span></span>
<span><span class="co">##              splicedAlignment: FALSE</span></span>
<span><span class="co">##              bisulfite       : dir</span></span>
<span><span class="co">##              snpFile         : none</span></span>
<span><span class="co">##              geneAnnotation  : none</span></span>
<span><span class="co">##  Aligner   : Rbowtie v1.43.0 (parameters: -k 2 --best --strata -v 2)</span></span>
<span><span class="co">##  Genome    : /Users/runner/work/QuasR/QuasR/vignettes/extdata/hg19sub.fa (file)</span></span>
<span><span class="co">## </span></span>
<span><span class="co">##  Reads     : 1 file, 1 sample (fasta format):</span></span>
<span><span class="co">##    1. bis_1_1.fa.bz2  Sample1</span></span>
<span><span class="co">## </span></span>
<span><span class="co">##  Genome alignments: directory: same as reads</span></span>
<span><span class="co">##    1. bis_1_1_10c532e8cdb2f.bam</span></span>
<span><span class="co">## </span></span>
<span><span class="co">##  Aux. alignments: none</span></span></code></pre>
<p>The resulting alignments are not different from those of non-Bis-seq
experiments, apart from the fact that they may contain many C-to-T (or
A-to-G) mismatches that are not counted as mismatches when aligning the
reads. The number of alignments in specific genomic regions could be
quantified using <code>qCount</code> as with ChIP-seq or RNA-seq
experiments. For quantification of methylation the <code>qMeth</code>
function is used:</p>
<div class="sourceCode" id="cb184"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">meth</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/qMeth.html">qMeth</a></span><span class="op">(</span><span class="va">proj4</span>, mode <span class="op">=</span> <span class="st">"CpGcomb"</span>, collapseBySample <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="va">meth</span></span></code></pre></div>
<pre><code><span><span class="co">## GRanges object with 3110 ranges and 2 metadata columns:</span></span>
<span><span class="co">##          seqnames      ranges strand | Sample1_T Sample1_M</span></span>
<span><span class="co">##             &lt;Rle&gt;   &lt;IRanges&gt;  &lt;Rle&gt; | &lt;integer&gt; &lt;integer&gt;</span></span>
<span><span class="co">##      [1]     chr1       19-20      * |         1         1</span></span>
<span><span class="co">##      [2]     chr1       21-22      * |         1         1</span></span>
<span><span class="co">##      [3]     chr1       54-55      * |         3         1</span></span>
<span><span class="co">##      [4]     chr1       57-58      * |         3         0</span></span>
<span><span class="co">##      [5]     chr1       80-81      * |         6         5</span></span>
<span><span class="co">##      ...      ...         ...    ... .       ...       ...</span></span>
<span><span class="co">##   [3106]     chr3 44957-44958      * |         8         7</span></span>
<span><span class="co">##   [3107]     chr3 44977-44978      * |         5         3</span></span>
<span><span class="co">##   [3108]     chr3 44981-44982      * |         4         3</span></span>
<span><span class="co">##   [3109]     chr3 44989-44990      * |         1         1</span></span>
<span><span class="co">##   [3110]     chr3 44993-44994      * |         1         1</span></span>
<span><span class="co">##   -------</span></span>
<span><span class="co">##   seqinfo: 3 sequences from an unspecified genome</span></span></code></pre>
<p>By default, <code>qMeth</code> quantifies methylation for all
cytosines in CpG contexts, combining the data from plus and minus
strands (<code>mode="CpGcomb"</code>). The results are returned as a
<code>GRanges</code> object with coordinates of each CpG, and two
metadata columns for each input sequence file in the
<code>qProject</code> object. These two columns contain the total number
of aligned reads that overlap a given CpG (C-to-C matches or C-to-T
mismatches, suffix <code>_T</code> in the column name), and the number
of read alignments that had a C-to-C match at that position (methylated
events, suffix <code>_M</code>).</p>
<p>Independent of the number of alignments, the returned object will
list all CpGs in the target genome including the ones that have zero
coverage, unless you set <code>keepZero=FALSE</code>:</p>
<div class="sourceCode" id="cb186"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">chrs</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/Biostrings/man/XStringSet-io.html" class="external-link">readDNAStringSet</a></span><span class="op">(</span><span class="va">genomeFile</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/sum.html" class="external-link">sum</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/Biostrings/man/matchPattern.html" class="external-link">vcountPattern</a></span><span class="op">(</span><span class="st">"CG"</span>,<span class="va">chrs</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] 3110</span></span></code></pre>
<div class="sourceCode" id="cb188"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="fu"><a href="../reference/qMeth.html">qMeth</a></span><span class="op">(</span><span class="va">proj4</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] 3110</span></span></code></pre>
<div class="sourceCode" id="cb190"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="fu"><a href="../reference/qMeth.html">qMeth</a></span><span class="op">(</span><span class="va">proj4</span>, keepZero <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] 2929</span></span></code></pre>
<p>The fraction methylation can easily be obtained as the ratio between
<code>_M</code> and <code>_T</code> columns:</p>
<div class="sourceCode" id="cb192"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">percMeth</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Vector-class.html" class="external-link">mcols</a></span><span class="op">(</span><span class="va">meth</span><span class="op">)</span><span class="op">[</span>,<span class="fl">2</span><span class="op">]</span> <span class="op">*</span> <span class="fl">100</span> <span class="op">/</span> <span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Vector-class.html" class="external-link">mcols</a></span><span class="op">(</span><span class="va">meth</span><span class="op">)</span><span class="op">[</span>,<span class="fl">1</span><span class="op">]</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/summary.html" class="external-link">summary</a></span><span class="op">(</span><span class="va">percMeth</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##    Min. 1st Qu.  Median    Mean 3rd Qu.    Max.    NA's </span></span>
<span><span class="co">##     0.0    75.0    90.9    75.4   100.0   100.0     181</span></span></code></pre>
<div class="sourceCode" id="cb194"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">axisTrack</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/Gviz/man/GenomeAxisTrack-class.html" class="external-link">GenomeAxisTrack</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="va">dTrack1</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/Gviz/man/DataTrack-class.html" class="external-link">DataTrack</a></span><span class="op">(</span>range <span class="op">=</span> <span class="va">gr1</span>, name <span class="op">=</span> <span class="st">"H3K4me3"</span>, type <span class="op">=</span> <span class="st">"h"</span><span class="op">)</span></span>
<span><span class="va">dTrack2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/Gviz/man/DataTrack-class.html" class="external-link">DataTrack</a></span><span class="op">(</span>range <span class="op">=</span> <span class="va">meth</span>, data <span class="op">=</span> <span class="va">percMeth</span>,</span>
<span>                     name <span class="op">=</span> <span class="st">"Methylation"</span>, type <span class="op">=</span> <span class="st">"p"</span><span class="op">)</span></span>
<span><span class="va">txTrack</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/Gviz/man/GeneRegionTrack-class.html" class="external-link">GeneRegionTrack</a></span><span class="op">(</span><span class="va">txdb</span>, name <span class="op">=</span> <span class="st">"Transcripts"</span>, showId <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/Gviz/man/plotTracks.html" class="external-link">plotTracks</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="va">axisTrack</span>, <span class="va">dTrack1</span>, <span class="va">dTrack2</span>, <span class="va">txTrack</span><span class="op">)</span>,</span>
<span>           chromosome <span class="op">=</span> <span class="st">"chr3"</span>, extend.left <span class="op">=</span> <span class="fl">1000</span><span class="op">)</span></span></code></pre></div>
<p><img src="QuasR_files/figure-html/Bis_visualize-1.png" width="768"></p>
<p>If <code>qMeth</code> is called without a <code>query</code>
argument, it will by default return methylation states for each C or CpG
in the genome. Using a <code>query</code> argument it is possible to
restrict the analysis to specific genomic regions, and if using in
addition <code>collapseByQueryRegion=TRUE</code>, the single base
methylation states will further be combined for all C’s that are
contained in the same query region:</p>
<div class="sourceCode" id="cb195"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/qMeth.html">qMeth</a></span><span class="op">(</span><span class="va">proj4</span>, query <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/GenomicRanges/man/GRanges-class.html" class="external-link">GRanges</a></span><span class="op">(</span><span class="st">"chr1"</span>,<span class="fu"><a href="https://rdrr.io/pkg/IRanges/man/IRanges-constructor.html" class="external-link">IRanges</a></span><span class="op">(</span>start <span class="op">=</span> <span class="fl">31633</span>, width <span class="op">=</span> <span class="fl">2</span><span class="op">)</span><span class="op">)</span>,</span>
<span>      collapseBySample <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## GRanges object with 1 range and 2 metadata columns:</span></span>
<span><span class="co">##       seqnames      ranges strand | Sample1_T Sample1_M</span></span>
<span><span class="co">##          &lt;Rle&gt;   &lt;IRanges&gt;  &lt;Rle&gt; | &lt;integer&gt; &lt;integer&gt;</span></span>
<span><span class="co">##   [1]     chr1 31633-31634      * |        10         2</span></span>
<span><span class="co">##   -------</span></span>
<span><span class="co">##   seqinfo: 3 sequences from an unspecified genome</span></span></code></pre>
<div class="sourceCode" id="cb197"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/qMeth.html">qMeth</a></span><span class="op">(</span><span class="va">proj4</span>, query <span class="op">=</span> <span class="va">promRegSel</span>, collapseByQueryRegion <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>      collapseBySample <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## GRanges object with 12 ranges and 2 metadata columns:</span></span>
<span><span class="co">##        seqnames      ranges strand | Sample1_T Sample1_M</span></span>
<span><span class="co">##           &lt;Rle&gt;   &lt;IRanges&gt;  &lt;Rle&gt; | &lt;numeric&gt; &lt;numeric&gt;</span></span>
<span><span class="co">##    [1]     chr1 31629-33128      + |       426        74</span></span>
<span><span class="co">##    [2]     chr1   6452-7951      - |       388       244</span></span>
<span><span class="co">##    [3]     chr1 14013-15512      - |       627       560</span></span>
<span><span class="co">##    [4]     chr1 31882-33381      - |       522       232</span></span>
<span><span class="co">##    [5]     chr2   1795-3294      + |       997       539</span></span>
<span><span class="co">##    ...      ...         ...    ... .       ...       ...</span></span>
<span><span class="co">##    [8]     chr3   1276-2775      + |       715       253</span></span>
<span><span class="co">##    [9]     chr3 19069-20568      + |       253       204</span></span>
<span><span class="co">##   [10]     chr3 26692-28191      + |       934       818</span></span>
<span><span class="co">##   [11]     chr3 26834-28333      + |       934       777</span></span>
<span><span class="co">##   [12]     chr3 13102-14601      - |       307       287</span></span>
<span><span class="co">##   -------</span></span>
<span><span class="co">##   seqinfo: 3 sequences from an unspecified genome</span></span></code></pre>
<p>Finally, <code>qMeth</code> allows the retrieval of methylation
states for individual molecules (per alignment). This is done by using a
<code>query</code> containing a single genomic region (typically small,
such as a PCR amplicon) and setting
<code>reportLevel="alignment"</code>. In that case, the return value of
<code>qMeth</code> will be a list (over samples) of lists (with four
elements giving the identities of alignment, C nucleotide, strand and
the methylation state). See the documentation of <code>qMeth</code> for
more details.</p>
</div>
<div class="section level3">
<h3 id="alleleSpecificAnalysis">Allele-specific analysis<a class="anchor" aria-label="anchor" href="#alleleSpecificAnalysis"></a>
</h3>
<p>All experiment types supported by <em><a href="https://bioconductor.org/packages/3.19/QuasR" class="external-link">QuasR</a></em>
(ChIP-seq, RNA-seq and Bis-seq; only alignments to the genome, but not
to auxiliaries) can also be analyzed in an allele-specific manner. For
this, a file containing genomic location and the two alleles of known
sequence polymorphisms has to be provided to the <code>snpFile</code>
argument of <code>qAlign</code>. The file is in tab-delimited text
format without a header and contains four columns with chromosome name,
position, reference allele and alternative allele.</p>
<p>Below is an example of a SNP file, also available from
<code>system.file(package="QuasR", "extdata", "hg19sub_snp.txt")</code>:</p>
<pre>
chr1    3199    C   T
chr1    3277    C   T
chr1    4162    C   T
chr1    4195    C   T
...
</pre>
<!--
`chr1` |  `8596` | `G` | `A`  
`chr1` | `18443` | `G` | `A`  
`chr1` | `18981` | `C` | `T`  
`chr1` | `19341` | `G` | `A`  
`...`  |         |     |  
-->
<p>For a given locus, either reference or alternative allele may but
does not have to be identical to the sequence of the reference genome
(<code>genomeFile</code> in this case). To avoid an alignment bias, all
reads are aligned separately to each of the two new genomes, which
<em><a href="https://bioconductor.org/packages/3.19/QuasR" class="external-link">QuasR</a></em>
generates by <em>injecting</em> the SNPs listed in <code>snpFile</code>
into the reference genome. Finally, the two alignment files are
combined, only retaining the best alignment for each read. While this
procedure takes more than twice as long as aligning against a single
genome, it has the advantage to correctly align reads even in regions of
high SNP density and has been shown to produce more accurate
results.</p>
<p>While combining alignments, each read is classified into one of three
groups (stored in the <code>bam</code> files under the <code>XV</code>
tag):</p>
<ul>
<li>
<strong>R</strong>: the read aligned better to the
<strong>reference</strong> genome<br>
</li>
<li>
<strong>U</strong>: the read aligned equally well to both genomes
(<strong>unknown</strong> origin)<br>
</li>
<li>
<strong>A</strong>: the read aligned better to the
<strong>alternative</strong> genome</li>
</ul>
<p>Using these alignment classifications, the <code>qCount</code> and
<code>qMeth</code> functions will produce three counts instead of a
single count; one for each class. The column names will be suffixed by
<code>_R</code>, <code>_U</code> and <code>_A</code>.</p>
<p>The examples below use data provided with the <em><a href="https://bioconductor.org/packages/3.19/QuasR" class="external-link">QuasR</a></em>
package, which is first copied to the current working directory, into a
subfolder called <code>"extdata"</code>:</p>
<div class="sourceCode" id="cb199"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/files.html" class="external-link">file.copy</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span>package <span class="op">=</span> <span class="st">"QuasR"</span>, <span class="st">"extdata"</span><span class="op">)</span>, <span class="st">"."</span>, recursive <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] TRUE</span></span></code></pre>
<p>The example below aligns the same reads that were also used in the
ChIP-seq workflow (section @ref(ChIPseq)), but this time using a
<code>snpFile</code>:</p>
<div class="sourceCode" id="cb201"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sampleFile</span> <span class="op">&lt;-</span> <span class="st">"extdata/samples_chip_single.txt"</span></span>
<span><span class="va">genomeFile</span> <span class="op">&lt;-</span> <span class="st">"extdata/hg19sub.fa"</span></span>
<span><span class="va">snpFile</span> <span class="op">&lt;-</span> <span class="st">"extdata/hg19sub_snp.txt"</span></span>
<span><span class="va">proj1SNP</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/qAlign.html">qAlign</a></span><span class="op">(</span><span class="va">sampleFile</span>, genome <span class="op">=</span> <span class="va">genomeFile</span>, snpFile <span class="op">=</span> <span class="va">snpFile</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## alignment files missing - need to:</span></span></code></pre>
<pre><code><span><span class="co">##     create alignment index for the genome</span></span></code></pre>
<pre><code><span><span class="co">##     create 2 genomic alignment(s)</span></span></code></pre>
<pre><code><span><span class="co">## Reading and processing the SNP file: /Users/runner/work/QuasR/QuasR/vignettes/extdata/hg19sub_snp.txt</span></span></code></pre>
<pre><code><span><span class="co">## Creating the genome fasta file containing the SNPs: /Users/runner/work/QuasR/QuasR/vignettes/extdata/hg19sub_snp.txt.hg19sub.fa.R.fa</span></span></code></pre>
<pre><code><span><span class="co">## Creating the genome fasta file containing the SNPs: /Users/runner/work/QuasR/QuasR/vignettes/extdata/hg19sub_snp.txt.hg19sub.fa.A.fa</span></span></code></pre>
<pre><code><span><span class="co">## Creating a .fai file for the snp genome: /Users/runner/work/QuasR/QuasR/vignettes/extdata/hg19sub_snp.txt.hg19sub.fa.R.fa</span></span></code></pre>
<pre><code><span><span class="co">## Creating a .fai file for the snp genome: /Users/runner/work/QuasR/QuasR/vignettes/extdata/hg19sub_snp.txt.hg19sub.fa.A.fa</span></span></code></pre>
<pre><code><span><span class="co">## Creating an Rbowtie index for /Users/runner/work/QuasR/QuasR/vignettes/extdata/hg19sub_snp.txt.hg19sub.fa.R.fa</span></span></code></pre>
<pre><code><span><span class="co">## Finished creating index</span></span></code></pre>
<pre><code><span><span class="co">## Creating an Rbowtie index for /Users/runner/work/QuasR/QuasR/vignettes/extdata/hg19sub_snp.txt.hg19sub.fa.A.fa</span></span></code></pre>
<pre><code><span><span class="co">## Finished creating index</span></span></code></pre>
<pre><code><span><span class="co">## Testing the compute nodes...OK</span></span>
<span><span class="co">## Loading QuasR on the compute nodes...preparing to run on 1 nodes...done</span></span>
<span><span class="co">## Available cores:</span></span>
<span><span class="co">## Mac-1698842773251.local: 1</span></span>
<span><span class="co">## Performing genomic alignments for 2 samples. See progress in the log file:</span></span>
<span><span class="co">## /Users/runner/work/QuasR/QuasR/vignettes/QuasR_log_10c53226942f2.txt</span></span>
<span><span class="co">## Genomic alignments have been created successfully</span></span></code></pre>
<div class="sourceCode" id="cb215"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">proj1SNP</span></span></code></pre></div>
<pre><code><span><span class="co">## Project: qProject</span></span>
<span><span class="co">##  Options   : maxHits         : 1</span></span>
<span><span class="co">##              paired          : no</span></span>
<span><span class="co">##              splicedAlignment: FALSE</span></span>
<span><span class="co">##              bisulfite       : no</span></span>
<span><span class="co">##              snpFile         : /Users/runner/work/QuasR/Quas.../hg19sub_snp.txt</span></span>
<span><span class="co">##              geneAnnotation  : none</span></span>
<span><span class="co">##  Aligner   : Rbowtie v1.43.0 (parameters: -k 2 --best --strata -v 2)</span></span>
<span><span class="co">##  Genome    : /Users/runner/work/QuasR/QuasR/vignettes/extdata/hg19sub.fa (file)</span></span>
<span><span class="co">## </span></span>
<span><span class="co">##  Reads     : 2 files, 2 samples (fastq format):</span></span>
<span><span class="co">##    1. chip_1_1.fq.bz2  Sample1 (phred33)</span></span>
<span><span class="co">##    2. chip_2_1.fq.bz2  Sample2 (phred33)</span></span>
<span><span class="co">## </span></span>
<span><span class="co">##  Genome alignments: directory: same as reads</span></span>
<span><span class="co">##    1. chip_1_1_10c532caa2f84.bam</span></span>
<span><span class="co">##    2. chip_2_1_10c5359159a04.bam</span></span>
<span><span class="co">## </span></span>
<span><span class="co">##  Aux. alignments: none</span></span></code></pre>
<p>Instead of one count per promoter region and sample,
<code>qCount</code> now returns three (<code>promRegSel</code> was
generated in the ChIP-seq example workflow):</p>
<div class="sourceCode" id="cb217"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="fu"><a href="../reference/qCount.html">qCount</a></span><span class="op">(</span><span class="va">proj1</span>, <span class="va">promRegSel</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## counting alignments...done</span></span></code></pre>
<pre><code><span><span class="co">##                 width Sample1 Sample2</span></span>
<span><span class="co">## ENSG00000176022  1500     157     701</span></span>
<span><span class="co">## ENSG00000186891  1500      22       5</span></span>
<span><span class="co">## ENSG00000186827  1500      10       3</span></span>
<span><span class="co">## ENSG00000078808  1500      73     558</span></span>
<span><span class="co">## ENSG00000171863  1500      94     339</span></span>
<span><span class="co">## ENSG00000252531  1500      59       9</span></span></code></pre>
<div class="sourceCode" id="cb220"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="fu"><a href="../reference/qCount.html">qCount</a></span><span class="op">(</span><span class="va">proj1SNP</span>, <span class="va">promRegSel</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## counting alignments...done</span></span></code></pre>
<pre><code><span><span class="co">##                 width Sample1_R Sample1_U Sample1_A Sample2_R Sample2_U</span></span>
<span><span class="co">## ENSG00000176022  1500         0       133         0         0       559</span></span>
<span><span class="co">## ENSG00000186891  1500         4        16         0         0         5</span></span>
<span><span class="co">## ENSG00000186827  1500         2         8         0         0         2</span></span>
<span><span class="co">## ENSG00000078808  1500         0        59         0         0       432</span></span>
<span><span class="co">## ENSG00000171863  1500         4        78         0         8       263</span></span>
<span><span class="co">## ENSG00000252531  1500         3        50         2         0         6</span></span>
<span><span class="co">##                 Sample2_A</span></span>
<span><span class="co">## ENSG00000176022         0</span></span>
<span><span class="co">## ENSG00000186891         0</span></span>
<span><span class="co">## ENSG00000186827         0</span></span>
<span><span class="co">## ENSG00000078808         0</span></span>
<span><span class="co">## ENSG00000171863         0</span></span>
<span><span class="co">## ENSG00000252531         0</span></span></code></pre>
<p>The example below illustrates use of a <code>snpFile</code> for
Bis-seq experiments. Similarly as for <code>qCount</code>, the count
types are labeled by <code>R</code>, <code>U</code> and <code>A</code>.
These labels are added to the total and methylated column suffixes
<code>_T</code> and <code>_M</code>, resulting in a total of six instead
of two counts per feature and sample:</p>
<div class="sourceCode" id="cb223"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sampleFile</span> <span class="op">&lt;-</span> <span class="st">"extdata/samples_bis_single.txt"</span></span>
<span><span class="va">genomeFile</span> <span class="op">&lt;-</span> <span class="st">"extdata/hg19sub.fa"</span></span>
<span><span class="va">proj4SNP</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/qAlign.html">qAlign</a></span><span class="op">(</span><span class="va">sampleFile</span>, <span class="va">genomeFile</span>,</span>
<span>                   snpFile <span class="op">=</span> <span class="va">snpFile</span>, bisulfite <span class="op">=</span> <span class="st">"dir"</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## alignment files missing - need to:</span></span></code></pre>
<pre><code><span><span class="co">##     create alignment index for the genome</span></span></code></pre>
<pre><code><span><span class="co">##     create 1 genomic alignment(s)</span></span></code></pre>
<pre><code><span><span class="co">## Creating an RbowtieCtoT index for /Users/runner/work/QuasR/QuasR/vignettes/extdata/hg19sub_snp.txt.hg19sub.fa.R.fa</span></span></code></pre>
<pre><code><span><span class="co">## Finished creating index</span></span></code></pre>
<pre><code><span><span class="co">## Creating an RbowtieCtoT index for /Users/runner/work/QuasR/QuasR/vignettes/extdata/hg19sub_snp.txt.hg19sub.fa.A.fa</span></span></code></pre>
<pre><code><span><span class="co">## Finished creating index</span></span></code></pre>
<pre><code><span><span class="co">## Testing the compute nodes...OK</span></span>
<span><span class="co">## Loading QuasR on the compute nodes...preparing to run on 1 nodes...done</span></span>
<span><span class="co">## Available cores:</span></span>
<span><span class="co">## Mac-1698842773251.local: 1</span></span>
<span><span class="co">## Performing genomic alignments for 1 samples. See progress in the log file:</span></span>
<span><span class="co">## /Users/runner/work/QuasR/QuasR/vignettes/QuasR_log_10c537fe6df0c.txt</span></span>
<span><span class="co">## Genomic alignments have been created successfully</span></span></code></pre>
<div class="sourceCode" id="cb232"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="fu"><a href="../reference/qMeth.html">qMeth</a></span><span class="op">(</span><span class="va">proj4SNP</span>, mode <span class="op">=</span> <span class="st">"CpGcomb"</span>, collapseBySample <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## GRanges object with 6 ranges and 6 metadata columns:</span></span>
<span><span class="co">##       seqnames    ranges strand | Sample1_TR Sample1_MR Sample1_TU Sample1_MU</span></span>
<span><span class="co">##          &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt; |  &lt;integer&gt;  &lt;integer&gt;  &lt;integer&gt;  &lt;integer&gt;</span></span>
<span><span class="co">##   [1]     chr1     19-20      * |          0          0          1          1</span></span>
<span><span class="co">##   [2]     chr1     21-22      * |          0          0          1          1</span></span>
<span><span class="co">##   [3]     chr1     54-55      * |          0          0          3          1</span></span>
<span><span class="co">##   [4]     chr1     57-58      * |          0          0          3          0</span></span>
<span><span class="co">##   [5]     chr1     80-81      * |          0          0          6          5</span></span>
<span><span class="co">##   [6]     chr1   103-104      * |          0          0          6          5</span></span>
<span><span class="co">##       Sample1_TA Sample1_MA</span></span>
<span><span class="co">##        &lt;integer&gt;  &lt;integer&gt;</span></span>
<span><span class="co">##   [1]          0          0</span></span>
<span><span class="co">##   [2]          0          0</span></span>
<span><span class="co">##   [3]          0          0</span></span>
<span><span class="co">##   [4]          0          0</span></span>
<span><span class="co">##   [5]          0          0</span></span>
<span><span class="co">##   [6]          0          0</span></span>
<span><span class="co">##   -------</span></span>
<span><span class="co">##   seqinfo: 3 sequences from an unspecified genome</span></span></code></pre>
</div>
</div>
<div class="section level2">
<h2 id="description-of-individual-quasr-functions">Description of Individual <em>QuasR</em> Functions<a class="anchor" aria-label="anchor" href="#description-of-individual-quasr-functions"></a>
</h2>
<p>Please refer to the <em><a href="https://bioconductor.org/packages/3.19/QuasR" class="external-link">QuasR</a></em>
reference manual or the function documentation (e.g. using
<code><a href="../reference/qAlign.html">?qAlign</a></code>) for a complete description of <em><a href="https://bioconductor.org/packages/3.19/QuasR" class="external-link">QuasR</a></em>
functions. The descriptions provided below are meant to give an overview
over all functions and summarize the purpose of each one.</p>
<div class="section level3">
<h3 id="preprocessReads">
<code>preprocessReads</code><a class="anchor" aria-label="anchor" href="#preprocessReads"></a>
</h3>
<p>The <code>preprocessReads</code> function can be used to prepare the
input sequences before alignment to the reference genome, for example to
filter out low quality reads unlikely to produce informative alignments.
When working with paired-end experiments, the paired reads are expected
to be contained in identical order in two separate files. For this
reason, both reads of a pair are filtered out if any of the two reads
fulfills the filtering criteria. The following types of filtering tasks
can be performed (in the order as listed):</p>
<ol style="list-style-type: decimal">
<li>
<strong>Truncate reads</strong>: remove nucleotides from the start
and/or end of each read.<br>
</li>
<li>
<strong>Trim adapters</strong>: remove nucleotides at the beginning
and/or end of each read that match to a defined (adapter) sequence. The
adapter trimming is done by calling <code>trimLRPatterns</code> from the
<em><a href="https://bioconductor.org/packages/3.19/Biostrings" class="external-link">Biostrings</a></em>
package <span class="citation">(Pages et al., n.d.)</span>.<br>
</li>
<li>
<strong>Filter out low quality reads</strong>: Filter out reads that
fulfill any of the filtering criteria (contain more than
<code>nBases</code> <code>N</code> bases, are shorter than
<code>minLength</code> or have a dinucleotide complexity of less than
<code>complexity</code>-times the average complexity of the human genome
sequence).</li>
</ol>
<p>The dinucleotide complexity is calculated in bits as Shannon entropy
using the following formula <span class="math inline">\(-\sum_i f_i
\cdot \log_2 f_i\)</span>, where <span class="math inline">\(f_i\)</span> is the frequency of dinucleotide
<span class="math inline">\(i\)</span> (<span class="math inline">\(i=1,
2, ..., 16\)</span>).</p>
</div>
<div class="section level3">
<h3 id="qAlign">
<code>qAlign</code><a class="anchor" aria-label="anchor" href="#qAlign"></a>
</h3>
<p><code>qAlign</code> is the function that generates alignment files in
<code>bam</code> format, for all input sequence files listed in
<code>sampleFile</code> (see section @ref(sampleFile)), against the
reference genome (<code>genome</code> argument), and for reads that do
not match to the reference genome, against one or several auxiliary
target sequences (<code>auxiliaryFile</code>, see section
@ref(auxiliaryFile)).</p>
<p>The reference genome can be provided either as a <code>fasta</code>
sequence file or as a <code>BSgenome</code> package name (see section
@ref(genome)). If a <code>BSgenome</code> package is not found in the
installed packages, <code>qAlign</code> will abort with a description
how the missing package can be installed from Bioconductor.</p>
<p>The alignment program is set by <code>aligner</code>, and parameters
by <code>maxHits</code>, <code>paired</code>,
<code>splicedAlignment</code> and <code>alignmentParameter</code>.
Currently, <code>aligner</code> can only be set to
<code>"Rbowtie"</code>, which is a wrapper for <em>bowtie</em> <span class="citation">(Langmead et al. 2009)</span> and <em>SpliceMap</em>
<span class="citation">(Au et al. 2010)</span>, or
<code>"Rhisat2"</code>, which is a wrapper for <em>HISAT2</em> <span class="citation">(Kim, Langmead, and Salzberg 2015)</span>. When
<code>aligner="Rbowtie"</code>, <em>SpliceMap</em> will be used if
<code>splicedAlignment=TRUE</code> (not recommended anymore except for
reproducing older analyses). However, it is recommended to create
spliced alignment using
<code>splicedAlignment=TRUE, aligner="Rhisat2"</code>, which will use
the <em>HISAT2</em> aligner and typically leads to more sensistive
alignments and shorter alignment times compared to <em>SpliceMap</em>.
The alignment strategy is furthermore affected by the parameters
<code>snpFile</code> (alignments to variant genomes containing sequence
polymorphisms) and <code>bisulfite</code> (alignment of
bisulfite-converted reads). Finally, <code>clObj</code> can be used to
enable parallelized alignment, sorting and conversion to
<code>bam</code> format.</p>
<p>For each input sequence file listed in <code>sampleFile</code>, one
<code>bam</code> file with alignments to the reference genome will be
generated, and an additional one for each auxiliary sequence file listed
in <code>auxiliaryFile</code>. By default, these <code>bam</code> files
are stored at the same location as the sequence files, unless a
different location is specified under <code>alignmentsDir</code>. If
compatible alignment files are found at this location, they will not be
regenerated, which allows re-use of the same sequencing samples in
multiple analysis projects by listing them in several project-specific
<code>sampleFile</code>s.</p>
<p>The alignment process produces temporary files, such as decompressed
input sequence files or raw alignment files before conversion to
<code>bam</code> format, which can be several times the size of the
input sequence files. These temporary files are stored in the directory
specified by <code>cacheDir</code>, which defaults to the
<strong>R</strong> process temporary directory returned by
<code><a href="https://rdrr.io/r/base/tempfile.html" class="external-link">tempdir()</a></code>. The location of <code><a href="https://rdrr.io/r/base/tempfile.html" class="external-link">tempdir()</a></code> can be
set using environment variables (see <code><a href="https://rdrr.io/r/base/tempfile.html" class="external-link">?tempdir</a></code>).</p>
<p><code>qAlign</code> returns a <code>qProject</code> object that
contains all file names and paths, as well as all alignment parameters
necessary for further analysis (see section @ref(qProject) for methods
to access the information contained in a <code>qProject</code>
object).</p>
</div>
<div class="section level3">
<h3 id="qProject">
<code>qProject</code> class<a class="anchor" aria-label="anchor" href="#qProject"></a>
</h3>
<p>The <code>qProject</code> objects are returned by <code>qAlign</code>
and contain all information about a sequencing experiment needed for
further analysis. It is the main argument passed to the functions that
start with a <em>q</em> letter, such as <code>qCount</code>,
<code>qQCReport</code> and <code>qExportWig</code>. Some information
inside of a <code>qProject</code> object can be accessed by specific
methods (in the examples below, <code>x</code> is a
<code>qProject</code> object):</p>
<ul>
<li>
<code>length(x)</code> gets the number of input files.<br>
</li>
<li>
<code>genome(x)</code> gets the reference genome as a
<code>character(1)</code>. The type of genome is stored as an attribute
in <code>attr(genome(x),"genomeFormat")</code>: <code>"BSgenome"</code>
indicates that <code>genome(x)</code> refers to the name of a
<em>BSgenome</em> package, <code>"file"</code> indicates that it
contains the path and file name of a genome in <code>fasta</code>
format.<br>
</li>
<li>
<code>auxiliaries(x)</code> gets a <code>data.frame</code> with
auxiliary target sequences. The <code>data.frame</code> has one row per
auxiliary target file, and two columns “FileName” and “AuxName”.<br>
</li>
<li>
<code>alignments(x)</code> gets a list with two elements
<code>"genome"</code> and <code>"aux"</code>. <code>"genome"</code>
contains a <code>data.frame</code> with <code>length(x)</code> rows and
two columns <code>"FileName"</code> (containing the path to bam files
with genomic alignments) and <code>"SampleName"</code>.
<code>"aux"</code> contains a <code>data.frame</code> with one row per
auxiliary target file (with auxiliary names as row names), and
<code>length(x)</code> columns (one per input sequence file).<br>
</li>
<li>
<code>x[i]</code> returns a <code>qProject</code> object instance
with <code>i</code> input files, where <code>i</code> can be an
<code>NA</code>-free logical, numeric, or character vector.</li>
</ul>
</div>
<div class="section level3">
<h3 id="qQCReport">
<code>qQCReport</code><a class="anchor" aria-label="anchor" href="#qQCReport"></a>
</h3>
<p>The <code>qQCReport</code> function samples a random subset of
sequences and alignments from each sample or input file and generates a
series of diagnostic plots for estimating data quality. The available
plots vary depending on the types of available input (fasta, fastq, bam
files or <code>qProject</code> object; paired-end or single-end). The
plots below show the currently available plots as produced by the
ChIP-seq example in section @ref(ChIPseq) (except for the fragment size
distributions which are based on an unspliced alignment of paired-end
RNA seq reads):</p>
<ul>
<li><p><strong>Quality score boxplot</strong> shows the distribution of
base quality values as a box plot for each position in the input
sequence. The background color (green, orange or red) indicates ranges
of high, intermediate and low qualities.
<img src="QuasR_files/figure-html/qcplotsFig1-1.png" width="768"></p></li>
<li><p><strong>Nucleotide frequency</strong> plot shows the frequency of
A, C, G, T and N bases by position in the read.
<img src="QuasR_files/figure-html/qcplotsFig2-1.png" width="768"></p></li>
<li><p><strong>Duplication level</strong> plot shows for each sample the
fraction of reads observed at different duplication levels (e.g. once,
two-times, three-times, etc.). In addition, the most frequent sequences
are listed.
<img src="QuasR_files/figure-html/qcplotsFig3-1.png" width="768"></p></li>
<li><p><strong>Mapping statistics</strong> shows fractions of reads that
were (un)mappable to the reference genome.
<img src="QuasR_files/figure-html/qcplotsFig4-1.png" width="768"></p></li>
<li><p><strong>Library complexity</strong> shows fractions of unique
read(-pair) alignment positions. Please note that this measure is not
independent from the total number of reads in a library, and is best
compared between libraries of similar sizes.
<img src="QuasR_files/figure-html/qcplotsFig5-1.png" width="768"></p></li>
<li><p><strong>Mismatch frequency</strong> shows the frequency and
position (relative to the read sequence) of mismatches in the alignments
against the reference genome.
<img src="QuasR_files/figure-html/qcplotsFig6-1.png" width="768"></p></li>
<li><p><strong>Mismatch types</strong> shows the frequency of read bases
that caused mismatches in the alignments to the reference genome,
separately for each genome base.
<img src="QuasR_files/figure-html/qcplotsFig7-1.png" width="768"></p></li>
<li><p><strong>Fragment size</strong> shows the distribution of fragment
sizes inferred from aligned read pairs.<br><img src="QuasR_files/figure-html/qcplotsFig8-1.png" width="768"></p></li>
</ul>
</div>
<div class="section level3">
<h3 id="alignmentStats">
<code>alignmentStats</code><a class="anchor" aria-label="anchor" href="#alignmentStats"></a>
</h3>
<p><code>alignmentStats</code> is comparable to the
<code>idxstats</code> function from Samtools; it returns the size of the
target sequence, as well as the number of mapped and unmapped reads that
are contained in an indexed <code>bam</code> file. The function works
for arguments of type <code>qProject</code>, as well as on a string with
one or several <code>bam</code> file names. There is however a small
difference in the two that is illustrated in the following example,
which uses the <code>qProject</code> object from the ChIP-seq
workflow:</p>
<div class="sourceCode" id="cb234"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># using bam files</span></span>
<span><span class="fu"><a href="../reference/alignmentStats.html">alignmentStats</a></span><span class="op">(</span><span class="fu"><a href="../reference/qProject-class.html">alignments</a></span><span class="op">(</span><span class="va">proj1</span><span class="op">)</span><span class="op">$</span><span class="va">genome</span><span class="op">$</span><span class="va">FileName</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##                            seqlength mapped unmapped</span></span>
<span><span class="co">## chip_1_1_10c5340447364.bam     95000   2339      258</span></span>
<span><span class="co">## chip_2_1_10c534df3cb32.bam     95000   3609      505</span></span></code></pre>
<div class="sourceCode" id="cb236"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/alignmentStats.html">alignmentStats</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/unlist.html" class="external-link">unlist</a></span><span class="op">(</span><span class="fu"><a href="../reference/qProject-class.html">alignments</a></span><span class="op">(</span><span class="va">proj1</span><span class="op">)</span><span class="op">$</span><span class="va">aux</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##                            seqlength mapped unmapped</span></span>
<span><span class="co">## chip_1_1_10c5338b3c97c.bam      5386    251        0</span></span>
<span><span class="co">## chip_2_1_10c53237108f9.bam      5386    493        0</span></span></code></pre>
<div class="sourceCode" id="cb238"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># using a qProject object</span></span>
<span><span class="fu"><a href="../reference/alignmentStats.html">alignmentStats</a></span><span class="op">(</span><span class="va">proj1</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##                 seqlength mapped unmapped</span></span>
<span><span class="co">## Sample1:genome      95000   2339      258</span></span>
<span><span class="co">## Sample2:genome      95000   3609      505</span></span>
<span><span class="co">## Sample1:phiX174      5386    251        7</span></span>
<span><span class="co">## Sample2:phiX174      5386    493       12</span></span></code></pre>
<p>If calling <code>alignmentStats</code> on the bam files directly as
in the first two expressions of the above example, the returned numbers
correspond exactly to what you would obtain by the <code>idxstats</code>
function from Samtools, only that the latter would report them
separately for each target sequence, while <code>alignmentStats</code>
sums them for each <code>bam</code> file. These numbers correctly state
that there are zero unmapped reads in the auxiliary <code>bam</code>
files. However, if calling <code>alignmentStats</code> on a
<code>qProject</code> object, it will report 7 and 12 unmapped reads in
the auxiliary <code>bam</code> files. This is because
<code>alignmentStats</code> is aware that unmapped reads are removed
from auxiliary <code>bam</code> files by <em><a href="https://bioconductor.org/packages/3.19/QuasR" class="external-link">QuasR</a></em>, but
can be calculated from the total number of reads to be aligned to the
auxiliary target, which equals the number of unmapped reads in the
corresponding genomic <code>bam</code> file.</p>
</div>
<div class="section level3">
<h3 id="qExportWig">
<code>qExportWig</code><a class="anchor" aria-label="anchor" href="#qExportWig"></a>
</h3>
<p><code>qExportWig</code> creates fixed-step wig files (see (<a href="http://genome.ucsc.edu/goldenPath/help/wiggle.html" class="external-link uri">http://genome.ucsc.edu/goldenPath/help/wiggle.html</a>) for
format definition) from the genomic alignments contained in a
<code>qProject</code> object. The <code>combine</code> argument controls
if several input files are combined into a single multi-track wig file,
or if they are exported as individual wig files. Alignments of single
read experiments can be shifted towards there 3’-end using
<code>shift</code>; paired-end alignments are automatically shifted by
half the insert size. The resolution of the created wig file is defines
by the <code>binsize</code> argument, and if <code>scaling=TRUE</code>,
multiple alignment files in the <code>qProject</code> object are scaled
by their total number of reads.</p>
</div>
<div class="section level3">
<h3 id="qCount">
<code>qCount</code><a class="anchor" aria-label="anchor" href="#qCount"></a>
</h3>
<p><code>qCount</code> is the workhorse for counting alignments that
overlap query regions. Usage and details on parameters can be obtained
from the <code>qCount</code> function documentation. Two aspects that
are of special importance are also discussed here:</p>
<div class="section level4">
<h4 id="determination-of-overlap">Determination of overlap<a class="anchor" aria-label="anchor" href="#determination-of-overlap"></a>
</h4>
<p>How an alignment overlap with a query region is defined can be
controlled by the following arguments of <code>qCount</code>:</p>
<ul>
<li>
<code>selectReadPosition</code> specifies the read base that serves
as a reference for overlaps with query regions. The alignment position
of that base, eventually after shifting (see below), needs to be
contained in the query region for an overlap.
<code>selectReadPosition</code> can be set to <code>"start"</code> (the
default) or <code>"end"</code> , which refer to the biological start
(5’-end) and end (3’-end) of the read. For example, the
<code>"start"</code> of a read aligned to the plus strand is the
leftmost base in the alignment (the one with the lowest coordinate), and
the <code>"end"</code> of a read aligned to the minus strand is also its
leftmost base in the alignment.<br>
</li>
<li>
<code>shift</code> allows shifting of alignments towards their
3’-end prior to overlap determination and counting. This can be helpful
to increase resolution of ChIP-seq experiments by moving alignments by
half the immuno-precipitated fragment size towards the middle of
fragments. <code>shift</code> can either contain <code>"integer"</code>
values that specify the shift size, or for paired-end experiments, it
can be set to the keyword <code>"halfInsert"</code>, which will estimate
the true fragment size from the distance between aligned read pairs and
shift the alignments accordingly.<br>
</li>
<li>
<code>orientation</code> controls the interpretation of alignment
strand relative to the strand of the query region. The default value
<code>"any"</code> will count all overlapping alignments, irrespective
of the strand. This setting is for example used in an unstranded RNA-seq
experiment where both sense and antisense reads are generated from an
mRNA. A value of <code>"same"</code> will only count the alignments on
the same strand as the query region (e.g. in a stranded RNA-seq
experiment), and <code>"opposite"</code> will only count the alignments
on the opposite strand from the query region (e.g. to quantify
anti-sense transcription in a stranded RNA-seq experiment).<br>
</li>
<li>
<code>useRead</code> only applies to paired-end experiments and
allows to quantify either all alignments (<code>useRead="any"</code>),
or only the first (<code>useRead="first"</code>) or last
(<code>useRead="last"</code>) read from each read pair or read group.
Note that for <code>useRead="any"</code> (the default), an alignment
pair that is fully contained within a query region will contribute two
counts to the value of that region.<br>
</li>
<li>
<code>includeSpliced</code>: When set to <code>FALSE</code>, spliced
alignments will be excluded from the quantification. This could be
useful for example to avoid redundant counting of reads when the spliced
alignments are quantified separately using
<code>reportLevel="junction"</code>.</li>
</ul>
</div>
<div class="section level4">
<h4 id="running-modes-of-qcount">Running modes of <code>qCount</code><a class="anchor" aria-label="anchor" href="#running-modes-of-qcount"></a>
</h4>
<p>The features to be quantified are specified by the <code>query</code>
argument. At the same time, the type of <code>query</code> selects the
mode of quantification. <code>qCount</code> supports three different
types of <code>query</code> arguments and implements three corresponding
quantification types, which primarily differ in the way they deal with
redundancy, such as query bases that are contained in more than one
query region. A fourth quantification mode allows counting of alignments
supporting exon-exon juctions:</p>
<ul>
<li>
<code>GRanges</code> query: Overlapping alignments are counted
separately for each coordinate region in the query object. If multiple
regions have identical names, their counts will be summed, counting each
alignment only once even if it overlaps more than one of these regions.
Alignments may however be counted more than once if they overlap
multiple regions with different names. This mode is for example used to
quantify ChIP-seq alignments in promoter regions (see section
@ref(ChIPseq)), or gene expression levels in an RNA-seq experiment
(using a ‘query’ with exon regions named by gene).<br>
</li>
<li>
<code>GRangesList</code> query: Alignments are counted and summed
for each list element in the query object if they overlap with any of
the regions contained in the list element. The order of the list
elements defines a hierarchy for quantification: Alignment will only be
counted for the first element (the one with the lowest index in the
query) that they overlap, but not for any potential further list
elements containing overlapping regions. This mode can be used to
hierarchically and uniquely count (assign) each alignment to a one of
several groups of regions (the elements in the query), for example to
estimate the fractions of different classes of RNA in an RNA-seq
experiment (rRNA, tRNA, snRNA, snoRNA, mRNA, etc.)<br>
</li>
<li>
<code>TxDb</code> query: Used to extract regions from annotation and
report alignment counts depending on the value of the
<code>reportLevel</code> argument. If <code>reportLevel="exon"</code>,
alignments overlapping each exon in the query are counted. If
<code>reportLevel="gene"</code>, alignment counts for all exons of a
gene will be summed, counting each alignment only once even if it
overlaps multiple annotated exons of a gene. These are useful to
calculate exon or gene expression levels in RNA-seq experiments based on
the annotation in a <code>TxDb</code> object. If
<code>reportLevel="promoter"</code>, the <code>promoters</code> function
from package <em><a href="https://bioconductor.org/packages/3.19/GenomicFeatures" class="external-link">GenomicFeatures</a></em>
is used with default arguments to extract promoter regions around
transcript start sites, e.g. to quantify alignments inf a ChIP-seq
experiment.<br>
</li>
<li>any of the above or <code>NULL</code> for
<code>reportLevel="junction"</code>: The <code>query</code> argument is
ignored if <code>reportLevel</code> is set to <code>"junction"</code>,
and <code>qCount</code> will count the number of alignments supporting
each exon-exon junction detected in any of the samples in
<code>proj</code>. The arguments <code>selectReadPosition</code>,
<code>shift</code>, <code>orientation</code>, <code>useRead</code> and
<code>mask</code> will have no effect in this quantification mode.</li>
</ul>
</div>
</div>
<div class="section level3">
<h3 id="qProfile">
<code>qProfile</code><a class="anchor" aria-label="anchor" href="#qProfile"></a>
</h3>
<p>The <code>qProfile</code> function differs from <code>qCount</code>
in that it returns alignments counts relative to their position in the
query region. Except for <code>upstream</code> and
<code>downstream</code>, the arguments of <code>qProfile</code> and
<code>qCount</code> are the same. This section will describe these two
additional arguments; more details on the other arguments are available
in section @ref(qCount) and from the <code>qProfile</code> function
documentation.</p>
<p>The regions to be profiled are anchored by the biological start
position, which are aligned at position zero in the return value. The
biological start position is defined as <code>start(query)</code> for
regions on the plus strand and <code>end(query)</code> for regions on
the minus strand. The anchor positions are extended to the left and
right sides by the number of bases indicated in the
<code>upstream</code> and <code>downstream</code> arguments.</p>
<ul>
<li>
<code>upstream</code> indicates the number of bases upstream of the
anchor position, which is on the left side of the anchor point for
regions on the plus strand and on the right side for regions on the
minus strand.<br>
</li>
<li>
<code>downstream</code> indicates the number of bases downstream of
the anchor position, which is on the left side of the anchor point for
regions on the plus strand and on the left side for regions on the minus
strand.</li>
</ul>
<p>Be aware that query regions with a <code>*</code> strand are handled
the same way as regions on the plus strand.</p>
</div>
<div class="section level3">
<h3 id="qMeth">
<code>qMeth</code><a class="anchor" aria-label="anchor" href="#qMeth"></a>
</h3>
<p><code>qMeth</code> is used exclusively for Bis-seq experiments. In
contrast to <code>qCount</code>, which counts the number of read
alignments per query region, <code>qMeth</code> quantifies the number of
C and T bases per cytosine in query regions, in order to determine
methylation status.</p>
<p><code>qMeth</code> can be run in one of four modes, controlled by the
<code>mode</code> argument:</p>
<ul>
<li>
<code>CpGcomb</code>: Only C’s in CpG context are considered. It is
assumed that methylation status of the CpG base-pair on both strands is
identical. Therefore, the total and methylated counts obtained for the C
at position <span class="math inline">\(i\)</span> and the C on the
opposite strand at position <span class="math inline">\(i+1\)</span> are
summed.<br>
</li>
<li>
<code>CpG</code>: As with <code>CpGcomb</code>, only C’s in CpG
context are quantified. However, counts from opposite strand are not
summed, resulting in separate output values for C’s on both
strands.<br>
</li>
<li>
<code>allC</code>: All C’s contained in query regions are
quantified, keeping C’s from either strand separate. While this mode
allows quantification of non-CpG methylation, it should be used with
care, as the large result could use up available memory. In that case, a
possible work-around is to divide the region of interest (e.g. the
genome) into several regions (e.g. chromosomes) and call
<code>qMeth</code> separately for each region.<br>
</li>
<li>
<code>var</code>: In this mode, only alignments on the opposite
strand from C’s are analysed in order to collect evidence for sequence
polymorphisms. Methylated C’s are hot-spots for C-to-T transitions,
which in a Bis-seq experiment cannot be discriminated from completely
unmethylated C’s. The information is however contained in alignments to
the G from the opposite strand: Reads containing a G are consistent with
a non-mutated C, and reads with an A support the presence of a sequence
polymorphism. <code>qMeth(..., mode="var")</code> returns counts for
total and matching bases for all C’s on both strands. A low fraction of
matching bases is an indication of a mutation and can be used as a basis
to identify mutated positions in the studied genome relative to the
reference genome. Such positions should not be included in the
quantification of methylation.</li>
</ul>
<p>When using <code>qMeth</code> in a allele-specific quantification
(see also section @ref(alleleSpecificAnalysis), cytosines (or CpGs) that
overlap a sequence polymorphism will not be quantified.</p>
</div>
</div>
<div class="section level2">
<h2 id="session-information">Session information<a class="anchor" aria-label="anchor" href="#session-information"></a>
</h2>
<p>The output in this vignette was produced under:</p>
<pre><code><span><span class="co">## R Under development (unstable) (2023-10-31 r85450)</span></span>
<span><span class="co">## Platform: x86_64-apple-darwin20</span></span>
<span><span class="co">## Running under: macOS Monterey 12.6.9</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Matrix products: default</span></span>
<span><span class="co">## BLAS:   /Library/Frameworks/R.framework/Versions/4.4-x86_64/Resources/lib/libRblas.0.dylib </span></span>
<span><span class="co">## LAPACK: /Library/Frameworks/R.framework/Versions/4.4-x86_64/Resources/lib/libRlapack.dylib;  LAPACK version 3.11.0</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## locale:</span></span>
<span><span class="co">## [1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## time zone: UTC</span></span>
<span><span class="co">## tzcode source: internal</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## attached base packages:</span></span>
<span><span class="co">##  [1] grid      stats4    parallel  stats     graphics  grDevices utils    </span></span>
<span><span class="co">##  [8] datasets  methods   base     </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## other attached packages:</span></span>
<span><span class="co">##  [1] Gviz_1.47.0            GenomicFeatures_1.55.1 AnnotationDbi_1.65.0  </span></span>
<span><span class="co">##  [4] Biobase_2.63.0         Rsamtools_2.19.0       BSgenome_1.71.0       </span></span>
<span><span class="co">##  [7] rtracklayer_1.63.0     BiocIO_1.13.0          Biostrings_2.71.1     </span></span>
<span><span class="co">## [10] XVector_0.43.0         QuasR_1.43.1           Rbowtie_1.43.0        </span></span>
<span><span class="co">## [13] GenomicRanges_1.55.1   GenomeInfoDb_1.39.0    IRanges_2.37.0        </span></span>
<span><span class="co">## [16] S4Vectors_0.41.1       BiocGenerics_0.49.0    BiocStyle_2.31.0      </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## loaded via a namespace (and not attached):</span></span>
<span><span class="co">##   [1] RColorBrewer_1.1-3          rstudioapi_0.15.0          </span></span>
<span><span class="co">##   [3] jsonlite_1.8.7              magrittr_2.0.3             </span></span>
<span><span class="co">##   [5] rmarkdown_2.25              fs_1.6.3                   </span></span>
<span><span class="co">##   [7] zlibbioc_1.49.0             ragg_1.2.6                 </span></span>
<span><span class="co">##   [9] vctrs_0.6.4                 memoise_2.0.1              </span></span>
<span><span class="co">##  [11] RCurl_1.98-1.12             base64enc_0.1-3            </span></span>
<span><span class="co">##  [13] htmltools_0.5.6.1           S4Arrays_1.3.0             </span></span>
<span><span class="co">##  [15] progress_1.2.2              curl_5.1.0                 </span></span>
<span><span class="co">##  [17] SparseArray_1.3.0           Formula_1.2-5              </span></span>
<span><span class="co">##  [19] sass_0.4.7                  bslib_0.5.1                </span></span>
<span><span class="co">##  [21] htmlwidgets_1.6.2           desc_1.4.2                 </span></span>
<span><span class="co">##  [23] cachem_1.0.8                GenomicAlignments_1.39.0   </span></span>
<span><span class="co">##  [25] igraph_1.5.1                lifecycle_1.0.3            </span></span>
<span><span class="co">##  [27] pkgconfig_2.0.3             Matrix_1.6-1.1             </span></span>
<span><span class="co">##  [29] R6_2.5.1                    fastmap_1.1.1              </span></span>
<span><span class="co">##  [31] GenomeInfoDbData_1.2.11     MatrixGenerics_1.15.0      </span></span>
<span><span class="co">##  [33] digest_0.6.33               colorspace_2.1-0           </span></span>
<span><span class="co">##  [35] ShortRead_1.61.0            rprojroot_2.0.3            </span></span>
<span><span class="co">##  [37] textshaping_0.3.7           Hmisc_5.1-1                </span></span>
<span><span class="co">##  [39] RSQLite_2.3.2               hwriter_1.3.2.1            </span></span>
<span><span class="co">##  [41] filelock_1.0.2              fansi_1.0.5                </span></span>
<span><span class="co">##  [43] httr_1.4.7                  abind_1.4-5                </span></span>
<span><span class="co">##  [45] compiler_4.4.0              bit64_4.0.5                </span></span>
<span><span class="co">##  [47] backports_1.4.1             htmlTable_2.4.2            </span></span>
<span><span class="co">##  [49] BiocParallel_1.37.0         DBI_1.1.3                  </span></span>
<span><span class="co">##  [51] highr_0.10                  biomaRt_2.59.0             </span></span>
<span><span class="co">##  [53] Rhisat2_1.19.0              rappdirs_0.3.3             </span></span>
<span><span class="co">##  [55] DelayedArray_0.29.0         rjson_0.2.21               </span></span>
<span><span class="co">##  [57] tools_4.4.0                 foreign_0.8-85             </span></span>
<span><span class="co">##  [59] nnet_7.3-19                 glue_1.6.2                 </span></span>
<span><span class="co">##  [61] restfulr_0.0.15             checkmate_2.3.0            </span></span>
<span><span class="co">##  [63] cluster_2.1.4               generics_0.1.3             </span></span>
<span><span class="co">##  [65] gtable_0.3.4                ensembldb_2.27.0           </span></span>
<span><span class="co">##  [67] data.table_1.14.8           hms_1.1.3                  </span></span>
<span><span class="co">##  [69] xml2_1.3.5                  utf8_1.2.4                 </span></span>
<span><span class="co">##  [71] pillar_1.9.0                stringr_1.5.0              </span></span>
<span><span class="co">##  [73] dplyr_1.1.3                 BiocFileCache_2.11.1       </span></span>
<span><span class="co">##  [75] lattice_0.22-5              bit_4.0.5                  </span></span>
<span><span class="co">##  [77] deldir_1.0-9                biovizBase_1.51.0          </span></span>
<span><span class="co">##  [79] tidyselect_1.2.0            knitr_1.45                 </span></span>
<span><span class="co">##  [81] gridExtra_2.3               bookdown_0.36              </span></span>
<span><span class="co">##  [83] ProtGenerics_1.35.0         SummarizedExperiment_1.33.0</span></span>
<span><span class="co">##  [85] xfun_0.40                   matrixStats_1.0.0          </span></span>
<span><span class="co">##  [87] stringi_1.7.12              lazyeval_0.2.2             </span></span>
<span><span class="co">##  [89] yaml_2.3.7                  evaluate_0.22              </span></span>
<span><span class="co">##  [91] codetools_0.2-19            interp_1.1-4               </span></span>
<span><span class="co">##  [93] GenomicFiles_1.39.0         tibble_3.2.1               </span></span>
<span><span class="co">##  [95] BiocManager_1.30.22         cli_3.6.1                  </span></span>
<span><span class="co">##  [97] rpart_4.1.21                systemfonts_1.0.5          </span></span>
<span><span class="co">##  [99] munsell_0.5.0               jquerylib_0.1.4            </span></span>
<span><span class="co">## [101] dichromat_2.0-0.1           Rcpp_1.0.11                </span></span>
<span><span class="co">## [103] dbplyr_2.4.0                png_0.1-8                  </span></span>
<span><span class="co">## [105] RUnit_0.4.32                XML_3.99-0.14              </span></span>
<span><span class="co">## [107] pkgdown_2.0.7               ggplot2_3.4.4              </span></span>
<span><span class="co">## [109] blob_1.2.4                  prettyunits_1.2.0          </span></span>
<span><span class="co">## [111] latticeExtra_0.6-30         jpeg_0.1-10                </span></span>
<span><span class="co">## [113] AnnotationFilter_1.27.0     SGSeq_1.37.0               </span></span>
<span><span class="co">## [115] bitops_1.0-7                VariantAnnotation_1.49.0   </span></span>
<span><span class="co">## [117] scales_1.2.1                purrr_1.0.2                </span></span>
<span><span class="co">## [119] crayon_1.5.2                rlang_1.1.1                </span></span>
<span><span class="co">## [121] KEGGREST_1.43.0</span></span></code></pre>
</div>
<div class="section level2">
<h2 class="unnumbered" id="references">References<a class="anchor" aria-label="anchor" href="#references"></a>
</h2>
<div id="refs" class="references csl-bib-body hanging-indent">
<div id="ref-DESeq" class="csl-entry">
Anders, Simon, and Wolfgang Huber. 2010. <span>“Differential Expression
Analysis for Sequence Count Data.”</span> <em>Genome Biology</em> 11:
R106. <a href="https://doi.org/10.1186/gb-2010-11-10-r106" class="external-link">https://doi.org/10.1186/gb-2010-11-10-r106</a>.
</div>
<div id="ref-DEXSeq" class="csl-entry">
Anders, Simon, Alejandro Reyes, and Wolfgang Huber. 2012.
<span>“Detecting Differential Usage of Exons from <span>RNA</span>-Seq
Data.”</span> <em>Genome Research</em> 22: 2008–17. <a href="https://doi.org/10.1101/gr.133744.111" class="external-link">https://doi.org/10.1101/gr.133744.111</a>.
</div>
<div id="ref-SpliceMap" class="csl-entry">
Au, K. F., H. Jiang, L. Lin, Y. Xing, and W. H. Wong. 2010.
<span>“Detection of Splice Junctions from Paired-End
<span>RNA</span>-Seq Data by <span>SpliceMap</span>.”</span> <em>Nucleic
Acids Research</em> 38 (14): 4570–78.
</div>
<div id="ref-Dalgaard" class="csl-entry">
Dalgaard, P. 2002. <em>Introductory <span>S</span>tatistics with
<span>R</span></em>. Springer.
</div>
<div id="ref-QuasR" class="csl-entry">
Gaidatzis, D., A. Lerch, F. Hahne, and M. B. Stadler. 2015.
<span>“<span>QuasR</span>: <span>Q</span>uantify and
<span>A</span>nnotate <span>S</span>hort <span>R</span>eads in
<span>R</span>.”</span> <em>Bioinformatics</em> 31 (7): 1130–32. <a href="https://doi.org/10.1093/bioinformatics/btu781" class="external-link">https://doi.org/10.1093/bioinformatics/btu781</a>.
</div>
<div id="ref-Rbowtie" class="csl-entry">
Hahne, F., A. Lerch, and M. B. Stadler. 2012. <span>“Rbowtie: A
<span>R</span> Wrapper for Bowtie and <span>SpliceMap</span> Short Read
Aligners.”</span>
</div>
<div id="ref-baySeq" class="csl-entry">
Hardcastle, Thomas J, and Krystyna A Kelly. 2010. <span>“<span class="nocase">baySeq</span>: Empirical Bayesian Methods for Identifying
Differential Expression in Sequence Count Data.”</span> <em>BMC
Bioinformatics</em> 11: 422. <a href="https://doi.org/10.1186/1471-2105-11-422" class="external-link">https://doi.org/10.1186/1471-2105-11-422</a>.
</div>
<div id="ref-hisat2" class="csl-entry">
Kim, D., B. Langmead, and S. L. Salzberg. 2015. <span>“HISAT: A Fast
Spliced Aligner with Low Memory Requirements.”</span> <em>Nature
Methods</em> 12: 357–60. <a href="https://doi.org/10.1038/nmeth.3317" class="external-link">https://doi.org/10.1038/nmeth.3317</a>.
</div>
<div id="ref-bowtie" class="csl-entry">
Langmead, B., C. Trapnell, M. Pop, and S. L. Salzberg. 2009.
<span>“Ultrafast and Memory-Efficient Alignment of Short
<span>DNA</span> Sequences to the Human Genome.”</span> <em>Genome
Biology</em> 10 (3): R25.
</div>
<div id="ref-IRanges" class="csl-entry">
Lawrence, Michael, Wolfgang Huber, Hervé Pagès, Patrick Aboyoun, Marc
Carlson, Robert Gentleman, Martin T. Morgan, and Vincent J. Carey. 2013.
<span>“Software for <span>C</span>omputing and <span>A</span>nnotating
<span>G</span>enomic <span>R</span>anges.”</span> <em>PLoS Comput
Biol</em> 9: e1003118. <a href="https://doi.org/doi:10.1371/journal.pcbi.1003118" class="external-link">https://doi.org/doi:10.1371/journal.pcbi.1003118</a>.
</div>
<div id="ref-DESeq2" class="csl-entry">
Love, M. I., W. Huber, and S. Anders. 2014. <span>“Moderated Estimation
of Fold Change and Dispersion for <span class="nocase">RNA-seq</span>
Data with <span>DESeq2</span>.”</span> <em>Genome Biology</em> 15: 550.
<a href="https://doi.org/10.1186/s13059-014-0550-8" class="external-link">https://doi.org/10.1186/s13059-014-0550-8</a>.
</div>
<div id="ref-ShortRead" class="csl-entry">
Morgan, Martin, Simon Anders, Michael Lawrence, Patrick Aboyoun, Hervé
Pagès, and Robert Gentleman. 2009. <span>“<span>ShortRead</span>: A
Bioconductor Package for Input, Quality Assessment and Exploration of
High-Throughput Sequence Data.”</span> <em>Bioinformatics</em> 25:
2607–8. <a href="https://doi.org/10.1093/bioinformatics/btp450" class="external-link">https://doi.org/10.1093/bioinformatics/btp450</a>.
</div>
<div id="ref-Biostrings" class="csl-entry">
Pages, H., P. Aboyoun, R. Gentleman, and S. DebRoy. n.d. <em>Biostrings:
<span>S</span>tring Objects Representing Biological Sequences, and
Matching Algorithms</em>.
</div>
<div id="ref-edgeR" class="csl-entry">
Robinson, Mark D, Davis J McCarthy, and Gordon K Smyth. 2010.
<span>“<span class="nocase">edgeR</span>: A Bioconductor Package for
Differential Expression Analysis of Digital Gene Expression
Data.”</span> <em>Bioinformatics</em> 26: 139–40.
</div>
<div id="ref-TCC" class="csl-entry">
Sun, Jianqiang, Tomoaki Nishiyama, Kentaro Shimizu1, and Koji Kadota.
2013. <span>“<span>TCC</span>: An <span>R</span> Package for Comparing
Tag Count Data with Robust Normalization Strategies.”</span> <em>BMC
Bioinformatics</em> 14: 219. <a href="https://doi.org/doi:10.1186/1471-2105-14-219" class="external-link">https://doi.org/doi:10.1186/1471-2105-14-219</a>.
</div>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Anita Lerch, Adam Alexander Thil SMITH, Charlotte Soneson, Dimos Gaidatzis, Michael Stadler.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
