<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>An introduction to QuasR • QuasR</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="An introduction to QuasR">
<meta property="og:description" content="QuasR">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">QuasR</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">1.29.6</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fas fa fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../articles/QuasR.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right"></ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><script src="QuasR_files/accessible-code-block-0.0.1/empty-anchor.js"></script><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>An introduction to QuasR</h1>
                        <h4 class="author">Michael Stadler</h4>
                        <h4 class="author">Dimos Gaidatzis</h4>
                        <h4 class="author">Charlotte Soneson</h4>
                        <h4 class="author">Anita Lerch</h4>
            
            <h4 class="date">12 October, 2020</h4>
      
      
      <div class="hidden name"><code>QuasR.Rmd</code></div>

    </div>

    
    
<!--
\DeclareRobustCommand{\QuasRlogo}{\raisebox{-8mm}{\includegraphics[width=45mm]{QuasR-logo.pdf}}}% use \DeclareRobustCommand instead of \newcommand to avoid premature expansion of \QuasRlogo within by 'titling' package
-->
<p><img src="QuasR-logo.png" style="position:absolute;top:0px;right:0px;"></p>
<!--

-->
<div id="introduction" class="section level1">
<h1 class="hasAnchor">
<a href="#introduction" class="anchor"></a>Introduction</h1>
<p>The <em><a href="https://bioconductor.org/packages/3.12/QuasR">QuasR</a></em> package (short for <i>Qu</i>antify and <i>a</i>nnotate <i>s</i>hort reads in <i>R</i>) integrates the functionality of several <strong>R</strong> packages (such as <em><a href="https://bioconductor.org/packages/3.12/IRanges">IRanges</a></em> <span class="citation">(Lawrence et al. 2013)</span> and <em><a href="https://bioconductor.org/packages/3.12/Rsamtools">Rsamtools</a></em>) and external software (e.g. <code>bowtie</code>, through the <em><a href="https://bioconductor.org/packages/3.12/Rbowtie">Rbowtie</a></em> package, and <code>HISAT2</code>, through the <em><a href="https://bioconductor.org/packages/3.12/Rhisat2">Rhisat2</a></em> package). The package aims to cover the whole analysis workflow of typical high throughput sequencing experiments, starting from the raw sequence reads, over pre-processing and alignment, up to quantification. A single <strong>R</strong> script can contain all steps of a complete analysis, making it simple to document, reproduce or share the workflow containing all relevant details.</p>
<p>The current <em><a href="https://bioconductor.org/packages/3.12/QuasR">QuasR</a></em> release supports the analysis of single read and paired-end ChIP-seq (chromatin immuno-precipitation combined with sequencing), RNA-seq (gene expression profiling by sequencing of RNA) and Bis-seq (measurement of DNA methylation by sequencing of bisulfite-converted genomic DNA) experiments. It has been successfully used with data from Illumina, 454 Life Technologies and SOLiD sequencers, the latter by using bam files created externally of <em><a href="https://bioconductor.org/packages/3.12/QuasR">QuasR</a></em>.</p>
</div>
<div id="preliminaries" class="section level1">
<h1 class="hasAnchor">
<a href="#preliminaries" class="anchor"></a>Preliminaries</h1>
<div id="citing-quasr" class="section level2">
<h2 class="hasAnchor">
<a href="#citing-quasr" class="anchor"></a>Citing <em><a href="https://bioconductor.org/packages/3.12/QuasR">QuasR</a></em>
</h2>
<p>If you use <em><a href="https://bioconductor.org/packages/3.12/QuasR">QuasR</a></em> <span class="citation">(Gaidatzis et al. 2015)</span> in your work, you can cite it as follows:</p>
<div class="sourceCode" id="cb1"><pre class="downlit">
<span class="fu"><a href="https://rdrr.io/r/utils/citation.html">citation</a></span><span class="op">(</span><span class="st">"QuasR"</span><span class="op">)</span></pre></div>
<pre><code>## 
## Please use the QuasR reference below to cite the software itself. If
## you were using qAlign with Rbowtie as aligner, it can be cited as
## Langmead et al. (2009) (unspliced alignments) or Au et al. (2010)
## (spliced alignments). If you were using qAlign with Rhisat2 as aligner,
## it can be cited as Kim et al. (2015).
## 
##   Gaidatzis D, Lerch A, Hahne F, Stadler MB. QuasR: Quantification and
##   annotation of short reads in R. Bioinformatics 31(7):1130-1132
##   (2015).
## 
##   Langmead B, Trapnell C, Pop M, Salzberg SL. Ultrafast and
##   memory-efficient alignment of short DNA sequences to the human
##   genome. Genome Biology 10(3):R25 (2009).
## 
##   Au KF, Jiang H, Lin L, Xing Y, Wong WH. Detection of splice junctions
##   from paired-end RNA-seq data by SpliceMap. Nucleic Acids Research,
##   38(14):4570-8 (2010).
## 
##   Kim D, Langmead B, Salzberg SL. HISAT: a fast spliced aligner with
##   low memory requirements. Nat Methods, 12(4):357-60 (2015).
## 
## This free open-source software implements academic research by the
## authors and co-workers. If you use it, please support the project by
## citing the appropriate journal articles.
## To see these entries in BibTeX format, use 'print(&lt;citation&gt;,
## bibtex=TRUE)', 'toBibtex(.)', or set
## 'options(citation.bibtex.max=999)'.</code></pre>
</div>
<div id="installation" class="section level2">
<h2 class="hasAnchor">
<a href="#installation" class="anchor"></a>Installation</h2>
<p><em><a href="https://bioconductor.org/packages/3.12/QuasR">QuasR</a></em> is a package for the <strong>R</strong> computing environment and it is assumed that you have already installed <strong>R</strong>. See the <strong>R</strong> project at (<a href="http://www.r-project.org" class="uri">http://www.r-project.org</a>). To install the latest version of <em><a href="https://bioconductor.org/packages/3.12/QuasR">QuasR</a></em>, you will need to be using the latest version of <strong>R</strong>. <em><a href="https://bioconductor.org/packages/3.12/QuasR">QuasR</a></em> is part of the Bioconductor project at (<a href="http://www.bioconductor.org" class="uri">http://www.bioconductor.org</a>). To get <em><a href="https://bioconductor.org/packages/3.12/QuasR">QuasR</a></em> together with its dependencies you can use</p>
<div class="sourceCode" id="cb3"><pre class="downlit">
<span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="kw"><a href="https://rdrr.io/r/base/library.html">require</a></span><span class="op">(</span><span class="st">"BiocManager"</span><span class="op">)</span><span class="op">)</span>
    <span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html">install.packages</a></span><span class="op">(</span><span class="st">"BiocManager"</span><span class="op">)</span>
<span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/BiocManager/man/install.html">install</a></span><span class="op">(</span><span class="st">"QuasR"</span><span class="op">)</span></pre></div>
<p>Bioconductor works on a 6-monthly official release cycle. As with other Bioconductor packages, there are always two versions of <em><a href="https://bioconductor.org/packages/3.12/QuasR">QuasR</a></em>. Most users will use the current official release version, which will be installed by <code><a href="https://rdrr.io/pkg/BiocManager/man/install.html">BiocManager::install</a></code> if you are using the current release version of <strong>R</strong>. There is also a development version of <em><a href="https://bioconductor.org/packages/3.12/QuasR">QuasR</a></em> that includes new features due for the next official release. The development version will be installed if you are using the development version of <strong>Bioconductor</strong> (see <code>version = "devel"</code> in <em><a href="https://bioconductor.org/packages/3.12/BiocManager">BiocManager</a></em>). The official release version always has an even second number (for example 1.20.1), whereas the developmental version has an odd second number (for example 1.21.4).</p>
</div>
<div id="loading-of-quasr-and-other-required-packages" class="section level2">
<h2 class="hasAnchor">
<a href="#loading-of-quasr-and-other-required-packages" class="anchor"></a>Loading of <em>QuasR</em> and other required packages</h2>
<p>In order to run the code examples in this vignette, the <em><a href="https://bioconductor.org/packages/3.12/QuasR">QuasR</a></em> package and a few additional packages need to be loaded:</p>
<div class="sourceCode" id="cb4"><pre class="downlit">
<span class="fu"><a href="https://rdrr.io/r/base/message.html">suppressPackageStartupMessages</a></span><span class="op">(</span><span class="op">{</span>
    <span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">QuasR</span><span class="op">)</span>
    <span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">BSgenome</span><span class="op">)</span>
    <span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="http://bioconductor.org/packages/Rsamtools">Rsamtools</a></span><span class="op">)</span>
    <span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">rtracklayer</span><span class="op">)</span>
    <span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">GenomicFeatures</span><span class="op">)</span>
    <span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/ivanek/Gviz">Gviz</a></span><span class="op">)</span>
<span class="op">}</span><span class="op">)</span></pre></div>
<pre><code>## Warning: multiple methods tables found for 'rowRanges'</code></pre>
</div>
<div id="how-to-get-help" class="section level2">
<h2 class="hasAnchor">
<a href="#how-to-get-help" class="anchor"></a>How to get help</h2>
<p>Most questions about <em><a href="https://bioconductor.org/packages/3.12/QuasR">QuasR</a></em> will hopefully be answered by the documentation or references. If you’ve run into a question which isn’t addressed by the documentation, or you’ve found a conflict between the documentation and software itself, then there is an active support community which can offer help.</p>
<p>The authors of the package (maintainer: Michael Stadler <a href="mailto:michael.stadler@fmi.ch" class="email">michael.stadler@fmi.ch</a>) always appreciate receiving reports of bugs in the package functions or in the documentation. The same goes for well-considered suggestions for improvements.</p>
<p>Any other questions or problems concerning <em><a href="https://bioconductor.org/packages/3.12/QuasR">QuasR</a></em> should be posted to the Bioconductor support site (<a href="https://support.bioconductor.org" class="uri">https://support.bioconductor.org</a>). Users posting to the support site for the first time should read the helpful posting guide at (<a href="https://support.bioconductor.org/info/faq/" class="uri">https://support.bioconductor.org/info/faq/</a>). Note that each function in <em><a href="https://bioconductor.org/packages/3.12/QuasR">QuasR</a></em> has it’s own help page, as described in the section @ref(introToR). Posting etiquette requires that you read the relevant help page carefully before posting a problem to the site.</p>
</div>
</div>
<div id="quick-start" class="section level1">
<h1 class="hasAnchor">
<a href="#quick-start" class="anchor"></a>Quick Start</h1>
<div id="introToR" class="section level2">
<h2 class="hasAnchor">
<a href="#introToR" class="anchor"></a>A brief introduction to <strong>R</strong>
</h2>
<p>If you already use <strong>R</strong> and know about its command line interface, just skip this section and continue with section @ref(sampleQuasRsession).</p>
<p>The structure of this vignette and in particular this section is based on the excellent user guide of the <em><a href="https://bioconductor.org/packages/3.12/limma">limma</a></em> package, which we would like to hereby acknowledge. <strong>R</strong> is a program for statistical computing. It is a command-driven language meaning that you have to type commands into it rather than pointing and clicking using a mouse. In this guide it will be assumed that you have successfully downloaded and installed <strong>R</strong> from (<a href="http://www.r-project.org" class="uri">http://www.r-project.org</a>) as well as <em><a href="https://bioconductor.org/packages/3.12/QuasR">QuasR</a></em> (see section @ref(installation)). A good way to get started is to type</p>
<div class="sourceCode" id="cb6"><pre class="downlit">
<span class="fu"><a href="https://rdrr.io/r/utils/help.start.html">help.start</a></span><span class="op">(</span><span class="op">)</span></pre></div>
<p>at the <strong>R</strong> prompt or, if you’re using <strong>R</strong> for Windows, to follow the drop-down menu items <em>Help</em> <span class="math inline">\(\succ\)</span> <em>Html help</em>. Following the links <em>Packages</em> <span class="math inline">\(\succ\)</span> <em><a href="https://bioconductor.org/packages/3.12/QuasR">QuasR</a></em> from the html help page will lead you to the contents page of help topics for functions in <em><a href="https://bioconductor.org/packages/3.12/QuasR">QuasR</a></em>.</p>
<p>Before you can use any <em><a href="https://bioconductor.org/packages/3.12/QuasR">QuasR</a></em> commands you have to load the package by typing</p>
<div class="sourceCode" id="cb7"><pre class="downlit">
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">QuasR</span><span class="op">)</span></pre></div>
<p>at the <strong>R</strong> prompt. You can get help on any function in any loaded package by typing <code><a href="https://rdrr.io/r/utils/Question.html">?</a></code> and the function name at the <strong>R</strong> prompt, for example</p>
<div class="sourceCode" id="cb8"><pre class="downlit">
<span class="op">?</span><span class="va">preprocessReads</span></pre></div>
<p>or equivalently</p>
<div class="sourceCode" id="cb9"><pre class="downlit">
<span class="fu"><a href="https://rdrr.io/r/utils/help.html">help</a></span><span class="op">(</span><span class="st">"preprocessReads"</span><span class="op">)</span></pre></div>
<p>for detailed help on the <code>preprocessReads</code> function. The individual function help pages are especially important for listing all the arguments which a function will accept and what values the arguments can take.</p>
<p>A key to understanding <strong>R</strong> is to appreciate that anything that you create in <strong>R</strong> is an <em>object</em>. Objects might include data sets, variables, functions, anything at all. For example</p>
<div class="sourceCode" id="cb10"><pre class="downlit">
<span class="va">x</span> <span class="op">&lt;-</span> <span class="fl">2</span></pre></div>
<p>will create a variable <code>x</code> and will assign it the value 2. At any stage of your <strong>R</strong> session you can type</p>
<div class="sourceCode" id="cb11"><pre class="downlit">
<span class="fu"><a href="https://rdrr.io/r/base/ls.html">ls</a></span><span class="op">(</span><span class="op">)</span></pre></div>
<p>to get a list of all the objects you have created. You can see the contents of any object by typing the name of the object at the prompt. The following command will print out the contents of <code>x</code>:</p>
<div class="sourceCode" id="cb12"><pre class="downlit">
<span class="va">x</span></pre></div>
<p>We hope that you can use <em><a href="https://bioconductor.org/packages/3.12/QuasR">QuasR</a></em> without having to spend a lot of time learning about the <strong>R</strong> language itself but a little knowledge in this direction will be very helpful, especially when you want to do something not explicitly provided for in <em><a href="https://bioconductor.org/packages/3.12/QuasR">QuasR</a></em> or in the other Bioconductor packages. For more details about the <strong>R</strong> language see <em>An Introduction to R</em> which is available from the online help. For more background on using <strong>R</strong> for statistical analysis see <span class="citation">(Dalgaard 2002)</span>.</p>
</div>
<div id="sampleQuasRsession" class="section level2">
<h2 class="hasAnchor">
<a href="#sampleQuasRsession" class="anchor"></a>Sample <em>QuasR</em> session</h2>
<p>This is a quick overview of what an analysis could look like for users preferring to jump right into an analysis. The example uses data that is provided with the <em><a href="https://bioconductor.org/packages/3.12/QuasR">QuasR</a></em> package, which is first copied to the current working directory, into a subfolder called <code>"extdata"</code>:</p>
<div class="sourceCode" id="cb13"><pre class="downlit">
<span class="fu"><a href="https://rdrr.io/r/base/files.html">file.copy</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span><span class="op">(</span>package<span class="op">=</span><span class="st">"QuasR"</span>, <span class="st">"extdata"</span><span class="op">)</span>, <span class="st">"."</span>, recursive<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span></pre></div>
<pre><code>## [1] TRUE</code></pre>
<p>The sequence files to be analyzed are listed in <code>sampleFile</code> (see section @ref(sampleFile) for details). The sequence reads will be aligned using <code>bowtie</code> <span class="citation">(Langmead et al. 2009)</span> (from the <em><a href="https://bioconductor.org/packages/3.12/Rbowtie">Rbowtie</a></em> package <span class="citation">(Hahne, Lerch, and Stadler 2012)</span>) to a small reference genome (consisting of three short segments from the hg19 human genome assembly, available in full for example in the <em><a href="https://bioconductor.org/packages/3.12/BSgenome.Hsapiens.UCSC.hg19">BSgenome.Hsapiens.UCSC.hg19</a></em> package). Make sure that you have sufficient disk space, both in your <strong>R</strong> temporary directory (<code><a href="https://rdrr.io/r/base/tempfile.html">tempdir()</a></code>) as well as to store the resulting alignments (see section @ref(qAlign)).</p>
<div class="sourceCode" id="cb15"><pre class="downlit">
<span class="va">sampleFile</span> <span class="op">&lt;-</span> <span class="st">"extdata/samples_chip_single.txt"</span>
<span class="va">genomeFile</span> <span class="op">&lt;-</span> <span class="st">"extdata/hg19sub.fa"</span>

<span class="va">proj</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/qAlign.html">qAlign</a></span><span class="op">(</span><span class="va">sampleFile</span>, <span class="va">genomeFile</span><span class="op">)</span></pre></div>
<pre><code>## Creating .fai file for: /Users/runner/work/QuasR/QuasR/vignettes/extdata/hg19sub.fa</code></pre>
<pre><code>## alignment files missing - need to:</code></pre>
<pre><code>##     create alignment index for the genome</code></pre>
<pre><code>##     create 2 genomic alignment(s)</code></pre>
<pre><code>## Creating an Rbowtie index for /Users/runner/work/QuasR/QuasR/vignettes/extdata/hg19sub.fa</code></pre>
<pre><code>## Finished creating index</code></pre>
<pre><code>## Testing the compute nodes...OK
## Loading QuasR on the compute nodes...OK
## Available cores:
## Mac-1602493909147.local: 1
## Performing genomic alignments for 2 samples. See progress in the log file:
## /Users/runner/work/QuasR/QuasR/vignettes/QuasR_log_17d55137ec1e.txt
## Genomic alignments have been created successfully</code></pre>
<div class="sourceCode" id="cb23"><pre class="downlit">
<span class="va">proj</span></pre></div>
<pre><code>## Project: qProject
##  Options   : maxHits         : 1
##              paired          : no
##              splicedAlignment: FALSE
##              bisulfite       : no
##              snpFile         : none
##              geneAnnotation  : none
##  Aligner   : Rbowtie v1.29.1 (parameters: -m 1 --best --strata)
##  Genome    : /Users/runner/work/QuasR/QuasR/vignettes/extdata/hg19sub.fa (file)
## 
##  Reads     : 2 files, 2 samples (fastq format):
##    1. chip_1_1.fq.bz2  Sample1 (phred33)
##    2. chip_2_1.fq.bz2  Sample2 (phred33)
## 
##  Genome alignments: directory: same as reads
##    1. chip_1_1_17d52e6ecf3a.bam
##    2. chip_2_1_17d56ce700a6.bam
## 
##  Aux. alignments: none</code></pre>
<p>The <code>proj</code> object keeps track of all the information of a sequencing experiment, for example where sequence and alignment files are stored, and what aligner and reference genome was used to generate the alignments.</p>
<p>Now that the alignments have been generated, further analyses can be performed. A quality control report is saved to the <code>"extdata/qc_report.pdf"</code> file using the <code>qQCReport</code> function.</p>
<div class="sourceCode" id="cb25"><pre class="downlit">
<span class="fu"><a href="../reference/qQCReport.html">qQCReport</a></span><span class="op">(</span><span class="va">proj</span>, <span class="st">"extdata/qc_report.pdf"</span><span class="op">)</span></pre></div>
<pre><code>## collecting quality control data</code></pre>
<pre><code>## creating QC plots</code></pre>
<p>The number of alignments per promoter region is quantified using <code>qCount</code>. Genomic coordinates for promoter regions are imported from a gtf file (<code>annotFile</code>) into the <code>GRanges</code>-object with the name <code>promReg</code>:</p>
<div class="sourceCode" id="cb28"><pre class="downlit">
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">rtracklayer</span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">GenomicFeatures</span><span class="op">)</span>
<span class="va">annotFile</span> <span class="op">&lt;-</span> <span class="st">"extdata/hg19sub_annotation.gtf"</span>
<span class="va">txStart</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/rtracklayer/man/GFFFile-class.html">import.gff</a></span><span class="op">(</span><span class="va">annotFile</span>, format<span class="op">=</span><span class="st">"gtf"</span>, feature.type<span class="op">=</span><span class="st">"start_codon"</span><span class="op">)</span>
<span class="va">promReg</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/GenomicFeatures/man/transcripts.html">promoters</a></span><span class="op">(</span><span class="va">txStart</span>, upstream<span class="op">=</span><span class="fl">500</span>, downstream<span class="op">=</span><span class="fl">500</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">promReg</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu">mcols</span><span class="op">(</span><span class="va">promReg</span><span class="op">)</span><span class="op">$</span><span class="va">transcript_name</span>

<span class="va">promCounts</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/qCount.html">qCount</a></span><span class="op">(</span><span class="va">proj</span>, query<span class="op">=</span><span class="va">promReg</span><span class="op">)</span></pre></div>
<pre><code>## counting alignments...done</code></pre>
<div class="sourceCode" id="cb30"><pre class="downlit">
<span class="va">promCounts</span></pre></div>
<pre><code>##                width Sample1 Sample2
## TNFRSF18-003    1000      20       4
## TNFRSF18-002    1000      20       4
## TNFRSF18-001    1000      20       4
## TNFRSF4-001     1000       5       2
## SDF4-007        1000       8       2
## SDF4-001        1000       8       2
## SDF4-002        1000       8       2
## SDF4-201        1000       8       2
## B3GALT6-001     1000      25     274
## RPS7-001        1000     121     731
## RPS7-008        1000     121     731
## RPS7-009        1000     121     731
## RPS7-005        1000     121     731
## C3orf10-201     1000     176     496
## C3orf10-001     1000     176     496
## AC034193.1-201  1000       5       2
## VHL-001         1000      61     336
## VHL-002         1000      61     336
## VHL-201         1000      61     336</code></pre>
</div>
</div>
<div id="quasr-overview" class="section level1">
<h1 class="hasAnchor">
<a href="#quasr-overview" class="anchor"></a><em>QuasR</em> Overview</h1>
<p>The following scheme shows the major components of <em><a href="https://bioconductor.org/packages/3.12/QuasR">QuasR</a></em> and their relationships: <img src="QuasR-scheme.png"></p>
<p><em><a href="https://bioconductor.org/packages/3.12/QuasR">QuasR</a></em> works with data (sequences and alignments, reference genome, etc.) that are stored as files on your storage (the gray cylinder on the lower left of Figure above, see section @ref(fileStorageLocations) for details on storage locations). <em><a href="https://bioconductor.org/packages/3.12/QuasR">QuasR</a></em> does not need a database management system, or these files to be named and organized according to a specific scheme.</p>
<p>In order to keep track of directory paths during an analysis, <em><a href="https://bioconductor.org/packages/3.12/QuasR">QuasR</a></em> makes use of a <code>qProject</code> object that is returned by the <code>qAlign</code> function, which at the minimum requires two inputs: the name of a samples text file (see section @ref(sampleFile) for details), and the reference genome for the alignments (see section @ref(genome)).</p>
<p>The <code>qProject</code> object is the main argument passed to subsequent functions such as <code>qQCReport</code> and <code>qCount</code>. The <code>qProject</code> object contains all necessary information on the current project and eliminates the need to repeatedly enter the same information. All functions that work on <code>qProject</code> objects can be recognized by their names starting with the letter <em>q</em>.</p>
<p>Read quantification (apart from quantification of methylation which has its own function <code>qMeth</code>) is done using the <code>qCount</code> function: It counts the alignments in regions of interest (e.g. promoters, genes, exons, etc.) and produces a count table (regions in rows, samples in columns) for further visualization and analysis. The count table can also be used as input to a statistical analysis using packages such as <em><a href="https://bioconductor.org/packages/3.12/edgeR">edgeR</a></em> <span class="citation">(Robinson, McCarthy, and Smyth 2010)</span>, <em><a href="https://bioconductor.org/packages/3.12/DESeq">DESeq</a></em> <span class="citation">(Anders and Huber 2010)</span>, <em><a href="https://bioconductor.org/packages/3.12/DESeq2">DESeq2</a></em> <span class="citation">(Love, Huber, and Anders 2014)</span>, <em><a href="https://bioconductor.org/packages/3.12/TCC">TCC</a></em> <span class="citation">(Sun et al. 2013)</span>, <em><a href="https://bioconductor.org/packages/3.12/DEXSeq">DEXSeq</a></em> <span class="citation">(Anders, Reyes, and Huber 2012)</span> or <em><a href="https://bioconductor.org/packages/3.12/baySeq">baySeq</a></em> <span class="citation">(Hardcastle and Kelly 2010)</span>.</p>
<p>In summary, a typical <em><a href="https://bioconductor.org/packages/3.12/QuasR">QuasR</a></em> analysis consists of the following steps (some of them are optional):</p>
<ul>
<li>
<code>preprocessReads</code> (optional): Remove adapters from start or end of reads, filter out reads of low quality, short length or low complexity (section @ref(preProcessing)).<br>
</li>
<li>Prepare <em>samples file</em>: List sequence files or alignments, provide sample names (section @ref(sampleFile)).<br>
</li>
<li>Prepare <em>auxiliary file</em> (optional): List additional reference sequences for alignment of reads not matching the reference genome (section @ref(auxiliaryFile)).<br>
</li>
<li>
<code>qAlign</code>: Create <code>qProject</code> object and specify project parameters. Also download BSgenome package, create aligner indices and align reads if not already existing (section @ref(qAlign)).<br>
</li>
<li>
<code>qQCReport</code> (optional): Create quality control report with plots on sequence qualities and alignment statistics (section @ref(qQCReport)).<br>
</li>
<li>
<code>qExportWig</code> (optional): Export genomic alignments as wiggle tracks for genome browser visualization (section @ref(qExportWig)).<br>
</li>
<li>
<code>qCount</code>: Quantify alignments in regions of interest (section @ref(qCount)).</li>
</ul>
<p>Recurrent example tasks that may be part of any typical analysis are described in section @ref(exampleTasks). Example workflows for specific experiment types (ChIP-seq, RNA-seq and Bis-seq) are described in section @ref(exampleWorkflows).</p>
<div id="fileStorageLocations" class="section level2">
<h2 class="hasAnchor">
<a href="#fileStorageLocations" class="anchor"></a>File storage locations</h2>
<p>Apart from <code>qExportWig</code> and <code>qQCReport</code>, which generate wig files and pdf reports, <code>qAlign</code> is the only function in <em><a href="https://bioconductor.org/packages/3.12/QuasR">QuasR</a></em> that stores files on the disk (see section @ref(qAlign) for details). All files generated by <code>qAlign</code> are listed here by type, together with their default location and how locations can be changed.</p>
<ul>
<li>
<em>Temporary files</em> (default: <code><a href="https://rdrr.io/r/base/tempfile.html">tempdir()</a></code>): Temporary files include reference genomes in <code>fasta</code> format, decompressed input sequence files, and temporary alignments in text format, and can require a large amount of disk space. By default, these files will be written to the temporary directory of the <strong>R</strong> process (as reported by <code><a href="https://rdrr.io/r/base/tempfile.html">tempdir()</a></code>). If using <code>clObj</code> for parallel processing, this may be the <code><a href="https://rdrr.io/r/base/tempfile.html">tempdir()</a></code> from the cluster node(s). An alternative location can be set using the <code>TMPDIR</code> environment variable (see <code><a href="https://rdrr.io/r/base/tempfile.html">?tempdir</a></code>).<br>
</li>
<li>
<em>Alignment files (<code>bam</code> format)</em> (default: same directory as the input sequence files): Alignments against reference genome and auxiliary targets are stored in <code>bam</code> format in the same directory that also contains the input sequence file (listed in <code>sampleFile</code>). Please note that if the input sequence file corresponds to a symbolic link, <em><a href="https://bioconductor.org/packages/3.12/QuasR">QuasR</a></em> will follow the link and use the directory of the original file instead. An alternative directory can be specified with the <code>alignmentsDir</code> argument from <code>qAlign</code>, which will store all <code>bam</code> files in that directory even if the input sequence files are located in different directories.<br>
</li>
<li>
<em>Alignment index files</em> (default: depends on <code>genome</code> and <code>snpFile</code> arguments): Many alignment tools including <code>bowtie</code> require an index of the reference sequence to perform alignments. If necessary, <code>qAlign</code> will build this index automatically and store it in a default location that depends on the <code>genome</code> argument:
<ul>
<li>
<code>BSgenome</code>: If <code>genome</code> is the name of a <em><a href="https://bioconductor.org/packages/3.12/BSgenome">BSgenome</a></em> package (such as <code>"BSgenome.Hsapiens.UCSC.hg19"</code>), the index will be stored as a new <strong>R</strong> package in the default library path (as reported by <code>.libPaths()[1]</code>, see <code><a href="https://rdrr.io/r/utils/install.packages.html">?install.packages</a></code> for details). The name of this index package will be the name of the original <em><a href="https://bioconductor.org/packages/3.12/BSgenome">BSgenome</a></em> package with a suffix for the index type, for example <code>"BSgenome.Hsapiens.UCSC.hg19.Rbowtie"</code>.<br>
</li>
<li>
<code>fasta</code>: If <code>genome</code> refers to a reference genome file in <code>fasta</code> format, the index will be stored in a subdirectory at the same location. Similarly, the indices for files listed in <code>auxiliaryFile</code> are store at the location of these files. For example, the <code>Rbowtie</code> index for the genome at <code>"./genome/mm9.fa"</code> is stored in <code>"./genome/mm9.fa.Rbowtie"</code>.<br>
</li>
<li>
<em>Allele-specific analysis</em>: A special case is the allele-specific analysis, where reference and alternative alleles listed in <code>snpFile</code> (e.g. <code>"./mySNPs.tab"</code>) are injected into the <code>genome</code> (e.g. <code>"BSgenome.Mmusculus.UCSC.mm9"</code>) to create two variant genomes to be indexed. These indices are saved at the location of the <code>snpFile</code> in a directory named after <code>snpFile</code>, <code>genome</code> and the index type (e.g. <code>"./mySNPs.tab.BSgenome.Mmusculus.UCSC.mm9.A.fa.Rbowtie"</code>).</li>
</ul>
</li>
</ul>
</div>
</div>
<div id="exampleTasks" class="section level1">
<h1 class="hasAnchor">
<a href="#exampleTasks" class="anchor"></a>Example tasks</h1>
<div id="sampleFile" class="section level2">
<h2 class="hasAnchor">
<a href="#sampleFile" class="anchor"></a>Create a sample file</h2>
<p>The sample file is a tab-delimited text file with two or three columns. The first row contains the column names: For a single read experiment, these are ‘FileName’ and ‘SampleName’; for a paired-end experiment, these are ‘FileName1’, ‘FileName2’ and ‘SampleName’. If the first row does not contain the correctly spelled column names, <em><a href="https://bioconductor.org/packages/3.12/QuasR">QuasR</a></em> will not accept the samples file. Subsequent rows contain the input sequence files.</p>
<p>Here are examples of such sample files for a single read experiment:</p>
<pre>
FileName    SampleName
chip_1_1.fq.bz2 Sample1
chip_2_1.fq.bz2 Sample2
</pre>
<!--
------------------- | -------------
`FileName`          | `SampleName`  
`chip_1_1.fq.bz2`   | `Sample1`  
`chip_2_1.fq.bz2`   | `Sample2`  
-->
<p>and for a paired-end experiment:</p>
<pre>
FileName1   FileName2   SampleName
rna_1_1.fq.bz2  rna_1_2.fq.bz2  Sample1
rna_2_1.fq.bz2  rna_2_2.fq.bz2  Sample2
</pre>
<!--
`FileName1`        | `FileName2`        | `SampleName`  
`rna_1_1.fq.bz2`   | `rna_1_2.fq.bz2`   | `Sample1`  
`rna_2_1.fq.bz2`   | `rna_2_2.fq.bz2`   | `Sample2`  
-->
<p>These example files are also contained in the <em><a href="https://bioconductor.org/packages/3.12/QuasR">QuasR</a></em> package and may be used as templates. The path of the files can be determined using:</p>
<div class="sourceCode" id="cb32"><pre class="downlit">
<span class="va">sampleFile1</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span><span class="op">(</span>package<span class="op">=</span><span class="st">"QuasR"</span>, <span class="st">"extdata"</span>,
                           <span class="st">"samples_chip_single.txt"</span><span class="op">)</span>
<span class="va">sampleFile2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span><span class="op">(</span>package<span class="op">=</span><span class="st">"QuasR"</span>, <span class="st">"extdata"</span>,
                           <span class="st">"samples_rna_paired.txt"</span><span class="op">)</span></pre></div>
<p>The columns <em>FileName</em> for single-read, or <em>FileName1</em> and <em>FileName2</em> for paired-end experiments contain paths and names to files containing the sequence data. The paths can be absolute or relative to the location of the sample file. This allows combining files from different directories in a single analysis. For each input sequence file, <code>qAlign</code> will create one alignment file and by default store it in the same directory as the sequence file. Already existing alignment files with identical parameters will not be re-created, so that it is easy to reuse the same sequence files in multiple projects without unnecessarily copying sequence files or recreating alignments.</p>
<p>The <em>SampleName</em> column contains sample names for each sequence file. The same name can be used on several lines to indicate multiple sequence files that belong to the same sample (<code>qCount</code> can use this information to automatically combine counts for one sample from multiple files).</p>
<p>Three file formats are supported for input files (but cannot be mixed within a single sample file):</p>
<ul>
<li>
<strong>fasta</strong> files have names that end with ‘.fa’, ‘.fna’ or ‘.fasta’. They contain only sequences (and no base qualities) and will thus by default be aligned on the basis of mismatches (the best alignment is the one with fewest mismatches).<br>
</li>
<li>
<strong>fastq</strong> files have names that end with ‘.fq’ or ‘.fastq’. They contain sequences and corresponding base qualities and will be aligned by default using these qualities. The encoding scheme of base qualities is automatically detected for each individual fastq file.<br>
</li>
<li>
<strong>bam</strong> files have names that end with ‘.bam’. They can be used if the sequence reads have already been aligned outside of <em><a href="https://bioconductor.org/packages/3.12/QuasR">QuasR</a></em>, and <em><a href="https://bioconductor.org/packages/3.12/QuasR">QuasR</a></em> will only be used for downstream analysis based on the alignments contained in the <code>bam</code> files. This makes it possible to use alignment tools that are not available within <em><a href="https://bioconductor.org/packages/3.12/QuasR">QuasR</a></em>, but making use of this option comes with a risk and should only be used by experienced users. For example, it cannot be guaranteed any more that certain assumptions made by <code>qCount</code> are fulfilled by the external aligner (see below). When using external <code>bam</code> files, we recommend to use files which contain only one alignment per read. This may also include multi-hit reads, for which one of the alignments is randomly selected. This allows <em><a href="https://bioconductor.org/packages/3.12/QuasR">QuasR</a></em> to count the total number of reads by counting the total number of alignments. Furthermore, if the <code>bam</code> files also contain the unmapped reads, <em><a href="https://bioconductor.org/packages/3.12/QuasR">QuasR</a></em> will be able to calculate the fraction of mapped reads. For bisulfite samples we require ungapped alignments stored in unpaired or paired <em>ff</em> orientation (even if the input reads are <em>fr</em>). For allele-specific <code>bam</code> files, <em><a href="https://bioconductor.org/packages/3.12/QuasR">QuasR</a></em> requires an additional tag for each alignment called <code>XV</code> of type <code>A</code> (printable character) with the possible values <code>R</code> (Reference), <code>U</code> (Unknown) and <code>A</code> (Alternative).</li>
</ul>
<p><strong>fasta</strong> and <strong>fastq</strong> files can be compressed with gzip, bzip2 or xz (file extensions ‘.gz’, ‘.bz2’ or ‘xz’, respectively) and will be automatically decompressed when necessary.</p>
<div id="working-only-with-bam-files-after-performing-alignments" class="section level3">
<h3 class="hasAnchor">
<a href="#working-only-with-bam-files-after-performing-alignments" class="anchor"></a>Working only with <code>bam</code> files after performing alignments</h3>
<p>Once alignments have been created, most analyses will only require the <code>bam</code> files and will not access the original raw sequence files anymore. However, re-creating a <code>qProject</code> object by a later identical call to <code>qAlign</code> will still need access to the raw sequences to verify consistency between raw data and alignments. It may be desirable to remove this dependency, for example to archive or move away the raw sequence files and to reclaim used disk space.</p>
<p>This can be achieved using the following procedure involving two sequential calls to <code>qAlign</code>. First, <code>qAlign</code> is called with the orignial sample file (<code>sampleFile1</code>) that lists the raw sequence files, and subsequently with a second sample file (<code>sampleFile2</code>) that lists the <code>bam</code> files generated in the first call. Such a second sample file can be easily generated given the <code>qProject</code> object (<code>proj1</code>) returned by the first call:</p>
<div class="sourceCode" id="cb33"><pre class="downlit">
<span class="va">sampleFile1</span> <span class="op">&lt;-</span> <span class="st">"samples_fastq.txt"</span>
<span class="va">sampleFile2</span> <span class="op">&lt;-</span> <span class="st">"samples_bam.txt"</span>

<span class="va">proj1</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/qAlign.html">qAlign</a></span><span class="op">(</span><span class="va">sampleFile1</span>, <span class="va">genomeFile</span><span class="op">)</span>

<span class="fu"><a href="https://rdrr.io/r/utils/write.table.html">write.table</a></span><span class="op">(</span><span class="fu"><a href="../reference/qProject-class.html">alignments</a></span><span class="op">(</span><span class="va">proj1</span><span class="op">)</span><span class="op">$</span><span class="va">genome</span>, <span class="va">sampleFile2</span>, sep<span class="op">=</span><span class="st">"\t"</span>, row.names<span class="op">=</span><span class="cn">FALSE</span><span class="op">)</span>

<span class="va">proj2</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/qAlign.html">qAlign</a></span><span class="op">(</span><span class="va">sampleFile2</span>, <span class="va">genomeFile</span><span class="op">)</span></pre></div>
<p>The analysis can now be exclusively based on the <code>bam</code> files using <code>sampleFile2</code> and <code>proj2</code>.</p>
</div>
<div id="consistency-of-samples-within-a-project" class="section level3">
<h3 class="hasAnchor">
<a href="#consistency-of-samples-within-a-project" class="anchor"></a>Consistency of samples within a project</h3>
<p>The sample file implicitly defines the type of samples contained in the project: <em>single read</em> or <em>paired-end read</em>, sequences <em>with</em> or <em>without</em> qualities. This type will have a profound impact on the downstream analysis. For example, it controls whether alignments will be performed in single or paired-end mode, either with or without base qualities. That will also determine availability of certain options for quality control and quantification in <code>qQCReport</code> and <code>qCount</code>. For consistency, it is therefore required that all samples within a project have the same type; it is not possible to mix both single and paired-end read samples, or <strong>fasta</strong> and <strong>fastq</strong> files in a single project (sample file). If necessary, it may be possible to analyse different types of files in separate <em><a href="https://bioconductor.org/packages/3.12/QuasR">QuasR</a></em> projects and combine the derived results at the end.</p>
</div>
</div>
<div id="auxiliaryFile" class="section level2">
<h2 class="hasAnchor">
<a href="#auxiliaryFile" class="anchor"></a>Create an auxiliary file (optional)</h2>
<p>By default <em><a href="https://bioconductor.org/packages/3.12/QuasR">QuasR</a></em> aligns reads only to the reference genome. However, it may be interesting to align non-matching reads to further targets, for example to identify contamination from vectors or a different species, or in order to quantify spike-in material not contained in the reference genome. In <em><a href="https://bioconductor.org/packages/3.12/QuasR">QuasR</a></em>, such supplementary reference files are called <em>auxiliary</em> references and can be specified to <code>qAlign</code> using the <code>auxiliaryFile</code> argument (see section @ref(qAlign) for details). The format of the auxiliary file is similar to the one of the sample file described in section @ref(sampleFile): It contains two columns with column names ‘FileName’ and ‘AuxName’ in the first row. Additional rows contain names and files of one or several auxiliary references in <code>fasta</code> format.</p>
<p>An example auxiliary file looks like this:</p>
<pre>
FileName    AuxName
NC_001422.1.fa  phiX174
</pre>
<!--
`FileName`        | `AuxName`  
`NC_001422.1.fa`  | `phiX174`  
-->
<p>and is available from your <em><a href="https://bioconductor.org/packages/3.12/QuasR">QuasR</a></em> installation at</p>
<div class="sourceCode" id="cb34"><pre class="downlit">
<span class="va">auxFile</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span><span class="op">(</span>package<span class="op">=</span><span class="st">"QuasR"</span>, <span class="st">"extdata"</span>, <span class="st">"auxiliaries.txt"</span><span class="op">)</span></pre></div>
</div>
<div id="genome" class="section level2">
<h2 class="hasAnchor">
<a href="#genome" class="anchor"></a>Select the reference genome</h2>
<p>Sequence reads are primarily aligned against the reference genome. If necessary, <em><a href="https://bioconductor.org/packages/3.12/QuasR">QuasR</a></em> will create an aligner index for the genome. The reference genome can be provided in one of two different formats:</p>
<ul>
<li>
<p><strong>a string</strong>, referring to the name of a <em><a href="https://bioconductor.org/packages/3.12/BSgenome">BSgenome</a></em> package:</p>
<div class="sourceCode" id="cb35"><pre class="downlit">
<span class="fu"><a href="https://rdrr.io/pkg/BSgenome/man/available.genomes.html">available.genomes</a></span><span class="op">(</span><span class="op">)</span></pre></div>
<pre><code>##   [1] "BSgenome.Alyrata.JGI.v1"                  
##   [2] "BSgenome.Amellifera.BeeBase.assembly4"    
##   [3] "BSgenome.Amellifera.UCSC.apiMel2"         
##   [4] "BSgenome.Amellifera.UCSC.apiMel2.masked"  
##   [5] "BSgenome.Aofficinalis.NCBI.V1"            
##   [6] "BSgenome.Athaliana.TAIR.04232008"         
##   [7] "BSgenome.Athaliana.TAIR.TAIR9"            
##   [8] "BSgenome.Btaurus.UCSC.bosTau3"            
##   [9] "BSgenome.Btaurus.UCSC.bosTau3.masked"     
##  [10] "BSgenome.Btaurus.UCSC.bosTau4"            
##  [11] "BSgenome.Btaurus.UCSC.bosTau4.masked"     
##  [12] "BSgenome.Btaurus.UCSC.bosTau6"            
##  [13] "BSgenome.Btaurus.UCSC.bosTau6.masked"     
##  [14] "BSgenome.Btaurus.UCSC.bosTau8"            
##  [15] "BSgenome.Btaurus.UCSC.bosTau9"            
##  [16] "BSgenome.Carietinum.NCBI.v1"              
##  [17] "BSgenome.Celegans.UCSC.ce10"              
##  [18] "BSgenome.Celegans.UCSC.ce11"              
##  [19] "BSgenome.Celegans.UCSC.ce2"               
##  [20] "BSgenome.Celegans.UCSC.ce6"               
##  [21] "BSgenome.Cfamiliaris.UCSC.canFam2"        
##  [22] "BSgenome.Cfamiliaris.UCSC.canFam2.masked" 
##  [23] "BSgenome.Cfamiliaris.UCSC.canFam3"        
##  [24] "BSgenome.Cfamiliaris.UCSC.canFam3.masked" 
##  [25] "BSgenome.Cjacchus.UCSC.calJac3"           
##  [26] "BSgenome.Dmelanogaster.UCSC.dm2"          
##  [27] "BSgenome.Dmelanogaster.UCSC.dm2.masked"   
##  [28] "BSgenome.Dmelanogaster.UCSC.dm3"          
##  [29] "BSgenome.Dmelanogaster.UCSC.dm3.masked"   
##  [30] "BSgenome.Dmelanogaster.UCSC.dm6"          
##  [31] "BSgenome.Drerio.UCSC.danRer10"            
##  [32] "BSgenome.Drerio.UCSC.danRer11"            
##  [33] "BSgenome.Drerio.UCSC.danRer5"             
##  [34] "BSgenome.Drerio.UCSC.danRer5.masked"      
##  [35] "BSgenome.Drerio.UCSC.danRer6"             
##  [36] "BSgenome.Drerio.UCSC.danRer6.masked"      
##  [37] "BSgenome.Drerio.UCSC.danRer7"             
##  [38] "BSgenome.Drerio.UCSC.danRer7.masked"      
##  [39] "BSgenome.Dvirilis.Ensembl.dvircaf1"       
##  [40] "BSgenome.Ecoli.NCBI.20080805"             
##  [41] "BSgenome.Gaculeatus.UCSC.gasAcu1"         
##  [42] "BSgenome.Gaculeatus.UCSC.gasAcu1.masked"  
##  [43] "BSgenome.Ggallus.UCSC.galGal3"            
##  [44] "BSgenome.Ggallus.UCSC.galGal3.masked"     
##  [45] "BSgenome.Ggallus.UCSC.galGal4"            
##  [46] "BSgenome.Ggallus.UCSC.galGal4.masked"     
##  [47] "BSgenome.Ggallus.UCSC.galGal5"            
##  [48] "BSgenome.Ggallus.UCSC.galGal6"            
##  [49] "BSgenome.Hsapiens.1000genomes.hs37d5"     
##  [50] "BSgenome.Hsapiens.NCBI.GRCh38"            
##  [51] "BSgenome.Hsapiens.UCSC.hg17"              
##  [52] "BSgenome.Hsapiens.UCSC.hg17.masked"       
##  [53] "BSgenome.Hsapiens.UCSC.hg18"              
##  [54] "BSgenome.Hsapiens.UCSC.hg18.masked"       
##  [55] "BSgenome.Hsapiens.UCSC.hg19"              
##  [56] "BSgenome.Hsapiens.UCSC.hg19.masked"       
##  [57] "BSgenome.Hsapiens.UCSC.hg38"              
##  [58] "BSgenome.Hsapiens.UCSC.hg38.masked"       
##  [59] "BSgenome.Mdomestica.UCSC.monDom5"         
##  [60] "BSgenome.Mfascicularis.NCBI.5.0"          
##  [61] "BSgenome.Mfuro.UCSC.musFur1"              
##  [62] "BSgenome.Mmulatta.UCSC.rheMac10"          
##  [63] "BSgenome.Mmulatta.UCSC.rheMac2"           
##  [64] "BSgenome.Mmulatta.UCSC.rheMac2.masked"    
##  [65] "BSgenome.Mmulatta.UCSC.rheMac3"           
##  [66] "BSgenome.Mmulatta.UCSC.rheMac3.masked"    
##  [67] "BSgenome.Mmulatta.UCSC.rheMac8"           
##  [68] "BSgenome.Mmusculus.UCSC.mm10"             
##  [69] "BSgenome.Mmusculus.UCSC.mm10.masked"      
##  [70] "BSgenome.Mmusculus.UCSC.mm8"              
##  [71] "BSgenome.Mmusculus.UCSC.mm8.masked"       
##  [72] "BSgenome.Mmusculus.UCSC.mm9"              
##  [73] "BSgenome.Mmusculus.UCSC.mm9.masked"       
##  [74] "BSgenome.Osativa.MSU.MSU7"                
##  [75] "BSgenome.Ppaniscus.UCSC.panPan1"          
##  [76] "BSgenome.Ppaniscus.UCSC.panPan2"          
##  [77] "BSgenome.Ptroglodytes.UCSC.panTro2"       
##  [78] "BSgenome.Ptroglodytes.UCSC.panTro2.masked"
##  [79] "BSgenome.Ptroglodytes.UCSC.panTro3"       
##  [80] "BSgenome.Ptroglodytes.UCSC.panTro3.masked"
##  [81] "BSgenome.Ptroglodytes.UCSC.panTro5"       
##  [82] "BSgenome.Ptroglodytes.UCSC.panTro6"       
##  [83] "BSgenome.Rnorvegicus.UCSC.rn4"            
##  [84] "BSgenome.Rnorvegicus.UCSC.rn4.masked"     
##  [85] "BSgenome.Rnorvegicus.UCSC.rn5"            
##  [86] "BSgenome.Rnorvegicus.UCSC.rn5.masked"     
##  [87] "BSgenome.Rnorvegicus.UCSC.rn6"            
##  [88] "BSgenome.Scerevisiae.UCSC.sacCer1"        
##  [89] "BSgenome.Scerevisiae.UCSC.sacCer2"        
##  [90] "BSgenome.Scerevisiae.UCSC.sacCer3"        
##  [91] "BSgenome.Sscrofa.UCSC.susScr11"           
##  [92] "BSgenome.Sscrofa.UCSC.susScr3"            
##  [93] "BSgenome.Sscrofa.UCSC.susScr3.masked"     
##  [94] "BSgenome.Tgondii.ToxoDB.7.0"              
##  [95] "BSgenome.Tguttata.UCSC.taeGut1"           
##  [96] "BSgenome.Tguttata.UCSC.taeGut1.masked"    
##  [97] "BSgenome.Tguttata.UCSC.taeGut2"           
##  [98] "BSgenome.Vvinifera.URGI.IGGP12Xv0"        
##  [99] "BSgenome.Vvinifera.URGI.IGGP12Xv2"        
## [100] "BSgenome.Vvinifera.URGI.IGGP8X"</code></pre>
<div class="sourceCode" id="cb37"><pre class="downlit">
<span class="va">genomeName</span> <span class="op">&lt;-</span> <span class="st">"BSgenome.Hsapiens.UCSC.hg19"</span></pre></div>
</li>
</ul>
<p>In this example, the BSgenome package <code>"BSgenome.Hsapiens.UCSC.hg19"</code> refers to an unmasked genome; alignment index and alignments will be performed on the full unmasked genome sequence (recommended). If using a masked genome (e.g. <code>"BSgenome.Hsapiens.UCSC.hg19.masked"</code>), masked regions will be replaced with <code>"N"</code> bases, and this hard-masked version of the genome will be used for creating the alignment index and further alignments.</p>
<ul>
<li>
<p><strong>a file name</strong>, referring to a sequence file containing one or several reference sequences (e.g. chromosomes) in <code>fasta</code> format:</p>
<div class="sourceCode" id="cb38"><pre class="downlit">
<span class="va">genomeFile</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span><span class="op">(</span>package<span class="op">=</span><span class="st">"QuasR"</span>, <span class="st">"extdata"</span>, <span class="st">"hg19sub.fa"</span><span class="op">)</span></pre></div>
</li>
</ul>
</div>
<div id="preProcessing" class="section level2">
<h2 class="hasAnchor">
<a href="#preProcessing" class="anchor"></a>Sequence data pre-processing</h2>
<p>The <code>preprocessReads</code> function can be used to prepare the input sequence files prior to alignment. The function takes one or several sequence files (or pairs of files for a paired-end experiment) in <code>fasta</code> or <code>fastq</code> format as input and produces the same number of output files with the processed reads.</p>
<p>In the following example, we truncate the reads by removing the three bases from the 3’-end (the right side), remove the adapter sequence <code>AAAAAAAAAA</code> from the 5’-end (the left side) and filter out reads that, after truncation and adapter removal, are shorter than 14 bases or contain more than 2 <code>N</code> bases:</p>
<div class="sourceCode" id="cb39"><pre class="downlit">
<span class="va">td</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/tempfile.html">tempdir</a></span><span class="op">(</span><span class="op">)</span>
<span class="va">infiles</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span><span class="op">(</span>package<span class="op">=</span><span class="st">"QuasR"</span>, <span class="st">"extdata"</span>,
                       <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"rna_1_1.fq.bz2"</span>,<span class="st">"rna_2_1.fq.bz2"</span><span class="op">)</span><span class="op">)</span>
<span class="va">outfiles</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html">file.path</a></span><span class="op">(</span><span class="va">td</span>, <span class="fu"><a href="https://rdrr.io/r/base/basename.html">basename</a></span><span class="op">(</span><span class="va">infiles</span><span class="op">)</span><span class="op">)</span>
<span class="va">res</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/preprocessReads.html">preprocessReads</a></span><span class="op">(</span>filename <span class="op">=</span> <span class="va">infiles</span>,
                       outputFilename <span class="op">=</span> <span class="va">outfiles</span>,
                       truncateEndBases <span class="op">=</span> <span class="fl">3</span>,
                       Lpattern <span class="op">=</span> <span class="st">"AAAAAAAAAA"</span>,
                       minLength <span class="op">=</span> <span class="fl">14</span>, 
                       nBases <span class="op">=</span> <span class="fl">2</span><span class="op">)</span></pre></div>
<pre><code>##   filtering /Users/runner/work/_temp/Library/QuasR/extdata/rna_1_1.fq.bz2</code></pre>
<pre><code>##   filtering /Users/runner/work/_temp/Library/QuasR/extdata/rna_2_1.fq.bz2</code></pre>
<div class="sourceCode" id="cb42"><pre class="downlit">
<span class="va">res</span></pre></div>
<pre><code>##                  rna_1_1.fq.bz2 rna_2_1.fq.bz2
## totalSequences             3002           3000
## matchTo5pAdapter            466            463
## matchTo3pAdapter              0              0
## tooShort                    107             91
## tooManyN                      0              0
## lowComplexity                 0              0
## totalPassed                2895           2909</code></pre>
<div class="sourceCode" id="cb44"><pre class="downlit">
<span class="fu"><a href="https://rdrr.io/r/base/unlink.html">unlink</a></span><span class="op">(</span><span class="va">outfiles</span><span class="op">)</span></pre></div>
<p><code>preprocessReads</code> returns a matrix with a summary of the pre-processing. The matrix contains one column per (pair of) input sequence files, and contains the total number of reads (<code>totalSequences</code>), the number of reads that matched to the five prime or three prime adapters (<code>matchTo5pAdapter</code> and <code>matchTo3pAdapter</code>), the number of reads that were too short (<code>tooShort</code>), contained too many non-base characters (<code>tooManyN</code>) or were of low sequence complexity (<code>lowComplexity</code>, deactivated by default). Finally, the number of reads that passed the filtering steps is reported in the last row (<code>totalPassed</code>).</p>
<p>In the example below we process paired-end reads, removing all pairs with one or several <code>N</code> bases. Even if only one sequence in a pair fulfills the filtering criteria, both reads in the pair are removed, thereby preserving the matching order of the sequences in the two files:</p>
<div class="sourceCode" id="cb45"><pre class="downlit">
<span class="va">td</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/tempfile.html">tempdir</a></span><span class="op">(</span><span class="op">)</span>
<span class="va">infiles1</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span><span class="op">(</span>package<span class="op">=</span><span class="st">"QuasR"</span>, <span class="st">"extdata"</span>, <span class="st">"rna_1_1.fq.bz2"</span><span class="op">)</span>
<span class="va">infiles2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span><span class="op">(</span>package<span class="op">=</span><span class="st">"QuasR"</span>, <span class="st">"extdata"</span>, <span class="st">"rna_1_2.fq.bz2"</span><span class="op">)</span>
<span class="va">outfiles1</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html">file.path</a></span><span class="op">(</span><span class="va">td</span>, <span class="fu"><a href="https://rdrr.io/r/base/basename.html">basename</a></span><span class="op">(</span><span class="va">infiles1</span><span class="op">)</span><span class="op">)</span>
<span class="va">outfiles2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html">file.path</a></span><span class="op">(</span><span class="va">td</span>, <span class="fu"><a href="https://rdrr.io/r/base/basename.html">basename</a></span><span class="op">(</span><span class="va">infiles2</span><span class="op">)</span><span class="op">)</span>
<span class="va">res</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/preprocessReads.html">preprocessReads</a></span><span class="op">(</span>filename<span class="op">=</span><span class="va">infiles1</span>,
                       filenameMate<span class="op">=</span><span class="va">infiles2</span>,
                       outputFilename<span class="op">=</span><span class="va">outfiles1</span>,
                       outputFilenameMate<span class="op">=</span><span class="va">outfiles2</span>,
                       nBases<span class="op">=</span><span class="fl">0</span><span class="op">)</span></pre></div>
<pre><code>##   filtering /Users/runner/work/_temp/Library/QuasR/extdata/rna_1_1.fq.bz2 and
##     /Users/runner/work/_temp/Library/QuasR/extdata/rna_1_2.fq.bz2</code></pre>
<div class="sourceCode" id="cb47"><pre class="downlit">
<span class="va">res</span></pre></div>
<pre><code>##                  rna_1_1.fq.bz2:rna_1_2.fq.bz2
## totalSequences                            3002
## matchTo5pAdapter                            NA
## matchTo3pAdapter                            NA
## tooShort                                     0
## tooManyN                                     3
## lowComplexity                                0
## totalPassed                               2999</code></pre>
<p>More details on the <code>preprocessReads</code> function can be found in the function documentation (see <code><a href="../reference/preprocessReads.html">?preprocessReads</a></code>) or in the section @ref(preprocessReads).</p>
</div>
</div>
<div id="exampleWorkflows" class="section level1">
<h1 class="hasAnchor">
<a href="#exampleWorkflows" class="anchor"></a>Example workflows</h1>
<div id="ChIPseq" class="section level2">
<h2 class="hasAnchor">
<a href="#ChIPseq" class="anchor"></a>ChIP-seq: Measuring protein-DNA binding and chromatin modifications</h2>
<p>Here we show an exemplary single-end ChIP-seq workflow using a small number of reads from a histone 3 lysine 4 trimethyl (H3K4me3) ChIP-seq experiment. This histone modification is known to locate to genomic regions with a high density of CpG dinucleotides (so called CpG islands); about 60% of mammalian genes have such a CpG island close to their transcript start site. All necessary files are included in the <em><a href="https://bioconductor.org/packages/3.12/QuasR">QuasR</a></em> package, and we start the example workflow by copying those files into the current working directly, into a subfolder called <code>"extdata"</code>:</p>
<div class="sourceCode" id="cb49"><pre class="downlit">
<span class="fu"><a href="https://rdrr.io/r/base/files.html">file.copy</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span><span class="op">(</span>package<span class="op">=</span><span class="st">"QuasR"</span>, <span class="st">"extdata"</span><span class="op">)</span>, <span class="st">"."</span>, recursive<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span></pre></div>
<pre><code>## [1] TRUE</code></pre>
<div id="align-reads-using-the-qalign-function" class="section level3">
<h3 class="hasAnchor">
<a href="#align-reads-using-the-qalign-function" class="anchor"></a>Align reads using the <code>qAlign</code> function</h3>
<p>We assume that the sequence reads have already been pre-processed as described in section @ref(preProcessing). Also, a sample file (section @ref(sampleFile)) that lists all sequence files to be analyzed has been prepared. A <code>fasta</code> file with the reference genome sequence(s) is also available (section @ref(genome)), as well as an auxiliary file for alignment of reads that failed to match the reference genome (section @ref(auxiliaryFile)).</p>
<p>By default, newly generated <code>bam</code> files will be stored at the location of the input sequence files, which should be writable and have sufficient capacity (an alternative location can be specified using the <code>alignmentsDir</code> argument). Make also sure that you have sufficient temporary disk space for intermediate files in <code><a href="https://rdrr.io/r/base/tempfile.html">tempdir()</a></code> (see section @ref(qAlign)). We start by aligning the reads using <code>qAlign</code>:</p>
<div class="sourceCode" id="cb51"><pre class="downlit">
<span class="va">sampleFile</span> <span class="op">&lt;-</span> <span class="st">"extdata/samples_chip_single.txt"</span>
<span class="va">auxFile</span> <span class="op">&lt;-</span> <span class="st">"extdata/auxiliaries.txt"</span>
<span class="va">genomeFile</span> <span class="op">&lt;-</span> <span class="st">"extdata/hg19sub.fa"</span>

<span class="va">proj1</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/qAlign.html">qAlign</a></span><span class="op">(</span><span class="va">sampleFile</span>, genome<span class="op">=</span><span class="va">genomeFile</span>, auxiliaryFile<span class="op">=</span><span class="va">auxFile</span><span class="op">)</span></pre></div>
<pre><code>## alignment files missing - need to:</code></pre>
<pre><code>##     create 2 auxiliary alignment(s)</code></pre>
<pre><code>## Creating an Rbowtie index for /Users/runner/work/QuasR/QuasR/vignettes/extdata/NC_001422.1.fa</code></pre>
<pre><code>## Finished creating index</code></pre>
<pre><code>## Testing the compute nodes...OK
## Loading QuasR on the compute nodes...OK
## Available cores:</code></pre>
<pre><code>## nodeNames
## Mac-1602493909147.local 
##                       1</code></pre>
<pre><code>## Performing auxiliary alignments for 2 samples. See progress in the log file:
## /Users/runner/work/QuasR/QuasR/vignettes/QuasR_log_17d541c8f484.txt
## Auxiliary alignments have been created successfully</code></pre>
<div class="sourceCode" id="cb59"><pre class="downlit">
<span class="va">proj1</span></pre></div>
<pre><code>## Project: qProject
##  Options   : maxHits         : 1
##              paired          : no
##              splicedAlignment: FALSE
##              bisulfite       : no
##              snpFile         : none
##              geneAnnotation  : none
##  Aligner   : Rbowtie v1.29.1 (parameters: -m 1 --best --strata)
##  Genome    : /Users/runner/work/QuasR/QuasR/vignettes/extdata/hg19sub.fa (file)
## 
##  Reads     : 2 files, 2 samples (fastq format):
##    1. chip_1_1.fq.bz2  Sample1 (phred33)
##    2. chip_2_1.fq.bz2  Sample2 (phred33)
## 
##  Genome alignments: directory: same as reads
##    1. chip_1_1_17d52e6ecf3a.bam
##    2. chip_2_1_17d56ce700a6.bam
## 
##  Aux. alignments: 1 file, directory: same as reads
##    a. /Users/runner/work/QuasR/QuasR/vignettes/extdata/NC_001422.1.fa    phiX174
##      1. chip_1_1_17d5702d27d9.bam
##      2. chip_2_1_17d524935118.bam</code></pre>
<p><code>qAlign</code> will build alignment indices if they do not yet exist (by default, if the genome and auxiliary sequences are given in the form of <code>fasta</code> files, they will be stored in the same folder). The <code>qProject</code> object (<code>proj1</code>) returned by <code>qAlign</code> now contains all information about the ChIP-seq experiment: the (optional) project name, the project options, aligner package, reference genome, and at the bottom the sequence and alignment files. For each input sequence file, there will be one <code>bam</code> file with alignments against the reference genome, and one for each auxiliary target sequence with alignments of reads without genome hits. Our <code>auxFile</code> contains a single auxiliary target sequence, so we expect two <code>bam</code> files per input sequence file:</p>
<div class="sourceCode" id="cb61"><pre class="downlit">
<span class="fu"><a href="https://rdrr.io/r/base/list.files.html">list.files</a></span><span class="op">(</span><span class="st">"extdata"</span>, pattern<span class="op">=</span><span class="st">".bam$"</span><span class="op">)</span></pre></div>
<pre><code>## [1] "chip_1_1_17d52e6ecf3a.bam"     "chip_1_1_17d5702d27d9.bam"    
## [3] "chip_2_1_17d524935118.bam"     "chip_2_1_17d56ce700a6.bam"    
## [5] "phiX_paired_withSecondary.bam"</code></pre>
<p>The <code>bam</code> file names consist of the base name of the sequence file with an added random string. The random suffix makes sure that newly generated alignment files do not overwrite existing ones, for example of the same reads aligned against an alternative reference genome. Each alignment file is accompanied by two additional files with suffixes <code>.bai</code> and <code>.txt</code>:</p>
<div class="sourceCode" id="cb63"><pre class="downlit">
<span class="fu"><a href="https://rdrr.io/r/base/list.files.html">list.files</a></span><span class="op">(</span><span class="st">"extdata"</span>, pattern<span class="op">=</span><span class="st">"^chip_1_1_"</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">3</span><span class="op">]</span></pre></div>
<pre><code>## [1] "chip_1_1_17d52e6ecf3a.bam"     "chip_1_1_17d52e6ecf3a.bam.bai"
## [3] "chip_1_1_17d52e6ecf3a.bam.txt"</code></pre>
<p>The <code>.bai</code> file is the <code>bam</code> index used for fast access by genomic coordinate. The <code>.txt</code> file contains all the parameters used to generate the corresponding <code>bam</code> file. Before new alignments are generated, <code>qAlign</code> will look for available <code>.txt</code> files in default locations (the directory containing the input sequence file, or the value of <code>alignmentsDir</code>), and read their contents to determine if a compatible <code>bam</code> file already exists. A compatible <code>bam</code> file is one with the same reads and genome, aligned using the same aligner and identical alignment parameters. If a compatible <code>bam</code> file is not found, or the <code>.txt</code> file is missing, <code>qAlign</code> will generate a new <code>bam</code> file. It is therefore recommended not to delete the <code>.txt</code> file - without it, the corresponding <code>bam</code> file will become unusable for <em><a href="https://bioconductor.org/packages/3.12/QuasR">QuasR</a></em>.</p>
</div>
<div id="create-a-quality-control-report" class="section level3">
<h3 class="hasAnchor">
<a href="#create-a-quality-control-report" class="anchor"></a>Create a quality control report</h3>
<p><em><a href="https://bioconductor.org/packages/3.12/QuasR">QuasR</a></em> can produce a quality control report in the form of a series of diagnostic plots with details on sequences and alignments (see QuasR scheme figure above). The plots are generated by calling the <code>qQCReport</code> function with the <code>qProject</code> object as argument. <code>qQCReport</code> uses <em><a href="https://bioconductor.org/packages/3.12/ShortRead">ShortRead</a></em> <span class="citation">(Morgan et al. 2009)</span> internally to obtain some of the quality metrics, and some of the plots are inspired by the FastQC quality control tool by Simon Andrews (<a href="http://www.bioinformatics.bbsrc.ac.uk/projects/fastqc/" class="uri">http://www.bioinformatics.bbsrc.ac.uk/projects/fastqc/</a>). The plots will be stored into a multipage PDF document defined by the <code>pdfFilename</code> argument, or else shown as individual plot windows on the current graphics device. In order to keep the running time reasonably short, some quality metrics are obtained from a random sub-sample of the sequences or alignments.</p>
<pre><code>## collecting quality control data</code></pre>
<pre><code>## creating QC plots</code></pre>
<div class="sourceCode" id="cb67"><pre class="downlit">
<span class="fu"><a href="../reference/qQCReport.html">qQCReport</a></span><span class="op">(</span><span class="va">proj1</span>, pdfFilename<span class="op">=</span><span class="st">"extdata/qc_report.pdf"</span><span class="op">)</span></pre></div>
<pre><code>## collecting quality control data</code></pre>
<pre><code>## creating QC plots</code></pre>
<p>Currently available plots are described in section @ref(qQCReport) and following.</p>
</div>
<div id="alignment-statistics" class="section level3">
<h3 class="hasAnchor">
<a href="#alignment-statistics" class="anchor"></a>Alignment statistics</h3>
<p>The <code>alignmentStats</code> gets the number of (un-)mapped reads for each sequence file in a <code>qProject</code> object, by reading the <code>bam</code> file indices, and returns them as a <code>data.frame</code>. The function also works for arguments of type <code>character</code> with one or several <code>bam</code> file names (for details see section @ref(alignmentStats)).</p>
<div class="sourceCode" id="cb70"><pre class="downlit">
<span class="fu"><a href="../reference/alignmentStats.html">alignmentStats</a></span><span class="op">(</span><span class="va">proj1</span><span class="op">)</span></pre></div>
<pre><code>##                 seqlength mapped unmapped
## Sample1:genome      95000   2339      258
## Sample2:genome      95000   3609      505
## Sample1:phiX174      5386    251        7
## Sample2:phiX174      5386    493       12</code></pre>
</div>
<div id="export-genome-wig-file-from-alignments" class="section level3">
<h3 class="hasAnchor">
<a href="#export-genome-wig-file-from-alignments" class="anchor"></a>Export genome wig file from alignments</h3>
<p>For visualization in a genome browser, alignment coverage along the genome can be exported to a (compressed) wig file using the <code>qExportWig</code> function. The created fixedStep wig file (see (<a href="http://genome.ucsc.edu/goldenPath/help/wiggle.html" class="uri">http://genome.ucsc.edu/goldenPath/help/wiggle.html</a>) for details on the wig format) will contain one track per sample in the <code>qProject</code> object. The resolution is defined using the <code>binsize</code> argument, and if <code>scaling</code> is set to <code>TRUE</code>, read counts per bin are scaled by the total number of aligned reads in each sample to improve comparability:</p>
<div class="sourceCode" id="cb72"><pre class="downlit">
<span class="fu"><a href="../reference/qExportWig.html">qExportWig</a></span><span class="op">(</span><span class="va">proj1</span>, binsize<span class="op">=</span><span class="fl">100L</span>, scaling<span class="op">=</span><span class="cn">TRUE</span>, collapseBySample<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span></pre></div>
<pre><code>## collecting mapping statistics for scaling...done
## start creating wig files...
##   Sample1.wig.gz (Sample1)
##   Sample2.wig.gz (Sample2)
## done</code></pre>
</div>
<div id="count-alignments-using-qcount" class="section level3">
<h3 class="hasAnchor">
<a href="#count-alignments-using-qcount" class="anchor"></a>Count alignments using <code>qCount</code>
</h3>
<p>Alignments are quantified using <code>qCount</code>, for example using a <code>GRanges</code> object as a query. In our H3K4me3 ChIP-seq example, we expect the reads to occur around the transcript start site of genes. We can therefore construct suitable query regions using genomic intervals around the start sites of known genes. In the code below, this is achieved with help from the <em><a href="https://bioconductor.org/packages/3.12/GenomicFeatures">GenomicFeatures</a></em> package: We first create a <code>TxDb</code> object from a <code>.gtf</code> file with gene annotation. With the <code>promoters</code> function, we can then create the <code>GRanges</code> object with regions to be quantified. Finally, because most genes consist of multiple overlapping transcripts, we select the first transcript for each gene:</p>
<div class="sourceCode" id="cb74"><pre class="downlit">
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">GenomicFeatures</span><span class="op">)</span>
<span class="va">annotFile</span> <span class="op">&lt;-</span> <span class="st">"extdata/hg19sub_annotation.gtf"</span>
<span class="va">chrLen</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/Rsamtools/man/scanFa.html">scanFaIndex</a></span><span class="op">(</span><span class="va">genomeFile</span><span class="op">)</span>
<span class="va">chrominfo</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html">data.frame</a></span><span class="op">(</span>chrom<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/character.html">as.character</a></span><span class="op">(</span><span class="fu">seqnames</span><span class="op">(</span><span class="va">chrLen</span><span class="op">)</span><span class="op">)</span>,
                        length<span class="op">=</span><span class="fu">width</span><span class="op">(</span><span class="va">chrLen</span><span class="op">)</span>,
                        is_circular<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/rep.html">rep</a></span><span class="op">(</span><span class="cn">FALSE</span>, <span class="fu"><a href="https://rdrr.io/r/base/length.html">length</a></span><span class="op">(</span><span class="va">chrLen</span><span class="op">)</span><span class="op">)</span><span class="op">)</span>
<span class="va">txdb</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/GenomicFeatures/man/makeTxDbFromGFF.html">makeTxDbFromGFF</a></span><span class="op">(</span>file<span class="op">=</span><span class="va">annotFile</span>, format<span class="op">=</span><span class="st">"gtf"</span>,
                        chrominfo<span class="op">=</span><span class="va">chrominfo</span>,
                        dataSource<span class="op">=</span><span class="st">"Ensembl"</span>,
                        organism<span class="op">=</span><span class="st">"Homo sapiens"</span><span class="op">)</span></pre></div>
<pre><code>## Import genomic features from the file as a GRanges object ... OK
## Prepare the 'metadata' data frame ... OK
## Make the TxDb object ... OK</code></pre>
<div class="sourceCode" id="cb76"><pre class="downlit">
<span class="va">promReg</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/GenomicFeatures/man/transcripts.html">promoters</a></span><span class="op">(</span><span class="va">txdb</span>, upstream<span class="op">=</span><span class="fl">1000</span>, downstream<span class="op">=</span><span class="fl">500</span>,
                     columns<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"gene_id"</span>,<span class="st">"tx_id"</span><span class="op">)</span><span class="op">)</span>
<span class="va">gnId</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html">sapply</a></span><span class="op">(</span><span class="fu">mcols</span><span class="op">(</span><span class="va">promReg</span><span class="op">)</span><span class="op">$</span><span class="va">gene_id</span>, <span class="va">paste</span>, collapse<span class="op">=</span><span class="st">","</span><span class="op">)</span>
<span class="va">promRegSel</span> <span class="op">&lt;-</span> <span class="va">promReg</span><span class="op">[</span> <span class="fu"><a href="https://rdrr.io/r/base/match.html">match</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/unique.html">unique</a></span><span class="op">(</span><span class="va">gnId</span><span class="op">)</span>, <span class="va">gnId</span><span class="op">)</span> <span class="op">]</span>
<span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">promRegSel</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/unique.html">unique</a></span><span class="op">(</span><span class="va">gnId</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">promRegSel</span><span class="op">)</span></pre></div>
<pre><code>## GRanges object with 6 ranges and 2 metadata columns:
##                   seqnames      ranges strand |         gene_id     tx_id
##                      &lt;Rle&gt;   &lt;IRanges&gt;  &lt;Rle&gt; | &lt;CharacterList&gt; &lt;integer&gt;
##   ENSG00000176022     chr1 31629-33128      + | ENSG00000176022         1
##   ENSG00000186891     chr1   6452-7951      - | ENSG00000186891         2
##   ENSG00000186827     chr1 14013-15512      - | ENSG00000186827         6
##   ENSG00000078808     chr1 31882-33381      - | ENSG00000078808         9
##   ENSG00000171863     chr2   1795-3294      + | ENSG00000171863        17
##   ENSG00000252531     chr2   7160-8659      + | ENSG00000252531        26
##   -------
##   seqinfo: 3 sequences from an unspecified genome</code></pre>
<p>Using <code>promRegSel</code> object as query, we can now count the alignment per sample in each of the promoter windows.</p>
<div class="sourceCode" id="cb78"><pre class="downlit">
<span class="va">cnt</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/qCount.html">qCount</a></span><span class="op">(</span><span class="va">proj1</span>, <span class="va">promRegSel</span><span class="op">)</span></pre></div>
<pre><code>## counting alignments...done</code></pre>
<div class="sourceCode" id="cb80"><pre class="downlit">
<span class="va">cnt</span></pre></div>
<pre><code>##                 width Sample1 Sample2
## ENSG00000176022  1500     157     701
## ENSG00000186891  1500      22       5
## ENSG00000186827  1500      10       3
## ENSG00000078808  1500      73     558
## ENSG00000171863  1500      94     339
## ENSG00000252531  1500      59       9
## ENSG00000247886  1500     172     971
## ENSG00000254999  1500     137     389
## ENSG00000238642  1500       8       3
## ENSG00000134086  1500       9      18
## ENSG00000238345  1500      13      25
## ENSG00000134075  1500       7       3</code></pre>
<p>The counts returned by <code>qCount</code> are the raw number of alignments per sample and region, without any normalization for the query region length, or the total number of aligned reads in a sample. As expected, we can find H3K4me3 signal at promoters of a subset of the genes with CpG island promoters, which we can visualize with help of the <em><a href="https://bioconductor.org/packages/3.12/Gviz">Gviz</a></em> package:</p>
<div class="sourceCode" id="cb82"><pre class="downlit">
<span class="va">gr1</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/rtracklayer/man/export.html">import</a></span><span class="op">(</span><span class="st">"Sample1.wig.gz"</span><span class="op">)</span></pre></div>
<pre><code>## Warning in asMethod(object): NAs introduced by coercion</code></pre>
<div class="sourceCode" id="cb84"><pre class="downlit">
<span class="va">gr2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/rtracklayer/man/export.html">import</a></span><span class="op">(</span><span class="st">"Sample2.wig.gz"</span><span class="op">)</span></pre></div>
<pre><code>## Warning in asMethod(object): NAs introduced by coercion</code></pre>
<div class="sourceCode" id="cb86"><pre class="downlit">
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/ivanek/Gviz">Gviz</a></span><span class="op">)</span>
<span class="va">axisTrack</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/Gviz/man/GenomeAxisTrack-class.html">GenomeAxisTrack</a></span><span class="op">(</span><span class="op">)</span>
<span class="va">dTrack1</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/Gviz/man/DataTrack-class.html">DataTrack</a></span><span class="op">(</span>range<span class="op">=</span><span class="va">gr1</span>, name<span class="op">=</span><span class="st">"Sample 1"</span>, type<span class="op">=</span><span class="st">"h"</span><span class="op">)</span>
<span class="va">dTrack2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/Gviz/man/DataTrack-class.html">DataTrack</a></span><span class="op">(</span>range<span class="op">=</span><span class="va">gr2</span>, name<span class="op">=</span><span class="st">"Sample 2"</span>, type<span class="op">=</span><span class="st">"h"</span><span class="op">)</span>
<span class="va">txTrack</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/Gviz/man/GeneRegionTrack-class.html">GeneRegionTrack</a></span><span class="op">(</span><span class="va">txdb</span>, name<span class="op">=</span><span class="st">"Transcripts"</span>, showId<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/pkg/Gviz/man/plotTracks.html">plotTracks</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span><span class="va">axisTrack</span>, <span class="va">dTrack1</span>, <span class="va">dTrack2</span>, <span class="va">txTrack</span><span class="op">)</span>,
           chromosome<span class="op">=</span><span class="st">"chr3"</span>, extend.left<span class="op">=</span><span class="fl">1000</span><span class="op">)</span></pre></div>
<p><img src="QuasR_files/figure-html/ChIP_visualize-1.png" width="768"></p>
</div>
<div id="create-a-genomic-profile-for-a-set-of-regions-using-qprofile" class="section level3">
<h3 class="hasAnchor">
<a href="#create-a-genomic-profile-for-a-set-of-regions-using-qprofile" class="anchor"></a>Create a genomic profile for a set of regions using <code>qProfile</code>
</h3>
<p>Given a set of anchor positions in the genome, <code>qProfile</code> calculates the number of nearby alignments relative to the anchor position, for example to generate a average profile. The neighborhood around anchor positions can be specified by the <code>upstream</code> and <code>downstream</code> argument. Alignments that are upstream of an anchor position will have a negative relative position, and downstream alignments a positive. The anchor positions are all aligned at position zero in the return value.</p>
<p>Anchor positions will be provided to <code>qProfile</code> using the <code>query</code> argument, which takes a <code>GRanges</code> object. The anchor positions correspond to <code><a href="https://rdrr.io/r/stats/start.html">start()</a></code> for regions on <code><a href="https://rdrr.io/r/base/Arithmetic.html">+</a></code> or <code><a href="https://rdrr.io/r/base/Arithmetic.html">*</a></code> strands, and to <code><a href="https://rdrr.io/r/stats/start.html">end()</a></code> for regions on the <code><a href="https://rdrr.io/r/base/Arithmetic.html">-</a></code> strand. As mentioned above, we expect H3K4me3 ChIP-seq alignments to be enriched around the transcript start site of genes. We can therefore construct a suitable <code>query</code> object from the start sites of known genes. In the code below, start sites (<code>start_codon</code>) are imported from a <code>.gtf</code> file with the help of the <em><a href="https://bioconductor.org/packages/3.12/rtracklayer">rtracklayer</a></em> package. In addition, <code>strand</code> and <code>gene_name</code> are also selected for import. Duplicated start sites, e.g. from genes with multiple transcripts, are removed. Finally, all regions are given the name <code>TSS</code>, because <code>qProfile</code> combines regions with identical names into a single profile.</p>
<div class="sourceCode" id="cb87"><pre class="downlit">
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">rtracklayer</span><span class="op">)</span>
<span class="va">annotationFile</span> <span class="op">&lt;-</span> <span class="st">"extdata/hg19sub_annotation.gtf"</span>
<span class="va">tssRegions</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/rtracklayer/man/GFFFile-class.html">import.gff</a></span><span class="op">(</span><span class="va">annotationFile</span>, format<span class="op">=</span><span class="st">"gtf"</span>,
                         feature.type<span class="op">=</span><span class="st">"start_codon"</span>,
                         colnames<span class="op">=</span><span class="st">"gene_name"</span><span class="op">)</span>
<span class="va">tssRegions</span> <span class="op">&lt;-</span> <span class="va">tssRegions</span><span class="op">[</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/duplicated.html">duplicated</a></span><span class="op">(</span><span class="va">tssRegions</span><span class="op">)</span><span class="op">]</span>
<span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">tssRegions</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/rep.html">rep</a></span><span class="op">(</span><span class="st">"TSS"</span>, <span class="fu"><a href="https://rdrr.io/r/base/length.html">length</a></span><span class="op">(</span><span class="va">tssRegions</span><span class="op">)</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">tssRegions</span><span class="op">)</span></pre></div>
<pre><code>## GRanges object with 6 ranges and 1 metadata column:
##       seqnames      ranges strand |   gene_name
##          &lt;Rle&gt;   &lt;IRanges&gt;  &lt;Rle&gt; | &lt;character&gt;
##   TSS     chr1   6949-6951      - |    TNFRSF18
##   TSS     chr1 14505-14507      - |     TNFRSF4
##   TSS     chr1 29171-29173      - |        SDF4
##   TSS     chr1 32659-32661      + |     B3GALT6
##   TSS     chr2   3200-3202      + |        RPS7
##   TSS     chr3   2386-2388      + |     C3orf10
##   -------
##   seqinfo: 3 sequences from an unspecified genome; no seqlengths</code></pre>
<p>Alignments around the <code>tssRegions</code> coordinates are counted in a window defined by the <code>upstream</code> and <code>downstream</code> arguments, which specify the number of bases to include around each anchor position. For <code>query</code> regions on <code><a href="https://rdrr.io/r/base/Arithmetic.html">+</a></code> or <code><a href="https://rdrr.io/r/base/Arithmetic.html">*</a></code> strands, upstream refers to the left side of the anchor position (lower coordinates), while for regions on the <code><a href="https://rdrr.io/r/base/Arithmetic.html">-</a></code> strand, upstream refers to the right side (higher coordinates). The following example creates separate profiles for alignments on the <em>same</em> and on the <em>opposite</em> strand of the regions in <code>query</code>.</p>
<div class="sourceCode" id="cb89"><pre class="downlit">
<span class="va">prS</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/qProfile.html">qProfile</a></span><span class="op">(</span><span class="va">proj1</span>, <span class="va">tssRegions</span>, upstream<span class="op">=</span><span class="fl">3000</span>, downstream<span class="op">=</span><span class="fl">3000</span>, 
                orientation<span class="op">=</span><span class="st">"same"</span><span class="op">)</span></pre></div>
<pre><code>## profiling alignments...done</code></pre>
<div class="sourceCode" id="cb91"><pre class="downlit">
<span class="va">prO</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/qProfile.html">qProfile</a></span><span class="op">(</span><span class="va">proj1</span>, <span class="va">tssRegions</span>, upstream<span class="op">=</span><span class="fl">3000</span>, downstream<span class="op">=</span><span class="fl">3000</span>, 
                orientation<span class="op">=</span><span class="st">"opposite"</span><span class="op">)</span></pre></div>
<pre><code>## profiling alignments...done</code></pre>
<div class="sourceCode" id="cb93"><pre class="downlit">
<span class="fu"><a href="https://rdrr.io/r/base/lapply.html">lapply</a></span><span class="op">(</span><span class="va">prS</span>, <span class="st">"["</span>, , <span class="fl">1</span><span class="op">:</span><span class="fl">10</span><span class="op">)</span></pre></div>
<pre><code>## $coverage
## -3000 -2999 -2998 -2997 -2996 -2995 -2994 -2993 -2992 -2991 
##     8     8     8     8     8     8     8     8     8     8 
## 
## $Sample1
## -3000 -2999 -2998 -2997 -2996 -2995 -2994 -2993 -2992 -2991 
##     1     0     0     0     0     0     0     0     0     0 
## 
## $Sample2
## -3000 -2999 -2998 -2997 -2996 -2995 -2994 -2993 -2992 -2991 
##     0     0     0     2     0     0     1     1     1     0</code></pre>
<p>The counts returned by <code>qProfile</code> are the raw number of alignments per sample and position, without any normalization for the number of query regions or the total number of alignments in a sample per position. To obtain the average number of alignments, we divide the alignment counts by the number of <code>query</code> regions that covered a given relative position around the anchor sites. This coverage is available as the first element in the return value. The shift between <em>same</em> and <em>opposite</em> strand alignments is indicative for the average length of the sequenced ChIP fragments.</p>
<div class="sourceCode" id="cb95"><pre class="downlit">
<span class="va">prCombS</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/do.call.html">do.call</a></span><span class="op">(</span><span class="st">"+"</span>, <span class="va">prS</span><span class="op">[</span><span class="op">-</span><span class="fl">1</span><span class="op">]</span><span class="op">)</span> <span class="op">/</span><span class="va">prS</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span>
<span class="va">prCombO</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/do.call.html">do.call</a></span><span class="op">(</span><span class="st">"+"</span>, <span class="va">prO</span><span class="op">[</span><span class="op">-</span><span class="fl">1</span><span class="op">]</span><span class="op">)</span> <span class="op">/</span><span class="va">prO</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span>

<span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html">plot</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/numeric.html">as.numeric</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span><span class="op">(</span><span class="va">prCombS</span><span class="op">)</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/stats/filter.html">filter</a></span><span class="op">(</span><span class="va">prCombS</span><span class="op">[</span><span class="fl">1</span>,<span class="op">]</span>, <span class="fu"><a href="https://rdrr.io/r/base/rep.html">rep</a></span><span class="op">(</span><span class="fl">1</span><span class="op">/</span><span class="fl">100</span>,<span class="fl">100</span><span class="op">)</span><span class="op">)</span>, 
     type<span class="op">=</span><span class="st">'l'</span>, xlab<span class="op">=</span><span class="st">"Position relative to TSS"</span>, ylab<span class="op">=</span><span class="st">"Mean no. of alignments"</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/graphics/lines.html">lines</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/numeric.html">as.numeric</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span><span class="op">(</span><span class="va">prCombO</span><span class="op">)</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/stats/filter.html">filter</a></span><span class="op">(</span><span class="va">prCombO</span><span class="op">[</span><span class="fl">1</span>,<span class="op">]</span>, <span class="fu"><a href="https://rdrr.io/r/base/rep.html">rep</a></span><span class="op">(</span><span class="fl">1</span><span class="op">/</span><span class="fl">100</span>,<span class="fl">100</span><span class="op">)</span><span class="op">)</span>, 
      type<span class="op">=</span><span class="st">'l'</span>, col<span class="op">=</span><span class="st">"red"</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/graphics/legend.html">legend</a></span><span class="op">(</span>title<span class="op">=</span><span class="st">"strand"</span>, legend<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"same as query"</span>,<span class="st">"opposite of query"</span><span class="op">)</span>, 
       x<span class="op">=</span><span class="st">"topleft"</span>, col<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"black"</span>,<span class="st">"red"</span><span class="op">)</span>, lwd<span class="op">=</span><span class="fl">1.5</span>, bty<span class="op">=</span><span class="st">"n"</span>, title.adj<span class="op">=</span><span class="fl">0.1</span><span class="op">)</span></pre></div>
<p><img src="QuasR_files/figure-html/ChIP_visualizeProfile-1.png" width="768"></p>
</div>
<div id="using-a-bsgenome-package-as-reference-genome" class="section level3">
<h3 class="hasAnchor">
<a href="#using-a-bsgenome-package-as-reference-genome" class="anchor"></a>Using a <em><a href="https://bioconductor.org/packages/3.12/BSgenome">BSgenome</a></em> package as reference genome</h3>
<p><em><a href="https://bioconductor.org/packages/3.12/QuasR">QuasR</a></em> also allows using of <em><a href="https://bioconductor.org/packages/3.12/BSgenome">BSgenome</a></em> packages instead of a <code>fasta</code> file as reference genome (see section @ref(genome)). To use a <em><a href="https://bioconductor.org/packages/3.12/BSgenome">BSgenome</a></em>, the <code>genome</code> argument of <code>qAlign</code> is set to a string matching the name of a <em><a href="https://bioconductor.org/packages/3.12/BSgenome">BSgenome</a></em> package, for example <code>"BSgenome.Hsapiens.UCSC.hg19"</code>. If that package is not already installed, <code>qAlign</code> will check if it is available from <a href="http://bioconductor.org/" class="uri">http://bioconductor.org/</a> and download it automatically. The corresponding alignment index will be saved as a new package, named after the original <em><a href="https://bioconductor.org/packages/3.12/BSgenome">BSgenome</a></em> package and the aligner used to build the index, for example <code>BSgenome.Hsapiens.UCSC.hg19.Rbowtie</code>.</p>
<p>The code example below illustrates the use of a <em><a href="https://bioconductor.org/packages/3.12/BSgenome">BSgenome</a></em> reference genome for the same example data as above. Running it for the first time will take several hours in order to build the aligner index:</p>
<div class="sourceCode" id="cb96"><pre class="downlit">
<span class="fu"><a href="https://rdrr.io/r/base/files.html">file.copy</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span><span class="op">(</span>package<span class="op">=</span><span class="st">"QuasR"</span>, <span class="st">"extdata"</span><span class="op">)</span>, <span class="st">"."</span>, recursive<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span>

<span class="va">sampleFile</span> <span class="op">&lt;-</span> <span class="st">"extdata/samples_chip_single.txt"</span>
<span class="va">auxFile</span> <span class="op">&lt;-</span> <span class="st">"extdata/auxiliaries.txt"</span>

<span class="fu"><a href="https://rdrr.io/pkg/BSgenome/man/available.genomes.html">available.genomes</a></span><span class="op">(</span><span class="op">)</span> <span class="co"># list available genomes</span>
<span class="va">genomeName</span> <span class="op">&lt;-</span> <span class="st">"BSgenome.Hsapiens.UCSC.hg19"</span>

<span class="va">proj1</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/qAlign.html">qAlign</a></span><span class="op">(</span><span class="va">sampleFile</span>, genome<span class="op">=</span><span class="va">genomeName</span>, auxiliaryFile<span class="op">=</span><span class="va">auxFile</span><span class="op">)</span>
<span class="va">proj1</span></pre></div>
</div>
</div>
<div id="RNAseq" class="section level2">
<h2 class="hasAnchor">
<a href="#RNAseq" class="anchor"></a>RNA-seq: Gene expression profiling</h2>
<p>In <em><a href="https://bioconductor.org/packages/3.12/QuasR">QuasR</a></em>, an analysis workflow for an RNA-seq dataset is very similar to the one described above for a ChIP-seq experiment. The major difference is that here reads are aligned using <code><a href="../reference/qAlign.html">qAlign(..., splicedAlignment=TRUE, aligner="Rhisat2")</a></code>, which will cause <code>qAlign</code> to align reads with the <code>HISAT2</code> aligner <span class="citation">(Kim, Langmead, and Salzberg 2015)</span> (via the <em><a href="https://bioconductor.org/packages/3.12/Rhisat2">Rhisat2</a></em> package), rather than with <code>bowtie</code> <span class="citation">(Langmead et al. 2009)</span>. Before the <em><a href="https://bioconductor.org/packages/3.12/Rhisat2">Rhisat2</a></em> package was available (introduced in Bioconductor 3.9), <code>qAlign(... splicedAlignment=TRUE)</code> aligned reads using <code>SpliceMap</code> <span class="citation">(Au et al. 2010)</span>, which is not recommended now but still possible in order to reproduce old results. Spliced paired-end alignments are also supported; the <code>splicedAlignment</code> argument can be freely combined with the <code>paired</code> argument. In addition, HISAT2 also allows the specification of known splice sites, which can help in the read alignment. This is done by specifying the argument <code>geneAnnotation</code> in <code><a href="../reference/qAlign.html">qAlign()</a></code>, to either a <code>.gtf</code> file or a <code>sqlite</code> database generated by exporting a <code>TxDb</code> object.</p>
<div id="spliced-alignment-of-rna-seq-reads" class="section level3">
<h3 class="hasAnchor">
<a href="#spliced-alignment-of-rna-seq-reads" class="anchor"></a>Spliced alignment of RNA-seq reads</h3>
<p>We start the example workflow by copying the example data files into the current working directly, into a subfolder called <code>"extdata"</code>, and then create spliced alignments using <code>qAlign</code>:</p>
<div class="sourceCode" id="cb97"><pre class="downlit">
<span class="fu"><a href="https://rdrr.io/r/base/files.html">file.copy</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span><span class="op">(</span>package<span class="op">=</span><span class="st">"QuasR"</span>, <span class="st">"extdata"</span><span class="op">)</span>, <span class="st">"."</span>, recursive<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span></pre></div>
<pre><code>## [1] TRUE</code></pre>
<div class="sourceCode" id="cb99"><pre class="downlit">
<span class="va">sampleFile</span> <span class="op">&lt;-</span> <span class="st">"extdata/samples_rna_paired.txt"</span>
<span class="va">genomeFile</span> <span class="op">&lt;-</span> <span class="st">"extdata/hg19sub.fa"</span>

<span class="va">proj2</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/qAlign.html">qAlign</a></span><span class="op">(</span><span class="va">sampleFile</span>, genome<span class="op">=</span><span class="va">genomeFile</span>, splicedAlignment<span class="op">=</span><span class="cn">TRUE</span>, aligner<span class="op">=</span><span class="st">"Rhisat2"</span><span class="op">)</span></pre></div>
<pre><code>## alignment files missing - need to:</code></pre>
<pre><code>##     create alignment index for the genome</code></pre>
<pre><code>##     create 2 genomic alignment(s)</code></pre>
<pre><code>## Creating an Rhisat2 index for /Users/runner/work/QuasR/QuasR/vignettes/extdata/hg19sub.fa</code></pre>
<pre><code>## Finished creating index</code></pre>
<pre><code>## Testing the compute nodes...OK
## Loading QuasR on the compute nodes...OK
## Available cores:
## Mac-1602493909147.local: 1
## Performing genomic alignments for 2 samples. See progress in the log file:
## /Users/runner/work/QuasR/QuasR/vignettes/QuasR_log_17d54cdd8628.txt
## Genomic alignments have been created successfully</code></pre>
<div class="sourceCode" id="cb106"><pre class="downlit">
<span class="va">proj2</span></pre></div>
<pre><code>## Project: qProject
##  Options   : maxHits         : 1
##              paired          : fr
##              splicedAlignment: TRUE
##              bisulfite       : no
##              snpFile         : none
##              geneAnnotation  : none
##  Aligner   : Rhisat2 v1.5.2 (parameters: -k 2)
##  Genome    : /Users/runner/work/QuasR/QuasR/vignettes/extdata/hg19sub.fa (file)
## 
##  Reads     : 2 pairs of files, 2 samples (fastq format):
##    1. rna_1_1.fq.bz2  rna_1_2.fq.bz2  Sample1 (phred33)
##    2. rna_2_1.fq.bz2  rna_2_2.fq.bz2  Sample2 (phred33)
## 
##  Genome alignments: directory: same as reads
##    1. rna_1_1_17d56392d384.bam
##    2. rna_2_1_17d53c7cb22e.bam
## 
##  Aux. alignments: none</code></pre>
<p>Aligning the reads with <code>splicedAlignment=TRUE</code> will allow to also align reads that cross exon junctions, and thus have a large deletion (the intron) relative to the reference genome.</p>
<div class="sourceCode" id="cb108"><pre class="downlit">
<span class="va">proj2unspl</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/qAlign.html">qAlign</a></span><span class="op">(</span><span class="va">sampleFile</span>, genome<span class="op">=</span><span class="va">genomeFile</span>,
                     splicedAlignment<span class="op">=</span><span class="cn">FALSE</span><span class="op">)</span></pre></div>
<pre><code>## alignment files missing - need to:</code></pre>
<pre><code>##     create 2 genomic alignment(s)</code></pre>
<pre><code>## Testing the compute nodes...OK
## Loading QuasR on the compute nodes...OK
## Available cores:
## Mac-1602493909147.local: 1
## Performing genomic alignments for 2 samples. See progress in the log file:
## /Users/runner/work/QuasR/QuasR/vignettes/QuasR_log_17d53378e425.txt
## Genomic alignments have been created successfully</code></pre>
<div class="sourceCode" id="cb112"><pre class="downlit">
<span class="fu"><a href="../reference/alignmentStats.html">alignmentStats</a></span><span class="op">(</span><span class="va">proj2</span><span class="op">)</span></pre></div>
<pre><code>##                seqlength mapped unmapped
## Sample1:genome     95000   5961        6
## Sample2:genome     95000   5914        2</code></pre>
<div class="sourceCode" id="cb114"><pre class="downlit">
<span class="fu"><a href="../reference/alignmentStats.html">alignmentStats</a></span><span class="op">(</span><span class="va">proj2unspl</span><span class="op">)</span></pre></div>
<pre><code>##                seqlength mapped unmapped
## Sample1:genome     95000   2258     3746
## Sample2:genome     95000   2652     3348</code></pre>
</div>
<div id="quantification-of-gene-and-exon-expression" class="section level3">
<h3 class="hasAnchor">
<a href="#quantification-of-gene-and-exon-expression" class="anchor"></a>Quantification of gene and exon expression</h3>
<p>As with ChIP-seq experiments, <code>qCount</code> is used to quantify alignments. For quantification of gene or exon expression levels, <code>qCount</code> can be called with a query of type <code>TxDb</code>, such as the one we constructed in the ChIP-seq workflow above from a <code>.gtf</code> file. The argument <code>reportLevel</code> can be used to control if annotated exonic regions should be quantified independently (<code>reportLevel="exon"</code>) or non-redundantly combined per gene (<code>reportLevel="gene"</code>):</p>
<div class="sourceCode" id="cb116"><pre class="downlit">
<span class="va">geneLevels</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/qCount.html">qCount</a></span><span class="op">(</span><span class="va">proj2</span>, <span class="va">txdb</span>, reportLevel<span class="op">=</span><span class="st">"gene"</span><span class="op">)</span></pre></div>
<pre><code>## extracting gene regions from TxDb...done
## counting alignments...done
## collapsing counts by query name...done</code></pre>
<div class="sourceCode" id="cb118"><pre class="downlit">
<span class="va">exonLevels</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/qCount.html">qCount</a></span><span class="op">(</span><span class="va">proj2</span>, <span class="va">txdb</span>, reportLevel<span class="op">=</span><span class="st">"exon"</span><span class="op">)</span></pre></div>
<pre><code>## extracting exon regions from TxDb...done
## counting alignments...done</code></pre>
<div class="sourceCode" id="cb120"><pre class="downlit">
<span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">geneLevels</span><span class="op">)</span></pre></div>
<pre><code>##                 width Sample1 Sample2
## ENSG00000078808  4697     710    1083
## ENSG00000134075   589    1173    1303
## ENSG00000134086  4213     279     295
## ENSG00000171863  5583    2924    2224
## ENSG00000176022  2793      62     344
## ENSG00000186827  1721      37       8</code></pre>
<div class="sourceCode" id="cb122"><pre class="downlit">
<span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">exonLevels</span><span class="op">)</span></pre></div>
<pre><code>##    width Sample1 Sample2
## 1   2793      62     344
## 10   187       3       0
## 11   307       3       0
## 12   300      11       2
## 13   493      19       2
## 14   129       7       0</code></pre>
</div>
<div id="calculation-of-rpkm-expression-values" class="section level3">
<h3 class="hasAnchor">
<a href="#calculation-of-rpkm-expression-values" class="anchor"></a>Calculation of RPKM expression values</h3>
<p>The values returned by <code>qCount</code> are the number of alignments. Sometimes it is required to normalize for the length of query regions, or the size of the libraries. For example, gene expression levels in the form of <em>RPKM</em> values (reads per kilobase of transcript and million mapped reads) can be obtained as follows:</p>
<div class="sourceCode" id="cb124"><pre class="downlit">
<span class="va">geneRPKM</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/t.html">t</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/t.html">t</a></span><span class="op">(</span><span class="va">geneLevels</span><span class="op">[</span>,<span class="op">-</span><span class="fl">1</span><span class="op">]</span> <span class="op">/</span><span class="va">geneLevels</span><span class="op">[</span>,<span class="fl">1</span><span class="op">]</span> <span class="op">*</span><span class="fl">1000</span><span class="op">)</span>
              <span class="op">/</span><span class="fu"><a href="https://rdrr.io/r/base/colSums.html">colSums</a></span><span class="op">(</span><span class="va">geneLevels</span><span class="op">[</span>,<span class="op">-</span><span class="fl">1</span><span class="op">]</span><span class="op">)</span> <span class="op">*</span><span class="fl">1e6</span><span class="op">)</span>
<span class="va">geneRPKM</span></pre></div>
<pre><code>##                 Sample1 Sample2
## ENSG00000078808   21350   31786
## ENSG00000134075  281287  304966
## ENSG00000134086    9354    9653
## ENSG00000171863   73974   54915
## ENSG00000176022    3135   16979
## ENSG00000186827    3037     641
## ENSG00000186891    2681     201
## ENSG00000238345       0       0
## ENSG00000238642       0       0
## ENSG00000247886       0       0
## ENSG00000252531    6066    1691
## ENSG00000254999  213296  222826</code></pre>
<p>Please note the RPKM values in our example are higher than what you would usually get for a real RNA-seq dataset. The values here are artificially scaled up because our example data contains reads only for a small number of genes.</p>
</div>
<div id="analysis-of-alternative-splicing-quantification-of-exon-exon-junctions" class="section level3">
<h3 class="hasAnchor">
<a href="#analysis-of-alternative-splicing-quantification-of-exon-exon-junctions" class="anchor"></a>Analysis of alternative splicing: Quantification of exon-exon junctions</h3>
<p>Exon-exon junctions can be quantified by setting <code>reportLevel="junction"</code>. In this case, <code>qCount</code> will ignore the <code>query</code> argument and scan all alignments for any detected splices, which are returned as a <code>GRanges</code> object: The region start and end coordinates correspond to the first and last bases of the intron, and the counts are returned in the <code>mcols()</code> of the <code>GRanges</code> object. Alignments that are identically spliced but reside on opposite strands will be quantified separately. In an unstranded RNA-seq experiment, this may give rise to two separate counts for the same intron, one each for the supporting alignments on plus and minus strands.</p>
<div class="sourceCode" id="cb126"><pre class="downlit">
<span class="va">exonJunctions</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/qCount.html">qCount</a></span><span class="op">(</span><span class="va">proj2</span>, <span class="cn">NULL</span>, reportLevel<span class="op">=</span><span class="st">"junction"</span><span class="op">)</span></pre></div>
<pre><code>## counting junctions...done</code></pre>
<div class="sourceCode" id="cb128"><pre class="downlit">
<span class="va">exonJunctions</span></pre></div>
<pre><code>## GRanges object with 46 ranges and 2 metadata columns:
##        seqnames      ranges strand |   Sample1   Sample2
##           &lt;Rle&gt;   &lt;IRanges&gt;  &lt;Rle&gt; | &lt;numeric&gt; &lt;numeric&gt;
##    [1]     chr1 12213-12321      + |         3         0
##    [2]     chr1 13085-13371      - |         1         0
##    [3]     chr1 18069-18837      + |         9        16
##    [4]     chr1 18069-18837      - |         7         4
##    [5]     chr1 18185-18837      - |         2         0
##    ...      ...         ...    ... .       ...       ...
##   [42]     chr1 14166-14362      + |         0         1
##   [43]     chr1 19308-23623      - |         0         2
##   [44]     chr1 29327-32271      + |         0         2
##   [45]     chr1 29327-32271      - |         0         1
##   [46]     chr3   2504-5589      - |         0         3
##   -------
##   seqinfo: 3 sequences from an unspecified genome; no seqlengths</code></pre>
<p>About half of the exon-exon junctions detected in this sample dataset correspond to known introns; they tend to be the ones with higher coverage:</p>
<div class="sourceCode" id="cb130"><pre class="downlit">
<span class="va">knownIntrons</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/unlist.html">unlist</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/GenomicFeatures/man/transcriptsBy.html">intronsByTranscript</a></span><span class="op">(</span><span class="va">txdb</span><span class="op">)</span><span class="op">)</span>
<span class="va">isKnown</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/GenomicRanges/man/findOverlaps-methods.html">overlapsAny</a></span><span class="op">(</span><span class="va">exonJunctions</span>, <span class="va">knownIntrons</span>, type<span class="op">=</span><span class="st">"equal"</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/table.html">table</a></span><span class="op">(</span><span class="va">isKnown</span><span class="op">)</span></pre></div>
<pre><code>## isKnown
## FALSE  TRUE 
##    25    21</code></pre>
<div class="sourceCode" id="cb132"><pre class="downlit">
<span class="fu"><a href="https://rdrr.io/r/base/tapply.html">tapply</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/colSums.html">rowSums</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/matrix.html">as.matrix</a></span><span class="op">(</span><span class="fu">mcols</span><span class="op">(</span><span class="va">exonJunctions</span><span class="op">)</span><span class="op">)</span><span class="op">)</span>,
       <span class="va">isKnown</span>, <span class="va">summary</span><span class="op">)</span></pre></div>
<pre><code>## $`FALSE`
##    Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
##       1       3       7      47      31     340 
## 
## $`TRUE`
##    Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
##     1.0     2.0    16.0    50.6    91.0   210.0</code></pre>
<p>When quantifying exon junctions, only spliced alignments will be included in the quantification. It is also possible to only include unspliced alignments in the quantification, for example when counting exon body alignments that complement the exon junction alignments. This can be done using the <code>includeSpliced</code> argument from <code>qCount</code>:</p>
<div class="sourceCode" id="cb134"><pre class="downlit">
<span class="va">exonBodyLevels</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/qCount.html">qCount</a></span><span class="op">(</span><span class="va">proj2</span>, <span class="va">txdb</span>, reportLevel<span class="op">=</span><span class="st">"exon"</span>, includeSpliced<span class="op">=</span><span class="cn">FALSE</span><span class="op">)</span></pre></div>
<pre><code>## extracting exon regions from TxDb...done
## counting alignments...done</code></pre>
<div class="sourceCode" id="cb136"><pre class="downlit">
<span class="fu"><a href="https://rdrr.io/r/base/summary.html">summary</a></span><span class="op">(</span><span class="va">exonLevels</span> <span class="op">-</span> <span class="va">exonBodyLevels</span><span class="op">)</span></pre></div>
<pre><code>##      width      Sample1       Sample2   
##  Min.   :0   Min.   :  0   Min.   :  0  
##  1st Qu.:0   1st Qu.:  0   1st Qu.:  0  
##  Median :0   Median :  3   Median :  1  
##  Mean   :0   Mean   : 42   Mean   : 35  
##  3rd Qu.:0   3rd Qu.: 48   3rd Qu.: 50  
##  Max.   :0   Max.   :818   Max.   :647</code></pre>
<pre><code>## collecting quality control data</code></pre>
<pre><code>## creating QC plots</code></pre>
</div>
</div>
<div id="smrna-seq-small-rna-and-mirna-expression-profiling" class="section level2">
<h2 class="hasAnchor">
<a href="#smrna-seq-small-rna-and-mirna-expression-profiling" class="anchor"></a>smRNA-seq: small RNA and miRNA expression profiling</h2>
<p>Expression profiling of miRNAs differs only slightly from the profiling of mRNAs. There are a few details that need special care, which are outlined in this section.</p>
<div id="preprocessing-of-small-rna-mirna-reads" class="section level3">
<h3 class="hasAnchor">
<a href="#preprocessing-of-small-rna-mirna-reads" class="anchor"></a>Preprocessing of small RNA (miRNA) reads</h3>
<p>Again, we start the example workflow by copying the example data files into the current working directly, into a subfolder called <code>"extdata"</code>.</p>
<div class="sourceCode" id="cb140"><pre class="downlit">
<span class="fu"><a href="https://rdrr.io/r/base/files.html">file.copy</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span><span class="op">(</span>package<span class="op">=</span><span class="st">"QuasR"</span>, <span class="st">"extdata"</span><span class="op">)</span>, <span class="st">"."</span>, recursive<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span></pre></div>
<pre><code>## [1] TRUE</code></pre>
<p>As a next step, we need to remove library adapter sequences from short RNA reads. Most sequencing experiments generate reads that are longer than the average length of a miRNA (22nt). Therefore, the read sequence will run through the miRNA into the library adapter sequence and would not match when aligned in full to the reference genome.</p>
<p>We can remove those adapter sequences using <code>preprocessReads</code> (see section @ref(preprocessReads) for more details), which for each input sequence file will generate an output sequence file containing appropriately truncated sequences. In the example below, we get the input sequence filenames from <code>sampleFile</code>, and also prepare an updated <code>sampleFile2</code> that refers to newly generated processed sequence files:</p>
<div class="sourceCode" id="cb142"><pre class="downlit">
<span class="co"># prepare sample file with processed reads filenames</span>
<span class="va">sampleFile</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html">file.path</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"samples_mirna.txt"</span><span class="op">)</span>
<span class="va">sampleFile</span></pre></div>
<pre><code>## [1] "extdata/samples_mirna.txt"</code></pre>
<div class="sourceCode" id="cb144"><pre class="downlit">
<span class="va">sampleFile2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/grep.html">sub</a></span><span class="op">(</span><span class="st">".txt"</span>, <span class="st">"_processed.txt"</span>, <span class="va">sampleFile</span><span class="op">)</span>
<span class="va">sampleFile2</span></pre></div>
<pre><code>## [1] "extdata/samples_mirna_processed.txt"</code></pre>
<div class="sourceCode" id="cb146"><pre class="downlit">
<span class="va">tab</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html">read.delim</a></span><span class="op">(</span><span class="va">sampleFile</span>, header<span class="op">=</span><span class="cn">TRUE</span>, as.is<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span>
<span class="va">tab</span></pre></div>
<pre><code>##     FileName SampleName
## 1 mirna_1.fa     miRNAs</code></pre>
<div class="sourceCode" id="cb148"><pre class="downlit">
<span class="va">tab2</span> <span class="op">&lt;-</span> <span class="va">tab</span>
<span class="va">tab2</span><span class="op">$</span><span class="va">FileName</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/grep.html">sub</a></span><span class="op">(</span><span class="st">".fa"</span>, <span class="st">"_processed.fa"</span>, <span class="va">tab</span><span class="op">$</span><span class="va">FileName</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/utils/write.table.html">write.table</a></span><span class="op">(</span><span class="va">tab2</span>, <span class="va">sampleFile2</span>, sep<span class="op">=</span><span class="st">"\t"</span>, quote<span class="op">=</span><span class="cn">FALSE</span>, row.names<span class="op">=</span><span class="cn">FALSE</span><span class="op">)</span>
<span class="va">tab2</span></pre></div>
<pre><code>##               FileName SampleName
## 1 mirna_1_processed.fa     miRNAs</code></pre>
<div class="sourceCode" id="cb150"><pre class="downlit">
<span class="co"># remove adapters</span>
<span class="va">oldwd</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/getwd.html">setwd</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/basename.html">dirname</a></span><span class="op">(</span><span class="va">sampleFile</span><span class="op">)</span><span class="op">)</span>
<span class="va">res</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/preprocessReads.html">preprocessReads</a></span><span class="op">(</span><span class="va">tab</span><span class="op">$</span><span class="va">FileName</span>,
                       <span class="va">tab2</span><span class="op">$</span><span class="va">FileName</span>,
                       Rpattern<span class="op">=</span><span class="st">"TGGAATTCTCGGGTGCCAAGG"</span><span class="op">)</span></pre></div>
<pre><code>##   filtering mirna_1.fa</code></pre>
<div class="sourceCode" id="cb152"><pre class="downlit">
<span class="va">res</span></pre></div>
<pre><code>##                  mirna_1.fa
## totalSequences         1000
## matchTo5pAdapter          0
## matchTo3pAdapter       1000
## tooShort                  0
## tooManyN                  0
## lowComplexity             0
## totalPassed            1000</code></pre>
<div class="sourceCode" id="cb154"><pre class="downlit">
<span class="fu"><a href="https://rdrr.io/r/base/getwd.html">setwd</a></span><span class="op">(</span><span class="va">oldwd</span><span class="op">)</span></pre></div>
<p>The miRNA reads in <code>mirna_1.fa</code> are by the way synthetic sequences and do not correspond to any existing miRNAs. As you can see above from the return value of <code>preprocessReads</code>, all reads matched to the 3’-adapter and were therefore truncated, reducing their length to roughly the expected 22nt:</p>
<div class="sourceCode" id="cb155"><pre class="downlit">
<span class="co"># get read lengths</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">Biostrings</span><span class="op">)</span>
<span class="va">oldwd</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/getwd.html">setwd</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/basename.html">dirname</a></span><span class="op">(</span><span class="va">sampleFile</span><span class="op">)</span><span class="op">)</span>
<span class="va">lens</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/Biostrings/man/XStringSet-io.html">fasta.seqlengths</a></span><span class="op">(</span><span class="va">tab</span><span class="op">$</span><span class="va">FileName</span>, nrec<span class="op">=</span><span class="fl">1e5</span><span class="op">)</span>
<span class="va">lens2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/Biostrings/man/XStringSet-io.html">fasta.seqlengths</a></span><span class="op">(</span><span class="va">tab2</span><span class="op">$</span><span class="va">FileName</span>, nrec<span class="op">=</span><span class="fl">1e5</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/getwd.html">setwd</a></span><span class="op">(</span><span class="va">oldwd</span><span class="op">)</span>
<span class="co"># plot length distribution</span>
<span class="va">lensTab</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/cbind.html">rbind</a></span><span class="op">(</span>raw<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/tabulate.html">tabulate</a></span><span class="op">(</span><span class="va">lens</span>,<span class="fl">50</span><span class="op">)</span>,
                 processed<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/tabulate.html">tabulate</a></span><span class="op">(</span><span class="va">lens2</span>,<span class="fl">50</span><span class="op">)</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span><span class="op">(</span><span class="va">lensTab</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fl">1</span><span class="op">:</span><span class="fl">50</span>
<span class="fu"><a href="https://rdrr.io/r/graphics/barplot.html">barplot</a></span><span class="op">(</span><span class="va">lensTab</span><span class="op">/</span><span class="fu"><a href="https://rdrr.io/r/base/colSums.html">rowSums</a></span><span class="op">(</span><span class="va">lensTab</span><span class="op">)</span><span class="op">*</span><span class="fl">100</span>,
        xlab<span class="op">=</span><span class="st">"Read length (nt)"</span>, ylab<span class="op">=</span><span class="st">"Percent of reads"</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/graphics/legend.html">legend</a></span><span class="op">(</span>x<span class="op">=</span><span class="st">"topleft"</span>, bty<span class="op">=</span><span class="st">"n"</span>, fill<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/grDevices/gray.colors.html">gray.colors</a></span><span class="op">(</span><span class="fl">2</span><span class="op">)</span>, legend<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span><span class="op">(</span><span class="va">lensTab</span><span class="op">)</span><span class="op">)</span></pre></div>
<p><img src="QuasR_files/figure-html/miRNA_lengthes-1.png" width="576"></p>
</div>
<div id="alignment-of-small-rna-mirna-reads" class="section level3">
<h3 class="hasAnchor">
<a href="#alignment-of-small-rna-mirna-reads" class="anchor"></a>Alignment of small RNA (miRNA) reads</h3>
<p>Next, we create alignments using <code>qAlign</code>. In contrast to the general RNA-seq workflow (section @ref(RNAseq)), alignment time can be reduced by using the default unspliced alignment (<code>splicedAlignment=FALSE</code>). Importantly, we need to set <code>maxHits=50</code> or similar to also align reads that perfectly match the genome multiple times. This is required because of the miRNAs that are encoded by multiple genes. Reads from such miRNAs would not be aligned and thus their expression would be underestimated if using the default <code>maxHits=1</code>. An example of such a multiply-encoded miRNA is mmu-miR-669a-5p, which has twelve exact copies in the mm10 genome assembly according to mirBase19.</p>
<div class="sourceCode" id="cb156"><pre class="downlit">
<span class="va">proj3</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/qAlign.html">qAlign</a></span><span class="op">(</span><span class="va">sampleFile2</span>, <span class="va">genomeFile</span>, maxHits<span class="op">=</span><span class="fl">50</span><span class="op">)</span></pre></div>
<pre><code>## alignment files missing - need to:</code></pre>
<pre><code>##     create 1 genomic alignment(s)</code></pre>
<pre><code>## Testing the compute nodes...OK
## Loading QuasR on the compute nodes...OK
## Available cores:
## Mac-1602493909147.local: 1
## Performing genomic alignments for 1 samples. See progress in the log file:
## /Users/runner/work/QuasR/QuasR/vignettes/QuasR_log_17d5313ae4f5.txt
## Genomic alignments have been created successfully</code></pre>
<div class="sourceCode" id="cb160"><pre class="downlit">
<span class="fu"><a href="../reference/alignmentStats.html">alignmentStats</a></span><span class="op">(</span><span class="va">proj3</span><span class="op">)</span></pre></div>
<pre><code>##               seqlength mapped unmapped
## miRNAs:genome     95000   1000        0</code></pre>
<p>A more detailed picture of the experiments’ quality can be obtained using <code><a href="../reference/qQCReport.html">qQCReport(proj3, "qcreport.pdf")</a></code> or similar (see also section @ref(qQCReport)).</p>
</div>
<div id="quantification-of-small-rna-mirna-reads" class="section level3">
<h3 class="hasAnchor">
<a href="#quantification-of-small-rna-mirna-reads" class="anchor"></a>Quantification of small RNA (miRNA) reads</h3>
<p>As with other experiment types, miRNAs are quantified using <code>qCount</code>. For this purpose, we first construct a query <code>GRanges</code> object with the genomic locations of mature miRNAs. The locations can be obtained from the <em><a href="https://bioconductor.org/packages/3.12/mirbase.db">mirbase.db</a></em> package, or directly from the species-specific gff files provided by the mirBase database (e.g. (<a href="ftp://mirbase.org/pub/mirbase/19/genomes/mmu.gff3" class="uri">ftp://mirbase.org/pub/mirbase/19/genomes/mmu.gff3</a>)). For the purpose of this example, the <em><a href="https://bioconductor.org/packages/3.12/QuasR">QuasR</a></em> package provides a small gff file (<code>"mirbaseXX_qsr.gff3"</code>) that is formatted as the ones available from mirBase. The gff file contains both the locations of pre-miRNAs (hairpin precursors), as well as mature miRNAs. The two can be discriminated by their <code>"type"</code>:</p>
<div class="sourceCode" id="cb162"><pre class="downlit">
<span class="va">mirs</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/rtracklayer/man/export.html">import</a></span><span class="op">(</span><span class="st">"extdata/mirbaseXX_qsr.gff3"</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">mirs</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">mirs</span><span class="op">$</span><span class="va">Name</span>
<span class="va">preMirs</span> <span class="op">&lt;-</span> <span class="va">mirs</span><span class="op">[</span> <span class="va">mirs</span><span class="op">$</span><span class="va">type</span><span class="op">==</span><span class="st">"miRNA_primary_transcript"</span> <span class="op">]</span>
<span class="va">matureMirs</span> <span class="op">&lt;-</span> <span class="va">mirs</span><span class="op">[</span> <span class="va">mirs</span><span class="op">$</span><span class="va">type</span><span class="op">==</span><span class="st">"miRNA"</span> <span class="op">]</span></pre></div>
<p>Please note that the name attribute of the <code>GRanges</code> object must be set appropriately, so that <code>qCount</code> can identify a single mature miRNA sequence that is encoded by multiple loci (see below) by their identical names. In this example, there are no multiply-encoded mature miRNAs, but in a real sample, you can detect them for example with <code><a href="https://rdrr.io/r/base/table.html">table(names(mirs))</a></code>.</p>
<p>The <code>preMirs</code> and <code>matureMirs</code> could now be used as <code>query</code> in <code>qCount</code>. In practise however, miRNA seem to not always be processed with high accuracy. Many miRNA reads that start one or two bases earlier or later can be observed in real data, and also their length may vary for a few bases. This is the case for the synthetic miRNAs used in this example, whose lengthes and start positions have been sampled from a read data set:</p>
<div class="sourceCode" id="cb163"><pre class="downlit">
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="http://bioconductor.org/packages/Rsamtools">Rsamtools</a></span><span class="op">)</span>
<span class="va">alns</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/Rsamtools/man/scanBam.html">scanBam</a></span><span class="op">(</span><span class="fu"><a href="../reference/qProject-class.html">alignments</a></span><span class="op">(</span><span class="va">proj3</span><span class="op">)</span><span class="op">$</span><span class="va">genome</span><span class="op">$</span><span class="va">FileName</span>,
                param<span class="op">=</span><span class="fu"><a href="https://rdrr.io/pkg/Rsamtools/man/ScanBamParam-class.html">ScanBamParam</a></span><span class="op">(</span>what<span class="op">=</span><span class="fu"><a href="https://rdrr.io/pkg/Rsamtools/man/ScanBamParam-class.html">scanBamWhat</a></span><span class="op">(</span><span class="op">)</span>, which<span class="op">=</span><span class="va">preMirs</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span><span class="op">)</span><span class="op">)</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span>
<span class="va">alnsIR</span> <span class="op">&lt;-</span> <span class="fu">IRanges</span><span class="op">(</span>start<span class="op">=</span><span class="va">alns</span><span class="op">$</span><span class="va">pos</span> <span class="op">-</span> <span class="fu"><a href="https://rdrr.io/r/stats/start.html">start</a></span><span class="op">(</span><span class="va">preMirs</span><span class="op">)</span>, width<span class="op">=</span><span class="va">alns</span><span class="op">$</span><span class="va">qwidth</span><span class="op">)</span>
<span class="va">mp</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/graphics/barplot.html">barplot</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/vector.html">as.vector</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/GenomicRanges/man/coverage-methods.html">coverage</a></span><span class="op">(</span><span class="va">alnsIR</span><span class="op">)</span><span class="op">)</span>, names.arg<span class="op">=</span><span class="fl">1</span><span class="op">:</span><span class="fu"><a href="https://rdrr.io/r/base/Extremes.html">max</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/stats/start.html">end</a></span><span class="op">(</span><span class="va">alnsIR</span><span class="op">)</span><span class="op">)</span>,
              xlab<span class="op">=</span><span class="st">"Relative position in pre-miRNA"</span>,
              ylab<span class="op">=</span><span class="st">"Alignment coverage"</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/graphics/rect.html">rect</a></span><span class="op">(</span>xleft<span class="op">=</span><span class="va">mp</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/stats/start.html">start</a></span><span class="op">(</span><span class="va">matureMirs</span><span class="op">)</span><span class="op">-</span><span class="fu"><a href="https://rdrr.io/r/stats/start.html">start</a></span><span class="op">(</span><span class="va">preMirs</span><span class="op">)</span><span class="op">+</span><span class="fl">1</span>,<span class="fl">1</span><span class="op">]</span>, ybottom<span class="op">=</span><span class="op">-</span><span class="fu"><a href="https://rdrr.io/r/graphics/par.html">par</a></span><span class="op">(</span><span class="st">'cxy'</span><span class="op">)</span><span class="op">[</span><span class="fl">2</span><span class="op">]</span>,
     xright<span class="op">=</span><span class="va">mp</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/stats/start.html">end</a></span><span class="op">(</span><span class="va">matureMirs</span><span class="op">)</span><span class="op">-</span><span class="fu"><a href="https://rdrr.io/r/stats/start.html">start</a></span><span class="op">(</span><span class="va">preMirs</span><span class="op">)</span><span class="op">+</span><span class="fl">1</span>,<span class="fl">1</span><span class="op">]</span>, ytop<span class="op">=</span><span class="fl">0</span>,
     col<span class="op">=</span><span class="st">"#CCAA0088"</span>, border<span class="op">=</span><span class="cn">NA</span>, xpd<span class="op">=</span><span class="cn">NA</span><span class="op">)</span></pre></div>
<p><img src="QuasR_files/figure-html/miRNA_coverage-1.png" width="576"></p>
<p>By default, <code>qCount</code> will count alignments that have their 5’-end within the query region (see <code>selectReadPosition</code> argument). The 5’-end correspond to the lower (left) coordinate for alignments on the plus strand, and to the higher (right) coordinate for alignments on the minus strand. In order not to miss miRNAs that have a couple of extra or missing bases, we therefore construct a query window around the 5’-end of each mature miRNA, by adding three bases up- and downstream:</p>
<div class="sourceCode" id="cb164"><pre class="downlit">
<span class="va">matureMirsExtended</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/GenomicRanges/man/intra-range-methods.html">resize</a></span><span class="op">(</span><span class="va">matureMirs</span>, width<span class="op">=</span><span class="fl">1L</span>, fix<span class="op">=</span><span class="st">"start"</span><span class="op">)</span> <span class="op">+</span> <span class="fl">3L</span></pre></div>
<p>The resulting extended query is then used to quantify mature miRNAs. Multiple-encoded miRNAs will be represented by multiple ranges in <code>matureMirs</code> and <code>matureMirsExtended</code>, which have identical names. <code>qCount</code> will automatically sum all alignments from any of those regions and return a single number per sample and unique miRNA name.</p>
<div class="sourceCode" id="cb165"><pre class="downlit">
<span class="co"># quantify mature miRNAs</span>
<span class="va">cnt</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/qCount.html">qCount</a></span><span class="op">(</span><span class="va">proj3</span>, <span class="va">matureMirsExtended</span>, orientation<span class="op">=</span><span class="st">"same"</span><span class="op">)</span></pre></div>
<pre><code>## counting alignments...done</code></pre>
<div class="sourceCode" id="cb167"><pre class="downlit">
<span class="va">cnt</span></pre></div>
<pre><code>##                 width miRNAs
## qsr-miR-9876-5p     7     13
## qsr-miR-9876-3p     7    984</code></pre>
<div class="sourceCode" id="cb169"><pre class="downlit">
<span class="co"># quantify pre-miRNAs</span>
<span class="va">cnt</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/qCount.html">qCount</a></span><span class="op">(</span><span class="va">proj3</span>, <span class="va">preMirs</span>, orientation<span class="op">=</span><span class="st">"same"</span><span class="op">)</span></pre></div>
<pre><code>## counting alignments...done</code></pre>
<div class="sourceCode" id="cb171"><pre class="downlit">
<span class="va">cnt</span></pre></div>
<pre><code>##              width miRNAs
## qsr-mir-9876    75   1000</code></pre>
</div>
</div>
<div id="bis-seq-measuring-dna-methylation" class="section level2">
<h2 class="hasAnchor">
<a href="#bis-seq-measuring-dna-methylation" class="anchor"></a>Bis-seq: Measuring DNA methylation</h2>
<p>Sequencing of bisulfite-converted genomic DNA allows detection of methylated cytosines, which in mammalian genomes typically occur in the context of CpG dinucleotides. The treatment of DNA with bisulfite induces deamination of non-methylated cytosines, converting them to uracils. Sequencing and aligning of such bisulfite-converted DNA results in C-to-T mismatches. Both alignment of converted reads, as well as the interpretation of the alignments for calculation of methylation levels require specific approaches and are supported in <em><a href="https://bioconductor.org/packages/3.12/QuasR">QuasR</a></em> by <code>qAlign</code> (<code>bisulfite</code> argument, section @ref(qAlign)) and <code>qMeth</code> (section @ref(qMeth)), respectively.</p>
<p>We start the analysis by copying the example data files into the current working directly, into a subfolder called <code>"extdata"</code>. Then, bisulfite-specific alignment is selected in <code>qAlign</code> by setting <code>bisulfite</code> to <code>"dir"</code> for a directional experiment, or to <code>"undir"</code> for an undirectional Bis-seq experiment:</p>
<div class="sourceCode" id="cb173"><pre class="downlit">
<span class="fu"><a href="https://rdrr.io/r/base/files.html">file.copy</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span><span class="op">(</span>package<span class="op">=</span><span class="st">"QuasR"</span>, <span class="st">"extdata"</span><span class="op">)</span>, <span class="st">"."</span>, recursive<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span></pre></div>
<pre><code>## [1] TRUE</code></pre>
<div class="sourceCode" id="cb175"><pre class="downlit">
<span class="va">sampleFile</span> <span class="op">&lt;-</span> <span class="st">"extdata/samples_bis_single.txt"</span>
<span class="va">genomeFile</span> <span class="op">&lt;-</span> <span class="st">"extdata/hg19sub.fa"</span>

<span class="va">proj4</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/qAlign.html">qAlign</a></span><span class="op">(</span><span class="va">sampleFile</span>, <span class="va">genomeFile</span>, bisulfite<span class="op">=</span><span class="st">"dir"</span><span class="op">)</span></pre></div>
<pre><code>## alignment files missing - need to:</code></pre>
<pre><code>##     create alignment index for the genome</code></pre>
<pre><code>##     create 1 genomic alignment(s)</code></pre>
<pre><code>## Creating an RbowtieCtoT index for /Users/runner/work/QuasR/QuasR/vignettes/extdata/hg19sub.fa</code></pre>
<pre><code>## Finished creating index</code></pre>
<pre><code>## Testing the compute nodes...OK
## Loading QuasR on the compute nodes...OK
## Available cores:
## Mac-1602493909147.local: 1
## Performing genomic alignments for 1 samples. See progress in the log file:
## /Users/runner/work/QuasR/QuasR/vignettes/QuasR_log_17d5784287a9.txt
## Genomic alignments have been created successfully</code></pre>
<div class="sourceCode" id="cb182"><pre class="downlit">
<span class="va">proj4</span></pre></div>
<pre><code>## Project: qProject
##  Options   : maxHits         : 1
##              paired          : no
##              splicedAlignment: FALSE
##              bisulfite       : dir
##              snpFile         : none
##              geneAnnotation  : none
##  Aligner   : Rbowtie v1.29.1 (parameters: -k 2 --best --strata -v 2)
##  Genome    : /Users/runner/work/QuasR/QuasR/vignettes/extdata/hg19sub.fa (file)
## 
##  Reads     : 1 file, 1 sample (fasta format):
##    1. bis_1_1.fa.bz2  Sample1
## 
##  Genome alignments: directory: same as reads
##    1. bis_1_1_17d557d8a5ed.bam
## 
##  Aux. alignments: none</code></pre>
<p>The resulting alignments are not different from those of non-Bis-seq experiments, apart from the fact that they may contain many C-to-T (or A-to-G) mismatches that are not counted as mismatches when aligning the reads. The number of alignments in specific genomic regions could be quantified using <code>qCount</code> as with ChIP-seq or RNA-seq experiments. For quantification of methylation the <code>qMeth</code> function is used:</p>
<div class="sourceCode" id="cb184"><pre class="downlit">
<span class="va">meth</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/qMeth.html">qMeth</a></span><span class="op">(</span><span class="va">proj4</span>, mode<span class="op">=</span><span class="st">"CpGcomb"</span>, collapseBySample<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span>
<span class="va">meth</span></pre></div>
<pre><code>## GRanges object with 3110 ranges and 2 metadata columns:
##          seqnames      ranges strand | Sample1_T Sample1_M
##             &lt;Rle&gt;   &lt;IRanges&gt;  &lt;Rle&gt; | &lt;integer&gt; &lt;integer&gt;
##      [1]     chr1       19-20      * |         1         1
##      [2]     chr1       21-22      * |         1         1
##      [3]     chr1       54-55      * |         3         1
##      [4]     chr1       57-58      * |         3         0
##      [5]     chr1       80-81      * |         6         5
##      ...      ...         ...    ... .       ...       ...
##   [3106]     chr3 44957-44958      * |         8         7
##   [3107]     chr3 44977-44978      * |         5         3
##   [3108]     chr3 44981-44982      * |         4         3
##   [3109]     chr3 44989-44990      * |         1         1
##   [3110]     chr3 44993-44994      * |         1         1
##   -------
##   seqinfo: 3 sequences from an unspecified genome</code></pre>
<p>By default, <code>qMeth</code> quantifies methylation for all cytosines in CpG contexts, combining the data from plus and minus strands (<code>mode="CpGcomb"</code>). The results are returned as a <code>GRanges</code> object with coordinates of each CpG, and two metadata columns for each input sequence file in the <code>qProject</code> object. These two columns contain the total number of aligned reads that overlap a given CpG (C-to-C matches or C-to-T mismatches, suffix <code>_T</code> in the column name), and the number of read alignments that had a C-to-C match at that position (methylated events, suffix <code>_M</code>).</p>
<p>Independent of the number of alignments, the returned object will list all CpGs in the target genome including the ones that have zero coverage, unless you set <code>keepZero=FALSE</code>:</p>
<div class="sourceCode" id="cb186"><pre class="downlit">
<span class="va">chrs</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/Biostrings/man/XStringSet-io.html">readDNAStringSet</a></span><span class="op">(</span><span class="va">genomeFile</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/sum.html">sum</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/Biostrings/man/matchPattern.html">vcountPattern</a></span><span class="op">(</span><span class="st">"CG"</span>,<span class="va">chrs</span><span class="op">)</span><span class="op">)</span></pre></div>
<pre><code>## [1] 3110</code></pre>
<div class="sourceCode" id="cb188"><pre class="downlit">
<span class="fu"><a href="https://rdrr.io/r/base/length.html">length</a></span><span class="op">(</span><span class="fu"><a href="../reference/qMeth.html">qMeth</a></span><span class="op">(</span><span class="va">proj4</span><span class="op">)</span><span class="op">)</span></pre></div>
<pre><code>## [1] 3110</code></pre>
<div class="sourceCode" id="cb190"><pre class="downlit">
<span class="fu"><a href="https://rdrr.io/r/base/length.html">length</a></span><span class="op">(</span><span class="fu"><a href="../reference/qMeth.html">qMeth</a></span><span class="op">(</span><span class="va">proj4</span>, keepZero<span class="op">=</span><span class="cn">FALSE</span><span class="op">)</span><span class="op">)</span></pre></div>
<pre><code>## [1] 2929</code></pre>
<p>The fraction methylation can easily be obtained as the ratio between <code>_M</code> and <code>_T</code> columns:</p>
<div class="sourceCode" id="cb192"><pre class="downlit">
<span class="va">percMeth</span> <span class="op">&lt;-</span> <span class="fu">mcols</span><span class="op">(</span><span class="va">meth</span><span class="op">)</span><span class="op">[</span>,<span class="fl">2</span><span class="op">]</span> <span class="op">*</span><span class="fl">100</span> <span class="op">/</span><span class="fu">mcols</span><span class="op">(</span><span class="va">meth</span><span class="op">)</span><span class="op">[</span>,<span class="fl">1</span><span class="op">]</span>
<span class="fu"><a href="https://rdrr.io/r/base/summary.html">summary</a></span><span class="op">(</span><span class="va">percMeth</span><span class="op">)</span></pre></div>
<pre><code>##    Min. 1st Qu.  Median    Mean 3rd Qu.    Max.    NA's 
##     0.0    75.0    90.9    75.4   100.0   100.0     181</code></pre>
<div class="sourceCode" id="cb194"><pre class="downlit">
<span class="va">axisTrack</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/Gviz/man/GenomeAxisTrack-class.html">GenomeAxisTrack</a></span><span class="op">(</span><span class="op">)</span>
<span class="va">dTrack1</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/Gviz/man/DataTrack-class.html">DataTrack</a></span><span class="op">(</span>range<span class="op">=</span><span class="va">gr1</span>, name<span class="op">=</span><span class="st">"H3K4me3"</span>, type<span class="op">=</span><span class="st">"h"</span><span class="op">)</span>
<span class="va">dTrack2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/Gviz/man/DataTrack-class.html">DataTrack</a></span><span class="op">(</span>range<span class="op">=</span><span class="va">meth</span>, data<span class="op">=</span><span class="va">percMeth</span>,
                     name<span class="op">=</span><span class="st">"Methylation"</span>, type<span class="op">=</span><span class="st">"p"</span><span class="op">)</span>
<span class="va">txTrack</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/Gviz/man/GeneRegionTrack-class.html">GeneRegionTrack</a></span><span class="op">(</span><span class="va">txdb</span>, name<span class="op">=</span><span class="st">"Transcripts"</span>, showId<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/pkg/Gviz/man/plotTracks.html">plotTracks</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span><span class="va">axisTrack</span>, <span class="va">dTrack1</span>, <span class="va">dTrack2</span>, <span class="va">txTrack</span><span class="op">)</span>,
           chromosome<span class="op">=</span><span class="st">"chr3"</span>, extend.left<span class="op">=</span><span class="fl">1000</span><span class="op">)</span></pre></div>
<p><img src="QuasR_files/figure-html/Bis_visualize-1.png" width="768"></p>
<p>If <code>qMeth</code> is called without a <code>query</code> argument, it will by default return methylation states for each C or CpG in the genome. Using a <code>query</code> argument it is possible to restrict the analysis to specific genomic regions, and if using in addition <code>collapseByQueryRegion=TRUE</code>, the single base methylation states will further be combined for all C’s that are contained in the same query region:</p>
<div class="sourceCode" id="cb195"><pre class="downlit">
<span class="fu"><a href="../reference/qMeth.html">qMeth</a></span><span class="op">(</span><span class="va">proj4</span>, query<span class="op">=</span><span class="fu"><a href="https://rdrr.io/pkg/GenomicRanges/man/GRanges-class.html">GRanges</a></span><span class="op">(</span><span class="st">"chr1"</span>,<span class="fu">IRanges</span><span class="op">(</span>start<span class="op">=</span><span class="fl">31633</span>,width<span class="op">=</span><span class="fl">2</span><span class="op">)</span><span class="op">)</span>,
      collapseBySample<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span></pre></div>
<pre><code>## GRanges object with 1 range and 2 metadata columns:
##       seqnames      ranges strand | Sample1_T Sample1_M
##          &lt;Rle&gt;   &lt;IRanges&gt;  &lt;Rle&gt; | &lt;integer&gt; &lt;integer&gt;
##   [1]     chr1 31633-31634      * |        10         2
##   -------
##   seqinfo: 3 sequences from an unspecified genome</code></pre>
<div class="sourceCode" id="cb197"><pre class="downlit">
<span class="fu"><a href="../reference/qMeth.html">qMeth</a></span><span class="op">(</span><span class="va">proj4</span>, query<span class="op">=</span><span class="va">promRegSel</span>, collapseByQueryRegion<span class="op">=</span><span class="cn">TRUE</span>,
      collapseBySample<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span></pre></div>
<pre><code>## GRanges object with 12 ranges and 2 metadata columns:
##        seqnames      ranges strand | Sample1_T Sample1_M
##           &lt;Rle&gt;   &lt;IRanges&gt;  &lt;Rle&gt; | &lt;numeric&gt; &lt;numeric&gt;
##    [1]     chr1 31629-33128      + |       426        74
##    [2]     chr1   6452-7951      - |       388       244
##    [3]     chr1 14013-15512      - |       627       560
##    [4]     chr1 31882-33381      - |       522       232
##    [5]     chr2   1795-3294      + |       997       539
##    ...      ...         ...    ... .       ...       ...
##    [8]     chr3   1276-2775      + |       715       253
##    [9]     chr3 19069-20568      + |       253       204
##   [10]     chr3 26692-28191      + |       934       818
##   [11]     chr3 26834-28333      + |       934       777
##   [12]     chr3 13102-14601      - |       307       287
##   -------
##   seqinfo: 3 sequences from an unspecified genome</code></pre>
<p>Finally, <code>qMeth</code> allows the retrieval of methylation states for individual molecules (per alignment). This is done by using a <code>query</code> containing a single genomic region (typically small, such as a PCR amplicon) and setting <code>reportLevel="alignment"</code>. In that case, the return value of <code>qMeth</code> will be a list (over samples) of lists (with four elements giving the identities of alignment, C nucleotide, strand and the methylation state). See the documentation of <code>qMeth</code> for more details.</p>
</div>
<div id="alleleSpecificAnalysis" class="section level2">
<h2 class="hasAnchor">
<a href="#alleleSpecificAnalysis" class="anchor"></a>Allele-specific analysis</h2>
<p>All experiment types supported by <em><a href="https://bioconductor.org/packages/3.12/QuasR">QuasR</a></em> (ChIP-seq, RNA-seq and Bis-seq; only alignments to the genome, but not to auxiliaries) can also be analyzed in an allele-specific manner. For this, a file containing genomic location and the two alleles of known sequence polymorphisms has to be provided to the <code>snpFile</code> argument of <code>qAlign</code>. The file is in tab-delimited text format without a header and contains four columns with chromosome name, position, reference allele and alternative allele.</p>
<p>Below is an example of a SNP file, also available from <code><a href="https://rdrr.io/r/base/system.file.html">system.file(package="QuasR", "extdata", "hg19sub_snp.txt")</a></code>:</p>
<pre>
chr1    3199    C   T
chr1    3277    C   T
chr1    4162    C   T
chr1    4195    C   T
...
</pre>
<!--
`chr1` |  `8596` | `G` | `A`  
`chr1` | `18443` | `G` | `A`  
`chr1` | `18981` | `C` | `T`  
`chr1` | `19341` | `G` | `A`  
`...`  |         |     |  
-->
<p>For a given locus, either reference or alternative allele may but does not have to be identical to the sequence of the reference genome (<code>genomeFile</code> in this case). To avoid an alignment bias, all reads are aligned separately to each of the two new genomes, which <em><a href="https://bioconductor.org/packages/3.12/QuasR">QuasR</a></em> generates by <em>injecting</em> the SNPs listed in <code>snpFile</code> into the reference genome. Finally, the two alignment files are combined, only retaining the best alignment for each read. While this procedure takes more than twice as long as aligning against a single genome, it has the advantage to correctly align reads even in regions of high SNP density and has been shown to produce more accurate results.</p>
<p>While combining alignments, each read is classified into one of three groups (stored in the <code>bam</code> files under the <code>XV</code> tag):</p>
<ul>
<li>
<strong>R</strong>: the read aligned better to the <strong>reference</strong> genome<br>
</li>
<li>
<strong>U</strong>: the read aligned equally well to both genomes (<strong>unknown</strong> origin)<br>
</li>
<li>
<strong>A</strong>: the read aligned better to the <strong>alternative</strong> genome</li>
</ul>
<p>Using these alignment classifications, the <code>qCount</code> and <code>qMeth</code> functions will produce three counts instead of a single count; one for each class. The column names will be suffixed by <code>_R</code>, <code>_U</code> and <code>_A</code>.</p>
<p>The examples below use data provided with the <em><a href="https://bioconductor.org/packages/3.12/QuasR">QuasR</a></em> package, which is first copied to the current working directory, into a subfolder called <code>"extdata"</code>:</p>
<div class="sourceCode" id="cb199"><pre class="downlit">
<span class="fu"><a href="https://rdrr.io/r/base/files.html">file.copy</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span><span class="op">(</span>package<span class="op">=</span><span class="st">"QuasR"</span>, <span class="st">"extdata"</span><span class="op">)</span>, <span class="st">"."</span>, recursive<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span></pre></div>
<pre><code>## [1] TRUE</code></pre>
<p>The example below aligns the same reads that were also used in the ChIP-seq workflow (section @ref(ChIPseq)), but this time using a <code>snpFile</code>:</p>
<div class="sourceCode" id="cb201"><pre class="downlit">
<span class="va">sampleFile</span> <span class="op">&lt;-</span> <span class="st">"extdata/samples_chip_single.txt"</span>
<span class="va">genomeFile</span> <span class="op">&lt;-</span> <span class="st">"extdata/hg19sub.fa"</span>
<span class="va">snpFile</span> <span class="op">&lt;-</span> <span class="st">"extdata/hg19sub_snp.txt"</span>
<span class="va">proj1SNP</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/qAlign.html">qAlign</a></span><span class="op">(</span><span class="va">sampleFile</span>, genome<span class="op">=</span><span class="va">genomeFile</span>, snpFile<span class="op">=</span><span class="va">snpFile</span><span class="op">)</span></pre></div>
<pre><code>## alignment files missing - need to:</code></pre>
<pre><code>##     create alignment index for the genome</code></pre>
<pre><code>##     create 2 genomic alignment(s)</code></pre>
<pre><code>## Reading and processing the SNP file: /Users/runner/work/QuasR/QuasR/vignettes/extdata/hg19sub_snp.txt</code></pre>
<pre><code>## Creating the genome fasta file containing the SNPs: /Users/runner/work/QuasR/QuasR/vignettes/extdata/hg19sub_snp.txt.hg19sub.fa.R.fa</code></pre>
<pre><code>## Creating the genome fasta file containing the SNPs: /Users/runner/work/QuasR/QuasR/vignettes/extdata/hg19sub_snp.txt.hg19sub.fa.A.fa</code></pre>
<pre><code>## Creating a .fai file for the snp genome: /Users/runner/work/QuasR/QuasR/vignettes/extdata/hg19sub_snp.txt.hg19sub.fa.R.fa</code></pre>
<pre><code>## Creating a .fai file for the snp genome: /Users/runner/work/QuasR/QuasR/vignettes/extdata/hg19sub_snp.txt.hg19sub.fa.A.fa</code></pre>
<pre><code>## Creating an Rbowtie index for /Users/runner/work/QuasR/QuasR/vignettes/extdata/hg19sub_snp.txt.hg19sub.fa.R.fa</code></pre>
<pre><code>## Finished creating index</code></pre>
<pre><code>## Creating an Rbowtie index for /Users/runner/work/QuasR/QuasR/vignettes/extdata/hg19sub_snp.txt.hg19sub.fa.A.fa</code></pre>
<pre><code>## Finished creating index</code></pre>
<pre><code>## Testing the compute nodes...OK
## Loading QuasR on the compute nodes...OK
## Available cores:
## Mac-1602493909147.local: 1
## Performing genomic alignments for 2 samples. See progress in the log file:
## /Users/runner/work/QuasR/QuasR/vignettes/QuasR_log_17d517b19714.txt
## Genomic alignments have been created successfully</code></pre>
<div class="sourceCode" id="cb215"><pre class="downlit">
<span class="va">proj1SNP</span></pre></div>
<pre><code>## Project: qProject
##  Options   : maxHits         : 1
##              paired          : no
##              splicedAlignment: FALSE
##              bisulfite       : no
##              snpFile         : /Users/runner/work/QuasR/Quas.../hg19sub_snp.txt
##              geneAnnotation  : none
##  Aligner   : Rbowtie v1.29.1 (parameters: -k 2 --best --strata -v 2)
##  Genome    : /Users/runner/work/QuasR/QuasR/vignettes/extdata/hg19sub.fa (file)
## 
##  Reads     : 2 files, 2 samples (fastq format):
##    1. chip_1_1.fq.bz2  Sample1 (phred33)
##    2. chip_2_1.fq.bz2  Sample2 (phred33)
## 
##  Genome alignments: directory: same as reads
##    1. chip_1_1_17d5c35ae33.bam 
##    2. chip_2_1_17d5183f9c88.bam
## 
##  Aux. alignments: none</code></pre>
<p>Instead of one count per promoter region and sample, <code>qCount</code> now returns three (<code>promRegSel</code> was generated in the ChIP-seq example workflow):</p>
<div class="sourceCode" id="cb217"><pre class="downlit">
<span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="fu"><a href="../reference/qCount.html">qCount</a></span><span class="op">(</span><span class="va">proj1</span>, <span class="va">promRegSel</span><span class="op">)</span><span class="op">)</span></pre></div>
<pre><code>## counting alignments...done</code></pre>
<pre><code>##                 width Sample1 Sample2
## ENSG00000176022  1500     157     701
## ENSG00000186891  1500      22       5
## ENSG00000186827  1500      10       3
## ENSG00000078808  1500      73     558
## ENSG00000171863  1500      94     339
## ENSG00000252531  1500      59       9</code></pre>
<div class="sourceCode" id="cb220"><pre class="downlit">
<span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="fu"><a href="../reference/qCount.html">qCount</a></span><span class="op">(</span><span class="va">proj1SNP</span>, <span class="va">promRegSel</span><span class="op">)</span><span class="op">)</span></pre></div>
<pre><code>## counting alignments...done</code></pre>
<pre><code>##                 width Sample1_R Sample1_U Sample1_A Sample2_R Sample2_U
## ENSG00000176022  1500         0       133         0         0       559
## ENSG00000186891  1500         4        16         0         0         5
## ENSG00000186827  1500         2         8         0         0         2
## ENSG00000078808  1500         0        59         0         0       432
## ENSG00000171863  1500         4        78         0         8       263
## ENSG00000252531  1500         3        50         2         0         6
##                 Sample2_A
## ENSG00000176022         0
## ENSG00000186891         0
## ENSG00000186827         0
## ENSG00000078808         0
## ENSG00000171863         0
## ENSG00000252531         0</code></pre>
<p>The example below illustrates use of a <code>snpFile</code> for Bis-seq experiments. Similarly as for <code>qCount</code>, the count types are labeled by <code>R</code>, <code>U</code> and <code>A</code>. These labels are added to the total and methylated column suffixes <code>_T</code> and <code>_M</code>, resulting in a total of six instead of two counts per feature and sample:</p>
<div class="sourceCode" id="cb223"><pre class="downlit">
<span class="va">sampleFile</span> <span class="op">&lt;-</span> <span class="st">"extdata/samples_bis_single.txt"</span>
<span class="va">genomeFile</span> <span class="op">&lt;-</span> <span class="st">"extdata/hg19sub.fa"</span>
<span class="va">proj4SNP</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/qAlign.html">qAlign</a></span><span class="op">(</span><span class="va">sampleFile</span>, <span class="va">genomeFile</span>,
                   snpFile<span class="op">=</span><span class="va">snpFile</span>, bisulfite<span class="op">=</span><span class="st">"dir"</span><span class="op">)</span></pre></div>
<pre><code>## alignment files missing - need to:</code></pre>
<pre><code>##     create alignment index for the genome</code></pre>
<pre><code>##     create 1 genomic alignment(s)</code></pre>
<pre><code>## Creating an RbowtieCtoT index for /Users/runner/work/QuasR/QuasR/vignettes/extdata/hg19sub_snp.txt.hg19sub.fa.R.fa</code></pre>
<pre><code>## Finished creating index</code></pre>
<pre><code>## Creating an RbowtieCtoT index for /Users/runner/work/QuasR/QuasR/vignettes/extdata/hg19sub_snp.txt.hg19sub.fa.A.fa</code></pre>
<pre><code>## Finished creating index</code></pre>
<pre><code>## Testing the compute nodes...OK
## Loading QuasR on the compute nodes...OK
## Available cores:
## Mac-1602493909147.local: 1
## Performing genomic alignments for 1 samples. See progress in the log file:
## /Users/runner/work/QuasR/QuasR/vignettes/QuasR_log_17d511bc37ce.txt
## Genomic alignments have been created successfully</code></pre>
<div class="sourceCode" id="cb232"><pre class="downlit">
<span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="fu"><a href="../reference/qMeth.html">qMeth</a></span><span class="op">(</span><span class="va">proj4SNP</span>, mode<span class="op">=</span><span class="st">"CpGcomb"</span>, collapseBySample<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span></pre></div>
<pre><code>## GRanges object with 6 ranges and 6 metadata columns:
##       seqnames    ranges strand | Sample1_TR Sample1_MR Sample1_TU Sample1_MU
##          &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt; |  &lt;integer&gt;  &lt;integer&gt;  &lt;integer&gt;  &lt;integer&gt;
##   [1]     chr1     19-20      * |          0          0          1          1
##   [2]     chr1     21-22      * |          0          0          1          1
##   [3]     chr1     54-55      * |          0          0          3          1
##   [4]     chr1     57-58      * |          0          0          3          0
##   [5]     chr1     80-81      * |          0          0          6          5
##   [6]     chr1   103-104      * |          0          0          6          5
##       Sample1_TA Sample1_MA
##        &lt;integer&gt;  &lt;integer&gt;
##   [1]          0          0
##   [2]          0          0
##   [3]          0          0
##   [4]          0          0
##   [5]          0          0
##   [6]          0          0
##   -------
##   seqinfo: 3 sequences from an unspecified genome</code></pre>
</div>
</div>
<div id="description-of-individual-quasr-functions" class="section level1">
<h1 class="hasAnchor">
<a href="#description-of-individual-quasr-functions" class="anchor"></a>Description of Individual <em>QuasR</em> Functions</h1>
<p>Please refer to the <em><a href="https://bioconductor.org/packages/3.12/QuasR">QuasR</a></em> reference manual or the function documentation (e.g. using <code><a href="../reference/qAlign.html">?qAlign</a></code>) for a complete description of <em><a href="https://bioconductor.org/packages/3.12/QuasR">QuasR</a></em> functions. The descriptions provided below are meant to give an overview over all functions and summarize the purpose of each one.</p>
<div id="preprocessReads" class="section level2">
<h2 class="hasAnchor">
<a href="#preprocessReads" class="anchor"></a><code>preprocessReads</code>
</h2>
<p>The <code>preprocessReads</code> function can be used to prepare the input sequences before alignment to the reference genome, for example to filter out low quality reads unlikely to produce informative alignments. When working with paired-end experiments, the paired reads are expected to be contained in identical order in two separate files. For this reason, both reads of a pair are filtered out if any of the two reads fulfills the filtering criteria. The following types of filtering tasks can be performed (in the order as listed):</p>
<ol style="list-style-type: decimal">
<li>
<strong>Truncate reads</strong>: remove nucleotides from the start and/or end of each read.<br>
</li>
<li>
<strong>Trim adapters</strong>: remove nucleotides at the beginning and/or end of each read that match to a defined (adapter) sequence. The adapter trimming is done by calling <code>trimLRPatterns</code> from the <em><a href="https://bioconductor.org/packages/3.12/Biostrings">Biostrings</a></em> package <span class="citation">(Pages et al., n.d.)</span>.<br>
</li>
<li>
<strong>Filter out low quality reads</strong>: Filter out reads that fulfill any of the filtering criteria (contain more than <code>nBases</code> <code>N</code> bases, are shorter than <code>minLength</code> or have a dinucleotide complexity of less than <code>complexity</code>-times the average complexity of the human genome sequence).</li>
</ol>
<p>The dinucleotide complexity is calculated in bits as Shannon entropy using the following formula <span class="math inline">\(-\sum_i f_i \cdot \log_2 f_i\)</span>, where <span class="math inline">\(f_i\)</span> is the frequency of dinucleotide <span class="math inline">\(i\)</span> (<span class="math inline">\(i=1, 2, ..., 16\)</span>).</p>
</div>
<div id="qAlign" class="section level2">
<h2 class="hasAnchor">
<a href="#qAlign" class="anchor"></a><code>qAlign</code>
</h2>
<p><code>qAlign</code> is the function that generates alignment files in <code>bam</code> format, for all input sequence files listed in <code>sampleFile</code> (see section @ref(sampleFile)), against the reference genome (<code>genome</code> argument), and for reads that do not match to the reference genome, against one or several auxiliary target sequences (<code>auxiliaryFile</code>, see section @ref(auxiliaryFile)).</p>
<p>The reference genome can be provided either as a <code>fasta</code> sequence file or as a <code>BSgenome</code> package name (see section @ref(genome)). If a <code>BSgenome</code> package is not found in the installed packages but available from Bioconductor, it will be automatically downloaded.</p>
<p>The alignment program is set by <code>aligner</code>, and parameters by <code>maxHits</code>, <code>paired</code>, <code>splicedAlignment</code> and <code>alignmentParameter</code>. Currently, <code>aligner</code> can only be set to <code>"Rbowtie"</code>, which is a wrapper for <em>bowtie</em> <span class="citation">(Langmead et al. 2009)</span> and <em>SpliceMap</em> <span class="citation">(Au et al. 2010)</span>, or <code>"Rhisat2"</code>, which is a wrapper for <em>HISAT2</em> <span class="citation">(Kim, Langmead, and Salzberg 2015)</span>. When <code>aligner="Rbowtie"</code>, <em>SpliceMap</em> will be used if <code>splicedAlignment=TRUE</code> (not recommended anymore except for reproducing older analyses). However, it is recommended to create spliced alignment using <code>splicedAlignment=TRUE, aligner="Rhisat2"</code>, which will use the <em>HISAT2</em> aligner and typically leads to more sensistive alignments and shorter alignment times compared to <em>SpliceMap</em>. The alignment strategy is furthermore affected by the parameters <code>snpFile</code> (alignments to variant genomes containing sequence polymorphisms) and <code>bisulfite</code> (alignment of bisulfite-converted reads). Finally, <code>clObj</code> can be used to enable parallelized alignment, sorting and conversion to <code>bam</code> format.</p>
<p>For each input sequence file listed in <code>sampleFile</code>, one <code>bam</code> file with alignments to the reference genome will be generated, and an additional one for each auxiliary sequence file listed in <code>auxiliaryFile</code>. By default, these <code>bam</code> files are stored at the same location as the sequence files, unless a different location is specified under <code>alignmentsDir</code>. If compatible alignment files are found at this location, they will not be regenerated, which allows re-use of the same sequencing samples in multiple analysis projects by listing them in several project-specific <code>sampleFile</code>s.</p>
<p>The alignment process produces temporary files, such as decompressed input sequence files or raw alignment files before conversion to <code>bam</code> format, which can be several times the size of the input sequence files. These temporary files are stored in the directory specified by <code>cacheDir</code>, which defaults to the <strong>R</strong> process temporary directory returned by <code><a href="https://rdrr.io/r/base/tempfile.html">tempdir()</a></code>. The location of <code><a href="https://rdrr.io/r/base/tempfile.html">tempdir()</a></code> can be set using environment variables (see <code><a href="https://rdrr.io/r/base/tempfile.html">?tempdir</a></code>).</p>
<p><code>qAlign</code> returns a <code>qProject</code> object that contains all file names and paths, as well as all alignment parameters necessary for further analysis (see section @ref(qProject) for methods to access the information contained in a <code>qProject</code> object).</p>
</div>
<div id="qProject" class="section level2">
<h2 class="hasAnchor">
<a href="#qProject" class="anchor"></a><code>qProject</code> class</h2>
<p>The <code>qProject</code> objects are returned by <code>qAlign</code> and contain all information about a sequencing experiment needed for further analysis. It is the main argument passed to the functions that start with a <em>q</em> letter, such as <code>qCount</code>, <code>qQCReport</code> and <code>qExportWig</code>. Some information inside of a <code>qProject</code> object can be accessed by specific methods (in the examples below, <code>x</code> is a <code>qProject</code> object):</p>
<ul>
<li>
<code><a href="https://rdrr.io/r/base/length.html">length(x)</a></code> gets the number of input files.<br>
</li>
<li>
<code>genome(x)</code> gets the reference genome as a <code><a href="https://rdrr.io/r/base/character.html">character(1)</a></code>. The type of genome is stored as an attribute in <code><a href="https://rdrr.io/r/base/attr.html">attr(genome(x),"genomeFormat")</a></code>: <code>"BSgenome"</code> indicates that <code>genome(x)</code> refers to the name of a <em>BSgenome</em> package, <code>"file"</code> indicates that it contains the path and file name of a genome in <code>fasta</code> format.<br>
</li>
<li>
<code><a href="../reference/qProject-class.html">auxiliaries(x)</a></code> gets a <code>data.frame</code> with auxiliary target sequences. The <code>data.frame</code> has one row per auxiliary target file, and two columns “FileName” and “AuxName”.<br>
</li>
<li>
<code><a href="../reference/qProject-class.html">alignments(x)</a></code> gets a list with two elements <code>"genome"</code> and <code>"aux"</code>. <code>"genome"</code> contains a <code>data.frame</code> with <code><a href="https://rdrr.io/r/base/length.html">length(x)</a></code> rows and two columns <code>"FileName"</code> (containing the path to bam files with genomic alignments) and <code>"SampleName"</code>. <code>"aux"</code> contains a <code>data.frame</code> with one row per auxiliary target file (with auxiliary names as row names), and <code><a href="https://rdrr.io/r/base/length.html">length(x)</a></code> columns (one per input sequence file).<br>
</li>
<li>
<code>x[i]</code> returns a <code>qProject</code> object instance with <code>i</code> input files, where <code>i</code> can be an <code>NA</code>-free logical, numeric, or character vector.</li>
</ul>
</div>
<div id="qQCReport" class="section level2">
<h2 class="hasAnchor">
<a href="#qQCReport" class="anchor"></a><code>qQCReport</code>
</h2>
<p>The <code>qQCReport</code> function samples a random subset of sequences and alignments from each sample or input file and generates a series of diagnostic plots for estimating data quality. The available plots vary depending on the types of available input (fasta, fastq, bam files or <code>qProject</code> object; paired-end or single-end). The plots below show the currently available plots as produced by the ChIP-seq example in section @ref(ChIPseq) (except for the fragment size distributions which are based on an unspliced alignment of paired-end RNA seq reads):</p>
<ul>
<li><p><strong>Quality score boxplot</strong> shows the distribution of base quality values as a box plot for each position in the input sequence. The background color (green, orange or red) indicates ranges of high, intermediate and low qualities. <img src="QuasR_files/figure-html/qcplotsFig1-1.png" width="768"></p></li>
<li><p><strong>Nucleotide frequency</strong> plot shows the frequency of A, C, G, T and N bases by position in the read. <img src="QuasR_files/figure-html/qcplotsFig2-1.png" width="768"></p></li>
<li><p><strong>Duplication level</strong> plot shows for each sample the fraction of reads observed at different duplication levels (e.g. once, two-times, three-times, etc.). In addition, the most frequent sequences are listed. <img src="QuasR_files/figure-html/qcplotsFig3-1.png" width="768"></p></li>
<li><p><strong>Mapping statistics</strong> shows fractions of reads that were (un)mappable to the reference genome. <img src="QuasR_files/figure-html/qcplotsFig4-1.png" width="768"></p></li>
<li><p><strong>Library complexity</strong> shows fractions of unique read(-pair) alignment positions. Please note that this measure is not independent from the total number of reads in a library, and is best compared between libraries of similar sizes. <img src="QuasR_files/figure-html/qcplotsFig5-1.png" width="768"></p></li>
<li><p><strong>Mismatch frequency</strong> shows the frequency and position (relative to the read sequence) of mismatches in the alignments against the reference genome. <img src="QuasR_files/figure-html/qcplotsFig6-1.png" width="768"></p></li>
<li><p><strong>Mismatch types</strong> shows the frequency of read bases that caused mismatches in the alignments to the reference genome, separately for each genome base. <img src="QuasR_files/figure-html/qcplotsFig7-1.png" width="768"></p></li>
<li><p><strong>Fragment size</strong> shows the distribution of fragment sizes inferred from aligned read pairs.<br><img src="QuasR_files/figure-html/qcplotsFig8-1.png" width="768"></p></li>
</ul>
</div>
<div id="alignmentStats" class="section level2">
<h2 class="hasAnchor">
<a href="#alignmentStats" class="anchor"></a><code>alignmentStats</code>
</h2>
<p><code>alignmentStats</code> is comparable to the <code>idxstats</code> function from Samtools; it returns the size of the target sequence, as well as the number of mapped and unmapped reads that are contained in an indexed <code>bam</code> file. The function works for arguments of type <code>qProject</code>, as well as on a string with one or several <code>bam</code> file names. There is however a small difference in the two that is illustrated in the following example, which uses the <code>qProject</code> object from the ChIP-seq workflow:</p>
<div class="sourceCode" id="cb234"><pre class="downlit">
<span class="co"># using bam files</span>
<span class="fu"><a href="../reference/alignmentStats.html">alignmentStats</a></span><span class="op">(</span><span class="fu"><a href="../reference/qProject-class.html">alignments</a></span><span class="op">(</span><span class="va">proj1</span><span class="op">)</span><span class="op">$</span><span class="va">genome</span><span class="op">$</span><span class="va">FileName</span><span class="op">)</span></pre></div>
<pre><code>##                           seqlength mapped unmapped
## chip_1_1_17d52e6ecf3a.bam     95000   2339      258
## chip_2_1_17d56ce700a6.bam     95000   3609      505</code></pre>
<div class="sourceCode" id="cb236"><pre class="downlit">
<span class="fu"><a href="../reference/alignmentStats.html">alignmentStats</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/unlist.html">unlist</a></span><span class="op">(</span><span class="fu"><a href="../reference/qProject-class.html">alignments</a></span><span class="op">(</span><span class="va">proj1</span><span class="op">)</span><span class="op">$</span><span class="va">aux</span><span class="op">)</span><span class="op">)</span></pre></div>
<pre><code>##                           seqlength mapped unmapped
## chip_1_1_17d5702d27d9.bam      5386    251        0
## chip_2_1_17d524935118.bam      5386    493        0</code></pre>
<div class="sourceCode" id="cb238"><pre class="downlit">
<span class="co"># using a qProject object</span>
<span class="fu"><a href="../reference/alignmentStats.html">alignmentStats</a></span><span class="op">(</span><span class="va">proj1</span><span class="op">)</span></pre></div>
<pre><code>##                 seqlength mapped unmapped
## Sample1:genome      95000   2339      258
## Sample2:genome      95000   3609      505
## Sample1:phiX174      5386    251        7
## Sample2:phiX174      5386    493       12</code></pre>
<p>If calling <code>alignmentStats</code> on the bam files directly as in the first two expressions of the above example, the returned numbers correspond exactly to what you would obtain by the <code>idxstats</code> function from Samtools, only that the latter would report them separately for each target sequence, while <code>alignmentStats</code> sums them for each <code>bam</code> file. These numbers correctly state that there are zero unmapped reads in the auxiliary <code>bam</code> files. However, if calling <code>alignmentStats</code> on a <code>qProject</code> object, it will report 7 and 12 unmapped reads in the auxiliary <code>bam</code> files. This is because <code>alignmentStats</code> is aware that unmapped reads are removed from auxiliary <code>bam</code> files by <em><a href="https://bioconductor.org/packages/3.12/QuasR">QuasR</a></em>, but can be calculated from the total number of reads to be aligned to the auxiliary target, which equals the number of unmapped reads in the corresponding genomic <code>bam</code> file.</p>
</div>
<div id="qExportWig" class="section level2">
<h2 class="hasAnchor">
<a href="#qExportWig" class="anchor"></a><code>qExportWig</code>
</h2>
<p><code>qExportWig</code> creates fixed-step wig files (see (<a href="http://genome.ucsc.edu/goldenPath/help/wiggle.html" class="uri">http://genome.ucsc.edu/goldenPath/help/wiggle.html</a>) for format definition) from the genomic alignments contained in a <code>qProject</code> object. The <code>combine</code> argument controls if several input files are combined into a single multi-track wig file, or if they are exported as individual wig files. Alignments of single read experiments can be shifted towards there 3’-end using <code>shift</code>; paired-end alignments are automatically shifted by half the insert size. The resolution of the created wig file is defines by the <code>binsize</code> argument, and if <code>scaling=TRUE</code>, multiple alignment files in the <code>qProject</code> object are scaled by their total number of reads.</p>
</div>
<div id="qCount" class="section level2">
<h2 class="hasAnchor">
<a href="#qCount" class="anchor"></a><code>qCount</code>
</h2>
<p><code>qCount</code> is the workhorse for counting alignments that overlap query regions. Usage and details on parameters can be obtained from the <code>qCount</code> function documentation. Two aspects that are of special importance are also discussed here:</p>
<div id="determination-of-overlap" class="section level3">
<h3 class="hasAnchor">
<a href="#determination-of-overlap" class="anchor"></a>Determination of overlap</h3>
<p>How an alignment overlap with a query region is defined can be controlled by the following arguments of <code>qCount</code>:</p>
<ul>
<li>
<code>selectReadPosition</code> specifies the read base that serves as a reference for overlaps with query regions. The alignment position of that base, eventually after shifting (see below), needs to be contained in the query region for an overlap. <code>selectReadPosition</code> can be set to <code>"start"</code> (the default) or <code>"end"</code> , which refer to the biological start (5’-end) and end (3’-end) of the read. For example, the <code>"start"</code> of a read aligned to the plus strand is the leftmost base in the alignment (the one with the lowest coordinate), and the <code>"end"</code> of a read aligned to the minus strand is also its leftmost base in the alignment.<br>
</li>
<li>
<code>shift</code> allows shifting of alignments towards their 3’-end prior to overlap determination and counting. This can be helpful to increase resolution of ChIP-seq experiments by moving alignments by half the immuno-precipitated fragment size towards the middle of fragments. <code>shift</code> can either contain <code>"integer"</code> values that specify the shift size, or for paired-end experiments, it can be set to the keyword <code>"halfInsert"</code>, which will estimate the true fragment size from the distance between aligned read pairs and shift the alignments accordingly.<br>
</li>
<li>
<code>orientation</code> controls the interpretation of alignment strand relative to the strand of the query region. The default value <code>"any"</code> will count all overlapping alignments, irrespective of the strand. This setting is for example used in an unstranded RNA-seq experiment where both sense and antisense reads are generated from an mRNA. A value of <code>"same"</code> will only count the alignments on the same strand as the query region (e.g. in a stranded RNA-seq experiment), and <code>"opposite"</code> will only count the alignments on the opposite strand from the query region (e.g. to quantify anti-sense transcription in a stranded RNA-seq experiment).<br>
</li>
<li>
<code>useRead</code> only applies to paired-end experiments and allows to quantify either all alignments (<code>useRead="any"</code>), or only the first (<code>useRead="first"</code>) or last (<code>useRead="last"</code>) read from each read pair or read group. Note that for <code>useRead="any"</code> (the default), an alignment pair that is fully contained within a query region will contribute two counts to the value of that region.<br>
</li>
<li>
<code>includeSpliced</code>: When set to <code>FALSE</code>, spliced alignments will be excluded from the quantification. This could be useful for example to avoid redundant counting of reads when the spliced alignments are quantified separately using <code>reportLevel="junction"</code>.</li>
</ul>
</div>
<div id="running-modes-of-qcount" class="section level3">
<h3 class="hasAnchor">
<a href="#running-modes-of-qcount" class="anchor"></a>Running modes of <code>qCount</code>
</h3>
<p>The features to be quantified are specified by the <code>query</code> argument. At the same time, the type of <code>query</code> selects the mode of quantification. <code>qCount</code> supports three different types of <code>query</code> arguments and implements three corresponding quantification types, which primarily differ in the way they deal with redundancy, such as query bases that are contained in more than one query region. A fourth quantification mode allows counting of alignments supporting exon-exon juctions:</p>
<ul>
<li>
<code>GRanges</code> query: Overlapping alignments are counted separately for each coordinate region in the query object. If multiple regions have identical names, their counts will be summed, counting each alignment only once even if it overlaps more than one of these regions. Alignments may however be counted more than once if they overlap multiple regions with different names. This mode is for example used to quantify ChIP-seq alignments in promoter regions (see section @ref(ChIPseq)), or gene expression levels in an RNA-seq experiment (using a ‘query’ with exon regions named by gene).<br>
</li>
<li>
<code>GRangesList</code> query: Alignments are counted and summed for each list element in the query object if they overlap with any of the regions contained in the list element. The order of the list elements defines a hierarchy for quantification: Alignment will only be counted for the first element (the one with the lowest index in the query) that they overlap, but not for any potential further list elements containing overlapping regions. This mode can be used to hierarchically and uniquely count (assign) each alignment to a one of several groups of regions (the elements in the query), for example to estimate the fractions of different classes of RNA in an RNA-seq experiment (rRNA, tRNA, snRNA, snoRNA, mRNA, etc.)<br>
</li>
<li>
<code>TxDb</code> query: Used to extract regions from annotation and report alignment counts depending on the value of the <code>reportLevel</code> argument. If <code>reportLevel="exon"</code>, alignments overlapping each exon in the query are counted. If <code>reportLevel="gene"</code>, alignment counts for all exons of a gene will be summed, counting each alignment only once even if it overlaps multiple annotated exons of a gene. These are useful to calculate exon or gene expression levels in RNA-seq experiments based on the annotation in a <code>TxDb</code> object. If <code>reportLevel="promoter"</code>, the <code>promoters</code> function from package <em><a href="https://bioconductor.org/packages/3.12/GenomicFeatures">GenomicFeatures</a></em> is used with default arguments to extract promoter regions around transcript start sites, e.g. to quantify alignments inf a ChIP-seq experiment.<br>
</li>
<li>any of the above or <code>NULL</code> for <code>reportLevel="junction"</code>: The <code>query</code> argument is ignored if <code>reportLevel</code> is set to <code>"junction"</code>, and <code>qCount</code> will count the number of alignments supporting each exon-exon junction detected in any of the samples in <code>proj</code>. The arguments <code>selectReadPosition</code>, <code>shift</code>, <code>orientation</code>, <code>useRead</code> and <code>mask</code> will have no effect in this quantification mode.</li>
</ul>
</div>
</div>
<div id="qProfile" class="section level2">
<h2 class="hasAnchor">
<a href="#qProfile" class="anchor"></a><code>qProfile</code>
</h2>
<p>The <code>qProfile</code> function differs from <code>qCount</code> in that it returns alignments counts relative to their position in the query region. Except for <code>upstream</code> and <code>downstream</code>, the arguments of <code>qProfile</code> and <code>qCount</code> are the same. This section will describe these two additional arguments; more details on the other arguments are available in section @ref(qCount) and from the <code>qProfile</code> function documentation.</p>
<p>The regions to be profiled are anchored by the biological start position, which are aligned at position zero in the return value. The biological start position is defined as <code><a href="https://rdrr.io/r/stats/start.html">start(query)</a></code> for regions on the plus strand and <code><a href="https://rdrr.io/r/stats/start.html">end(query)</a></code> for regions on the minus strand. The anchor positions are extended to the left and right sides by the number of bases indicated in the <code>upstream</code> and <code>downstream</code> arguments.</p>
<ul>
<li>
<code>upstream</code> indicates the number of bases upstream of the anchor position, which is on the left side of the anchor point for regions on the plus strand and on the right side for regions on the minus strand.<br>
</li>
<li>
<code>downstream</code> indicates the number of bases downstream of the anchor position, which is on the left side of the anchor point for regions on the plus strand and on the left side for regions on the minus strand.</li>
</ul>
<p>Be aware that query regions with a <code><a href="https://rdrr.io/r/base/Arithmetic.html">*</a></code> strand are handled the same way as regions on the plus strand.</p>
</div>
<div id="qMeth" class="section level2">
<h2 class="hasAnchor">
<a href="#qMeth" class="anchor"></a><code>qMeth</code>
</h2>
<p><code>qMeth</code> is used exclusively for Bis-seq experiments. In contrast to <code>qCount</code>, which counts the number of read alignments per query region, <code>qMeth</code> quantifies the number of C and T bases per cytosine in query regions, in order to determine methylation status.</p>
<p><code>qMeth</code> can be run in one of four modes, controlled by the <code>mode</code> argument:</p>
<ul>
<li>
<code>CpGcomb</code>: Only C’s in CpG context are considered. It is assumed that methylation status of the CpG base-pair on both strands is identical. Therefore, the total and methylated counts obtained for the C at position <span class="math inline">\(i\)</span> and the C on the opposite strand at position <span class="math inline">\(i+1\)</span> are summed.<br>
</li>
<li>
<code>CpG</code>: As with <code>CpGcomb</code>, only C’s in CpG context are quantified. However, counts from opposite strand are not summed, resulting in separate output values for C’s on both strands.<br>
</li>
<li>
<code>allC</code>: All C’s contained in query regions are quantified, keeping C’s from either strand separate. While this mode allows quantification of non-CpG methylation, it should be used with care, as the large result could use up available memory. In that case, a possible work-around is to divide the region of interest (e.g. the genome) into several regions (e.g. chromosomes) and call <code>qMeth</code> separately for each region.<br>
</li>
<li>
<code>var</code>: In this mode, only alignments on the opposite strand from C’s are analysed in order to collect evidence for sequence polymorphisms. Methylated C’s are hot-spots for C-to-T transitions, which in a Bis-seq experiment cannot be discriminated from completely unmethylated C’s. The information is however contained in alignments to the G from the opposite strand: Reads containing a G are consistent with a non-mutated C, and reads with an A support the presence of a sequence polymorphism. <code><a href="../reference/qMeth.html">qMeth(..., mode="var")</a></code> returns counts for total and matching bases for all C’s on both strands. A low fraction of matching bases is an indication of a mutation and can be used as a basis to identify mutated positions in the studied genome relative to the reference genome. Such positions should not be included in the quantification of methylation.</li>
</ul>
<p>When using <code>qMeth</code> in a allele-specific quantification (see also section @ref(alleleSpecificAnalysis), cytosines (or CpGs) that overlap a sequence polymorphism will not be quantified.</p>
</div>
</div>
<div id="session-information" class="section level1">
<h1 class="hasAnchor">
<a href="#session-information" class="anchor"></a>Session information</h1>
<p>The output in this vignette was produced under:</p>
<pre><code>## R version 4.0.2 (2020-06-22)
## Platform: x86_64-apple-darwin17.0 (64-bit)
## Running under: macOS Catalina 10.15.7
## 
## Matrix products: default
## BLAS:   /Library/Frameworks/R.framework/Versions/4.0/Resources/lib/libRblas.dylib
## LAPACK: /Library/Frameworks/R.framework/Versions/4.0/Resources/lib/libRlapack.dylib
## 
## locale:
## [1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8
## 
## attached base packages:
##  [1] grid      stats4    parallel  stats     graphics  grDevices utils    
##  [8] datasets  methods   base     
## 
## other attached packages:
##  [1] Gviz_1.33.2            GenomicFeatures_1.41.3 AnnotationDbi_1.51.3  
##  [4] Biobase_2.49.1         Rsamtools_2.5.3        BSgenome_1.57.7       
##  [7] rtracklayer_1.49.5     Biostrings_2.57.2      XVector_0.29.3        
## [10] QuasR_1.29.6           Rbowtie_1.29.1         GenomicRanges_1.41.6  
## [13] GenomeInfoDb_1.25.11   IRanges_2.23.10        S4Vectors_0.27.14     
## [16] BiocGenerics_0.35.4    BiocStyle_2.17.1      
## 
## loaded via a namespace (and not attached):
##   [1] colorspace_1.4-1            hwriter_1.3.2              
##   [3] ellipsis_0.3.1              rprojroot_1.3-2            
##   [5] biovizBase_1.37.0           htmlTable_2.1.0            
##   [7] base64enc_0.1-3             fs_1.5.0                   
##   [9] dichromat_2.0-0             rstudioapi_0.11            
##  [11] bit64_4.0.5                 xml2_1.3.2                 
##  [13] splines_4.0.2               knitr_1.30                 
##  [15] Formula_1.2-3               cluster_2.1.0              
##  [17] dbplyr_1.4.4                png_0.1-7                  
##  [19] BiocManager_1.30.10         compiler_4.0.2             
##  [21] httr_1.4.2                  backports_1.1.10           
##  [23] lazyeval_0.2.2              assertthat_0.2.1           
##  [25] Matrix_1.2-18               htmltools_0.5.0            
##  [27] prettyunits_1.1.1           tools_4.0.2                
##  [29] gtable_0.3.0                glue_1.4.2                 
##  [31] GenomeInfoDbData_1.2.4      dplyr_1.0.2                
##  [33] rappdirs_0.3.1              ShortRead_1.47.2           
##  [35] Rcpp_1.0.5                  pkgdown_1.6.1              
##  [37] vctrs_0.3.4                 xfun_0.18                  
##  [39] stringr_1.4.0               lifecycle_0.2.0            
##  [41] ensembldb_2.13.1            XML_3.99-0.5               
##  [43] zlibbioc_1.35.0             scales_1.1.1               
##  [45] VariantAnnotation_1.35.3    ProtGenerics_1.21.0        
##  [47] ragg_0.4.0                  hms_0.5.3                  
##  [49] MatrixGenerics_1.1.4        SummarizedExperiment_1.19.9
##  [51] AnnotationFilter_1.13.0     RColorBrewer_1.1-2         
##  [53] yaml_2.2.1                  curl_4.3                   
##  [55] memoise_1.1.0               gridExtra_2.3              
##  [57] ggplot2_3.3.2               biomaRt_2.45.5             
##  [59] rpart_4.1-15                latticeExtra_0.6-29        
##  [61] stringi_1.5.3               RSQLite_2.2.1              
##  [63] desc_1.2.0                  checkmate_2.0.0            
##  [65] BiocParallel_1.23.2         rlang_0.4.8                
##  [67] pkgconfig_2.0.3             systemfonts_0.3.2          
##  [69] GenomicFiles_1.25.0         matrixStats_0.57.0         
##  [71] bitops_1.0-6                evaluate_0.14              
##  [73] lattice_0.20-41             purrr_0.3.4                
##  [75] htmlwidgets_1.5.2           GenomicAlignments_1.25.3   
##  [77] bit_4.0.4                   tidyselect_1.1.0           
##  [79] magrittr_1.5                bookdown_0.20              
##  [81] R6_2.4.1                    generics_0.0.2             
##  [83] Hmisc_4.4-1                 DelayedArray_0.15.16       
##  [85] DBI_1.1.0                   pillar_1.4.6               
##  [87] foreign_0.8-80              survival_3.1-12            
##  [89] RCurl_1.98-1.2              nnet_7.3-14                
##  [91] tibble_3.0.3                crayon_1.3.4               
##  [93] BiocFileCache_1.13.1        rmarkdown_2.4              
##  [95] jpeg_0.1-8.1                progress_1.2.2             
##  [97] data.table_1.13.0           blob_1.2.1                 
##  [99] digest_0.6.25               textshaping_0.1.2          
## [101] openssl_1.4.3               munsell_0.5.0              
## [103] askpass_1.1</code></pre>
</div>
<div id="references" class="section level1 unnumbered">
<h1 class="hasAnchor">
<a href="#references" class="anchor"></a>References</h1>
<div id="refs" class="references">
<div id="ref-DESeq">
<p>Anders, Simon, and Wolfgang Huber. 2010. “Differential Expression Analysis for Sequence Count Data.” <em>Genome Biology</em> 11: R106. <a href="https://doi.org/10.1186/gb-2010-11-10-r106">https://doi.org/10.1186/gb-2010-11-10-r106</a>.</p>
</div>
<div id="ref-DEXSeq">
<p>Anders, Simon, Alejandro Reyes, and Wolfgang Huber. 2012. “Detecting Differential Usage of Exons from RNA-Seq Data.” <em>Genome Research</em> 22: 2008–17. <a href="https://doi.org/10.1101/gr.133744.111">https://doi.org/10.1101/gr.133744.111</a>.</p>
</div>
<div id="ref-SpliceMap">
<p>Au, K. F., H. Jiang, L. Lin, Y. Xing, and W. H. Wong. 2010. “Detection of Splice Junctions from Paired-End RNA-Seq Data by SpliceMap.” <em>Nucleic Acids Research</em> 38 (14): 4570–8.</p>
</div>
<div id="ref-Dalgaard">
<p>Dalgaard, P. 2002. <em>Introductory Statistics with R</em>. Springer.</p>
</div>
<div id="ref-QuasR">
<p>Gaidatzis, D., A. Lerch, F. Hahne, and M. B. Stadler. 2015. “QuasR: Quantify and Annotate Short Reads in R.” <em>Bioinformatics</em> 31 (7): 1130–2. <a href="https://doi.org/10.1093/bioinformatics/btu781">https://doi.org/10.1093/bioinformatics/btu781</a>.</p>
</div>
<div id="ref-Rbowtie">
<p>Hahne, F., A. Lerch, and M. B. Stadler. 2012. “Rbowtie: A R Wrapper for Bowtie and SpliceMap Short Read Aligners.”</p>
</div>
<div id="ref-baySeq">
<p>Hardcastle, Thomas J, and Krystyna A Kelly. 2010. “baySeq: Empirical Bayesian Methods for Identifying Differential Expression in Sequence Count Data.” <em>BMC Bioinformatics</em> 11: 422. <a href="https://doi.org/10.1186/1471-2105-11-422">https://doi.org/10.1186/1471-2105-11-422</a>.</p>
</div>
<div id="ref-hisat2">
<p>Kim, D., B. Langmead, and S. L. Salzberg. 2015. “HISAT: A Fast Spliced Aligner with Low Memory Requirements.” <em>Nature Methods</em> 12: 357–60. <a href="https://doi.org/10.1038/nmeth.3317">https://doi.org/10.1038/nmeth.3317</a>.</p>
</div>
<div id="ref-bowtie">
<p>Langmead, B., C. Trapnell, M. Pop, and S. L. Salzberg. 2009. “Ultrafast and Memory-Efficient Alignment of Short DNA Sequences to the Human Genome.” <em>Genome Biology</em> 10 (3): R25.</p>
</div>
<div id="ref-IRanges">
<p>Lawrence, Michael, Wolfgang Huber, Hervé Pagès, Patrick Aboyoun, Marc Carlson, Robert Gentleman, Martin T. Morgan, and Vincent J. Carey. 2013. “Software for Computing and Annotating Genomic Ranges.” <em>PLoS Comput Biol</em> 9: e1003118. <a href="https://doi.org/doi:10.1371/journal.pcbi.1003118">https://doi.org/doi:10.1371/journal.pcbi.1003118</a>.</p>
</div>
<div id="ref-DESeq2">
<p>Love, M. I., W. Huber, and S. Anders. 2014. “Moderated Estimation of Fold Change and Dispersion for RNA-seq Data with DESeq2.” <em>Genome Biology</em> 15: 550. <a href="https://doi.org/10.1186/s13059-014-0550-8">https://doi.org/10.1186/s13059-014-0550-8</a>.</p>
</div>
<div id="ref-ShortRead">
<p>Morgan, Martin, Simon Anders, Michael Lawrence, Patrick Aboyoun, Hervé Pagès, and Robert Gentleman. 2009. “ShortRead: A Bioconductor Package for Input, Quality Assessment and Exploration of High-Throughput Sequence Data.” <em>Bioinformatics</em> 25: 2607–8. <a href="https://doi.org/10.1093/bioinformatics/btp450">https://doi.org/10.1093/bioinformatics/btp450</a>.</p>
</div>
<div id="ref-Biostrings">
<p>Pages, H., P. Aboyoun, R. Gentleman, and S. DebRoy. n.d. <em>Biostrings: String Objects Representing Biological Sequences, and Matching Algorithms</em>.</p>
</div>
<div id="ref-edgeR">
<p>Robinson, Mark D, Davis J McCarthy, and Gordon K Smyth. 2010. “edgeR: A Bioconductor Package for Differential Expression Analysis of Digital Gene Expression Data.” <em>Bioinformatics</em> 26: 139–40.</p>
</div>
<div id="ref-TCC">
<p>Sun, Jianqiang, Tomoaki Nishiyama, Kentaro Shimizu1, and Koji Kadota. 2013. “TCC: An R Package for Comparing Tag Count Data with Robust Normalization Strategies.” <em>BMC Bioinformatics</em> 14: 219. <a href="https://doi.org/doi:10.1186/1471-2105-14-219">https://doi.org/doi:10.1186/1471-2105-14-219</a>.</p>
</div>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p>Developed by Michael Stadler.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
